--- a/GNUmakefile
+++ b/GNUmakefile
@@ -37,13 +37,18 @@ ifdef DEBUG
 CFLAGS += $(DEBUG_CFLAGS)
 endif
 
+INC_PATH = -I $(PLATFORM_INC_PATH)
+LIB_PATH = -L $(PLATFORM_LIB_PATH)
+
+LIBS = -lbget
+
 all: $(LIB) $(DYN)
 
 $(LIB): $(OBJS)
 	$(AR) rcs $(LIB) $(OBJS)
 
 $(DYN): $(OBJS)
-	$(CC) -shared -o $(DYN) $(OBJS)
+	$(CC) -shared -o $(DYN) $(OBJS) $(LIB_PATH) $(LIBS)
 
 nommap: CFLAGS += -D EZXML_NOMMAP
 nommap: all
@@ -61,7 +66,7 @@ $(TEST): $(OBJS)
 ezxml.o: ezxml.h ezxml.c
 
 .c.o:
-	$(CC) $(CFLAGS) -c -o $@ $<
+	$(CC) $(CFLAGS) $(INC_PATH) $(LIB_PATH) $(LIBS) -c -o $@ $<
 
 clean:
-	$(RM) $(OBJS) $(LIB) $(TEST) *~
\ No newline at end of file
+	$(RM) $(OBJS) $(LIB) $(TEST) *~
