--- a/ezxml.c
+++ b/ezxml.c
@@ -846,10 +846,13 @@ ezxml_t ezxml_new(const char *name)
 }
 
 // inserts an existing tag into an ezxml structure
-ezxml_t ezxml_insert(ezxml_t xml, ezxml_t dest, size_t off)
+ezxml_t ezxml_insert(ezxml_t xml, ezxml_t dest, size_t off,int flag)
 {
     ezxml_t cur, prev, head;
 
+    if(flag)
+    	free(((ezxml_root_t)xml)->ent);
+
     xml->next = xml->sibling = xml->ordered = NULL;
     xml->off = off;
     xml->parent = dest;
@@ -900,7 +903,7 @@ ezxml_t ezxml_add_child(ezxml_t xml, con
     child->attr = EZXML_NIL;
     child->txt = "";
 
-    return ezxml_insert(child, xml, off);
+    return ezxml_insert(child, xml, off,0);
 }
 
 // sets the character content for the given tag and returns the tag
--- a/ezxml.h
+++ b/ezxml.h
@@ -151,7 +151,7 @@ ezxml_t ezxml_set_flag(ezxml_t xml, shor
 ezxml_t ezxml_cut(ezxml_t xml);
 
 // inserts an existing tag into an ezxml structure
-ezxml_t ezxml_insert(ezxml_t xml, ezxml_t dest, size_t off);
+ezxml_t ezxml_insert(ezxml_t xml, ezxml_t dest, size_t off,int flag);
 
 // Moves an existing tag to become a subtag of dest at the given offset from
 // the start of dest's character content. Returns the moved tag.
