use script to create a better .hgignore from .gitignore files

diff --git a/.hgignore b/.hgignore
--- a/.hgignore
+++ b/.hgignore
@@ -1,57 +1,59 @@
-syntax:glob
-.svn
-.hgsvn
-msg.txt
+# auto-generated from .gitignore files by build-hgignore.py
+syntax: re
+^\.hgignore$
+\.o$                                                         # '*.o' in ''
+\.DS_Store$                                                  # '.DS_Store' in ''
+\..*\.swp$                                                   # '.*.swp' in ''
+^env$                                                        # '/env' in ''
+^dl$                                                         # '/dl' in ''
+^\.config$                                                   # '/.config' in ''
+^\.config\.old$                                              # '/.config.old' in ''
+^bin$                                                        # '/bin' in ''
+^build_dir$                                                  # '/build_dir' in ''
+^staging_dir$                                                # '/staging_dir' in ''
+^tmp$                                                        # '/tmp' in ''
+^logs$                                                       # '/logs' in ''
+^feeds$                                                      # '/feeds' in ''
+^feeds\.conf$                                                # '/feeds.conf' in ''
+^files$                                                      # '/files' in ''
+^package/feeds$                                              # '/package/feeds' in ''
+^package/openwrt-packages$                                   # '/package/openwrt-packages' in ''
+\.orig$                                                      # '*.orig' in ''
+\.rej$                                                       # '*.rej' in ''
+~$                                                           # '*~' in ''
+\.emacs\.desktop.*$                                          # '.emacs.desktop*' in ''
+TAGS.*~$                                                     # 'TAGS*~' in ''
+git-src$                                                     # 'git-src' in ''
+docs/.*\.log$                                                # '*.log' in 'docs'
+docs/.*\.aux$                                                # '*.aux' in 'docs'
+docs/.*\.toc$                                                # '*.toc' in 'docs'
+docs/.*\.out$                                                # '*.out' in 'docs'
+docs/.*\.lg$                                                 # '*.lg' in 'docs'
+docs/.*\.dvi$                                                # '*.dvi' in 'docs'
+docs/.*\.idv$                                                # '*.idv' in 'docs'
+docs/.*\.4ct$                                                # '*.4ct' in 'docs'
+docs/.*\.4tc$                                                # '*.4tc' in 'docs'
+docs/.*\.xref$                                               # '*.xref' in 'docs'
+docs/.*\.tmp$                                                # '*.tmp' in 'docs'
+docs/.*\.dvi$                                                # '*.dvi' in 'docs'
+docs/.*\.html$                                               # '*.html' in 'docs'
+docs/.*\.css$                                                # '*.css' in 'docs'
+docs/.*\.pdf$                                                # '*.pdf' in 'docs'
+scripts/config/.*\.o$                                        # '*.o' in 'scripts/config'
+scripts/config/conf.*$                                       # 'conf*' in 'scripts/config'
+# (ignored)                                                  # '!conf*.c' in 'scripts/config'
+# (ignored)                                                  # '!conf*.h' in 'scripts/config'
+scripts/config/mconf.*$                                      # 'mconf*' in 'scripts/config'
+# (ignored)                                                  # '!mconf*.c' in 'scripts/config'
+# (ignored)                                                  # '!mconf*.h' in 'scripts/config'
+scripts/config/mconf_check$                                  # 'mconf_check' in 'scripts/config'
+scripts/config/.*\..*\.c$                                    # '*.*.c' in 'scripts/config'
+scripts/config/lxdialog/lxdialog$                            # 'lxdialog' in 'scripts/config/lxdialog'
+scripts/config/lxdialog/.*\.o$                               # '*.o' in 'scripts/config/lxdialog'
 
-# These lines are suggested according to the svn:ignore property
-# Feel free to enable them by uncommenting them
-
-syntax:regexp
-
-^dl
-^.config.*
-^bin.*
-^build_dir
-^staging_dir
-^tmp
-^feeds
-^feeds.conf
-^toolchain/gdb/.*gdb
-^package/feeds
-^package/openwrt-packages
-^scripts/config/zconf.hash.c
-^scripts/config/conf
-^scripts/config/lex.zconf.c
-^scripts/config/mconf
-^scripts/config/zconf.tab.c
-^scripts/config/mconf_check
-^scripts/config/lxdialog/lxdialog
 
 # added manually:
+
 ^active_config
 ^logs
 
-syntax:glob
-
-docs/*.log
-docs/*.aux
-docs/*.toc
-docs/*.out
-docs/*.lg
-docs/*.dvi
-docs/*.idv
-docs/*.4ct
-docs/*.4tc
-docs/*.xref
-docs/*.tmp
-docs/*.dvi
-docs/*.html
-docs/*.css
-docs/*.dvi
-docs/*.pdf
-docs/tex4ht.fls
-
-# added manually:
-*.o
-active_config
-logs
