--- wpa_supplicant-0.5.11/Makefile	2010-08-25 23:58:36.415806362 +0530
+++ wpa_supplicant-0.5.11.patched/Makefile	2010-08-26 00:08:39.174555807 +0530
@@ -10,6 +10,9 @@
 # Include directories for CVS version
 #CFLAGS += -I. -I../utils -I../hostapd
 
+LFLAGS = $(IFX_LDFLAGS)
+CFLAGS += $(IFX_CFLAGS)
+
 ALL=wpa_supplicant wpa_passphrase wpa_cli
 
 all: verify_config $(ALL) dynamic_eap_methods
@@ -786,21 +789,21 @@
 dynamic_eap_methods: $(EAPDYN)
 
 wpa_supplicant: .config $(OBJS)
-	$(LDO) $(LDFLAGS) -o wpa_supplicant $(OBJS) $(LIBS) $(EXTRALIBS)
+	$(LDO) $(LDFLAGS) -o wpa_supplicant $(OBJS) $(LIBS) $(EXTRALIBS) $(LFLAGS)
 	if [ "$(STRIP)" ]; then cp wpa_supplicant wpa_supplicant_debug && $(STRIP) wpa_supplicant; fi
 
 eapol_test: .config $(OBJS_t)
-	$(LDO) $(LDFLAGS) -o eapol_test $(OBJS_t) $(LIBS)
+	$(LDO) $(LDFLAGS) -o eapol_test $(OBJS_t) $(LIBS) $(LFLAGS)
 
 preauth_test: .config $(OBJS_t2) 
-	$(LDO) $(LDFLAGS) -o preauth_test $(OBJS_t2) $(LIBS)
+	$(LDO) $(LDFLAGS) -o preauth_test $(OBJS_t2) $(LIBS) $(LFLAGS)
 
 wpa_passphrase: $(OBJS_p)
-	$(LDO) $(LDFLAGS) -o wpa_passphrase $(OBJS_p) $(LIBS_p)
+	$(LDO) $(LDFLAGS) -o wpa_passphrase $(OBJS_p) $(LIBS_p) $(LFLAGS)
 	if [ "$(STRIP)" ]; then cp wpa_passphrase wpa_passphrase_debug && $(STRIP) wpa_passphrase; fi
 
 wpa_cli: $(OBJS_c)
-	$(LDO) $(LDFLAGS) -o wpa_cli $(OBJS_c) $(LIBS_c)
+	$(LDO) $(LDFLAGS) -o wpa_cli $(OBJS_c) $(LIBS_c) $(LFLAGS)
 	if [ "$(STRIP)" ]; then cp wpa_cli wpa_cli_debug && $(STRIP) wpa_cli; fi
 
 OBJSa=asn1_test.o asn1.o x509v3.o common.o os_unix.o \
