--- a/lib/autofs.c
+++ b/lib/autofs.c
@@ -145,16 +145,17 @@
 	fds[0].fd = fdout;
 	fds[0].events = POLLIN;
 
-	while(1)
+	for(;;)
 	{
-		if(poll(fds, 2, 1000) == -1)
+		int res = poll(fds, 1, -1);
+		if(res == -1)
 		{
 			if (errno == EINTR)
 				continue;
 			log_printf("failed while trying to read packet from kernel\n");
 			return -1;
 		}
-		if(fds[0].revents & POLLIN)
+		else if ( res > 0 && fds[0].revents & POLLIN)
 			return fullread(pkt, sizeof(*pkt));
 	}
 }
