diff -Naur mountd-0.1/lib/mount.c mountd-0.1.ltq_patched/lib/mount.c
--- mountd-0.1/lib/mount.c	2009-10-03 16:24:57.000000000 +0530
+++ mountd-0.1.ltq_patched/lib/mount.c	2011-02-03 03:51:49.044752360 +0530
@@ -244,6 +244,7 @@
 		if(mount->fs == NTFS)
 		{
 			log_printf("ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
+			ret = system_printf("insmod /lib/modules/*/fuse.ko");
 			ret = system_printf("ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
 		}
 		exit(WEXITSTATUS(ret));
@@ -521,6 +522,7 @@
 			rmdir(tmp);
 			snprintf(tmp, 64, "%s%s", uci_path, mount->name);
 			unlink(tmp);
+			system_printf("rmmod fuse");
 			mount_dump_uci_state();
 		}
 	}
