diff -Naur mountd-0.1/lib/mount.c mountd-0.1.ltq_patched/lib/mount.c
--- mountd-0.1/lib/mount.c	2011-02-28 17:40:42.025279307 +0530
+++ mountd-0.1.ltq_patched/lib/mount.c	2011-02-28 18:25:37.222784933 +0530
@@ -53,6 +53,7 @@
 	"EXT3",
 	"FAT",
 	"HFSPLUS",
+	"EFI", /* added EFI to align char array with FS enumerations */
 	"NTFS"
 };
 
@@ -91,6 +92,8 @@
 		ucix_add_option(ctx, mountd, q->serial, "vendor", q->vendor);
 		ucix_add_option(ctx, mountd, q->serial, "model", q->model);
 		ucix_add_option(ctx, mountd, q->serial, "rev", q->rev);
+		snprintf(t, 64, "name%d", atoi(&q->dev[3])); /* print mounted directory name also */
+		ucix_add_option(ctx, mountd, q->serial, t, q->name);
 		snprintf(t, 64, "size%d", atoi(&q->dev[3]));
 		ucix_add_option(ctx, mountd, q->serial, t, q->size);
 		if(q->fs > MBR && q->fs <= NTFS)
