--- mountd-0.1/lib/mount.c	2011-03-05 17:31:50.752789653 +0530
+++ mountd-0.1.ltq_patched/lib/mount.c	2011-03-05 17:30:53.762790208 +0530
@@ -226,8 +226,8 @@
 	{
 		if(mount->fs == FAT)
 		{
-			log_printf("mount -t vfat -o rw,uid=1000,gid=1000 /dev/%s %s", mount->dev, tmp);
-			ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000 /dev/%s %s", mount->dev, tmp);
+			log_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
+			ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
 		}
 		if(mount->fs == EXT3)
 		{
