--- mountd-0.1/lib/mount.c	2011-04-09 12:58:00.505277394 +0530
+++ mountd-0.1.ltq_patched/lib/mount.c	2011-04-11 17:03:02.134039904 +0530
@@ -136,6 +136,9 @@
 	char tmp[64], tmp2[64];
 	if(fs <= MBR || fs > NTFS)
 		return;
+
+	system_printf("/etc/usb_led.sh &");
+
 	mount  = malloc(sizeof(struct mount));
 	INIT_LIST_HEAD(&mount->list);
 	strncpy(mount->vendor, vendor, 64);
@@ -514,6 +517,9 @@
 {
 	struct mount *mount = mount_find(0, dev);
 	char tmp[256];
+
+	system_printf("/etc/usb_led.sh &");
+
 	if(mount)
 	{
 		if(mount->mounted)
