Index: mountd-0.1/include/fs.h
===================================================================
--- mountd-0.1.orig/include/fs.h	2009-10-03 02:27:04.000000000 +0530
+++ mountd-0.1/include/fs.h	2012-07-27 16:41:37.000000000 +0530
@@ -7,5 +7,6 @@
 #define EFI			7
 #define NTFS		8
 #define EXTENDED	9
+#define EXT4		10
 
 int detect_fs(char *device);
Index: mountd-0.1/lib/fs.c
===================================================================
--- mountd-0.1.orig/lib/fs.c	2009-10-03 02:56:22.000000000 +0530
+++ mountd-0.1/lib/fs.c	2012-07-27 16:41:37.000000000 +0530
@@ -91,9 +91,14 @@
 		goto out;
 	if(get_le_short(buffer + 56) == 0xEF53)
 	{
-		if((get_le_long(buffer + 96) & 0x0008)
-			|| (get_le_long(buffer + 92) & 0x0004))
-			ret = EXT3;
+		if(get_le_long(buffer + 92) & 0x0004)
+		{
+			if ((get_le_long(buffer + 96) < 0x0000040)
+				&& (get_le_long(buffer + 100) < 0x0000008))
+				ret = EXT3;
+			else
+				ret = EXT4;
+		}
 		else
 			ret = EXT2;
 	}
Index: mountd-0.1/lib/mount.c
===================================================================
--- mountd-0.1.orig/lib/mount.c	2012-07-27 16:37:27.000000000 +0530
+++ mountd-0.1/lib/mount.c	2012-07-27 16:52:39.000000000 +0530
@@ -53,8 +53,10 @@
 	"EXT3",
 	"FAT",
 	"HFSPLUS",
-	"EFI", /* added EFI to align char array with FS enumerations */
-	"NTFS"
+	"",
+	"NTFS",
+	"",
+	"EXT4"
 };
 
 #define MAX_MOUNTED		32
@@ -96,7 +98,7 @@
 		ucix_add_option(ctx, mountd, q->serial, t, q->name);
 		snprintf(t, 64, "size%d", atoi(&q->dev[3]));
 		ucix_add_option(ctx, mountd, q->serial, t, q->size);
-		if(q->fs > MBR && q->fs <= NTFS)
+		if(q->fs > MBR && q->fs <= EXT4)
 		{
 			snprintf(t, 64, "fs%d", atoi(&q->dev[3]));
 			ucix_add_option(ctx, mountd, q->serial, t, fs_names[q->fs]);
@@ -134,7 +136,7 @@
 {
 	struct mount *mount;
 	char tmp[64], tmp2[64];
-	if(fs <= MBR || fs > NTFS)
+	if(fs <= MBR || fs > EXT4)
 		return;
 
 	system_printf("/etc/usb_led.sh &");
@@ -153,7 +155,7 @@
 	mount->mounted = 0;
 	mount->fs = fs;
 	list_add(&mount->list, &mounts);
-	if((!mount->ignore) && (mount->fs > MBR) && (mount->fs <= NTFS))
+	if((!mount->ignore) && (mount->fs > MBR) && (mount->fs <= EXT4))
 	{
 		log_printf("new mount : %s -> %s (%s)\n", name, dev, fs_names[mount->fs]);
 		snprintf(tmp, 64, "%s%s", uci_path, name);
@@ -232,6 +234,11 @@
 			log_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
 		}
+		if(mount->fs == EXT4)
+		{
+			log_printf("mount -t ext4 -o rw,defaults /dev/%s %s", mount->dev, tmp);
+			ret = system_printf("mount -t ext4 -o rw,defaults /dev/%s %s", mount->dev, tmp);
+		}
 		if(mount->fs == EXT3)
 		{
 			log_printf("mount -t ext3 -o rw,defaults /dev/%s %s", mount->dev, tmp);
