diff -Nuar mountd-0.1/include/fs.h modif_mountd-0.1/include/fs.h
--- mountd-0.1/include/fs.h	2012-08-31 18:36:50.401970000 +0530
+++ modif_mountd-0.1/include/fs.h	2012-08-31 17:25:48.024354000 +0530
@@ -4,9 +4,10 @@
 #define EXT3		4
 #define FAT			5
 #define HFSPLUS		6
-#define EFI			7
-#define NTFS		8
-#define EXTENDED	9
-#define EXT4		10
+#define HFSPLUSJOURNAL	7
+#define EFI			8
+#define NTFS		9
+#define EXTENDED	10
+#define EXT4		11
 
 int detect_fs(char *device);
diff -Nuar mountd-0.1/lib/fs.c modif_mountd-0.1/lib/fs.c
--- mountd-0.1/lib/fs.c	2012-08-31 18:36:50.404969000 +0530
+++ modif_mountd-0.1/lib/fs.c	2012-08-31 18:48:29.180961000 +0530
@@ -132,16 +132,25 @@
 	unsigned short magic;
 	unsigned int journal;
 	unsigned char *buffer = (unsigned char*)malloc(HFSPLUS_BUF_SIZE);
+	/* HFS+ partition starts with 1024 reserved bytes for boot code
+	 * followed by the Volume Header structure. The first 4 bytes of
+	 * the Volume Header structure are a signature and a version nuber.
+	 * The magic nubmer 0x482b (HFS) or 0x4858 (HFS+) should be checked. 
+	 */
+	if(lseek(fd, 1024, SEEK_SET) == -1 ) {
+		goto out;
+	}
+
 	if(get_buffer(fd, buffer, 1024, HFSPLUS_BUF_SIZE) != 0)
 		goto out;
 	magic = get_be_short(buffer);
 	journal = get_be_long(buffer + 4) & HFSPLUS_VOL_JOURNALED;
-	if(magic == 0x482B)
+	if(magic == 0x482B || magic == 0x4858)
 	{
 		if(!journal)
 			ret = HFSPLUS;
-	//	else
-	//		ret = HFSPLUSJOURNAL;
+		else
+			ret = HFSPLUSJOURNAL;
 	}
 out:
 	free(buffer);
diff -Nuar mountd-0.1/lib/mount.c modif_mountd-0.1/lib/mount.c
--- mountd-0.1/lib/mount.c	2012-08-31 18:36:50.410969000 +0530
+++ modif_mountd-0.1/lib/mount.c	2012-08-31 18:46:58.749055000 +0530
@@ -53,6 +53,7 @@
 	"EXT3",
 	"FAT",
 	"HFSPLUS",
+	"HFSPLUSJOURNAL",
 	"",
 	"NTFS",
 	"",
@@ -254,7 +255,7 @@
 			log_printf("mount -t ext2 -o rw,defaults /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t ext2 -o rw,defaults /dev/%s %s", mount->dev, tmp);
 		}
-		if(mount->fs == HFSPLUS)
+		if(mount->fs == HFSPLUS || mount->fs == HFSPLUSJOURNAL)
 		{
 			ret = system_printf("insmod /lib/modules/*/nls_utf8.ko; insmod /lib/modules/*/hfsplus.ko");
 			log_printf("mount -t hfsplus -o rw,defaults,uid=1000,gid=1000 /dev/%s %s", mount->dev, tmp);
