Index: mountd-0.1/lib/mount.c
===================================================================
--- mountd-0.1.orig/lib/mount.c	2013-02-06 18:02:50.000000000 +0530
+++ mountd-0.1/lib/mount.c	2013-02-06 18:02:57.000000000 +0530
@@ -241,6 +241,7 @@
 				insmod /lib/modules/*/jbd2.ko; insmod /lib/modules/*/crc16.ko; insmod /lib/modules/*/ext4.ko");
 			log_printf("mount -t ext4 -o rw,defaults /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t ext4 -o rw,defaults /dev/%s %s", mount->dev, tmp);
+			system_printf("chmod 777 %s", tmp);
 		}
 		if(mount->fs == EXT3)
 		{
@@ -248,18 +249,21 @@
 				insmod /lib/modules/*/jbd.ko; insmod /lib/modules/*/ext3.ko");
 			log_printf("mount -t ext3 -o rw,defaults /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t ext3 -o rw,defaults /dev/%s %s", mount->dev, tmp);
+			system_printf("chmod 777 %s", tmp);
 		}
 		if(mount->fs == EXT2)
 		{
 			ret = system_printf("insmod /lib/modules/*/mbcache.ko; insmod /lib/modules/*/ext2.ko");
 			log_printf("mount -t ext2 -o rw,defaults /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t ext2 -o rw,defaults /dev/%s %s", mount->dev, tmp);
+			system_printf("chmod 777 %s", tmp);
 		}
 		if(mount->fs == HFSPLUS || mount->fs == HFSPLUSJOURNAL)
 		{
 			ret = system_printf("insmod /lib/modules/*/nls_utf8.ko; insmod /lib/modules/*/hfsplus.ko");
-			log_printf("mount -t hfsplus -o rw,defaults,uid=1000,gid=1000 /dev/%s %s", mount->dev, tmp);
-			ret = system_printf("mount -t hfsplus -o rw,defaults,uid=1000,gid=1000 /dev/%s %s", mount->dev, tmp);
+			log_printf("mount -t hfsplus -o rw,defaults,uid=1000,gid=1000,umask=000,force /dev/%s %s", mount->dev, tmp);
+			ret = system_printf("mount -t hfsplus -o rw,defaults,uid=1000,gid=1000,umask=000,force /dev/%s %s", mount->dev, tmp);
+			system_printf("chmod 777 %s", tmp);
 		}
 		if(mount->fs == NTFS)
 		{
