--- a/lib/mount.c
+++ b/lib/mount.c
@@ -269,9 +269,16 @@ int mount_new(char *path, char *dev)
 		}
 		if(mount->fs == NTFS)
 		{
-			log_printf("ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
-			ret = system_printf("insmod /lib/modules/*/fuse.ko");
-			ret = system_printf("nice -n 19 ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
+			if (system_printf("[ -f /lib/modules/*/tntfs.ko ]") == 0) {
+				ret = system_printf("insmod /lib/modules/*/nls_utf8.ko; insmod /lib/modules/*/tntfs.ko");
+				log_printf("mount -t tntfs -o umask=00 /dev/%s %s", mount->dev, tmp);
+				ret = system_printf("mount -t tntfs -o umask=00 /dev/%s %s", mount->dev, tmp);
+				system_printf("chmod 777 %s", tmp);
+			} else {
+				log_printf("ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
+				ret = system_printf("insmod /lib/modules/*/fuse.ko");
+				ret = system_printf("nice -n 19 ntfs-3g /dev/%s %s -o force", mount->dev, tmp);
+			}
 		}
 		exit(WEXITSTATUS(ret));
 	}
@@ -559,6 +566,7 @@ void mount_dev_del(char *dev)
 			system_printf("rmmod ext2; rmmod ext3; rmmod ext4");
 			system_printf("rmmod jbd; rmmod jbd2; rmmod mbcache; rmmod crc16");
 			system_printf("rmmod hfsplus; rmmod nls_utf8");
+			system_printf("rmmod tntfs; rmmod nls_utf8");
 			mount_dump_uci_state();
 		}
 	}
