--- a/lib/mount.c
+++ b/lib/mount.c
@@ -234,6 +234,8 @@ int mount_new(char *path, char *dev)
 	{
 		if(mount->fs == FAT)
 		{
+			ret = system_printf("insmod /lib/modules/*/nls_iso*.ko; insmod /lib/modules/*/nls_cp*.ko; \
+				insmod /lib/modules/*/fat.ko; insmod /lib/modules/*/vfat.ko");
 			log_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
 			ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
 		}
@@ -563,6 +565,7 @@ void mount_dev_del(char *dev)
 			snprintf(tmp, 64, "%s/%s", uci_path, mount->serial);
 			rmdir(tmp); /* Remove intermediate directory */
 			system_printf("rmmod fuse");
+			system_printf("rmmod vfat; rmmod fat");
 			system_printf("rmmod ext2; rmmod ext3; rmmod ext4");
 			system_printf("rmmod jbd; rmmod jbd2; rmmod mbcache; rmmod crc16");
 			system_printf("rmmod hfsplus; rmmod nls_utf8");
