--- a/lib/mount.c
+++ b/lib/mount.c
@@ -234,10 +234,18 @@ int mount_new(char *path, char *dev)
 	{
 		if(mount->fs == FAT)
 		{
-			ret = system_printf("insmod /lib/modules/*/nls_iso*.ko; insmod /lib/modules/*/nls_cp*.ko; \
-				insmod /lib/modules/*/fat.ko; insmod /lib/modules/*/vfat.ko");
-			log_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
-			ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
+			if (system_printf("[ -f /lib/modules/*/tfat.ko ]") == 0) {
+				ret = system_printf("insmod /lib/modules/*/nls_utf8.ko; insmod /lib/modules/*/nls_cp*.ko; \
+					insmod /lib/modules/*/tfat.ko");
+				log_printf("mount -t tfat -o iostreaming /dev/%s %s", mount->dev, tmp);
+				ret = system_printf("mount -t tfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
+				//system_printf("chmod 777 %s", tmp);
+			} else {
+				ret = system_printf("insmod /lib/modules/*/nls_iso*.ko; insmod /lib/modules/*/nls_cp*.ko; \
+					insmod /lib/modules/*/fat.ko; insmod /lib/modules/*/vfat.ko");
+				log_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
+				ret = system_printf("mount -t vfat -o rw,uid=1000,gid=1000,umask=000 /dev/%s %s", mount->dev, tmp);
+			}
 		}
 		if(mount->fs == EXT4)
 		{
@@ -569,7 +577,7 @@ void mount_dev_del(char *dev)
 			system_printf("rmmod ext2; rmmod ext3; rmmod ext4");
 			system_printf("rmmod jbd; rmmod jbd2; rmmod mbcache; rmmod crc16");
 			system_printf("rmmod hfsplus; rmmod nls_utf8");
-			system_printf("rmmod tntfs; rmmod nls_utf8");
+			system_printf("rmmod tntfs; rmmod tfat; rmmod nls_utf8; nls_cp437");
 			mount_dump_uci_state();
 		}
 	}
