--- a/lib/log.c
+++ b/lib/log.c
@@ -24,7 +24,7 @@ void log_printf(char *fmt, ...)
 	va_end(ap);
 
 	if(daemonize)
-		syslog(10, p);
+		syslog(10, "%s", p);
 	else
-		printf(p);
+		printf("%s", p);
 }
--- a/lib/autofs.c
+++ b/lib/autofs.c
@@ -189,7 +189,7 @@ void autofs_init(void)
 	p = ucix_get_option(ctx, "mountd", "mountd", "path");
 	ucix_cleanup(ctx);
 	if(p)
-		snprintf(uci_path, 31, p);
+		snprintf(uci_path, 31, "%s", p);
 	else
 		snprintf(uci_path, 31, "/tmp/mounts/");
 	uci_path[31] = '\0';
