--- a/lib/mount.c
+++ b/lib/mount.c
@@ -165,6 +165,7 @@ void mount_add_list(char *name, char *de
 		snprintf(tmp2, 64, "../.run/mountd/%s", dev); /* Use releative symlink for anonymous ftp to access */
 		symlink(tmp2, tmp);
 		mount_new("/mnt/usb/.run/mountd/", dev);
+		system_printf("ACTION=mount DEVICE=%s FSNAME=%s MOUNTPATH=%s/%s/%s /sbin/hotplug-call mount", dev, fs_names[mount->fs], uci_path, serial, name);
 	}
 }
 
@@ -581,6 +582,7 @@ void mount_dev_del(char *dev)
 			mount_dump_uci_state();
 		}
 	}
+	system_printf("ACTION=umount DEVICE=%s FSNAME=%s MOUNTPATH=%s/%s/%s /sbin/hotplug-call mount", mount->dev, fs_names[mount->fs], uci_path, mount->serial, mount->name);
 }
 
 void mount_dump_list(void)
