--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,46 @@
+
+2016-03-08:
+	inserting/removing of vfat for 3.x kernel
+
+2016-03-08:
+	Support for tntfs fs
+
+2016-03-08:
+	Support for tfat fs
+
+2016-03-08:
+	Hotplug Support for mounts
+
+2016-02-24:
+	GCC hardening fixes
+
+2013-02-06:
+	providing rw permissions to all users
+
+2013-02-06:
+	Changing the mount location
+
+2012-08-31:
+	Support for hfsplus fs automount
+
+2012-07-27:
+	Support for ext4 fs
+
+2012-07-27:
+	Support for automatic insert and remove of modules
+
+2011-07-20:
+	lwered the priority(19) fro ntfs fs mounting
+
+2011-04-11:
+	Invoking usb_led script
+
+2011-03-05:
+	Added umask for mounting vfat
+
+2011-02-28:
+	added EFI to align char array with FS enumerations
+
+2011-02-03:
+	inserting fuso.ko
+
