--- netatalk-3.0.1/configure.ac	2012-09-28 14:46:56.000000000 +0530
+++ netatalk-3.0.1.new/configure.ac	2016-03-30 12:40:35.308051637 +0530
@@ -73,7 +73,7 @@
 AC_CHECK_MEMBERS(struct tm.tm_gmtoff,,, [#include <time.h>])
 
 dnl these tests have been comfirmed to be needed in 2011
-AC_CHECK_FUNCS(backtrace_symbols dirfd getusershell pread pwrite pselect)
+AC_CHECK_FUNCS(dirfd getusershell pread pwrite pselect)
 AC_CHECK_FUNCS(setlinebuf strlcat strlcpy strnlen)
 AC_CHECK_FUNCS(mmap utime getpagesize) dnl needed by tbd
 
