Index: lxc-2.0.8/src/lxc/utils.h
===================================================================
--- lxc-2.0.8.orig/src/lxc/utils.h
+++ lxc-2.0.8/src/lxc/utils.h
@@ -52,11 +52,7 @@ extern int mkdir_p(const char *dir, mode
 extern char *get_rundir(void);
 
 /* Define getline() if missing from the C library */
-#ifndef HAVE_GETLINE
-#ifdef HAVE_FGETLN
 #include <../include/getline.h>
-#endif
-#endif
 
 /* Define setns() if missing from the C library */
 #ifndef HAVE_SETNS
