--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,88 @@
+
+2016-02-13:
+	Fix busybox for gcc security flags
+
+2014-12-18:
+	Fix for unaligned access during gunzip directly from MTD flash partition
+
+2014-11-19:
+	Supports -p <seconds> to set PRETIMEOUT of Bootcore WDT (Basically configures TIMER2 timeout). Specific to Bootcore and Not applicable to InterAptiv
+
+2014-05-19:
+	Fix for upgrade board id failure
+
+2014-04-13:
+	Enhanced watchdog user application to support running on multiple VPEs. [-P CPU] option added to specify the CPU on which it runs
+
+2013-08-6:
+	Fix for vconfig command failure
+
+2013-07-31:
+	Support for increasing bootchartd timeout
+
+2013-07-15:
+	Feeds fixes and updates for buildsystem update, 
+
+2013-07-15:
+	Feeds fixes and updates for buildsystem update, 
+
+2013-07-05:
+	Fix for compilation if devm is disabbled
+
+2013-06-19:
+	Support for DHCP lease change notify to devm
+
+2013-05-16:
+	Fix dhcp relay binding and server binding to alias interfaces
+
+2013-05-16:
+	Fix dhcp relay binding and server binding to alias interfaces
+
+2012-12-26:
+	Extract ACS URL
+
+2012-10-30:
+	Support added for image upgrade in tftp
+
+2012-05-30:
+	Support for ACS discovery on DHCP
+
+2012-02-14:
+	Fix DHCP server padding issue
+
+2012-01-27:
+	Fix udhcpd unicast offer message
+
+2011-10-16:
+	Support for udhcp bootp server
+
+2011-09-19:
+	Support for ip neighbor
+
+2011-07-10:
+	Support for Combine dhcp functionality relay to server, 
+
+2011-07-10:
+	Support for Combine dhcp functionality relay to server, 
+
+2011-06-30:
+	dhcpd - send nak if client is in INIT-REBOOT state
+
+2011-05-02:
+	Support for RFC5969 IPv6 Rapid Deployment (6RD)
+
+2011-04-18:
+	Supporting DHCP User class, DHCP_TR111 device association, inetd enhancement
+
+2011-03-15:
+	IPv6 DSLIte changes
+
+2010-12-29:
+	927-busybox-split.patch added
+
+2010-12-20:
+	Support for inetd class support.
+
+2010-07-25:
+	Fix busybox compilation on FC15
+
