Index: busybox-1.17.1/applets/ifx_install.sh
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ busybox-1.17.1/applets/ifx_install.sh	2010-12-29 23:36:31.564650996 +0530
@@ -0,0 +1,49 @@
+#!/bin/bash
+
+prefix=$1
+if [ "$prefix" = "" ]; then
+    echo "No installation directory, aborting."
+    exit 1;
+fi
+if [ "$2" = "--hardlinks" ]; then
+    linkopts="-f"
+else
+    linkopts="-fs"
+fi
+h=`sort busybox.links | uniq`
+
+
+rm -f $prefix/bin/busybox2 || exit 1
+mkdir -p $prefix/bin || exit 1
+install -m 755 busybox $prefix/bin/busybox2 || exit 1
+
+for i in $h ; do
+	appdir=`dirname $i`
+	mkdir -p $prefix/$appdir || exit 1
+	if [ "$2" = "--hardlinks" ]; then
+	    bb_path="$prefix/bin/busybox2"
+	else
+	    case "$appdir" in
+		/)
+		    bb_path="bin/busybox2"
+		;;
+		/bin)
+		    bb_path="busybox2"
+		;;
+		/sbin)
+		    bb_path="../bin/busybox2"
+		;;
+		/usr/bin|/usr/sbin)
+		    bb_path="../../bin/busybox2"
+		;;
+		*)
+		echo "Unknown installation directory: $appdir"
+		exit 1
+		;;
+	    esac
+	fi
+	echo "  $prefix$i -> $bb_path"
+	ln $linkopts $bb_path $prefix$i || exit 1
+done
+
+exit 0
