Index: busybox-1.17.1/networking/udhcp/dhcpd.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/dhcpd.c	2011-06-30 23:21:51.233797438 +0530
+++ busybox-1.17.1/networking/udhcp/dhcpd.c	2011-06-30 23:23:40.252797470 +0530
@@ -831,11 +831,12 @@
 				send_ACK(&packet, lease->lease_nip);
 				break;
 			}
-			if (server_id_opt) {
-				/* client was talking specifically to us.
-				 * "No, we don't have this IP for you". */
+			if (server_id_opt    /* client is in SELECTING state */
+			|| requested_opt /* client is in INIT-REBOOT state */
+			) {
+				 /* "No, we don't have this IP for you". */
 				send_NAK(&packet);
-			}
+			} /* else: client is in RENEWING or REBINDING, do not answer */
 
 #ifndef CONFIG_FEATURE_DEVM_DEVICE  // TR-111 for gateway
 #if ENABLE_FEATURE_DHCP_TR111_DEVICE_ASSOCIATION
