Index: busybox-1.17.1/networking/udhcp/dhcpd.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/dhcpd.c	2011-07-10 22:31:55.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/dhcpd.c	2011-07-10 22:53:54.956797008 +0530
@@ -882,15 +882,6 @@
 				/* client requested or configured IP matches the lease.
 				 * ACK it, and bump lease expiration time. */
 				send_ACK(&packet, lease->lease_nip);
-				break;
-			}
-			if (server_id_opt    /* client is in SELECTING state */
-			|| requested_opt /* client is in INIT-REBOOT state */
-			) {
-				 /* "No, we don't have this IP for you". */
-				send_NAK(&packet);
-			} /* else: client is in RENEWING or REBINDING, do not answer */
-
 #ifndef CONFIG_FEATURE_DEVM_DEVICE  // TR-111 for gateway
 #if ENABLE_FEATURE_DHCP_TR111_DEVICE_ASSOCIATION
                         log1(
@@ -923,6 +914,15 @@
                         }
 #endif
 #endif
+				break;
+			}
+			if (server_id_opt    /* client is in SELECTING state */
+			|| requested_opt /* client is in INIT-REBOOT state */
+			) {
+				 /* "No, we don't have this IP for you". */
+				send_NAK(&packet);
+			} /* else: client is in RENEWING or REBINDING, do not answer */
+
 			break;
 
 		case DHCPDECLINE:
