Index: busybox-1.17.1/networking/udhcp/common.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/common.c	2011-10-16 11:38:08.656481157 +0530
+++ busybox-1.17.1/networking/udhcp/common.c	2011-10-16 12:05:05.882739656 +0530
@@ -14,6 +14,10 @@
 unsigned dhcp_verbose;
 #endif
 
+#if defined CONFIG_FEATURE_UDHCPD_BOOTP
+uint8_t state[1];
+#endif
+
 const uint8_t MAC_BCAST_ADDR[6] ALIGN2 = {
 	0xff, 0xff, 0xff, 0xff, 0xff, 0xff
 };
@@ -217,7 +221,9 @@
 		FILE_FIELD101  = FILE_FIELD  * 0x101,
 		SNAME_FIELD101 = SNAME_FIELD * 0x101,
 	};
-
+#if defined CONFIG_FEATURE_UDHCPD_BOOTP
+if (packet->cookie == htonl(DHCP_MAGIC)) {
+#endif
 	/* option bytes: [code][len][data1][data2]..[dataLEN] */
 	optionptr = packet->options;
 	rem = sizeof(packet->options);
@@ -264,6 +270,13 @@
 		}
 		optionptr += len;
 	}
+#if defined CONFIG_FEATURE_UDHCPD_BOOTP
+	}
+	if (code == DHCP_MESSAGE_TYPE && packet->cookie != htonl(DHCP_MAGIC)) {
+		state[0] = BOOTREQUEST;
+		return state;
+	}
+#endif
 
 	/* log3 because udhcpc uses it a lot - very noisy */
 	log3("Option 0x%02x not found", code);
Index: busybox-1.17.1/networking/udhcp/packet.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/packet.c	2011-10-16 11:38:08.406477041 +0530
+++ busybox-1.17.1/networking/udhcp/packet.c	2011-10-16 12:05:16.566493731 +0530
@@ -97,14 +97,21 @@
 		return bytes; /* returns -1 */
 	}
 
+#if defined CONFIG_FEATURE_UDHCPD_BOOTP
+	if ((packet->cookie) && (packet->cookie != htonl(DHCP_MAGIC))) {
+#else
 	if (packet->cookie != htonl(DHCP_MAGIC)) {
+#endif
 		bb_info_msg("Packet with bad magic, ignoring");
 		return -2;
 	}
 	log1("Received a packet");
 	udhcp_dump_packet(packet);
-
+#if defined CONFIG_FEATURE_UDHCPD_BOOTP
+	if (packet->cookie == htonl(DHCP_MAGIC) && packet->op == BOOTREQUEST) {
+#else
 	if (packet->op == BOOTREQUEST) {
+#endif
 		vendor = udhcp_get_option(packet, DHCP_VENDOR);
 		if (vendor) {
 #if 0
Index: busybox-1.17.1/networking/udhcp/Config.src
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/Config.src	2011-10-16 11:38:08.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/Config.src	2011-10-16 12:06:05.332738948 +0530
@@ -19,6 +19,13 @@
 	help
 	  This Lantiq enahanement comines relay functionality to server.
 
+config FEATURE_UDHCPD_BOOTP
+	bool "bootp support for server"
+	default n
+	depends on UDHCPD
+	help
+	  This Lantiq enahanement adds basic bootp support into dhcp server.
+
 config DHCPRELAY
 	bool "dhcprelay"
 	default y
