Index: busybox-1.17.1/networking/udhcp/dhcpd.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/dhcpd.c	2012-01-27 17:30:48.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/dhcpd.c	2012-01-27 17:32:19.193338593 +0530
@@ -92,18 +92,19 @@
 	// We should never unicast to dhcp_pkt->yiaddr!
 	// dhcp_pkt->ciaddr, OTOH, comes from client's request packet,
 	// and can be used.
-
-	if (force_broadcast
-	 || (dhcp_pkt->flags & htons(BROADCAST_FLAG))
-	 || dhcp_pkt->ciaddr == 0
-	) {
+	if ( !force_broadcast && !(dhcp_pkt->flags & htons(BROADCAST_FLAG)) ) {
+		if ( dhcp_pkt->ciaddr ) {
+			log1("Unicasting packet to client ciaddr");
+			ciaddr = dhcp_pkt->ciaddr;
+		} else {
+			log1("Unicasting packet to client yiaddr");
+			ciaddr = dhcp_pkt->yiaddr;
+		}
+		chaddr = dhcp_pkt->chaddr;
+	} else {
 		log1("Broadcasting packet to client");
 		ciaddr = INADDR_BROADCAST;
 		chaddr = MAC_BCAST_ADDR;
-	} else {
-		log1("Unicasting packet to client ciaddr");
-		ciaddr = dhcp_pkt->ciaddr;
-		chaddr = dhcp_pkt->chaddr;
 	}
 
 #if ENABLE_FEATURE_DHCP_USER_CLASS
