Index: busybox-1.17.1/networking/udhcp/common.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/common.c	2012-12-26 17:59:17.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/common.c	2012-12-26 18:01:14.121032000 +0530
@@ -911,7 +911,9 @@
 		}
 		else{
 			string_len = strlen(msg);
-			memcpy(string, (void *)msg, string_len);
+			memcpy(string, (void *)msg+2, (string_len-2));
+			string[(string_len-3)]='\0';
+			string_len=strlen(string);
 		}
 		if(string_len < 0){
 			log1( "Failure in forminmg string data");
