Index: busybox-1.17.1/networking/udhcp/dhcpd.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/dhcpd.c	2013-05-16 15:44:42.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/dhcpd.c	2013-05-16 15:53:12.000000000 +0530
@@ -481,11 +481,6 @@
 	int classindex=0;
 	int i;
 #endif
-#if ENABLE_FEATURE_UDHCPD_RELAY
-	/* Socket for relay mode */
-	int relay_socket = -1;
-	int relay_fd = -1;
-#endif
 
 #if ENABLE_FEATURE_UDHCP_PORT
 	SERVER_PORT = 67;
@@ -627,36 +622,19 @@
 			server_socket = 0;
 #else
 		if (server_socket < 0) {
-			server_socket = udhcp_listen_socket(/*INADDR_ANY,*/ SERVER_PORT,
-					server_config.interface);
-		}
-#endif
 #if ENABLE_FEATURE_UDHCPD_RELAY
-		/* If the relay mode is set for this node, then we listen on the relay interface */
-		if (head->server_config_global.dhcpmode == DHCP_MODE_RELAY)
-		{
-			if(relay_socket < 0)
-			{
-				if(relay_fd < 0)
-				{
-#if ENABLE_FEATURE_DHCP_RUN_FROM_INETD
-					relay_fd = 0;
+			if (head->server_config_global.dhcpmode == DHCP_MODE_RELAY) {
+				server_socket = udhcp_listen_socket(/*INADDR_ANY,*/ SERVER_PORT,NULL);
+			} else {
+				server_socket = udhcp_listen_socket(/*INADDR_ANY,*/ SERVER_PORT,server_config.interface);
+			}
 #else
-					relay_fd = udhcp_listen_socket(SERVER_PORT,server_config.relayif);
+			server_socket = udhcp_listen_socket(/*INADDR_ANY,*/ SERVER_PORT,
+					server_config.interface);
 #endif
-				}
-				relay_socket= relay_fd;
-			}
 		}
 #endif
 		max_sock = udhcp_sp_fd_set(&rfds, server_socket);
-#if ENABLE_FEATURE_UDHCPD_RELAY
-		/* udhcpd_sp_fd_set does not work for more than one socket so we set manually */
-		if ( relay_socket > 0 ) {
-			FD_SET(relay_socket, &rfds);
-			max_sock = relay_socket;
-		}
-#endif
 		if (server_config.auto_time) {
 			tv.tv_sec = timeout_end - monotonic_sec();
 			tv.tv_usec = 0;
Index: busybox-1.17.1/networking/udhcp/socket.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/socket.c	2010-07-06 07:55:54.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/socket.c	2013-05-16 15:44:42.000000000 +0530
@@ -95,8 +95,10 @@
 		bb_perror_msg_and_die("SO_BROADCAST");
 
 	/* NB: bug 1032 says this doesn't work on ethernet aliases (ethN:M) */
-	if (setsockopt_bindtodevice(fd, inf))
-		xfunc_die(); /* warning is already printed */
+	if (inf != NULL ){
+		if (setsockopt_bindtodevice(fd, inf))
+			xfunc_die(); /* warning is already printed */
+	}
 
 	memset(&addr, 0, sizeof(addr));
 	addr.sin_family = AF_INET;
