Index: busybox-1.17.1/networking/udhcp/socket.c
===================================================================
--- busybox-1.17.1.orig/networking/udhcp/socket.c	2013-05-16 15:53:38.000000000 +0530
+++ busybox-1.17.1/networking/udhcp/socket.c	2013-05-16 16:10:09.000000000 +0530
@@ -86,6 +86,7 @@
 {
 	int fd;
 	struct sockaddr_in addr;
+	char *colon;
 
 	log1("Opening listen socket on *:%d %s", port, inf);
 	fd = xsocket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
@@ -94,10 +95,15 @@
 	if (setsockopt_broadcast(fd) == -1)
 		bb_perror_msg_and_die("SO_BROADCAST");
 
-	/* NB: bug 1032 says this doesn't work on ethernet aliases (ethN:M) */
+    /* SO_BINDTODEVICE doesn't work on ethernet aliases (ethN:M) */
 	if (inf != NULL ){
+		colon = strrchr(inf, ':');
+		if (colon)
+			*colon = '\0';
 		if (setsockopt_bindtodevice(fd, inf))
 			xfunc_die(); /* warning is already printed */
+		if (colon)
+			*colon = ':';
 	}
 
 	memset(&addr, 0, sizeof(addr));
