Index: busybox-1.17.1/archival/libunarchive/decompress_unzip.c
===================================================================
--- busybox-1.17.1.orig/archival/libunarchive/decompress_unzip.c	2010-07-06 07:55:53.000000000 +0530
+++ busybox-1.17.1/archival/libunarchive/decompress_unzip.c	2014-12-18 14:27:02.000000000 +0530
@@ -129,8 +129,9 @@
 #define gunzip_bk           (S()gunzip_bk          )
 #define to_read             (S()to_read            )
 // #define bytebuffer_max   (S()bytebuffer_max     )
-// Both gunzip and unzip can use constant buffer size now (16k):
-#define bytebuffer_max      0x4000
+// Both gunzip and unzip can use constant buffer size now (1k):
+// fix for unaligned access during gunzip directly from MTD flash partition
+#define bytebuffer_max      0x400
 #define bytebuffer          (S()bytebuffer         )
 #define bytebuffer_offset   (S()bytebuffer_offset  )
 #define bytebuffer_size     (S()bytebuffer_size    )
