diff -ur ethtool-2.6.36/ethtool.c ethtool-2.6.36-ltq/ethtool.c
--- ethtool-2.6.36/ethtool.c	2010-11-16 20:19:45.000000000 +0100
+++ ethtool-2.6.36-ltq/ethtool.c	2011-07-15 15:13:22.687611000 +0200
@@ -86,7 +86,9 @@
 static int do_gdrv(int fd, struct ifreq *ifr);
 static int do_gset(int fd, struct ifreq *ifr);
 static int do_sset(int fd, struct ifreq *ifr);
+#if 0
 static int do_gregs(int fd, struct ifreq *ifr);
+#endif
 static int do_nway_rst(int fd, struct ifreq *ifr);
 static int do_geeprom(int fd, struct ifreq *ifr);
 static int do_seeprom(int fd, struct ifreq *ifr);
@@ -1654,6 +1656,7 @@
 	}
 }
 
+#if 0
 static struct {
 	const char *name;
 	int (*func)(struct ethtool_drvinfo *info, struct ethtool_regs *regs);
@@ -1726,6 +1729,7 @@
 	fprintf(stdout, "\n\n");
 	return 0;
 }
+#endif
 
 static int dump_eeprom(struct ethtool_drvinfo *info, struct ethtool_eeprom *ee)
 {
@@ -1735,13 +1739,13 @@
 		fwrite(ee->data, 1, ee->len, stdout);
 		return 0;
 	}
-
+#if 0
 	if (!strncmp("natsemi", info->driver, ETHTOOL_BUSINFO_LEN)) {
 		return natsemi_dump_eeprom(info, ee);
 	} else if (!strncmp("tg3", info->driver, ETHTOOL_BUSINFO_LEN)) {
 		return tg3_dump_eeprom(info, ee);
 	}
-
+#endif
 	fprintf(stdout, "Offset\t\tValues\n");
 	fprintf(stdout, "------\t\t------");
 	for (i = 0; i < ee->len; i++) {
@@ -1973,8 +1977,10 @@
 		return do_gset(fd, &ifr);
 	} else if (mode == MODE_SSET) {
 		return do_sset(fd, &ifr);
+#if 0
 	} else if (mode == MODE_GREGS) {
 		return do_gregs(fd, &ifr);
+#endif
 	} else if (mode == MODE_NWAY_RST) {
 		return do_nway_rst(fd, &ifr);
 	} else if (mode == MODE_GEEPROM) {
@@ -2626,6 +2632,7 @@
 	return 0;
 }
 
+#if 0
 static int do_gregs(int fd, struct ifreq *ifr)
 {
 	int err;
@@ -2663,6 +2670,7 @@
 
 	return 0;
 }
+#endif
 
 static int do_nway_rst(int fd, struct ifreq *ifr)
 {
diff -ur ethtool-2.6.36/Makefile.in ethtool-2.6.36-ltq/Makefile.in
--- ethtool-2.6.36/Makefile.in	2010-11-16 21:03:52.000000000 +0100
+++ ethtool-2.6.36-ltq/Makefile.in	2011-07-15 15:09:59.588392000 +0200
@@ -50,13 +50,13 @@
 CONFIG_CLEAN_VPATH_FILES =
 am__installdirs = "$(DESTDIR)$(sbindir)" "$(DESTDIR)$(man8dir)"
 PROGRAMS = $(sbin_PROGRAMS)
-am_ethtool_OBJECTS = ethtool.$(OBJEXT) amd8111e.$(OBJEXT) \
-	de2104x.$(OBJEXT) e100.$(OBJEXT) e1000.$(OBJEXT) igb.$(OBJEXT) \
-	fec_8xx.$(OBJEXT) ibm_emac.$(OBJEXT) ixgb.$(OBJEXT) \
-	ixgbe.$(OBJEXT) natsemi.$(OBJEXT) pcnet32.$(OBJEXT) \
-	realtek.$(OBJEXT) tg3.$(OBJEXT) marvell.$(OBJEXT) \
-	vioc.$(OBJEXT) smsc911x.$(OBJEXT) at76c50x-usb.$(OBJEXT) \
-	sfc.$(OBJEXT) stmmac.$(OBJEXT)
+am_ethtool_OBJECTS = ethtool.$(OBJEXT) # amd8111e.$(OBJEXT) \
+#   de2104x.$(OBJEXT) e100.$(OBJEXT) e1000.$(OBJEXT) igb.$(OBJEXT) \
+#   fec_8xx.$(OBJEXT) ibm_emac.$(OBJEXT) ixgb.$(OBJEXT) \
+#   ixgbe.$(OBJEXT) natsemi.$(OBJEXT) pcnet32.$(OBJEXT) \
+#   realtek.$(OBJEXT) tg3.$(OBJEXT) marvell.$(OBJEXT) \
+#   vioc.$(OBJEXT) smsc911x.$(OBJEXT) at76c50x-usb.$(OBJEXT) \
+#   sfc.$(OBJEXT) stmmac.$(OBJEXT)
 ethtool_OBJECTS = $(am_ethtool_OBJECTS)
 ethtool_LDADD = $(LDADD)
 DEFAULT_INCLUDES = -I.@am__isrc@
