diff -Naur zebra-0.94/lib/Makefile.am zebra-0.94.new/lib/Makefile.am
--- zebra-0.94/lib/Makefile.am	2010-02-09 12:02:18.581079000 +0530
+++ zebra-0.94.new/lib/Makefile.am	2010-02-09 12:03:40.853692000 +0530
@@ -21,7 +21,7 @@
 	print_version.c checksum.c vector.c linklist.c \
 	sockunion.c prefix.c thread.c if.c memory.c buffer.c table.c hash.c \
 	filter.c distribute.c stream.c str.c log.c plist.c \
-	sockopt.c smux.c md5.c keychain.c
+	sockopt.c smux.c md5.c keychain.c vty.c command.c
 else
 libzebra_a_SOURCES = \
 	version.c network.c pid_output.c getopt.c getopt1.c daemon.c \
