--- zebra/zebra/rt_netlink.c	2012-11-06 18:59:11.284038000 +0530
+++ zebra_patched/zebra/rt_netlink.c	2012-11-08 11:00:58.120813000 +0530
@@ -367,6 +367,17 @@
   /* Looking up interface name. */
   memset (tb, 0, sizeof tb);
   netlink_parse_rtattr (tb, IFLA_MAX, IFLA_RTA (ifi), len);
+  
+/* check for wireless messages to ignore */
+  if ((tb[IFLA_WIRELESS] != NULL) && (ifi->ifi_change == 0))
+  {
+	#if  !defined(RIPD_STANDALONE) || defined(DEBUG_RIPD_STANDALONE)
+  	if (IS_ZEBRA_DEBUG_KERNEL)
+        	zlog_debug ("%s: ignoring IFLA_WIRELESS message", __func__);
+	#endif
+  	return 0;
+  }
+
   if (tb[IFLA_IFNAME] == NULL)
     return -1;
   name = (char *)RTA_DATA(tb[IFLA_IFNAME]);
@@ -805,6 +816,16 @@
   /* Looking up interface name. */
   memset (tb, 0, sizeof tb);
   netlink_parse_rtattr (tb, IFLA_MAX, IFLA_RTA (ifi), len);
+
+  /* check for wireless messages to ignore */
+  if ((tb[IFLA_WIRELESS] != NULL) && (ifi->ifi_change == 0))
+  {
+  	#if  !defined(RIPD_STANDALONE) || defined(DEBUG_RIPD_STANDALONE)
+  	if (IS_ZEBRA_DEBUG_KERNEL)
+  		zlog_debug ("%s: ignoring IFLA_WIRELESS message", __func__);
+	#endif
+  	return 0;
+  }
   if (tb[IFLA_IFNAME] == NULL)
     return -1;
   name = (char *)RTA_DATA(tb[IFLA_IFNAME]);
