--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,19 @@
+
+2016-02-24:
+	gcc hardeing fixes added
+
+2013-05-20:
+	Fix for invalid memory access during interface down event handler
+
+2012-11-08:
+	Fix for ripd crash while bootup process
+
+2010-02-09:
+	002-open_zebra.patch added
+
+2009-11-13:
+	Adding ripd standalone option
+
 2002-07-07  Kunihiro Ishiguro  <kunihiro@ipinfusion.com>
 
 	* zebra-0.93 released.
