--- a/cli.c
+++ b/cli.c
@@ -17,7 +17,7 @@
 #include "libubus.h"
 
 static struct blob_buf b;
-static int timeout = 30;
+static int timeout = 302;
 static bool simple_output = false;
 static int verbose = 0;
 
@@ -373,7 +373,18 @@ int main(int argc, char **argv)
 	}
 
 	if (ret > 0 && !simple_output)
+	{
+		
+		fprintf(stderr, "\n ########################################### \n");
+
 		fprintf(stderr, "Command failed: %s\n", ubus_strerror(ret));
+		fprintf(stderr,"[%s %d ] @@@@@ UBUS MSG : \n \n",__func__, __LINE__); 
+		for(i=0; i<argc; i++)
+			fprintf(stderr, "%s\t ", argv[i]);
+
+		fprintf(stderr,"\n\n");
+		fprintf(stderr, "\n ########################################### \n");
+	}
 	else if (ret == -2)
 		usage(progname);
 
