diff --git a/shellinabox/launcher.c b/shellinabox/launcher.c
index ba54b36..c802e00 100644
--- a/shellinabox/launcher.c
+++ b/shellinabox/launcher.c
@@ -769,7 +769,7 @@ void closeAllFds(int *exceptFds, int num) {
 }
 
 #if !defined(HAVE_PTSNAME_R)
-static int ptsname_r(int fd, char *buf, size_t buflen) {
+int ptsname_r(int fd, char *buf, size_t buflen) {
   // It is unfortunate that ptsname_r is not universally available.
   // For the time being, this is not a big problem, as ShellInABox is
   // single-threaded (and so is the launcher process). But if this
diff --git a/libhttp/httpconnection.c b/libhttp/httpconnection.c
--- a/libhttp/httpconnection.c
+++ b/libhttp/httpconnection.c
@@ -220,20 +220,6 @@
   }
 }
 
-#ifndef HAVE_STRCASESTR
-static char *strcasestr(const char *haystack, const char *needle) {
-  // This algorithm is O(len(haystack)*len(needle)). Much better algorithms
-  // are available, but this code is much simpler and performance is not
-  // critical for our workloads.
-  int len = strlen(needle);
-  do {
-    if (!strncasecmp(haystack, needle, len)) {
-      return haystack;
-    }
-  } while (*haystack++);
-  return NULL;
-}
-#endif
 
 static int httpFinishCommand(struct HttpConnection *http) {
   int rc            = HTTP_DONE;
