--- WscRefImpl_Intel_1_0_6/src/Makefile	2010-08-25 21:30:41.654555891 +0530
+++ WscRefImpl_Intel_1_0_6.patched/src/Makefile	2010-08-25 21:26:33.444560059 +0530
@@ -1,5 +1,6 @@
+
 all: 
-	$(MAKE) -C linux
+	$(MAKE) -C linux IFX_CFLAGS="$(IFX_CFLAGS)"
 
 clean:
 	$(MAKE) -C linux clean


--- WscRefImpl_Intel_1_0_6/src/linux/Makefile	2010-08-25 21:30:41.644555127 +0530
+++ WscRefImpl_Intel_1_0_6.patched/src/linux/Makefile	2010-08-25 21:26:33.434560386 +0530
@@ -7,7 +7,7 @@
 #export CXX = g++
 #export CC = gcc
 #ifeq ($(BLDTYPE),debug)
-CXXFLAGS ?= -I/usr/src/linux/include
+CXXFLAGS ?= -I/usr/src/linux/include $(IFX_CFLAGS)
 CXXFLAGS += -MMD -Wall -g -D_REENTRANT -D_TUDEBUGTRACE -DPROVISION_IE -DMTLK_IF=1
 CFLAGS += -MMD -Wall -Wnested-externs -g -D_REENTRANT -D_TUDEBUGTRACE -DPROVISION_IE -DMTLK_IF=1
 STL_INCDIR ?= /usr/local/include
@@ -104,7 +104,7 @@
 	$(MAKE) wsccmd
 
 wsccmd: $(OBJS)
-	$(CXX) $(CXX_LDFLAGS) -g -o $@ $(OBJS) -lpthread -lssl -lcrypto $(LIBM) $(CXXLIBS) $(UUID_LIB)
+	$(CXX) $(CXX_LDFLAGS) -g -o $@ $(OBJS) -lpthread -lssl -lcrypto $(LIBM) $(CXXLIBS) $(UUID_LIB) $(IFX_LDFLAGS)
 #	$(LD) -r -o wsccmd $(OBJS) $(LIBS) -L$(STL_LIBDIR)
 #	$(LD) $(LDFLAGS) $(OBJS) $(LIBS) $(LIBPATH) -o wsccmd
 	if [ "$(STRIP)" ]; then cp wsccmd wsccmd_debug && $(STRIP) wsccmd; fi
