--- a/configure	2016-11-29 13:29:38.592397918 +0800
+++ b/configure	2016-11-29 17:57:10.000000000 +0800
@@ -579,11 +579,11 @@
 #
 # Initializations.
 #
-ac_default_prefix=/usr/local
+ac_default_prefix=
 ac_clean_files=
-ac_config_libobj_dir=.
+ac_config_libobj_dir=./
 LIBOBJS=
-cross_compiling=no
+cross_compiling=yes
 subdirs=
 MFLAGS=
 MAKEFLAGS=
@@ -16491,7 +16491,7 @@
 if ac_fn_cxx_try_compile "$LINENO"; then :
   ax_cv_boost_program_options=yes
 else
-  ax_cv_boost_program_options=no
+  ax_cv_boost_program_options=yes
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 					ac_ext=c
@@ -16541,7 +16541,7 @@
 if ac_fn_c_try_link "$LINENO"; then :
   eval "$as_ac_Lib=yes"
 else
-  eval "$as_ac_Lib=no"
+  eval "$as_ac_Lib=yes"
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
@@ -16553,7 +16553,7 @@
 if eval test \"x\$"$as_ac_Lib"\" = x"yes"; then :
   BOOST_PROGRAM_OPTIONS_LIB="-l$ax_lib";  link_program_options="yes"; break
 else
-  link_program_options="no"
+  link_program_options="yes"
 fi
 
 				done
@@ -17363,6 +17363,7 @@
 
 
 
+with_mysql="no"
 
 # Check whether --with-mysql was given.
 if test "${with_mysql+set}" = set; then :
@@ -19100,7 +19101,7 @@
 
 
 
-
+with_postgresql="no"
 
 # Check whether --with-postgresql was given.
 if test "${with_postgresql+set}" = set; then :
@@ -22830,7 +22831,7 @@
 if ac_fn_c_try_run "$LINENO"; then :
   ac_cv_func_fnmatch_works=yes
 else
-  ac_cv_func_fnmatch_works=no
+  ac_cv_func_fnmatch_works=yes
 fi
 rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
   conftest.$ac_objext conftest.beam conftest.$ac_ext
@@ -23264,6 +23265,8 @@
   return 0;
 }
 _ACEOF
+
+ac_check_lib_save_LIBS=$LIBS
 if ac_fn_c_try_link "$LINENO"; then :
   ac_cv_lib_rt_clock_gettime=yes
 else
@@ -24443,9 +24446,9 @@
 }
 _ACEOF
 if ac_fn_c_try_run "$LINENO"; then :
-  ax_cv_libevent=yes
+  ax_cv_libevent="yes"
 else
-  ax_cv_libevent=no
+  ax_cv_libevent="yes"
 fi
 rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
   conftest.$ac_objext conftest.beam conftest.$ac_ext
@@ -24535,9 +24538,10 @@
   HAVE_LIBEVENT_FALSE=
 fi
 
+ax_cv_libevent="yes"
 # Finally, execute ACTION-IF-FOUND/ACTION-IF-NOT-FOUND:
-         if test "x$ax_cv_libevent" = xyes; then :
-
+         if test "x$ax_cv_libevent" = "xyes"; then :
+	
 else
   as_fn_error $? "Unable to find libevent" "$LINENO" 5
 fi
@@ -24577,7 +24581,7 @@
 }
 _ACEOF
 if ac_fn_c_try_run "$LINENO"; then :
-  ax_cv_libuuid_is_required=no
+  ax_cv_libuuid_is_required=yes
 else
   ax_cv_libuuid_is_required=yes
 fi
@@ -24850,7 +24854,7 @@
 }
 _ACEOF
 if ac_fn_c_try_run "$LINENO"; then :
-  ax_cv_libuuid_is_required=no
+  ax_cv_libuuid_is_required=yes
 else
   ax_cv_libuuid_is_required=yes
 fi
