--- a/libtest/framework.cc	2016-11-29 20:33:33.328344188 +0800
+++ b/libtest/framework.cc	2016-11-29 20:34:46.072344035 +0800
@@ -182,10 +182,10 @@
 
 bool Framework::match(const char* arg)
 {
-  if (_wildcard.empty() == false and fnmatch(_wildcard.c_str(), arg, 0))
+  /*if (_wildcard.empty() == false and fnmatch(_wildcard.c_str(), arg, 0))
   {
     return true;
-  }
+  }*/
 
   return false;
 }
@@ -196,8 +196,9 @@
        iter != _collection.end() and (_signal.is_shutdown() == false);
        ++iter)
   {
-    if (_only_run.empty() == false and
-        fnmatch(_only_run.c_str(), (*iter)->name(), 0))
+    //if (_only_run.empty() == false and
+    //    fnmatch(_only_run.c_str(), (*iter)->name(), 0))
+    if (_only_run.empty() == false)
     {
       continue;
     }
