--- a/libntfs-3g/logging.c
+++ b/libntfs-3g/logging.c
@@ -439,7 +439,7 @@ int ntfs_log_handler_fprintf(const char 
 	}
 
 	if (col_prefix)
-		ret += fprintf(stream, col_prefix);
+		ret += fprintf(stream, "%s", col_prefix);
 
 	if ((ntfs_log.flags & NTFS_LOG_FLAG_ONLYNAME) &&
 	    (strchr(file, PATH_SEP)))		/* Abbreviate the filename */
@@ -464,7 +464,7 @@ int ntfs_log_handler_fprintf(const char 
 		ret += fprintf(stream, ": %s\n", strerror(olderr));
 
 	if (col_suffix)
-		ret += fprintf(stream, col_suffix);
+		ret += fprintf(stream, "%s", col_suffix);
 
 
 	fflush(stream);
--- a/src/ntfs-3g.c
+++ b/src/ntfs-3g.c
@@ -2210,7 +2210,7 @@ int main(int argc, char *argv[])
 	}
 
 	if (S_ISBLK(sbuf.st_mode) && (fstype == FSTYPE_FUSE))
-		ntfs_log_info(fuse26_kmod_msg);
+		ntfs_log_info("%s", fuse26_kmod_msg);
 	
 	if (!ctx->no_detach) {
 		if (daemon(0, ctx->debug))
--- a/src/utils.c
+++ b/src/utils.c
@@ -126,7 +126,7 @@ ntfs_volume *utils_mount_volume(const ch
 		else if (errno == EPERM)
 			ntfs_log_error("%s", hibernated_volume_msg);
 		else if (errno == EOPNOTSUPP) {
-			ntfs_log_error(unclean_journal_msg);
+			ntfs_log_error("%s", unclean_journal_msg);
 			ntfs_log_error(forced_mount_msg, volume, mntpoint,
 				       volume, mntpoint);
 		} else if (errno == EBUSY)
--- a/ChangeLog
+++ b/ChangeLog
@@ -947,3 +947,9 @@
 
 	initial CVS import
 
+2016-07-20 16:45  Intel UGW
+
+	* libntfs-3g/logging.c, src/ntfs-3g.c, src/utils.c, Changelog
+
+	fix to compile with gcc -Wformat -Wformat-security
+
