--- a/initd/early.c
+++ b/initd/early.c
@@ -65,13 +65,22 @@ early_mounts(void)
 	mount("proc", "/proc", "proc", MS_NOATIME, 0);
 	mount("sysfs", "/sys", "sysfs", MS_NOATIME, 0);
 	mount("none", "/sys/fs/cgroup", "cgroup", 0, 0);
+#ifdef PROCD_LTQ_SPECIFIC_INIT
+/* As ltq using ramfs, mounting tmpfs is commented*/
+   /* Mounintg /dev on ramfs, this is needed to mount /dev/pts/ on devpts*/
+    mount("/dev/ram", "/dev", "ramfs", MS_NOATIME, 0);
+    system("/sbin/makedevs -d /etc/device_table.txt /");
+#else
 	mount("tmpfs", "/dev", "tmpfs", MS_NOATIME, "mode=0755,size=512K");
 	mkdir("/dev/shm", 01777);
 	mkdir("/dev/pts", 0755);
+#endif
 	mount("devpts", "/dev/pts", "devpts", MS_NOATIME, "mode=600");
 	early_dev();
 
 	early_console("/dev/console");
+#ifndef PROCD_LTQ_SPECIFIC_INIT
+/* As ltq using ramfs, mounting tmpfs is commented*/
 	if (mount_zram_on_tmp())
 		mount("tmpfs", "/tmp", "tmpfs", MS_NOSUID | MS_NODEV | MS_NOATIME, NULL);
 	else
@@ -79,13 +88,21 @@ early_mounts(void)
 	mkdir("/tmp/run", 0777);
 	mkdir("/tmp/lock", 0777);
 	mkdir("/tmp/state", 0777);
+#endif
+
 	umask(oldumask);
 }
 
 static void
 early_env(void)
 {
+#ifdef PROCD_LTQ_SPECIFIC_EXPORT_PATH
+	/*Updating ltq specific env path*/
+    setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin:/opt/lantiq/bin:/opt/lantiq/sbin:/opt/lantiq/usr/bin:/opt/lantiq/usr/sbin", 1);
+    setenv("LD_LIBRARY_PATH", "/opt/lantiq/lib:/opt/lantiq/usr/lib:/lib:/opt/lantiq/servd/lib:/etc/lib", 1);
+#else
 	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+#endif
 }
 
 void
--- a/plug/coldplug.c
+++ b/plug/coldplug.c
@@ -40,8 +40,10 @@ static void udevtrigger_complete(struct 
 void procd_coldplug(void)
 {
 	char *argv[] = { "udevtrigger", NULL };
+#ifndef PROCD_LTQ_SPECIFIC_INIT
+/* Unmounting /dev causes to delete the deivces created in preinit under /dev,
+   to avoid it the below portion of code is commented*/
 	unsigned int oldumask = umask(0);
-
 	umount2("/dev/pts", MNT_DETACH);
 	umount2("/dev/", MNT_DETACH);
 	mount("tmpfs", "/dev", "tmpfs", 0, "mode=0755,size=512K");
@@ -49,6 +51,7 @@ void procd_coldplug(void)
 	mkdir("/dev/pts", 0755);
 	umask(oldumask);
 	mount("devpts", "/dev/pts", "devpts", 0, 0);
+#endif
 	udevtrigger.cb = udevtrigger_complete;
 	udevtrigger.pid = fork();
 	if (!udevtrigger.pid) {
