--- a/plug/hotplug.c
+++ b/plug/hotplug.c
@@ -226,7 +226,12 @@ static void handle_firmware(struct blob_
 		ERROR("Could not find firmware %s\n", path);
 		src = -1;
 		s.st_size = 0;
+#ifdef PROCD_LTQ_SPECIFIC_BYPASS_HANDLE_FIRMWARE
+	/* Return from this function if firmware notfound in the specified path */
+		goto end;
+#else
 		goto send_to_kernel;
+#endif
 	}
 
 	src = open(path, O_RDONLY);
@@ -276,7 +281,9 @@ send_to_kernel:
 	if (write(load, "0", 1) == -1)
 		ERROR("failed to write to %s\n", loadpath);
 	close(load);
-
+#ifdef PROCD_LTQ_SPECIFIC_BYPASS_HANDLE_FIRMWARE
+end:
+#endif
 	DEBUG(2, "Done loading %s\n", path);
 
 	exit(-1);
