--- a/service/instance.c
+++ b/service/instance.c
@@ -33,7 +33,6 @@
 #include "service.h"
 #include "instance.h"
 
-
 enum {
 	INSTANCE_ATTR_COMMAND,
 	INSTANCE_ATTR_ENV,
@@ -228,7 +227,11 @@ instance_run(struct service_instance *in
 	char **argv;
 	char *ld_preload;
 	int argc = 1; /* NULL terminated */
+#ifdef PROCD_LTQ_SPECIFIC_LOG_REDIRECTION
+	int rem;
+#else
 	int rem, _stdin;
+#endif
 	bool seccomp = !in->trace && !in->has_jail && in->seccomp;
 	bool setlbf = _stdout >= 0;
 
@@ -270,6 +273,8 @@ instance_run(struct service_instance *in
 
 	argv[argc] = NULL;
 
+#ifndef PROCD_LTQ_SPECIFIC_LOG_REDIRECTION
+/*TODO: going forward stderr and stdout optoins supported by procd to be used for log redirection*/
 	_stdin = open("/dev/null", O_RDONLY);
 
 	if (_stdout == -1)
@@ -290,6 +295,7 @@ instance_run(struct service_instance *in
 		dup2(_stderr, STDERR_FILENO);
 		closefd(_stderr);
 	}
+#endif
 
 	if (in->gid && setgid(in->gid)) {
 		ERROR("failed to set group id %d: %d (%s)\n", in->gid, errno, strerror(errno));
