diff --git a/src/Cedar/Admin.c b/src/Cedar/Admin.c
index 5e2ea80..3b0ab15 100644
--- a/src/Cedar/Admin.c
+++ b/src/Cedar/Admin.c
@@ -2218,10 +2218,10 @@ UINT StSetAcList(ADMIN *a, RPC_AC_LIST *t)
 		return ERR_NOT_SUPPORTED;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0 && LIST_NUM(t->o) >= 1)
-	{
-		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0 && LIST_NUM(t->o) >= 1)
+//	{
+//		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
+//	}
 
 	CHECK_RIGHT;
 	NO_SUPPORT_FOR_BRIDGE;
@@ -3664,10 +3664,10 @@ UINT StSetSysLog(ADMIN *a, SYSLOG_SETTING *t)
 
 	SERVER_ADMIN_ONLY;
 
-	if (GetGlobalServerFlag(GSF_DISABLE_SYSLOG) != 0 && t->SaveType != SYSLOG_NONE)
-	{
-		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_SYSLOG) != 0 && t->SaveType != SYSLOG_NONE)
+//	{
+//		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
+//	}
 
 	if (GetServerCapsBool(s, "b_support_syslog") == false)
 	{
@@ -5068,13 +5068,13 @@ UINT StSetUser(ADMIN *a, RPC_SET_USER *t)
 
 	CHECK_RIGHT;
 
-	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
-	{
-		if (t->AuthType == AUTHTYPE_USERCERT || t->AuthType == AUTHTYPE_RADIUS || t->AuthType == AUTHTYPE_ROOTCERT || t->AuthType == AUTHTYPE_NT)
-		{
-			return ERR_NOT_SUPPORTED_AUTH_ON_OPENSOURCE;
-		}
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
+//	{
+//		if (t->AuthType == AUTHTYPE_USERCERT || t->AuthType == AUTHTYPE_RADIUS || t->AuthType == AUTHTYPE_ROOTCERT || t->AuthType == AUTHTYPE_NT)
+//		{
+//			return ERR_NOT_SUPPORTED_AUTH_ON_OPENSOURCE;
+//		}
+//	}
 
 	if (StrCmpi(t->Name, "*") == 0)
 	{
@@ -5201,13 +5201,13 @@ UINT StCreateUser(ADMIN *a, RPC_SET_USER *t)
 
 	CHECK_RIGHT;
 
-	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
-	{
-		if (t->AuthType == AUTHTYPE_USERCERT || t->AuthType == AUTHTYPE_RADIUS || t->AuthType == AUTHTYPE_ROOTCERT || t->AuthType == AUTHTYPE_NT)
-		{
-			return ERR_NOT_SUPPORTED_AUTH_ON_OPENSOURCE;
-		}
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
+//	{
+//		if (t->AuthType == AUTHTYPE_USERCERT || t->AuthType == AUTHTYPE_RADIUS || t->AuthType == AUTHTYPE_ROOTCERT || t->AuthType == AUTHTYPE_NT)
+//		{
+//			return ERR_NOT_SUPPORTED_AUTH_ON_OPENSOURCE;
+//		}
+//	}
 
 	if (t->AuthType == AUTHTYPE_USERCERT)
 	{
@@ -7694,10 +7694,10 @@ UINT StSetHubRadius(ADMIN *a, RPC_RADIUS *t)
 		return ERR_NOT_SUPPORTED;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0 && IsEmptyStr(t->RadiusServerName) == false)
-	{
-		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0 && IsEmptyStr(t->RadiusServerName) == false)
+//	{
+//		return ERR_NOT_SUPPORTED_FUNCTION_ON_OPENSOURCE;
+//	}
 
 	CHECK_RIGHT;
 
diff --git a/src/Cedar/Cedar.c b/src/Cedar/Cedar.c
index 790e801..00ff67e 100644
--- a/src/Cedar/Cedar.c
+++ b/src/Cedar/Cedar.c
@@ -1422,7 +1422,7 @@ void CleanupCedar(CEDAR *c)
 
 	DeleteLock(c->OpenVPNPublicPortsLock);
 
-	DeleteLock(c->CurrentRegionLock);
+//	DeleteLock(c->CurrentRegionLock);
 
 	DeleteLock(c->CurrentTcpQueueSizeLock);
 	DeleteLock(c->QueueBudgetLock);
@@ -1691,7 +1691,7 @@ CEDAR *NewCedar(X *server_x, K *server_k)
 
 	c->CedarSuperLock = NewLock();
 
-	c->CurrentRegionLock = NewLock();
+//	c->CurrentRegionLock = NewLock();
 
 	StrCpy(c->OpenVPNDefaultClientOption, sizeof(c->OpenVPNDefaultClientOption), OVPN_DEF_CLIENT_OPTION_STRING);
 
diff --git a/src/Cedar/Cedar.h b/src/Cedar/Cedar.h
index 000cae1..fc8c9c7 100644
--- a/src/Cedar/Cedar.h
+++ b/src/Cedar/Cedar.h
@@ -1043,8 +1043,8 @@ typedef struct CEDAR
 	char CurrentDDnsFqdn[MAX_SIZE];	// FQDN of the current DDNS
 	char OpenVPNPublicPorts[MAX_SIZE];	// OpenVPN public UDP port list
 	LOCK *OpenVPNPublicPortsLock;	// Lock of OpenVPN public UDP port list
-	LOCK *CurrentRegionLock;		// Current region lock
-	char CurrentRegion[128];		// Current region
+//	LOCK *CurrentRegionLock;		// Current region lock
+//	char CurrentRegion[128];		// Current region
 	LOCK *CurrentTcpQueueSizeLock;	// Current TCP send queue size lock
 	UINT CurrentTcpQueueSize;		// Current TCP send queue size
 	COUNTER *CurrentActiveLinks;	// Current active cascade connections
diff --git a/src/Cedar/DDNS.c b/src/Cedar/DDNS.c
index b6281ee..feb8539 100644
--- a/src/Cedar/DDNS.c
+++ b/src/Cedar/DDNS.c
@@ -729,11 +729,11 @@ UINT DCRegister(DDNS_CLIENT *c, bool ipv6, DDNS_REGISTER_PARAM *p, char *replace
 				c->CurrentIPv4, c->CurrentIPv6,
 				c->CurrentAzureIp, c->CurrentAzureTimestamp, c->CurrentAzureSignature, c->AzureCertHash);
 
-			if (IsEmptyStr(current_region) == false)
-			{
-				// Update the current region
-				SiUpdateCurrentRegion(c->Cedar, current_region, false);
-			}
+//			if (IsEmptyStr(current_region) == false)
+//			{
+//				// Update the current region
+//				SiUpdateCurrentRegion(c->Cedar, current_region, false);
+//			}
 		}
 	}
 	Unlock(c->Lock);
diff --git a/src/Cedar/Hub.c b/src/Cedar/Hub.c
index 1f827a9..23b621a 100644
--- a/src/Cedar/Hub.c
+++ b/src/Cedar/Hub.c
@@ -920,10 +920,10 @@ bool IsIpDeniedByAcList(IP *ip, LIST *o)
 		return false;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0)
-	{
-		return false;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0)
+//	{
+//		return false;
+//	}
 
 	for (i = 0;i < LIST_NUM(o);i++)
 	{
@@ -955,10 +955,10 @@ bool IsIpMaskedByAc(IP *ip, AC *ac)
 		return false;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0)
-	{
-		return false;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_AC) != 0)
+//	{
+//		return false;
+//	}
 
 	if (IsIP4(ip))
 	{
diff --git a/src/Cedar/Logging.c b/src/Cedar/Logging.c
index e0fceaa..25a1fa3 100644
--- a/src/Cedar/Logging.c
+++ b/src/Cedar/Logging.c
@@ -1119,10 +1119,10 @@ bool PacketLog(HUB *hub, SESSION *src_session, SESSION *dest_session, PKT *packe
 
 	if (true)
 	{
-		if (GetGlobalServerFlag(GSF_DISABLE_DEEP_LOGGING) != 0)
-		{
-			no_log = true;
-		}
+//		if (GetGlobalServerFlag(GSF_DISABLE_DEEP_LOGGING) != 0)
+//		{
+//			no_log = true;
+//		}
 
 		if (hub->IsVgsHub)
 		{
diff --git a/src/Cedar/Protocol.c b/src/Cedar/Protocol.c
index 6d23e5c..b806213 100644
--- a/src/Cedar/Protocol.c
+++ b/src/Cedar/Protocol.c
@@ -1607,8 +1607,8 @@ bool ServerAccept(CONNECTION *c)
 			goto CLEANUP;
 		}
 
-		if (GetGlobalServerFlag(GSF_DISABLE_AC) == 0)
-		{
+//		if (GetGlobalServerFlag(GSF_DISABLE_AC) == 0)
+//		{
 			if (hub->HubDb != NULL && c->FirstSock != NULL)
 			{
 				IP ip;
@@ -1628,7 +1628,7 @@ bool ServerAccept(CONNECTION *c)
 					goto CLEANUP;
 				}
 			}
-		}
+//		}
 
 		Lock(hub->lock);
 		{
@@ -2022,10 +2022,10 @@ bool ServerAccept(CONNECTION *c)
 						{
 							// Attempt external authentication registered users
 							bool fail_ext_user_auth = false;
-							if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
-							{
-								fail_ext_user_auth = true;
-							}
+//							if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
+//							{
+//								fail_ext_user_auth = true;
+//							}
 
 							if (fail_ext_user_auth == false)
 							{
@@ -2044,10 +2044,10 @@ bool ServerAccept(CONNECTION *c)
 							bool b = false;
 							bool fail_ext_user_auth = false;
 
-							if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
-							{
-								fail_ext_user_auth = true;
-							}
+//							if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
+//							{
+//								fail_ext_user_auth = true;
+//							}
 
 							if (fail_ext_user_auth == false)
 							{
@@ -2085,8 +2085,8 @@ bool ServerAccept(CONNECTION *c)
 					break;
 
 				case CLIENT_AUTHTYPE_CERT:
-					if (GetGlobalServerFlag(GSF_DISABLE_CERT_AUTH) == 0)
-					{
+//					if (GetGlobalServerFlag(GSF_DISABLE_CERT_AUTH) == 0)
+//					{
 						// Certificate authentication
 						cert_size = PackGetDataSize(p, "cert");
 						if (cert_size >= 1 && cert_size <= 100000)
@@ -2134,17 +2134,17 @@ bool ServerAccept(CONNECTION *c)
 							}
 							Free(cert_buf);
 						}
-					}
-					else
-					{
-						// Certificate authentication is not supported in the open source version
-						HLog(hub, "LH_AUTH_CERT_NOT_SUPPORT_ON_OPEN_SOURCE", c->Name, username);
-						Unlock(hub->lock);
-						ReleaseHub(hub);
-						FreePack(p);
-						c->Err = ERR_AUTHTYPE_NOT_SUPPORTED;
-						goto CLEANUP;
-					}
+//					}
+//					else
+//					{
+//						// Certificate authentication is not supported in the open source version
+//						HLog(hub, "LH_AUTH_CERT_NOT_SUPPORT_ON_OPEN_SOURCE", c->Name, username);
+//						Unlock(hub->lock);
+//						ReleaseHub(hub);
+//						FreePack(p);
+//						c->Err = ERR_AUTHTYPE_NOT_SUPPORTED;
+//						goto CLEANUP;
+//					}
 					break;
 
 				default:
diff --git a/src/Cedar/Sam.c b/src/Cedar/Sam.c
index 7e6e04f..f7f3e31 100644
--- a/src/Cedar/Sam.c
+++ b/src/Cedar/Sam.c
@@ -187,10 +187,10 @@ bool SamAuthUserByPlainPassword(CONNECTION *c, HUB *hub, char *username, char *p
 		return false;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
-	{
-		return false;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH) != 0)
+//	{
+//		return false;
+//	}
 
 	h = hub;
 
@@ -349,10 +349,10 @@ bool SamAuthUserByCert(HUB *h, char *username, X *x)
 		return false;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_CERT_AUTH) != 0)
-	{
-		return false;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_CERT_AUTH) != 0)
+//	{
+//		return false;
+//	}
 
 	// Check expiration date
 	if (CheckXDateNow(x) == false)
diff --git a/src/Cedar/Server.c b/src/Cedar/Server.c
index dcac885..52be8a8 100644
--- a/src/Cedar/Server.c
+++ b/src/Cedar/Server.c
@@ -913,10 +913,10 @@ void SiWriteSysLog(SERVER *s, char *typestr, char *hubname, wchar_t *message)
 		return;
 	}
 
-	if (GetGlobalServerFlag(GSF_DISABLE_SYSLOG) != 0)
-	{
-		return;
-	}
+//	if (GetGlobalServerFlag(GSF_DISABLE_SYSLOG) != 0)
+//	{
+//		return;
+//	}
 
 	// Host name
 	GetMachineName(machinename, sizeof(machinename));
@@ -1358,25 +1358,25 @@ void GetServerCaps(SERVER *s, CAPSLIST *t)
 	Unlock(s->CapsCacheLock);
 }
 
-// Update the global server flags
-void UpdateGlobalServerFlags(SERVER *s, CAPSLIST *t)
-{
-	bool is_restricted = false;
-	// Validate arguments
-	if (s == NULL || t == NULL)
-	{
-		return;
-	}
-
-	is_restricted = SiIsEnterpriseFunctionsRestrictedOnOpenSource(s->Cedar);
-
-	SetGlobalServerFlag(GSF_DISABLE_PUSH_ROUTE, is_restricted);
-	SetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH, is_restricted);
-	SetGlobalServerFlag(GSF_DISABLE_CERT_AUTH, is_restricted);
-	SetGlobalServerFlag(GSF_DISABLE_DEEP_LOGGING, is_restricted);
-	SetGlobalServerFlag(GSF_DISABLE_AC, is_restricted);
-	SetGlobalServerFlag(GSF_DISABLE_SYSLOG, is_restricted);
-}
+//// Update the global server flags
+//void UpdateGlobalServerFlags(SERVER *s, CAPSLIST *t)
+//{
+//	bool is_restricted = false;
+//	// Validate arguments
+//	if (s == NULL || t == NULL)
+//	{
+//		return;
+//	}
+//
+//	is_restricted = SiIsEnterpriseFunctionsRestrictedOnOpenSource(s->Cedar);
+//
+//	SetGlobalServerFlag(GSF_DISABLE_PUSH_ROUTE, is_restricted);
+//	SetGlobalServerFlag(GSF_DISABLE_RADIUS_AUTH, is_restricted);
+//	SetGlobalServerFlag(GSF_DISABLE_CERT_AUTH, is_restricted);
+//	SetGlobalServerFlag(GSF_DISABLE_DEEP_LOGGING, is_restricted);
+//	SetGlobalServerFlag(GSF_DISABLE_AC, is_restricted);
+//	SetGlobalServerFlag(GSF_DISABLE_SYSLOG, is_restricted);
+//}
 
 // Set a global server flag
 void SetGlobalServerFlag(UINT index, UINT value)
@@ -1405,7 +1405,7 @@ UINT GetGlobalServerFlag(UINT index)
 // Main of the aquisition of Caps of the server
 void GetServerCapsMain(SERVER *s, CAPSLIST *t)
 {
-	bool is_restricted = false;
+//	bool is_restricted = false;
 
 	// Validate arguments
 	if (s == NULL || t == NULL)
@@ -1413,7 +1413,7 @@ void GetServerCapsMain(SERVER *s, CAPSLIST *t)
 		return;
 	}
 
-	is_restricted = SiIsEnterpriseFunctionsRestrictedOnOpenSource(s->Cedar);
+//	is_restricted = SiIsEnterpriseFunctionsRestrictedOnOpenSource(s->Cedar);
 
 	// Initialize
 	InitCapsList(t);
@@ -1559,7 +1559,8 @@ void GetServerCapsMain(SERVER *s, CAPSLIST *t)
 	AddCapsBool(t, "b_support_securenat", true);
 
 	// Pushing routing table function of SecureNAT Virtual DHCP Server is available
-	AddCapsBool(t, "b_suppport_push_route", !is_restricted);
+//	AddCapsBool(t, "b_suppport_push_route", !is_restricted);
+	AddCapsBool(t, "b_suppport_push_route", true);
 	AddCapsBool(t, "b_suppport_push_route_config", true);
 
 	if (s->ServerType != SERVER_TYPE_STANDALONE)
@@ -1800,7 +1801,7 @@ void GetServerCapsMain(SERVER *s, CAPSLIST *t)
 	AddCapsBool(t, "b_vpn4", true);
 
 
-	UpdateGlobalServerFlags(s, t);
+//	UpdateGlobalServerFlags(s, t);
 }
 
 // SYSLOG_SETTING
@@ -3333,9 +3334,9 @@ FOLDER *SiWriteConfigurationToCfg(SERVER *s)
 
 	root = CfgCreateFolder(NULL, TAG_ROOT);
 
-	SiGetCurrentRegion(s->Cedar, region, sizeof(region));
+//	SiGetCurrentRegion(s->Cedar, region, sizeof(region));
 
-	CfgAddStr(root, "Region", region);
+//	CfgAddStr(root, "Region", region);
 
 	CfgAddInt(root, "ConfigRevision", s->ConfigRevision);
 
@@ -10744,49 +10745,49 @@ FARM_CONTROLLER *SiStartConnectToController(SERVER *s)
 	return f;
 }
 
-// Get the current version
-void SiGetCurrentRegion(CEDAR *c, char *region, UINT region_size)
-{
-	ClearStr(region, region_size);
-	// Validate arguments
-	if (c == NULL || region == NULL)
-	{
-		return;
-	}
-
-	Lock(c->CurrentRegionLock);
-	{
-		StrCpy(region, region_size, c->CurrentRegion);
-	}
-	Unlock(c->CurrentRegionLock);
-
-	if (IsEmptyStr(region))
-	{
-		if (GetCurrentLangId() == SE_LANG_JAPANESE)
-		{
-			StrCpy(region, region_size, "JP");
-		}
-		else if (GetCurrentLangId() == SE_LANG_CHINESE_ZH)
-		{
-			StrCpy(region, region_size, "CN");
-		}
-	}
-}
-
-// Check the current region
-bool SiCheckCurrentRegion(CEDAR *c, char *r)
-{
-	char tmp[64];
-	// Validate arguments
-	if (c == NULL || r == NULL)
-	{
-		return false;
-	}
-
-	SiGetCurrentRegion(c, tmp, sizeof(tmp));
-
-	return (StrCmpi(r, tmp) == 0);
-}
+//// Get the current version
+//void SiGetCurrentRegion(CEDAR *c, char *region, UINT region_size)
+//{
+//	ClearStr(region, region_size);
+//	// Validate arguments
+//	if (c == NULL || region == NULL)
+//	{
+//		return;
+//	}
+//
+//	Lock(c->CurrentRegionLock);
+//	{
+//		StrCpy(region, region_size, c->CurrentRegion);
+//	}
+//	Unlock(c->CurrentRegionLock);
+//
+//	if (IsEmptyStr(region))
+//	{
+//		if (GetCurrentLangId() == SE_LANG_JAPANESE)
+//		{
+//			StrCpy(region, region_size, "JP");
+//		}
+//		else if (GetCurrentLangId() == SE_LANG_CHINESE_ZH)
+//		{
+//			StrCpy(region, region_size, "CN");
+//		}
+//	}
+//}
+//
+//// Check the current region
+//bool SiCheckCurrentRegion(CEDAR *c, char *r)
+//{
+//	char tmp[64];
+//	// Validate arguments
+//	if (c == NULL || r == NULL)
+//	{
+//		return false;
+//	}
+//
+//	SiGetCurrentRegion(c, tmp, sizeof(tmp));
+//
+//	return (StrCmpi(r, tmp) == 0);
+//}
 
 // Check whether some enterprise functions are restricted
 // 
@@ -10820,60 +10821,60 @@ bool SiCheckCurrentRegion(CEDAR *c, char *r)
 // Anyone, except Daiyuu Nobori, who understands and writes the C language
 // program can remove this restriction at his own risk.
 // 
-bool SiIsEnterpriseFunctionsRestrictedOnOpenSource(CEDAR *c)
-{
-	char region[128];
-	bool ret = false;
-	// Validate arguments
-	if (c == NULL)
-	{
-		return false;
-	}
-
-
-	SiGetCurrentRegion(c, region, sizeof(region));
-
-	if (StrCmpi(region, "JP") == 0 || StrCmpi(region, "CN") == 0)
-	{
-		ret = true;
-	}
-
-	return ret;
-}
-
-// Update the current region
-void SiUpdateCurrentRegion(CEDAR *c, char *region, bool force_update)
-{
-	bool changed = false;
-	// Validate arguments
-	if (c == NULL)
-	{
-		return;
-	}
-
-	if (IsEmptyStr(region) == false)
-	{
-		Lock(c->CurrentRegionLock);
-		{
-			if (StrCmpi(c->CurrentRegion, region) != 0)
-			{
-				StrCpy(c->CurrentRegion, sizeof(c->CurrentRegion), region);
-				changed = true;
-			}
-		}
-		Unlock(c->CurrentRegionLock);
-	}
-
-	if (force_update)
-	{
-		changed = true;
-	}
-
-	if (changed)
-	{
-		FlushServerCaps(c->Server);
-	}
-}
+//bool SiIsEnterpriseFunctionsRestrictedOnOpenSource(CEDAR *c)
+//{
+//	char region[128];
+//	bool ret = false;
+//	// Validate arguments
+//	if (c == NULL)
+//	{
+//		return false;
+//	}
+//
+//
+//	SiGetCurrentRegion(c, region, sizeof(region));
+//
+//	if (StrCmpi(region, "JP") == 0 || StrCmpi(region, "CN") == 0)
+//	{
+//		ret = true;
+//	}
+//
+//	return ret;
+//}
+//
+//// Update the current region
+//void SiUpdateCurrentRegion(CEDAR *c, char *region, bool force_update)
+//{
+//	bool changed = false;
+//	// Validate arguments
+//	if (c == NULL)
+//	{
+//		return;
+//	}
+//
+//	if (IsEmptyStr(region) == false)
+//	{
+//		Lock(c->CurrentRegionLock);
+//		{
+//			if (StrCmpi(c->CurrentRegion, region) != 0)
+//			{
+//				StrCpy(c->CurrentRegion, sizeof(c->CurrentRegion), region);
+//				changed = true;
+//			}
+//		}
+//		Unlock(c->CurrentRegionLock);
+//	}
+//
+//	if (force_update)
+//	{
+//		changed = true;
+//	}
+//
+//	if (changed)
+//	{
+//		FlushServerCaps(c->Server);
+//	}
+//}
 
 // Create a server
 SERVER *SiNewServer(bool bridge)
@@ -11035,7 +11036,7 @@ SERVER *SiNewServerEx(bool bridge, bool in_client_inner_server, bool relay_serve
 
 	SiInitDeadLockCheck(s);
 
-	SiUpdateCurrentRegion(s->Cedar, "", true);
+//	SiUpdateCurrentRegion(s->Cedar, "", true);
 
 	return s;
 }
diff --git a/src/Cedar/Server.h b/src/Cedar/Server.h
index 547df58..faf9b63 100644
--- a/src/Cedar/Server.h
+++ b/src/Cedar/Server.h
@@ -424,12 +424,12 @@ struct LOG_FILE
 
 // Global server flags
 #define	NUM_GLOBAL_SERVER_FLAGS			128
-#define	GSF_DISABLE_PUSH_ROUTE			1
-#define	GSF_DISABLE_RADIUS_AUTH			2
-#define	GSF_DISABLE_CERT_AUTH			3
-#define	GSF_DISABLE_DEEP_LOGGING		4
-#define	GSF_DISABLE_AC					5
-#define	GSF_DISABLE_SYSLOG				6
+//#define	GSF_DISABLE_PUSH_ROUTE			1
+//#define	GSF_DISABLE_RADIUS_AUTH			2
+//#define	GSF_DISABLE_CERT_AUTH			3
+//#define	GSF_DISABLE_DEEP_LOGGING		4
+//#define	GSF_DISABLE_AC					5
+//#define	GSF_DISABLE_SYSLOG				6
 #define	GSF_SHOW_OSS_MSG				7
 #define	GSF_LOCALBRIDGE_NO_DISABLE_OFFLOAD	8
 #define	GSF_DISABLE_SESSION_RECONNECT	9
@@ -711,7 +711,7 @@ void DestroyServerCapsCache(SERVER *s);
 
 void SetGlobalServerFlag(UINT index, UINT value);
 UINT GetGlobalServerFlag(UINT index);
-void UpdateGlobalServerFlags(SERVER *s, CAPSLIST *t);
+//void UpdateGlobalServerFlags(SERVER *s, CAPSLIST *t);
 
 
 bool IsAdminPackSupportedServerProduct(char *name);
@@ -739,10 +739,10 @@ void SiApplyAzureConfig(SERVER *s, DDNS_CLIENT_STATUS *ddns_status);
 void SiSetAzureEnable(SERVER *s, bool enabled);
 bool SiGetAzureEnable(SERVER *s);
 
-void SiUpdateCurrentRegion(CEDAR *c, char *region, bool force_update);
-void SiGetCurrentRegion(CEDAR *c, char *region, UINT region_size);
-bool SiIsEnterpriseFunctionsRestrictedOnOpenSource(CEDAR *c);
-bool SiCheckCurrentRegion(CEDAR *c, char *r);
+//void SiUpdateCurrentRegion(CEDAR *c, char *region, bool force_update);
+//void SiGetCurrentRegion(CEDAR *c, char *region, UINT region_size);
+//bool SiIsEnterpriseFunctionsRestrictedOnOpenSource(CEDAR *c);
+//bool SiCheckCurrentRegion(CEDAR *c, char *r);
 
 #endif	// SERVER_H
 
diff --git a/src/Cedar/Virtual.c b/src/Cedar/Virtual.c
index 2ce41b0..ee5b885 100644
--- a/src/Cedar/Virtual.c
+++ b/src/Cedar/Virtual.c
@@ -9647,8 +9647,8 @@ void VirtualDhcpServer(VH *v, PKT *p)
 				ret.DnsServer2 = v->DhcpDns2;
 				ret.Gateway = v->DhcpGateway;
 
-				if (GetGlobalServerFlag(GSF_DISABLE_PUSH_ROUTE) == 0)
-				{
+//				if (GetGlobalServerFlag(GSF_DISABLE_PUSH_ROUTE) == 0)
+//				{
 					Copy(&ret.ClasslessRoute, &v->PushRoute, sizeof(DHCP_CLASSLESS_ROUTE_TABLE));
 
 					if (IsIpcMacAddress(p->MacAddressSrc))
@@ -9697,7 +9697,7 @@ void VirtualDhcpServer(VH *v, PKT *p)
 							}
 						}
 					}
-				}
+//				}
 
 				if (opt->Opcode != DHCP_INFORM)
 				{
