diff --git a/rc/vpnbridge b/rc/vpnbridge
new file mode 100755
index 0000000..af5ea4d
--- /dev/null
+++ b/rc/vpnbridge
@@ -0,0 +1,10 @@
+#!/bin/sh /etc/rc.common
+START=60
+
+start(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnbridge start
+}
+
+stop(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnbridge stop
+}
diff --git a/rc/vpnclient b/rc/vpnclient
new file mode 100755
index 0000000..c223f74
--- /dev/null
+++ b/rc/vpnclient
@@ -0,0 +1,10 @@
+#!/bin/sh /etc/rc.common
+START=60
+
+start(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnclient start
+}
+
+stop(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnclient stop
+}
diff --git a/rc/vpnserver b/rc/vpnserver
new file mode 100755
index 0000000..0187b6e
--- /dev/null
+++ b/rc/vpnserver
@@ -0,0 +1,10 @@
+#!/bin/sh /etc/rc.common
+START=60
+
+start(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnserver start
+}
+
+stop(){
+	/usr/bin/env LANG=en_US.UTF-8 /usr/bin/vpnserver stop
+}
