diff -Nurb strace-4.5.19/configure strace-4.5.19-uClibc-0.9.30.2/configure
--- strace-4.5.19/configure	2009-10-21 19:41:11.000000000 +0200
+++ strace-4.5.19-uClibc-0.9.30.2/configure	2010-01-29 14:41:35.897738564 +0100
@@ -5880,7 +5880,7 @@
 do :
   as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
 ac_fn_c_check_header_compile "$LINENO" "$ac_header" "$as_ac_Header" "#include <stddef.h>
-#include <linux/socket.h>
+#include <sys/socket.h>
 "
 eval as_val=\$$as_ac_Header
    if test "x$as_val" = x""yes; then :
diff -Nurb strace-4.5.19/configure.ac strace-4.5.19-uClibc-0.9.30.2/configure.ac
--- strace-4.5.19/configure.ac	2009-10-12 21:54:43.000000000 +0200
+++ strace-4.5.19-uClibc-0.9.30.2/configure.ac	2010-01-29 14:42:02.880738326 +0100
@@ -258,7 +258,7 @@
 ], [], [])
 AC_CHECK_HEADERS([linux/icmp.h linux/in6.h linux/netlink.h linux/if_packet.h],
                  [], [], [#include <stddef.h>
-#include <linux/socket.h>])
+#include <sys/socket.h>])
 AC_CHECK_HEADERS([asm/sigcontext.h], [], [], [#include <signal.h>])
 AC_CHECK_TYPES([struct sigcontext_struct,
 		struct sigcontext],,, [#include <signal.h>])
