--- a/comgt.c
+++ b/comgt.c
@@ -1262,7 +1262,7 @@ void opengt(void) {
   else {
     if ((comfd = open(device, O_RDWR|O_EXCL|O_NONBLOCK|O_NOCTTY)) <0) { //O_NONBLOCK|O_NOCTTY)) <0) {//
       sprintf(msg,"Can't open device %s.\n",device);
-      printf(msg);
+      printf("%s", msg);
       ext(1);
     }
   }
@@ -1297,7 +1297,7 @@ void opendevice(void) {
   if(strcmp(device,"-")!=0) {
     if ((comfd = open(device, O_RDWR|O_EXCL|O_NONBLOCK|O_NOCTTY)) <0) { //O_NONBLOCK|O_NOCTTY)) <0) {//
       sprintf(msg,"Can't open device %s.\n",device);
-      printf(msg);
+      printf("%s", msg);
       ext(1);
     }
   }
