--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -4620,9 +4620,9 @@ install-plugin: installdirs lang.install
 # We keep the directory structure for files in config or c-family and .def
 # files. All other files are flattened to a single directory.
 	$(mkinstalldirs) $(DESTDIR)$(plugin_includedir)
-	headers=`echo $(PLUGIN_HEADERS) | tr ' ' '\012' | sort -u`; \
+	$(eval _headers=$(shell echo $(PLUGIN_HEADERS) | tr ' ' '\012' | sort -u))
 	srcdirstrip=`echo "$(srcdir)" | sed 's/[].[^$$\\*|]/\\\\&/g'`; \
-	for file in $$headers; do \
+	for file in $(_headers); do \
 	  if [ -f $$file ] ; then \
 	    path=$$file; \
 	  elif [ -f $(srcdir)/$$file ]; then \
