Index: mrd6-0.9.6/include/mrd/interface.h
===================================================================
--- mrd6-0.9.6.orig/include/mrd/interface.h	2007-01-07 07:44:32.000000000 +0530
+++ mrd6-0.9.6/include/mrd/interface.h	2010-07-02 17:56:51.462466000 +0530
@@ -112,6 +112,7 @@
 	interface(intfconf *, int indx, const char *name, int type, int mtu, int flags);
 	~interface();
 
+	void update(int indx, const char *name, int type, int mtu, int flags);  /*TAHI test mtu change fix */
 	void shutdown();
 
 	int index() const { return mif_index; }
Index: mrd6-0.9.6/src/interface.cpp
===================================================================
--- mrd6-0.9.6.orig/src/interface.cpp	2008-01-25 16:14:46.000000000 +0530
+++ mrd6-0.9.6/src/interface.cpp	2010-07-02 17:56:51.489470000 +0530
@@ -184,6 +184,15 @@
 	return owner_router()->log_router_desc(owner()->log());
 }
 
+void interface::update(int indx, const char *name, int type, int mtu, int flags) {
+
+	mif_index = indx;                                                                                                   
+        mif_name = name;                                                                                                  
+        mif_type = type;                                                                                                    
+        mif_mtu = mtu;                                                                                                      
+        mif_flags = flags;   
+}
+	 
 interface::interface(intfconf *cnf, int indx, const char *ifname, int type, int mtu, int flags)
 	: node(g_mrd->get_child("interface"), ifname) {
 
Index: mrd6-0.9.6/src/mrd.cpp
===================================================================
--- mrd6-0.9.6.orig/src/mrd.cpp	2010-07-02 17:54:19.000000000 +0530
+++ mrd6-0.9.6/src/mrd.cpp	2010-07-02 17:56:51.512465000 +0530
@@ -1478,8 +1478,16 @@
 	interface *intf;
 	if(index == 1)
 		return 0;
-	if ((intf = get_interface_by_index(index)))
+	if ((intf = get_interface_by_index(index))) {
+		intf->update(index, name, type, mtu, flags);
+		
+		if (should_log(VERBOSE)) {
+			log().xprintf("Updated %s interface %s with mtu %u.\n",
+				      intf->type_str(), intf->name(),
+				      (uint32_t)intf->mtu());
+		}
 		return intf;
+	}
 
 	bool enabled = !m_intfconfs.is_interface_disabled(name);
 
