Index: mrd6-0.9.6/src/linux/us_mfa.cpp
===================================================================
--- mrd6-0.9.6.orig/src/linux/us_mfa.cpp	2010-07-06 11:23:40.000000000 +0530
+++ mrd6-0.9.6/src/linux/us_mfa.cpp	2010-07-06 11:25:10.550471000 +0530
@@ -677,6 +677,9 @@
 	if ((hdr->ip6_dst.s6_addr[1] & ~0x3) == 0)
 		return;
 
+	if ((hdr->ip6_dst.s6_addr[1] & 0x0f) == 0) /* Drop pkts to multicast addr with scope = 0 */
+		return;
+
 	hdr->ip6_hlim--;
 
 	us_mfa_group *grp = match_group(hdr->ip6_dst);
