Index: mrd6-0.9.6/src/Makefile
===================================================================
--- mrd6-0.9.6.orig/src/Makefile	2010-07-07 23:52:31.000000000 +0530
+++ mrd6-0.9.6/src/Makefile	2010-07-07 23:56:51.220875000 +0530
@@ -114,7 +114,7 @@
 ifeq ($(STATIC_STDCXX),no)
 	LDFLAGS += -lstdc++
 else
-	LDFLAGS += `$(CXX) -print-file-name=libstdc++.a`
+	LDFLAGS += -L.   ##`$(CXX) -print-file-name=libstdc++.a`
 endif
 
 TEST_OBJECTS = $(addprefix $(OBJ_DIR)/,$(TEST_SOURCES:.cpp=.o))
