Index: mrd6-0.9.6/include/mrd/support/ptree.h
===================================================================
--- mrd6-0.9.6.orig/include/mrd/support/ptree.h	2007-03-21 14:21:27.000000000 +0530
+++ mrd6-0.9.6/include/mrd/support/ptree.h	2012-10-19 15:48:26.000000000 +0530
@@ -36,6 +36,7 @@
 #define _Ptree_h_
 
 #include <stdint.h>
+#include <stddef.h>
 #include <algorithm>
 #include <mrd/log.h>
 
