add kernel modules for support of soft-reset button

diff --git a/target/linux/lantiq/falcon/target.mk b/target/linux/lantiq/falcon/target.mk
--- a/target/linux/lantiq/falcon/target.mk
+++ b/target/linux/lantiq/falcon/target.mk
@@ -5,6 +5,7 @@ FEATURES:=squashfs jffs2
 DEVICE_TYPE:=other
 
 DEFAULT_PACKAGES+= kmod-ifxos gpon-base-files kmod-leds-gpio kmod-ledtrig-heartbeat \
+	kmod-input-core kmod-input-gpio-keys kmod-button-hotplug \
 	kmod-gpon-optic-drv gpon-optic-drv kmod-gpon-onu-drv gpon-onu-drv \
 	gpon-pe-firmware gpon-omci-api gpon-omci-onu gpon-luci
 
