extend list of kernel debugging options in menuconfig

diff --git a/config/Config-kernel.in b/config/Config-kernel.in
--- a/config/Config-kernel.in
+++ b/config/Config-kernel.in
@@ -103,6 +103,12 @@ config KERNEL_DEBUG_LL
 	help
 	  ARM low level debugging.
 
+config KERNEL_DEBUG_PINCTRL
+	bool "Compile the kernel with pinctrl debugging"
+		
+config KERNEL_DEBUG_GPIO
+	bool "Compile the kernel with gpio debugging"
+				
 config KERNEL_DYNAMIC_DEBUG 
 	bool "Compile the kernel with dynamic printk"
 	select KERNEL_DEBUG_FS
