do not add eth0 config when booting with nfs

diff --git a/target/linux/lantiq/base-files/etc/uci-defaults/02_network b/target/linux/lantiq/base-files/etc/uci-defaults/02_network
--- a/target/linux/lantiq/base-files/etc/uci-defaults/02_network
+++ b/target/linux/lantiq/base-files/etc/uci-defaults/02_network
@@ -58,7 +58,7 @@ EOF
 touch /etc/config/network
 
 ucidef_set_interface_loopback
-ucidef_set_interface_lan 'eth0'
+grep -q "nfsroot" /proc/cmdline || ucidef_set_interface_lan 'eth0'
 
 vpi=1
 vci=32
