Lantiq SVIP profile, device tree and image support

diff --git a/target/linux/lantiq/dts/EASY32002.dts b/target/linux/lantiq/dts/EASY32002.dts
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/dts/EASY32002.dts
@@ -0,0 +1,64 @@
+/dts-v1/;
+
+/include/ "svip.dtsi"
+
+/ {
+	model = "EASY32002 - Lantiq EASY32002 evalkit";
+
+	chosen {
+		// bootargs = "console=ttyLTQ0,115200 init=/etc/preinit";
+	};
+
+	memory@0 {
+		reg = <0x0 0x8000000>;  // 128M at 0x0
+	};
+
+	svip_reset@0 {
+		compatible = "lantiq,svip-reset";
+		svip-toggle-reset-pin = <0x120000f3 0x80>;
+	};
+};
+
+&ebu_cs0 {
+	nand-parts@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "gen_nand", "lantiq,nand-svip";
+		bank-width = <1>;
+		reg = <0x0 0x2000000>;
+		linux,mtd-name = "gen_nand";
+		bbt-use-flash;
+		
+		partition@0 {
+			label = "2nd Level Loader";
+			reg = <0x0 0x4000>;
+		};
+		
+		partition@4000 {
+			label = "uboot";
+			reg = <0x4000 0x4C000>;
+		};
+
+		partition@50000 {
+			label = "uboot_env";
+			reg = <0x50000 0x10000>;
+		};
+
+		partition@60000 {
+			label = "linux";
+			reg = <0x60000 0x800000>;
+		};
+		
+	};
+};
+
+&spi0 {
+	status = "okay";
+	gpio-xrq = <&port2 18 0>;
+
+	duslic_xs@0 {
+		compatible = "rohm,dh2228fv";
+		reg = <0>;
+		spi-max-frequency = <8192000>;
+	};
+};
diff --git a/target/linux/lantiq/dts/EASY33016.dts b/target/linux/lantiq/dts/EASY33016.dts
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/dts/EASY33016.dts
@@ -0,0 +1,65 @@
+/dts-v1/;
+
+/include/ "svip.dtsi"
+
+/ {
+	model = "EASY33016 - Lantiq EASY33016 evalkit";
+
+	chosen {
+		// bootargs = "console=ttyLTQ0,115200 init=/etc/preinit";
+	};
+
+	memory@0 {
+		reg = <0x0 0x1800000>;  // 24M at 0x0
+	};
+
+	svip_reset@0 {
+		compatible = "lantiq,svip-reset";
+		svip-toggle-reset-pin = <0x120000f3 0x88>;
+	};
+};
+
+&ebu_cs0 {
+	nand-parts@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "gen_nand", "lantiq,nand-svip";
+		bank-width = <1>;
+		reg = <0x0 0x2000000>;
+		linux,mtd-name = "gen_nand";
+		bbt-use-flash;
+		
+		partition@0 {
+			label = "2nd Level Loader";
+			reg = <0x0 0x4000>;
+		};
+		
+		partition@4000 {
+			label = "uboot";
+			reg = <0x4000 0x4C000>;
+		};
+
+		partition@50000 {
+			label = "uboot_env";
+			reg = <0x50000 0x10000>;
+		};
+
+		partition@60000 {
+			label = "linux";
+			reg = <0x60000 0x800000>;
+		};
+		
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	vinetic_xt@1 {
+		compatible = "lantiq,vinetic-xt";
+		reg = <1>;
+		spi-max-frequency = <1000000>;
+		spi-cpol;
+		spi-cpha;
+	};
+};
diff --git a/target/linux/lantiq/dts/EASY336.dts b/target/linux/lantiq/dts/EASY336.dts
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/dts/EASY336.dts
@@ -0,0 +1,56 @@
+/dts-v1/;
+
+/include/ "svip.dtsi"
+
+/ {
+	model = "EASY336 - Lantiq EASY336 evalkit";
+
+	chosen {
+		// bootargs = "console=ttyLTQ0,115200 init=/etc/preinit";
+	};
+
+	memory@0 {
+		reg = <0x0 0x1800000>;  // 24M at 0x0
+	};
+
+	svip_reset@0 {
+		compatible = "lantiq,svip-reset";
+		svip-toggle-reset-pin = <0x120000f3 1>;
+//		// for EASY33016 board it will be
+//		// svip-toggle-reset-pin = <0x120000f3 0x88888888>;
+	};
+};
+
+&ebu_cs0 {
+	nand-parts@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "gen_nand", "lantiq,nand-svip";
+		bank-width = <1>;
+		reg = <0x0 0x2000000>;
+		linux,mtd-name = "gen_nand";
+		bbt-use-flash;
+		
+		partition@0 {
+			label = "2nd Level Loader";
+			reg = <0x0 0x4000>;
+		};
+		
+		partition@4000 {
+			label = "uboot";
+			reg = <0x4000 0x4C000>;
+		};
+
+		partition@50000 {
+			label = "uboot_env";
+			reg = <0x50000 0x10000>;
+		};
+
+		partition@60000 {
+			label = "linux";
+			reg = <0x60000 0x800000>;
+		};
+		
+	};
+};
+
diff --git a/target/linux/lantiq/dts/EASY34016.dts b/target/linux/lantiq/dts/EASY34016.dts
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/dts/EASY34016.dts
@@ -0,0 +1,65 @@
+/dts-v1/;
+
+/include/ "svip.dtsi"
+
+/ {
+	model = "EASY34016 - Lantiq EASY34016 evalkit";
+
+	chosen {
+		// bootargs = "console=ttyLTQ0,115200 init=/etc/preinit";
+	};
+
+	memory@0 {
+		reg = <0x0 0x1800000>;  // 24M at 0x0
+	};
+
+	svip_reset@0 {
+		compatible = "lantiq,svip-reset";
+		svip-toggle-reset-pin = <0x120000f3 0x88>;
+	};
+};
+
+&ebu_cs0 {
+	nand-parts@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "gen_nand", "lantiq,nand-svip";
+		bank-width = <1>;
+		reg = <0x0 0x2000000>;
+		linux,mtd-name = "gen_nand";
+		bbt-use-flash;
+		
+		partition@0 {
+			label = "2nd Level Loader";
+			reg = <0x0 0x4000>;
+		};
+		
+		partition@4000 {
+			label = "uboot";
+			reg = <0x4000 0x4C000>;
+		};
+
+		partition@50000 {
+			label = "uboot_env";
+			reg = <0x50000 0x10000>;
+		};
+
+		partition@60000 {
+			label = "linux";
+			reg = <0x60000 0x800000>;
+		};
+		
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	vinetic_xs@1 {
+		compatible = "lantiq,vinetic-xt", "lantiq,vinetic-xs";
+		reg = <1>;
+		spi-max-frequency = <1000000>;
+		spi-cpol;
+		spi-cpha;
+	};
+};
diff --git a/target/linux/lantiq/dts/svip.dtsi b/target/linux/lantiq/dts/svip.dtsi
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/dts/svip.dtsi
@@ -0,0 +1,185 @@
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	compatible = "lantiq,svip";
+
+	cpus {
+		cpu@0 {
+			compatible = "mips,mips24KEc";
+		};
+	};
+
+	/* FPI (flexible peripheral interconnect) */
+	fpi@10000000 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "lantiq,fpi", "simple-bus";
+		reg = <0x10000000 0xE000000>;
+		ranges = <0x0 0x10000000 0xE000000>;
+
+		ebu_cs0: localbus@0 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			compatible = "lantiq,localbus", "simple-bus";
+			reg = <0x0 0x1000000>;
+			ranges = <0x0 0x0 0x1000000>;
+		};
+		serial0@4100100 {
+			compatible = "lantiq,asc";
+			reg = <0x4100100 0x100>;
+			interrupt-parent = <&icu0>;
+			interrupts = <8 10 11>;
+			line = <0>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&asc0_pins>;
+		};
+		serial1@4100200 {
+			status = "disabled";
+			compatible = "lantiq,asc";
+			reg = <0x4100200 0x100>;
+			interrupt-parent = <&icu0>;
+			interrupts = <15 17 18>;
+			line = <1>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&asc1_pins>;
+		};
+		spi0: spi0@4100300 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+			compatible = "lantiq,spi-lantiq-ssc", "lantiq,spi-xway-csi";
+			interrupts = <46 47 48>;
+			reg = <0x4100300 0x100>;
+			interrupt-parent = <&icu0>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&spi0_pins>;
+		};
+		ebu@4102000 {
+			compatible = "lantiq,ebu-svip";
+			reg = <0x4102000 0x100>;
+		};
+		dma: dma@4104000 {
+			compatible = "lantiq,dma-xway";
+			reg = <0x4104000 0x100>;
+			interrupt-parent = <&icu0>;
+			interrupts = <136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159>;
+			lantiq,desc-num = <256>;
+			lantiq,dma-hw-poll = <1>;
+			lantiq,dma-pkt-arb = <0>;
+		};
+		eth@8000000 {
+			compatible = "lantiq,svip-net";
+			reg = <0x8000000 0x200>;
+		};
+		sys0@C000000 {
+			compatible = "lantiq,sys0-svip";
+			reg = <0xC000000 0x800>;
+		};
+		sys1@C000800 {
+			compatible = "lantiq,sys1-svip";
+			reg = <0xC000800 0x100>;
+		};
+		port0: port@4100600 {
+			compatible = "lantiq,pinctrl-svip";
+			gpio-controller;
+			#gpio-cells = <2>;
+			reg = <0x4100600 0x100>;
+			lantiq,bank = <0>;
+
+			asc0_pins: asc0 {
+				asc0 {
+					lantiq,groups = "asc0";
+					lantiq,function = "asc";
+				};
+			};
+			asc1_pins: asc1 {
+				asc1 {
+					lantiq,groups = "asc1";
+					lantiq,function = "asc";
+				};
+			};
+			spi0_pins: spi0 {
+				spi0 {
+					lantiq,groups = "spi0";
+					lantiq,function = "spi";
+				};
+			};
+		};
+		port1: port@4108100 {
+			compatible = "lantiq,pinctrl-svip";
+			gpio-controller;
+			#gpio-cells = <2>;
+			reg = <0x4108100 0x100>;
+			lantiq,bank = <1>;
+		};
+		port2: port@4100800 {
+			compatible = "lantiq,pinctrl-svip";
+			gpio-controller;
+			#gpio-cells = <2>;
+			reg = <0x4100800 0x100>;
+			lantiq,bank = <2>;
+		};
+		port3: port@4100900 {
+			compatible = "lantiq,pinctrl-svip";
+			gpio-controller;
+			#gpio-cells = <2>;
+			reg = <0x4100900 0x100>;
+			lantiq,bank = <3>;
+		};
+	};
+
+	/* ICU0 resides on two busses: fpi and sbs2. Driver implementation (irq.c) requires exactly 6 memory regions to be specified, therefore icu0 can't be separated into two parts and placed in both fpi and sbs2 nodes. Driver implementation should be reviewed. */
+	icu0: icu@14106000 {
+		#interrupt-cells = <1>;
+		interrupt-controller;
+		compatible = "lantiq,icu";
+		reg =  <0x14106000 0x28
+			0x14106028 0x28
+			0x1E016000 0x28
+			0x1E016028 0x28
+			0x14106050 0x28
+			0x14106078 0x28>;
+	};
+
+	/* SBS2 (SRAM bus system 2) */
+	sbs2@1E000000 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "lantiq,sysb2", "simple-bus";
+		reg = <0x1E000000 0x1000000>;
+		ranges = <0x0 0x1E000000 0x1000000>;
+
+		port4: port@400 {
+			compatible = "lantiq,pinctrl-svip";
+			gpio-controller;
+			#gpio-cells = <2>;
+			reg = <0x400 0x100>;
+			lantiq,bank = <4>;
+		};
+		status@500 {
+			compatible = "lantiq,status-svip";
+			reg = <0x500 0x100>;
+		};
+		sys2@400000 {
+			compatible = "lantiq,sys2-svip";
+			reg = <0x400000 0x100>;
+		};
+	};
+
+	biu@1F800000 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "lantiq,svip";
+		reg = <0x1F800000 0x800000>;
+		ranges = <0x0 0x1F800000 0x800000>;
+
+		watchdog@803F0 {
+			compatible = "lantiq,wdt";
+			reg = <0x803F0 0x10>;
+		};
+	};
+
+	veth {
+		compatible = "lantiq,svip-ve-net";
+	};
+};
diff --git a/target/linux/lantiq/image/Makefile b/target/linux/lantiq/image/Makefile
--- a/target/linux/lantiq/image/Makefile
+++ b/target/linux/lantiq/image/Makefile
@@ -33,6 +33,14 @@ include $(INCLUDE_DIR)/image.mk
 include $(INCLUDE_DIR)/version.mk
 
 JFFS2_BLOCKSIZE = 64k 128k 256k
+ifeq ($(CONFIG_TARGET_lantiq_svip_be),y)
+JFFS2_BLOCKSIZE += 16k
+JFFS2OPTS += --no-cleanmarkers
+endif
+ifeq ($(CONFIG_TARGET_lantiq_svip_le),y)
+JFFS2_BLOCKSIZE += 16k
+JFFS2OPTS += --no-cleanmarkers
+endif
 KDIR_TMP:=$(KDIR)/tmp
 
 LOADER_MAKE := $(NO_TRACE_MAKE) -C lzma-loader KDIR=$(KDIR)
@@ -70,7 +78,7 @@ endef
 
 define MkImageLzma
 	mkimage -A mips -O linux -T kernel -a 0x80002000 -C lzma \
-		-e 0x80002000 -n $(IMAGE_NAME) \
+		-e 0x80002000 -n "$(IMAGE_NAME)" \
 		-d $(KDIR)/vmlinux$(2)-$(1).lzma $(KDIR)/uImage-$(1)$(2)
 endef
 
@@ -92,13 +100,14 @@ endef
 
 define PatchKernelGzip
 	cp $(KDIR)/vmlinux$(3) $(KDIR)/vmlinux$(3)-$(1)
-	$(STAGING_DIR_HOST)/bin/patch-cmdline $(KDIR)/vmlinux$(3)-$(1) '$(strip $(2))'
+	$(LINUX_DIR)/scripts/dtc/dtc -O dtb -o $(KDIR)/$(1).dtb ../dts/$(1).dts
+	$(STAGING_DIR_HOST)/bin/patch-dtb $(KDIR)/vmlinux$(3)-$(1) $(KDIR)/$(1).dtb
 	$(call CompressGzip,$(KDIR)/vmlinux$(3)-$(1),$(KDIR)/vmlinux$(3)-$(1).gzip)
 endef
 
 define MkImageGzip
 	mkimage -A mips -O linux -T kernel -a 0x80002000 -C gzip \
-		-e 0x80002000 -n $(IMAGE_NAME) \
+		-e 0x80002000 -n "$(IMAGE_NAME)" \
 		-d $(KDIR)/vmlinux$(2)-$(1).gzip $(KDIR)/uImage-$(1)$(2)
 endef
 
@@ -191,6 +200,11 @@ define Image/BuildEVA/jffs2-128k
 	cat $(KDIR)/$(2).eva.align.128k $(KDIR)/root.$(1) > $(BIN_DIR)/$(IMG_PREFIX)-$(2)-$(1).image.eva
 endef
 
+define Image/Build/jffs2-16k
+	dd if=$(KDIR)/uImage-$(2) of=$(KDIR)/uImage-$(2)-$(1) bs=16k conv=sync
+	cat $(KDIR)/uImage-$(2)-$(1) $(KDIR)/root.$(1) > $(BIN_DIR)/$(IMG_PREFIX)-$(2)-$(1).image
+endef
+
 define Image/Build/jffs2-64k
 	# FIXME: leave this line to allow foreach loops
 	dd if=$(KDIR)/uImage-$(2) of=$(KDIR)/uImage-$(2)-$(1) bs=64k conv=sync
@@ -466,6 +480,25 @@ ifeq ($(CONFIG_TARGET_lantiq_svip_be),y)
 
 Image/BuildKernel/Profile/EASY33016=$(call Image/BuildKernelGzip/Template,EASY33016)
 Image/Build/Profile/EASY33016=$(call Image/Build/$(1),$(1),EASY33016)
+Image/BuildKernel/Profile/EASY34016=$(call Image/BuildKernelGzip/Template,EASY34016)
+Image/Build/Profile/EASY34016=$(call Image/Build/$(1),$(1),EASY34016)
+Image/BuildKernel/Profile/EASY336=$(call Image/BuildKernelGzip/Template,EASY336)
+Image/Build/Profile/EASY336=$(call Image/Build/$(1),$(1),EASY336)
+Image/BuildKernel/Profile/EASY32002=$(call Image/BuildKernelGzip/Template,EASY32002)
+Image/Build/Profile/EASY32002=$(call Image/Build/$(1),$(1),EASY32002)
+
+endif
+
+ifeq ($(CONFIG_TARGET_lantiq_svip_le),y)
+
+Image/BuildKernel/Profile/EASY33016=$(call Image/BuildKernelGzip/Template,EASY33016)
+Image/Build/Profile/EASY33016=$(call Image/Build/$(1),$(1),EASY33016)
+Image/BuildKernel/Profile/EASY34016=$(call Image/BuildKernelGzip/Template,EASY34016)
+Image/Build/Profile/EASY34016=$(call Image/Build/$(1),$(1),EASY34016)
+Image/BuildKernel/Profile/EASY336=$(call Image/BuildKernelGzip/Template,EASY336)
+Image/Build/Profile/EASY336=$(call Image/Build/$(1),$(1),EASY336)
+Image/BuildKernel/Profile/EASY32002=$(call Image/BuildKernelGzip/Template,EASY32002)
+Image/Build/Profile/EASY32002=$(call Image/Build/$(1),$(1),EASY32002)
 
 endif
 
diff --git a/target/linux/lantiq/svip_be/profiles/001-lantiq.mk b/target/linux/lantiq/svip_be/profiles/001-lantiq.mk
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/svip_be/profiles/001-lantiq.mk
@@ -0,0 +1,46 @@
+define Profile/EASY336
+  NAME:=EASY336
+  PACKAGES:=
+endef
+
+define Profile/EASY336/Description
+	Lantiq EASY336 evalkit
+endef
+
+$(eval $(call Profile,EASY336))
+
+define Profile/EASY33016
+  NAME:=EASY33016
+  PACKAGES:= \
+	kmod-ifxos ifx-vos ifx-evtlog svip-switch \
+	voice-lib-tapi voice-lib-cli voice-tapi-cli \
+	voice-tapi voice-tapidemo-xt voice-vatests voice-vxt voice-xtbox
+endef
+
+define Profile/EASY33016/Description
+	Lantiq EASY33016 evalkit
+endef
+
+$(eval $(call Profile,EASY33016))
+
+define Profile/EASY34016
+  NAME:=EASY34016
+  PACKAGES:=
+endef
+
+define Profile/EASY34016/Description
+	Lantiq EASY34016 evalkit
+endef
+
+$(eval $(call Profile,EASY34016))
+
+define Profile/EASY32002
+  NAME:=EASY32002
+  PACKAGES:=
+endef
+
+define Profile/EASY32002/Description
+	Lantiq EASY32002 evalkit
+endef
+
+$(eval $(call Profile,EASY32002))
diff --git a/target/linux/lantiq/svip_le/profiles/001-lantiq.mk b/target/linux/lantiq/svip_le/profiles/001-lantiq.mk
new file mode 100644
--- /dev/null
+++ b/target/linux/lantiq/svip_le/profiles/001-lantiq.mk
@@ -0,0 +1,21 @@
+define Profile/EASY336
+  NAME:=EASY336
+  PACKAGES:=
+endef
+
+define Profile/EASY336/Description
+	Lantiq EASY336 evalkit
+endef
+
+$(eval $(call Profile,EASY336))
+
+define Profile/EASY32002
+  NAME:=EASY32002
+  PACKAGES:=
+endef
+
+define Profile/EASY32002/Description
+	Lantiq EASY32002 evalkit
+endef
+
+$(eval $(call Profile,EASY32002))
