# HG changeset patch
# Parent 8eed795db7f8006f66129a2f978fb873c79aa9a4

diff --git a/target/Makefile b/target/Makefile
--- a/target/Makefile
+++ b/target/Makefile
@@ -6,7 +6,7 @@
 #
 curdir:=target
 
-$(curdir)/builddirs:=linux ltq-sdk sdk imagebuilder toolchain
+$(curdir)/builddirs:=linux ltq-sdk ugw-sdk sdk imagebuilder toolchain
 $(curdir)/builddirs-default:=linux
 $(curdir)/builddirs-install:=$(if $(CONFIG_PACKAGE_ecos),,linux) $(if $(CONFIG_SDK),sdk) $(if $(CONFIG_IB),imagebuilder) $(if $(CONFIG_MAKE_TOOLCHAIN),toolchain)
 
