Index: vsftpd-3.0.0/Makefile
===================================================================
--- vsftpd-3.0.0.orig/Makefile	2013-01-21 17:20:25.000000000 +0530
+++ vsftpd-3.0.0/Makefile	2013-01-21 17:22:10.000000000 +0530
@@ -8,7 +8,7 @@
 	-D_FORTIFY_SOURCE=2 \
 	#-pedantic -Wconversion
 
-LIBS	=	-lcrypt -lnsl
+LIBS	=	-lcrypt -lnsl -lcrypto -lssl
 LINK	=
 LDFLAGS	=	-fPIE -pie -Wl,-z,relro -Wl,-z,now
 
Index: vsftpd-3.0.0/builddefs.h
===================================================================
--- vsftpd-3.0.0.orig/builddefs.h	2012-04-05 08:54:56.000000000 +0530
+++ vsftpd-3.0.0/builddefs.h	2013-01-21 17:22:37.000000000 +0530
@@ -3,7 +3,7 @@
 
 #undef VSF_BUILD_TCPWRAPPERS
 #define VSF_BUILD_PAM
-#undef VSF_BUILD_SSL
+#define VSF_BUILD_SSL
 
 #endif /* VSF_BUILDDEFS_H */
 
