diff -Naur ebtables-v2.0.9-2/extensions/ebt_imq.c ebtables-v2.0.9-2_mod/extensions/ebt_imq.c
--- ebtables-v2.0.9-2/extensions/ebt_imq.c	1970-01-01 05:30:00.000000000 +0530
+++ ebtables-v2.0.9-2_mod/extensions/ebt_imq.c	2011-09-28 12:01:44.679803214 +0530
@@ -0,0 +1,83 @@
+#include <stdio.h>
+#include <getopt.h>
+#include <stdlib.h>
+#include "../include/ebtables_u.h"
+#include <linux/netfilter_bridge/ebt_imq.h>
+
+#define IMQ_TODEV '1'
+
+static struct option opts[] =
+{
+	{ "todev"           , required_argument, 0, IMQ_TODEV },
+	{ 0 }
+};
+
+static void help(void)
+{
+  printf(
+    "IMQ options:\n"
+    "  --todev <N>         enqueue to imq<N>, defaults to 0\n");
+}
+
+static void init(struct ebt_entry_target *target)
+{
+  struct ebt_imq_info *imqinfo = (struct ebt_imq_info *)target->data;
+
+  imqinfo->todev = 0;
+}
+
+static int parse(int c, char **argv, int argc, const struct ebt_u_entry *entry,
+   unsigned int *flags, struct ebt_entry_target **target)
+{
+  struct ebt_imq_info *imqinfo = (struct ebt_imq_info *)(*target)->data;
+
+  switch(c) {
+  case IMQ_TODEV:
+    imqinfo->todev = atoi(optarg);
+  }
+
+  return 1;
+}
+
+static void final_check(const struct ebt_u_entry *entry,
+   const struct ebt_entry_target *target, const char *name,
+   unsigned int hookmask, unsigned int time)
+{
+}
+
+static void print(const struct ebt_u_entry *entry,
+   const struct ebt_entry_target *target)
+{
+  struct ebt_imq_info *imqinfo = (struct ebt_imq_info *)target->data;
+
+  printf("--todev %d", imqinfo->todev);
+}
+
+static int compare(const struct ebt_entry_target *t1,
+   const struct ebt_entry_target *t2)
+{
+  struct ebt_imq_info *imqinfo1 = (struct ebt_imq_info *)t1->data;
+  struct ebt_imq_info *imqinfo2 = (struct ebt_imq_info *)t2->data;
+
+  if (imqinfo1->todev != imqinfo2->todev)
+    return 0;
+
+  return 1;
+}
+
+static struct ebt_u_target imq_target =
+{
+	.name		= "IMQ",
+	.size		= sizeof(struct ebt_imq_info),
+	.help		= help,
+	.init		= init,
+	.parse		= parse,
+	.final_check	= final_check,
+	.print		= print,
+	.compare	= compare,
+	.extra_ops	= opts,
+};
+void _init(void)
+{
+	ebt_register_target(&imq_target);
+}
diff -Naur ebtables-v2.0.9-2/extensions/Makefile ebtables-v2.0.9-2_mod/extensions/Makefile
--- ebtables-v2.0.9-2/extensions/Makefile	2011-09-28 12:02:38.733803923 +0530
+++ ebtables-v2.0.9-2_mod/extensions/Makefile	2011-09-28 12:01:44.675803306 +0530
@@ -1,7 +1,7 @@
 #! /usr/bin/make
 
 EXT_FUNC+=802_3 nat arp arpreply ip ip6 standard log redirect vlan mark_m mark \
-          pkttype stp among limit ulog nflog
+          pkttype stp among limit ulog nflog imq
 EXT_TABLES+=filter nat broute
 EXT_OBJS+=$(foreach T,$(EXT_FUNC), extensions/ebt_$(T).o)
 EXT_OBJS+=$(foreach T,$(EXT_TABLES), extensions/ebtable_$(T).o)
diff -Naur ebtables-v2.0.9-2/include/linux/netfilter_bridge/ebt_imq.h ebtables-v2.0.9-2_mod/include/linux/netfilter_bridge/ebt_imq.h
--- ebtables-v2.0.9-2/include/linux/netfilter_bridge/ebt_imq.h	1970-01-01 05:30:00.000000000 +0530
+++ ebtables-v2.0.9-2_mod/include/linux/netfilter_bridge/ebt_imq.h	2011-09-28 12:01:44.680803206 +0530
@@ -0,0 +1,7 @@
+#ifndef __LINUX_BRIDGE_EBT_IMQ_H
+#define __LINUX_BRIDGE_EBT_IMQ_H
+struct ebt_imq_info
+{
+	unsigned int todev;
+};
+#endif
