Index: ebtables-v2.0.9-2/extensions/ebt_pkttype.c
===================================================================
--- ebtables-v2.0.9-2.orig/extensions/ebt_pkttype.c	2010-02-04 02:47:45.000000000 +0530
+++ ebtables-v2.0.9-2/extensions/ebt_pkttype.c	2012-06-17 14:59:49.000000000 +0530
@@ -12,6 +12,11 @@
 #include <getopt.h>
 #include <netdb.h>
 #include "../include/ebtables_u.h"
+/* BEGIN: Workaround for eglibc*/
+#ifndef __aligned_u64
+#define __aligned_u64 __u64 __attribute__((aligned(8)))
+#endif
+/* END: Workaround for eglibc */
 #include <linux/if_packet.h>
 #include <linux/netfilter_bridge/ebt_pkttype.h>
 
