Index: stunnel-4.54/src/Makefile.am
===================================================================
--- stunnel-4.54.orig/src/Makefile.am	2012-03-26 14:23:33.000000000 +0530
+++ stunnel-4.54/src/Makefile.am	2013-01-22 11:56:51.648851000 +0530
@@ -19,7 +19,9 @@
 libstunnel_la_LDFLAGS = -avoid-version
 
 # Red Hat "by design" bug #82369
-stunnel_CPPFLAGS = -I/usr/kerberos/include
+#stunnel_CPPFLAGS = -I/usr/kerberos/include
+
+stunnel_CPPFLAGS = 
 
 # Additional preprocesor definitions
 stunnel_CPPFLAGS += -I$(SSLDIR)/include
