--- a/ChangeLog
+++ b/ChangeLog
@@ -2,6 +2,16 @@ stunnel change log
 
 
 Version 4.55, 2013.03.03, urgency: HIGH:
+
+2016-04-18:
+	004-fix-ntp-validation-patch.patch added
+
+2016-04-18:
+	003-fix-cn-validation-patch.patch added
+
+2013-01-25:
+	002-fix-kerberos-patch.patch added
+
 * Security bugfix
   - OpenSSL updated to version 1.0.1e in Win32/Android builds.
   - Buffer overflow vulnerability fixed in the NTLM authentication
