Index: lighttpd-1.4.30/src/log.c
===================================================================
--- lighttpd-1.4.30.orig/src/log.c
+++ lighttpd-1.4.30/src/log.c
@@ -149,7 +149,7 @@ int log_error_open(server *srv) {
 	openlog("lighttpd", LOG_CONS | LOG_PID, LOG_DAEMON);
 #endif
 
-	srv->errorlog_mode = ERRORLOG_FD;
+	srv->errorlog_mode = ERRORLOG_SYSLOG;
 	srv->errorlog_fd = STDERR_FILENO;
 
 	if (srv->srvconf.errorlog_use_syslog) {
Index: lighttpd-1.4.30/src/mod_cgi.c
===================================================================
--- lighttpd-1.4.30.orig/src/mod_cgi.c
+++ lighttpd-1.4.30/src/mod_cgi.c
@@ -598,7 +598,7 @@ static handler_t cgi_connection_close(se
 #endif
 				return HANDLER_GO_ON;
 			} else {
-				log_error_write(srv, __FILE__, __LINE__, "sd", "cgi died, pid:", pid);
+				log_error_write(srv, __FILE__, __LINE__, "sd", "cgi child process died, pid:", pid);
 				return HANDLER_GO_ON;
 			}
 		}
@@ -1143,7 +1143,7 @@ static int cgi_create_env(server *srv, c
 					c->offset += r;
 					cq->bytes_out += r;
 				} else {
-					log_error_write(srv, __FILE__, __LINE__, "ss", "write() failed due to: ", strerror(errno)); 
+					log_error_write(srv, __FILE__, __LINE__, "ss", "pipe write() from cgi parent to child failed due to: ", strerror(errno)); 
 					con->http_status = 500;
 					break;
 				}
@@ -1310,7 +1310,7 @@ TRIGGER_FUNC(cgi_trigger) {
 				/* FIXME: what if we killed the CGI script with a kill(..., SIGTERM) ?
 				 */
 				if (WTERMSIG(status) != SIGTERM) {
-					log_error_write(srv, __FILE__, __LINE__, "sd", "cleaning up CGI: process died with signal", WTERMSIG(status));
+					log_error_write(srv, __FILE__, __LINE__, "sd", "cleaning up CGI: process died with signal", WTERMSIG(status),":",strsignal(WTERMSIG(status)));
 				}
 			} else {
 				log_error_write(srv, __FILE__, __LINE__, "s", "cleaning up CGI: ended unexpectedly");
Index: lighttpd-1.4.30/src/server.c
===================================================================
--- lighttpd-1.4.30.orig/src/server.c
+++ lighttpd-1.4.30/src/server.c
@@ -246,7 +246,7 @@ static server *server_init(void) {
 
 	/* use syslog */
 	srv->errorlog_fd = STDERR_FILENO;
-	srv->errorlog_mode = ERRORLOG_FD;
+	srv->errorlog_mode = ERRORLOG_SYSLOG;
 
 	srv->split_vals = array_init();
 
