diff -Naur a/block.c b/block.c
--- a/block.c
+++ b/block.c
@@ -483,8 +483,32 @@
 static struct probe_info* _probe_path(char *path)
 {
 	struct probe_info *pr;
-	char tmppath[64];
-
+//we used ubifs?bug!!!
+	char tmppath[64],result[64];
+	if (strcmp(path, "/dev/mtdblock12") == 0)
+	{
+		FILE* fp = fopen ("/sys/devices/virtual/mtd/mtd12/name", "r");
+			if (fp) {
+				if (fgets(result, sizeof(result), fp)!= NULL)
+				{
+					if (strstr(result,"data_vol"))
+						return NULL;
+				}
+				fclose(fp);
+			}
+	}
+	if (strcmp(path, "/dev/mtdblock14") == 0)
+	{
+		FILE* fp = fopen ("/sys/devices/virtual/mtd/mtd14/name", "r");
+			if (fp) {
+				if (fgets(result, sizeof(result), fp)!= NULL)
+				{
+					if (strstr(result,"data_vol"))
+						return NULL;
+				}
+				fclose(fp);
+			}
+	}
 	/* skip ubi device if ubiblock device is present */
 	if (path[5] == 'u' && path[6] == 'b' && path[7] == 'i' &&
 	    path[8] >= '0' && path[8] <= '9' ) {
@@ -505,7 +528,6 @@
 
 	if (glob(path, gl_flags, NULL, &gl) < 0)
 		return -1;
-
 	for (j = 0; j < gl.gl_pathc; j++) {
 		struct probe_info *pr = _probe_path(gl.gl_pathv[j]);
 		if (pr)
@@ -524,6 +546,7 @@
 		_cache_load("/dev/ubiblock*");
 		_cache_load("/dev/ubi[0-9]*");
 	}
+
 	_cache_load("/dev/loop*");
 	_cache_load("/dev/mmcblk*");
 	_cache_load("/dev/sd*");
@@ -1482,7 +1505,8 @@
 #ifdef UBIFS_EXTROOT
 	libubi_t libubi;
 #endif
-
+	ULOG_ERR("GRX500: Unsupported block extroot\n");
+	return -1;
 	if (!getenv("PREINIT"))
 		return -1;
 
@@ -1606,6 +1630,7 @@
 	struct probe_info *pr;
 
 	cache_load(1);
+
 	if (argc == 2) {
 		list_for_each_entry(pr, &devices, list)
 			print_block_info(pr);
@@ -1624,6 +1649,7 @@
 			ULOG_ERR("%s is not a block device\n", argv[i]);
 			continue;
 		}
+
 		pr = find_block_info(NULL, NULL, argv[i]);
 		if (pr)
 			print_block_info(pr);
@@ -1756,6 +1782,7 @@
 
 int main(int argc, char **argv)
 {
+
 	char *base = basename(*argv);
 
 	umask(0);
