# HG changeset patch
# Parent 3343efb18a2a4e97767d605c7fe145ff23bd640e

diff --git a/arch/mips/include/asm/uaccess.h b/arch/mips/include/asm/uaccess.h
--- a/arch/mips/include/asm/uaccess.h
+++ b/arch/mips/include/asm/uaccess.h
@@ -339,6 +339,7 @@ do {									\
 	long __gu_tmp;							\
 									\
 	__asm__ __volatile__(						\
+	"	.set	push					\n"	\
 	"       .set    eva                                     \n"     \
 	"1:     " insn "        %1, 0(%3)                          \n"     \
 	"2:							\n"	\
@@ -350,6 +351,7 @@ do {									\
 	"	.section __ex_table,\"a\"			\n"	\
 	"	"__UA_ADDR "\t1b, 3b				\n"	\
 	"	.previous					\n"	\
+	"	.set	pop					\n"	\
 	: "=r" (__gu_err), "=r" (__gu_tmp)				\
 	: "0" (0), "r" (addr), "i" (-EFAULT));                     \
 									\
@@ -396,6 +398,7 @@ do {									\
 	} __gu_tmp;							\
 									\
 	__asm__ __volatile__(						\
+	"	.set	push					\n"	\
 	"       .set    eva                                     \n"     \
 	"1:     lwe     %1, (%3)                                \n"     \
 	"2:     lwe     %D1, 4(%3)                              \n"     \
@@ -411,6 +414,7 @@ do {									\
 	"	" __UA_ADDR "	1b, 4b				\n"	\
 	"	" __UA_ADDR "	2b, 4b				\n"	\
 	"	.previous					\n"	\
+	"	.set	pop					\n"	\
 	: "=r" (__gu_err), "=&r" (__gu_tmp.l)				\
 	: "0" (0), "r" (addr), "i" (-EFAULT));                          \
 									\
@@ -571,6 +575,7 @@ do {									\
 #define __put_user_asm(insn, ptr)                                       \
 {									\
 	__asm__ __volatile__(						\
+	"	.set	push					\n"	\
 	"       .set        eva                                 \n"     \
 	"1:     " insn "        %2, 0(%3)         # __put_user_asm\n"   \
 	"2:							\n"	\
@@ -582,6 +587,7 @@ do {									\
 	"	.section	__ex_table,\"a\"		\n"	\
 	"	" __UA_ADDR "	1b, 3b				\n"	\
 	"	.previous					\n"	\
+	"	.set	pop					\n"	\
 	: "=r" (__pu_err)						\
 	: "0" (0), "r" (__pu_val), "r" (ptr),                           \
 	  "i" (-EFAULT));						\
@@ -613,6 +619,7 @@ do {									\
 #define __put_user_asm_ll32(ptr)                                        \
 {									\
 	__asm__ __volatile__(						\
+	"	.set	push					\n"	\
 	"       .set    eva                                     \n"     \
 	"1:     swe     %2, (%3)        # __put_user_asm_ll32   \n"     \
 	"2:     swe     %D2, 4(%3)                              \n"     \
@@ -626,6 +633,7 @@ do {									\
 	"	" __UA_ADDR "	1b, 4b				\n"	\
 	"	" __UA_ADDR "	2b, 4b				\n"	\
 	"	.previous"						\
+	"	.set	pop					\n"	\
 	: "=r" (__pu_err)						\
 	: "0" (0), "r" (__pu_val), "r" (ptr),				\
 	  "i" (-EFAULT));                                               \
