# HG changeset patch
# Parent 1652236e14c3c1014e78c3ab60b40b7230d82c5e

diff --git a/arch/mips/include/asm/mach-generic/dma-coherence.h b/arch/mips/include/asm/mach-generic/dma-coherence.h
--- a/arch/mips/include/asm/mach-generic/dma-coherence.h
+++ b/arch/mips/include/asm/mach-generic/dma-coherence.h
@@ -49,7 +49,11 @@ static inline int plat_dma_supported(str
 
 static inline int plat_device_is_coherent(struct device *dev)
 {
-	return coherentio;
+#ifdef CONFIG_DMA_COHERENT
+        return 1;
+#else
+	return coherentio > 0;
+#endif
 }
 
 #endif /* __ASM_MACH_GENERIC_DMA_COHERENCE_H */
