diff --git a/arch/mips/kernel/traps.c b/arch/mips/kernel/traps.c
--- a/arch/mips/kernel/traps.c
+++ b/arch/mips/kernel/traps.c
@@ -156,6 +156,7 @@ static void show_stacktrace(struct task_
 	long stackdata;
 	int i;
 	unsigned long __user *sp = (unsigned long __user *)regs->regs[29];
+	mm_segment_t old_fs = get_fs();
 
 	printk("Stack :");
 	i = 0;
@@ -167,11 +168,15 @@ static void show_stacktrace(struct task_
 			break;
 		}
 
+		set_fs(KERNEL_DS);
+
 		if (__get_user(stackdata, sp++)) {
 			printk(" (Bad stack address)");
 			break;
 		}
 
+		set_fs(old_fs);
+
 		printk(" %0*lx", field, stackdata);
 		i++;
 	}
@@ -182,6 +187,7 @@ static void show_stacktrace(struct task_
 void show_stack(struct task_struct *task, unsigned long *sp)
 {
 	struct pt_regs regs;
+	mm_segment_t old_fs = get_fs();
 	if (sp) {
 		regs.regs[29] = (unsigned long)sp;
 		regs.regs[31] = 0;
@@ -200,7 +206,9 @@ void show_stack(struct task_struct *task
 			prepare_frametrace(&regs);
 		}
 	}
+	set_fs(KERNEL_DS);
 	show_stacktrace(task, &regs);
+	set_fs(old_fs);
 }
 
 static void show_code(unsigned int __user *pc)
