# HG changeset patch
# Parent a315c2719e453ac7ba5c52af7c23f8d30d5a822b

diff --git a/net/core/dev.c b/net/core/dev.c
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4811,6 +4811,13 @@ int __dev_change_flags(struct net_device
 		dev_set_promiscuity(dev, inc);
 	}
 
+	/* hack handling for MULTIPATH, interface down and up shouldn't set multipath
+	 * on only Traffic Steering should do it, this is not clean.
+	 */
+        if ((old_flags ^ flags) & IFF_UP) {     /* Bit is different  ? */
+           dev->flags |= IFF_NOMULTIPATH;
+        }
+
 	/* NOTE: order of synchronization of IFF_PROMISC and IFF_ALLMULTI
 	   is important. Some (broken) drivers set IFF_PROMISC, when
 	   IFF_ALLMULTI is requested not asking us and not reporting.
