diff --git a/net/ipv4/tcp_input.c b/net/ipv4/tcp_input.c
--- a/net/ipv4/tcp_input.c
+++ b/net/ipv4/tcp_input.c
@@ -5219,8 +5219,9 @@ int tcp_rcv_established(struct sock *sk,
 	tp->rx_opt.saw_tstamp = 0;
 
 	/* MPTCP: force slowpath. */
-	if (mptcp(tp))
+	if (mptcp(tp)) {
 		goto slow_path;
+	}
 
 	/*	pred_flags is 0xS?10 << 16 + snd_wnd
 	 *	if header_prediction is to be made
@@ -5395,6 +5396,15 @@ step5:
 
 	/* Process urgent data. */
 	tcp_urg(sk, skb, th);
+#if 1
+       if (sock_flag(sk, SOCK_MPTCP) && !skb_cloned(skb)) {
+       	struct sk_buff *new_skb = skb_copy(skb, GFP_ATOMIC);  // FIXME: Check if GFP_DMA is needed
+               if (new_skb) {
+                       kfree_skb(skb);
+                       skb = new_skb;
+               }
+        }
+#endif
 
 	/* step 7: process the segment text */
 	tcp_data_queue(sk, skb);
