diff -ruN a/include/net/mptcp.h b/include/net/mptcp.h
--- a/include/net/mptcp.h	2016-09-16 13:11:36.000000000 +0200
+++ b/include/net/mptcp.h	2016-09-16 13:35:47.000000000 +0200
@@ -650,6 +650,10 @@
 extern int sysctl_mptcp_syn_retries;
 extern int sysctl_mptcp_master_load;
 extern int sysctl_mptcp_master_loadlimit;
+#ifdef CONFIG_MPTCP_FULLMESH_MF
+extern int sysctl_mptcp_fullmesh_tos_primary;
+extern int sysctl_mptcp_fullmesh_tos_secondary;
+#endif
 
 extern struct workqueue_struct *mptcp_wq;
 
diff -ruN a/net/mptcp/mptcp_ctrl.c b/net/mptcp/mptcp_ctrl.c
--- a/net/mptcp/mptcp_ctrl.c	2016-09-16 13:11:36.000000000 +0200
+++ b/net/mptcp/mptcp_ctrl.c	2016-09-16 13:35:47.000000000 +0200
@@ -1,3 +1,4 @@
+
 /*
  *	MPTCP implementation - MPTCP-control
  *
@@ -69,6 +70,10 @@
 int sysctl_mptcp_syn_retries __read_mostly = 3;
 int sysctl_mptcp_master_load __read_mostly;
 int sysctl_mptcp_master_loadlimit __read_mostly = 90;
+#ifdef CONFIG_MPTCP_FULLMESH_MF
+int sysctl_mptcp_fullmesh_tos_primary __read_mostly = 0x30;
+int sysctl_mptcp_fullmesh_tos_secondary __read_mostly = 0x38;
+#endif
 
 bool mptcp_init_failed __read_mostly;
 
@@ -156,6 +161,22 @@
 		.mode = 0644,
 		.proc_handler = &proc_dointvec
 	},
+#ifdef CONFIG_MPTCP_FULLMESH_MF
+	{
+		.procname = "mptcp_fullmesh_tos_primary",
+		.data = &sysctl_mptcp_fullmesh_tos_primary,
+		.maxlen = sizeof(int),
+		.mode = 0644,
+		.proc_handler = &proc_dointvec
+	},
+	{
+		.procname = "mptcp_fullmesh_tos_secondary",
+		.data = &sysctl_mptcp_fullmesh_tos_secondary,
+		.maxlen = sizeof(int),
+		.mode = 0644,
+		.proc_handler = &proc_dointvec
+	},
+#endif
 	{
 		.procname	= "mptcp_path_manager",
 		.mode		= 0644,
