diff --git a/net/mptcp/mptcp_fullmesh_mf.c b/net/mptcp/mptcp_fullmesh_mf.c
index d25c116..360155b 100644
--- a/net/mptcp/mptcp_fullmesh_mf.c
+++ b/net/mptcp/mptcp_fullmesh_mf.c
@@ -491,7 +491,7 @@ next_subflow:
 	lock_sock_nested(meta_sk, SINGLE_DEPTH_NESTING);
 
 	/* Create the additional subflows for the master interface */
-	if (master_tp && master_tp->mptcp->master && fmp->first_pair == 0) {
+	if (master_tp && master_tp->mptcp && master_tp->mptcp->master && fmp->first_pair == 0) {
 		struct mptcp_loc4 loc;
 		struct mptcp_rem4 rem;
 
@@ -549,7 +549,7 @@ next_subflow:
 	}
 
 #if IS_ENABLED(CONFIG_IPV6)
-	if (master_tp && master_tp->mptcp->master && fmp->first_pair == 0) {
+	if (master_tp && master_tp->mptcp && master_tp->mptcp->master && fmp->first_pair == 0) {
 			struct mptcp_loc6 loc;
 			struct mptcp_rem6 rem;
 
