diff --git a/drivers/pinctrl/pinctrl-xrx500.c b/drivers/pinctrl/pinctrl-xrx500.c
--- a/drivers/pinctrl/pinctrl-xrx500.c
+++ b/drivers/pinctrl/pinctrl-xrx500.c
@@ -328,6 +331,7 @@ static const unsigned pins_vcodec[] = {GPIO1, GPIO34, GPIO35, GPIO36};
 static const unsigned pins_clkout0[] = {GPIO8};
 static const unsigned pins_tdm[] = {GPIO28, GPIO29, GPIO30, GPIO31};
 static const unsigned pins_25MHz[] = {GPIO3};
+static const unsigned pins_gphyled[] = {GPIO5, GPIO6, GPIO9, GPIO14, GPIO34, GPIO35, GPIO36};
 
 static struct ltq_pin_group xrx500_grps[] = {
   GRP_MUX("spi0", SPI0, pins_spi0),
@@ -360,6 +363,7 @@ static struct ltq_pin_group xrx500_grps[] = {
 	GRP_MUX("25MHz", CGU, pins_25MHz),
 	GRP_MUX("vcodec", SLIC, pins_vcodec),
 	GRP_MUX("tdm", TDM, pins_tdm),
+	GRP_MUX("gphyled", GPIO, pins_gphyled),
 };
 
 static const char * const xrx500_spi0_grps[] = {"spi0", "spi0_cs1",
@@ -413,6 +416,7 @@ static const char * const xrx500_mdio_grps[] = {"mdio_l", "mdio_r"};
 static const char * const xrx500_vcodec_grps[] = {"vcodec", "clkout0"};
 static const char * const xrx500_25MHz_grps[] = {"25MHz"};
 static const char * const xrx500_tdm_grps[] = {"tdm"};
+static const char * const xrx500_gphyled_grps[] = {"gphyled"};
 
 static struct ltq_pmx_func xrx500_funcs[] = {
   {"spi0",    ARRAY_AND_SIZE(xrx500_spi0_grps)},
@@ -423,7 +426,8 @@ static struct ltq_pmx_func xrx500_funcs[] = {
	{"25MHz",	ARRAY_AND_SIZE(xrx500_25MHz_grps)},
	{"mdio",	ARRAY_AND_SIZE(xrx500_mdio_grps)},
	{"vcodec",	ARRAY_AND_SIZE(xrx500_vcodec_grps)},
-	{"tdm", 	ARRAY_AND_SIZE(xrx500_tdm_grps)}
+	{"tdm",     ARRAY_AND_SIZE(xrx500_tdm_grps)},
+	{"gphyled",    ARRAY_AND_SIZE(xrx500_gphyled_grps)}
 };
 
 /* ---------  pinconf related code --------- */
diff --git a/drivers/leds/Makefile b/drivers/leds/Makefile
--- a/drivers/leds/Makefile
+++ b/drivers/leds/Makefile
@@ -53,7 +53,7 @@ obj-$(CONFIG_LEDS_RENESAS_TPU)		+= leds-
 obj-$(CONFIG_LEDS_MAX8997)		+= leds-max8997.o
 obj-$(CONFIG_LEDS_LM355x)		+= leds-lm355x.o
 obj-$(CONFIG_LEDS_BLINKM)		+= leds-blinkm.o
-obj-$(CONFIG_LEDS_XRX500)		+= leds-xrx500.o leds-xrx500-proc.o
+obj-$(CONFIG_LEDS_XRX500)		+= leds-xrx500.o leds-xrx500-proc.o leds-k3c.o
 
 # LED SPI Drivers
 obj-$(CONFIG_LEDS_DAC124S085)		+= leds-dac124s085.o
diff --git a/drivers/leds/leds-k3c.c b/drivers/leds/leds-k3c.c
new file mode 100755
--- /dev/null
+++ b/drivers/leds/leds-k3c.c
@@ -0,0 +1,45 @@
+#include <linux/slab.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/of_platform.h>
+#include <linux/of_gpio.h>
+#include <linux/mutex.h>
+#include <linux/gpio.h>
+#include <linux/leds.h>
+#include <linux/io.h>
+#include <linux/clk.h>
+#include <linux/err.h>
+#include <lantiq_soc.h>
+#include <lantiq_timer.h>
+#include <linux/of_gpio.h>
+#include <linux/klogging.h>
+
+struct gpio_led leds[] = {
+    {"lan1", "lan1_trigger", 6, 0, 0, 0},
+    {"lan2", "lan2_trigger", 5, 0, 0, 0},
+    {"lan3", "lan3_trigger", 9, 0, 0, 0},
+    {"wan", "wan_trigger", 14, 0, 0, 0},
+    {"power", "power_trigger", 34, 0, 0, 0},
+    {"sys", "sys_trigger", 36, 1, 0, 0},
+    {"internet", "internet_trigger", 35, 1, 0, 0},
+};
+
+struct gpio_led_platform_data pdata = {
+    .num_leds = ARRAY_SIZE(leds),
+    .leds = leds,
+};
+
+
+struct platform_device k3cled = {
+    .name = "leds-gpio", 
+    .id = -1,
+    .dev = {
+        .platform_data = &pdata,
+    },
+};
+
+module_driver(k3cled, platform_device_register, platform_device_unregister);
+MODULE_AUTHOR("paldier");
+MODULE_DESCRIPTION("K3C LED driver");
+MODULE_LICENSE("GPL");
