--- a/drivers/i2c/busses/i2c-lantiq.c
+++ b/drivers/i2c/busses/i2c-lantiq.c
@@ -33,6 +33,8 @@
 #include <linux/io.h>
 #include <linux/of_irq.h>
 #include <linux/of_i2c.h>
+#include <linux/pinctrl/consumer.h>
+#include <linux/err.h>
 
 #include <lantiq_soc.h>
 #include "i2c-lantiq.h"
@@ -577,6 +579,7 @@ static int __devinit ltq_i2c_probe(struc
 	struct ltq_i2c *priv;
 	struct i2c_adapter *adap;
 	struct resource *mmres, irqres[4];
+	struct pinctrl *pinctrl;
 	int ret = 0;
 
 	dev_dbg(&pdev->dev, "probing\n");
@@ -588,6 +591,10 @@ static int __devinit ltq_i2c_probe(struc
 		return -ENODEV;
 	}
 
+	pinctrl = devm_pinctrl_get_select_default(&pdev->dev);
+	if (IS_ERR(pinctrl))
+		dev_warn(&pdev->dev, "pins are not configured from the driver\n");
+
 	/* allocate private data */
 	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv) {
