# HG changeset patch
# Parent 934460032e87250d36ca903501e7963986a8eeef

diff --git a/arch/mips/lantiq/irq.c b/arch/mips/lantiq/irq.c
--- a/arch/mips/lantiq/irq.c
+++ b/arch/mips/lantiq/irq.c
@@ -102,6 +102,16 @@ static void __iomem *ltq_icu_membase[MAX
 
 static void __iomem *ltq_eiu_membase;
 static struct irq_domain *ltq_domain;
+static DEFINE_SPINLOCK(ltq_eiu_lock);
+
+static void ltq_eiu_w32_mask(uint32_t clr, uint32_t set, uint32_t reg_off)
+{
+        unsigned long flags;
+
+        spin_lock_irqsave(&ltq_eiu_lock, flags);
+        ltq_eiu_w32((ltq_eiu_r32(reg_off) & ~(clr)) | (set), reg_off);
+        spin_unlock_irqrestore(&ltq_eiu_lock, flags);
+}
 
 int ltq_eiu_get_irq(int exin)
 {
@@ -223,8 +233,7 @@ static int ltq_eiu_settype(struct irq_da
 			if (edge)
 				irq_set_handler(d->hwirq, handle_edge_irq);
 
-			ltq_eiu_w32(ltq_eiu_r32(LTQ_EIU_EXIN_C) |
-				(val << (i * 4)), LTQ_EIU_EXIN_C);
+			ltq_eiu_w32_mask((7<< (i * 4)), (val << (i *4)), LTQ_EIU_EXIN_C);
 		}
 	}
 
