# HG changeset patch
# Parent 902da164ae4aa48c382a2b38220e91d606caf9f2

diff --git a/arch/mips/Makefile b/arch/mips/Makefile
--- a/arch/mips/Makefile
+++ b/arch/mips/Makefile
@@ -90,12 +90,17 @@ all-$(CONFIG_SYS_SUPPORTS_ZBOOT)+= vmlin
 cflags-y			+= -G 0 -mno-abicalls -fno-pic -pipe -mno-branch-likely
 cflags-y			+= -msoft-float
 LDFLAGS_vmlinux			+= -G 0 -static -n -nostdlib --gc-sections
-ifdef CONFIG_64BIT
+ifdef CONFIG_64BIT 
 KBUILD_AFLAGS_MODULE		+= -mlong-calls
 KBUILD_CFLAGS_MODULE		+= -mlong-calls
 else
-KBUILD_AFLAGS_MODULE		+= -mno-long-calls
-KBUILD_CFLAGS_MODULE		+= -mno-long-calls
+  ifdef CONFIG_SOC_GRX500
+    KBUILD_AFLAGS_MODULE		+= -mlong-calls
+    KBUILD_CFLAGS_MODULE		+= -mlong-calls
+  else
+    KBUILD_AFLAGS_MODULE		+= -mno-long-calls
+    KBUILD_CFLAGS_MODULE		+= -mno-long-calls
+  endif
 endif
 
 ifndef CONFIG_FUNCTION_TRACER
