# HG changeset patch
# Parent 067bea0ad12b816ac74a281b56806e10d067850d

diff --git a/drivers/mtd/nand/nand_base.c b/drivers/mtd/nand/nand_base.c
--- a/drivers/mtd/nand/nand_base.c
+++ b/drivers/mtd/nand/nand_base.c
@@ -813,7 +813,7 @@ static int nand_wait(struct mtd_info *mt
 {
 
 	int status, state = chip->state;
-	unsigned long timeo = (state == FL_ERASING ? 400 : 20);
+	unsigned long timeo = (state == FL_ERASING ? 400 : 50);
 
 	led_trigger_event(nand_led_trigger, LED_FULL);
 
