# HG changeset patch
# Parent 07910ad05be711e1b85c68cbd903173d67b8557c
diff --git a/drivers/net/ethernet/lantiq/Kconfig b/drivers/net/ethernet/lantiq/Kconfig
--- a/drivers/net/ethernet/lantiq/Kconfig
+++ b/drivers/net/ethernet/lantiq/Kconfig
@@ -3,7 +3,7 @@
 config NET_VENDOR_LANTIQ
         bool "Lantiq network devices"
         default y
-        depends on  SOC_XWAY
+        depends on  SOC_XWAY || SOC_GRX500
         ---help---
           If you have a network (Ethernet) card belonging to this class, say Y
           and read the Ethernet-HOWTO, available from
@@ -33,5 +33,6 @@
 	Lantiq framework for ethernet drivers
 
 source "drivers/net/ethernet/lantiq/switch-api/Kconfig"
+source "drivers/net/ethernet/lantiq/tmu/Kconfig"
 
 endif # NET_VENDOR_LANTIQ
diff --git a/drivers/net/ethernet/lantiq/Makefile b/drivers/net/ethernet/lantiq/Makefile
--- a/drivers/net/ethernet/lantiq/Makefile
+++ b/drivers/net/ethernet/lantiq/Makefile
@@ -1,3 +1,4 @@
 obj-$(CONFIG_LANTIQ_VRX318) += ltq_vrx318.o
 obj-$(CONFIG_LANTIQ_ETH_FRAMEWORK) += lantiq_eth_framework.o
 obj-$(CONFIG_LTQ_ETHSW_API) += switch-api/
+obj-$(CONFIG_LTQ_TMU) += tmu/
