# HG changeset patch
# Parent f8f9c5c8dbbc38eae5ebd6a32ad4481a085a6eb7

diff --git a/drivers/net/ethernet/lantiq/Kconfig b/drivers/net/ethernet/lantiq/Kconfig
--- a/drivers/net/ethernet/lantiq/Kconfig
+++ b/drivers/net/ethernet/lantiq/Kconfig
@@ -64,6 +64,21 @@ config XRX500_ETH_DRV_COC_SUPPORT
 		---help---
 		CoC support in the Ethernet driver of XRX500 series
 
+config HGU_BONDING
+ 	bool "Enable ethernet bonding for GRX500 HGU boards"
+	depends on LTQ_ETH_XRX500
+	default n
+	---help---
+	Lantiq HGU ethernet bonding enable
+
+config HGU_BOND_PORT
+	int "Bonded port number"
+	depends on HGU_BONDING
+	default 1
+	---help---
+	   This is the PAE port to be used as bond port along with the port 15 
+	   to provide 2Gbps ethernet bond port for downstream traffic.
+
 config SW_ROUTING_MODE
         bool "Enable the SW routing mode in ethernet driver"
 	depends on LTQ_ETH_XRX500
diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -67,6 +67,10 @@ static u8 g_my_ethaddr[MAX_ADDR_LEN * 2]
 
 static char wan_iface[IFNAMSIZ] = "eth1";
 
+#if defined(CONFIG_HGU_BONDING) && CONFIG_HGU_BONDING
+static struct net_device *eth_wan_dev = NULL;
+#endif
+
 static struct net_device *eth_dev[NUM_ETH_INF];
 static struct module g_ltq_eth_module[NUM_ETH_INF];
 static u32 g_rx_csum_offload = 0;
@@ -265,6 +269,15 @@ static void eth_rx(struct net_device *de
     skb->dev = dev;
     skb->protocol = eth_type_trans(skb, dev);
 
+#if defined(CONFIG_HGU_BONDING) && CONFIG_HGU_BONDING
+    if( priv->dp_port_id == CONFIG_HGU_BOND_PORT ) {
+        if(!eth_wan_dev){
+            eth_wan_dev=dev_get_by_name(&init_net, wan_iface);
+        }
+        skb->dev=eth_wan_dev;
+	dev=eth_wan_dev;
+    }
+#endif
 	if (dev->features & NETIF_F_RXCSUM) {
 		skb->ip_summed = CHECKSUM_UNNECESSARY;
 	}
