# HG changeset patch
# Parent b8e6a1bbfb3912b59734e56c1f4728f5ca343748

diff --git a/drivers/atm/lantiq/ltqmips_atm_core.h b/drivers/atm/lantiq/ltqmips_atm_core.h
--- a/drivers/atm/lantiq/ltqmips_atm_core.h
+++ b/drivers/atm/lantiq/ltqmips_atm_core.h
@@ -274,6 +274,7 @@ extern unsigned int ifx_atm_get_qsb_clk(
 
 #define IFX_REG_R32(_r)                    __raw_readl((volatile unsigned int *)(_r))
 #define IFX_REG_W32(_v, _r)               __raw_writel((_v), (volatile unsigned int *)(_r))
+#define IFX_REG_W32_MASK(_clr, _set, _r)   IFX_REG_W32((IFX_REG_R32((_r)) & ~(_clr)) | (_set), (_r))
 
 #define INT_NUM_IM0_IRL24 (INT_NUM_IM0_IRL0 + 24)
 #define INT_NUM_IM1_IRL29 (INT_NUM_IM1_IRL0 + 29)
diff --git a/drivers/atm/lantiq/ltqmips_atm_ppe_vr9.h b/drivers/atm/lantiq/ltqmips_atm_ppe_vr9.h
--- a/drivers/atm/lantiq/ltqmips_atm_ppe_vr9.h
+++ b/drivers/atm/lantiq/ltqmips_atm_ppe_vr9.h
@@ -182,6 +182,8 @@
 #define PDMA_ALIGNMENT                  4
 #define EMA_ALIGNMENT                   PDMA_ALIGNMENT
 
+#define INT_NUM_IM2_IRL24    (INT_NUM_IM2_IRL0 + 24)
+
 /*
  *  Mailbox IGU1 Interrupt
  */
diff --git a/drivers/atm/lantiq/ltqmips_atm_vr9.c b/drivers/atm/lantiq/ltqmips_atm_vr9.c
--- a/drivers/atm/lantiq/ltqmips_atm_vr9.c
+++ b/drivers/atm/lantiq/ltqmips_atm_vr9.c
@@ -45,15 +45,12 @@
 /*
  *  Chip Specific Head File
  */
-#include <common_routines.h>
-#include <ltq_regs.h>
-#include <ltq_pmu.h>
-#include <ltq_rcu.h>
+#include <linux/clk.h>
+#include <lantiq_soc.h>
+
 #include "ltqmips_atm_core.h"
 #include "ltqmips_atm_fw_vr9.h"
 
-
-
 /*
  * ####################################
  *              Definition
@@ -97,7 +94,8 @@ static inline void clear_share_buffer(vo
 
 static inline void init_pmu(void)
 {
-    //*PMU_PWDCR &= ~((1 << 29) | (1 << 22) | (1 << 21) | (1 << 19) | (1 << 18));
+/*
+    PMU_PWDCR &= ~((1 << 29) | (1 << 22) | (1 << 21) | (1 << 19) | (1 << 18));
     PPE_TOP_PMU_SETUP(IFX_PMU_ENABLE);
     PPE_SLL01_PMU_SETUP(IFX_PMU_ENABLE);
     PPE_TC_PMU_SETUP(IFX_PMU_ENABLE);
@@ -105,10 +103,12 @@ static inline void init_pmu(void)
     PPE_QSB_PMU_SETUP(IFX_PMU_ENABLE);
     PPE_TPE_PMU_SETUP(IFX_PMU_ENABLE);
     DSL_DFE_PMU_SETUP(IFX_PMU_ENABLE);
+*/
 }
 
 static inline void uninit_pmu(void)
 {
+/*
     PPE_SLL01_PMU_SETUP(IFX_PMU_DISABLE);
     PPE_TC_PMU_SETUP(IFX_PMU_DISABLE);
     PPE_EMA_PMU_SETUP(IFX_PMU_DISABLE);
@@ -116,6 +116,7 @@ static inline void uninit_pmu(void)
     PPE_TPE_PMU_SETUP(IFX_PMU_DISABLE);
     DSL_DFE_PMU_SETUP(IFX_PMU_DISABLE);
     PPE_TOP_PMU_SETUP(IFX_PMU_DISABLE);
+*/
 }
 
 static inline void reset_ppe(void)
@@ -303,7 +304,10 @@ void ifx_pp32_stop(int pp32)
 
 unsigned int ifx_atm_get_qsb_clk(void)
 {
-    unsigned int shift = (IFX_REG_R32(IFX_CGU_CLKFSR) >> 28) & 0x03;
+    unsigned int shift = ((ltq_cgu_r32(0x0010)) >> 28) & 0x03;
+	unsigned int fpi_clk_rate;
 
-    return ifx_get_fpi_hz() >> shift;
+	struct clk *fpi_clk = clk_get_fpi();
+	fpi_clk_rate = clk_get_rate(fpi_clk);
+	return (fpi_clk_rate >> shift);
 }
