# HG changeset patch
# Parent 251279cb807c7ebe195b98d50e393a4dce36e399
diff --git a/drivers/pinctrl/pinctrl-xway.c b/drivers/pinctrl/pinctrl-xway.c
--- a/drivers/pinctrl/pinctrl-xway.c
+++ b/drivers/pinctrl/pinctrl-xway.c
@@ -279,6 +279,7 @@
 static const unsigned pins_pci_req2[] = {GPIO31};
 static const unsigned pins_pci_req3[] = {GPIO3};
 static const unsigned pins_pci_req4[] = {GPIO37};
+static const unsigned pins_led[] = {GPIO1, GPIO9, GPIO5, GPIO33, GPIO11, GPIO12, GPIO15};
 
 static const unsigned ase_pins_jtag[] = {GPIO7, GPIO8, GPIO9, GPIO10, GPIO11};
 static const unsigned ase_pins_asc[] = {GPIO5, GPIO6};
@@ -362,6 +363,7 @@
 	GRP_MUX("nand wp", EBU, pins_nand_wp),
 	GRP_MUX("nand se", EBU, pins_nand_se),
 	GRP_MUX("arc_jtag", GPIO, pins_arc_jtag),
+	GRP_MUX("led", GPIO, pins_led),
 };
 
 static const struct ltq_pin_group ase_grps[] = {
@@ -404,6 +406,7 @@
 static const char * const xway_jtag_grps[] = {"jtag"};
 static const char * const xway_stp_grps[] = {"stp"};
 static const char * const xway_nmi_grps[] = {"nmi"};
+static const char * const xway_led_grps[] = {"led"};
 
 /* ar9/vr9/gr9 */
 static const char * const xrx_mdio_grps[] = {"mdio"};
@@ -466,6 +469,7 @@
 	{"ebu",		ARRAY_AND_SIZE(xrx_ebu_grps)},
 	{"mdio",	ARRAY_AND_SIZE(xrx_mdio_grps)},
 	{"arc_jtag",	ARRAY_AND_SIZE(xrx_arc_jtag_grps)},
+	{"led",		ARRAY_AND_SIZE(xway_led_grps)},
 };
 
 static const struct ltq_pmx_func ase_funcs[] = {
