add lantiq MEI driver entries to Kconfig and Makefile

--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -605,5 +605,27 @@ config TILE_SROM
 	  device appear much like a simple EEPROM, and knows
 	  how to partition a single ROM for multiple purposes.
 
+config IFXMIPS_DSL_CPE_MEI
+	bool
+	depends on LTQ_DSL_CPE_MEI_ARX
+	default y
+
+config LTQ_DSL_CPE_MEI_ARX
+	tristate "DSL CPE MEI ARX Low Level Driver"
+	default n
+	---help---
+      This is the DSL low level driver for ARX platforms.
+
+config DSL_MEI_CPE_DRV
+	bool
+	depends on LTQ_DSL_CPE_MEI_VRX
+	default y
+
+config LTQ_DSL_CPE_MEI_VRX
+	tristate "DSL CPE MEI VRX Low Level Driver"
+	default n
+	---help---
+      This is the DSL low level driver for VRX platforms.
+
 endmenu
