# HG changeset patch
# Parent 94a74d03b8cc6607c51c38e68587351da2b187ea

diff --git a/drivers/spi/ltq_sflash.c b/drivers/spi/ltq_sflash.c
--- a/drivers/spi/ltq_sflash.c
+++ b/drivers/spi/ltq_sflash.c
@@ -48,14 +48,18 @@
 #include <lantiq_ssc.h>
 #include "ltq_sflash.h"
 
+
 #define IFX_SFLASH_MODE                IFX_SSC_MODE_0
 #define IFX_SFLASH_PRIORITY            IFX_SSC_PRIO_LOW
 #define IFX_SFLASH_FRAGSIZE            264  /* must be more than 260, page size
 					* + cmd + addr, must be a multiple
 					* of DMA burst length */
 #define IFX_SFLASH_MAXFIFOSIZE         32 /* FIFO mode */
-
+#if defined(CONFIG_USE_PALLADIUM)
+#define IFX_SFLASH_BAUDRATE  			13156
+#else
 #define IFX_SFLASH_BAUDRATE            10000000 /*  20 MHz */
+#endif
 #define IFX_SFLASH_CS                  IFX_SSC_WHBGPOSTAT_OUT0_POS
 
 
@@ -458,7 +462,7 @@ static int ifx_sflash_rdsr(ifx_spi_dev_t
 static int ifx_sflash_sync(ifx_spi_dev_t *dev)
 {
 	int ret = 0;
-	u8 status;
+	u8 status = 0;
 	int count = 0;
 	unsigned long deadline;
 
@@ -693,6 +697,10 @@ static int  ifx_spi_flash_probe(ifx_spi_
 	IFX_SFLASH_PRINT("Vendor %02x Type %02x sig %02x\n",
 		 pdev->manufacturer_id, pdev->device_id1,
 		 pdev->device_id2);
+
+	printk("Vendor %02x Type %02x sig %02x\n",
+		 pdev->manufacturer_id, pdev->device_id1,
+		 pdev->device_id2);
 	for (i = 0; i < ARRAY_SIZE(flash_manufacturers); ++i) {
 		if (pdev->manufacturer_id == flash_manufacturers[i].id)
 			break;
@@ -836,7 +844,8 @@ ifx_spi_flash_write(struct mtd_info *mtd
 		ifx_spi_write(addr, mem, len_this_lp);
 		total += len_this_lp;
 	}
-	*retlen = len;
+	if (retlen)
+		*retlen = len;
 	up(&ifx_sflash_sem);
 	return 0;
 }
@@ -915,6 +924,8 @@ static INLINE IFX_SSC_HANDLE ifx_spi_fla
 	if (of_machine_is_compatible("lantiq,vr9")
 		|| of_machine_is_compatible("lantiq,ar9"))
 		ssc_cfg.cs_data = IFX_SSC_WHBGPOSTAT_OUT3_POS;
+	else if (of_machine_is_compatible("lantiq,grx500"))
+		ssc_cfg.cs_data = IFX_SSC_WHBGPOSTAT_OUT1_POS;
 	else
 		ssc_cfg.cs_data = IFX_SSC_WHBGPOSTAT_OUT0_POS;
 	ssc_cfg.fragSize = IFX_SFLASH_FRAGSIZE;
