# HG changeset patch
# Parent 9c02b637a02bcb22f88b17c841dca02af97e6673

diff --git a/arch/mips/mm/dma-default.c b/arch/mips/mm/dma-default.c
--- a/arch/mips/mm/dma-default.c
+++ b/arch/mips/mm/dma-default.c
@@ -69,7 +69,7 @@ static gfp_t massage_gfp_flags(const str
 	else
 #endif
 #if defined(CONFIG_ZONE_DMA) && !defined(CONFIG_ZONE_DMA32)
-	     if (dev->coherent_dma_mask < DMA_BIT_MASK(sizeof(phys_addr_t) * 8))
+	     if (dev->coherent_dma_mask < DMA_BIT_MASK(64))
 		dma_flag = __GFP_DMA;
 	else
 #endif
