# HG changeset patch
# Parent cbf5ab19877f8cf3712024108b3e94ee693c506d

diff --git a/include/uapi/linux/if_ether.h b/include/uapi/linux/if_ether.h
--- a/include/uapi/linux/if_ether.h
+++ b/include/uapi/linux/if_ether.h
@@ -31,8 +31,14 @@
 #define ETH_ALEN	6		/* Octets in one ethernet addr	 */
 #define ETH_HLEN	14		/* Total octets in header.	 */
 #define ETH_ZLEN	60		/* Min. octets in frame sans FCS */
+/* Mini-Jumbo Frame support */
+#ifdef CONFIG_LTQ_MINI_JUMBO_FRAME_SUPPORT
+#define ETH_DATA_LEN	1586		/* Max. octets in payload	 */
+#define ETH_FRAME_LEN	1600		/* Max. octets in frame sans FCS */
+#else
 #define ETH_DATA_LEN	1500		/* Max. octets in payload	 */
 #define ETH_FRAME_LEN	1514		/* Max. octets in frame sans FCS */
+#endif
 #define ETH_FCS_LEN	4		/* Octets in the FCS		 */
 
 /*
