# HG changeset patch
# Parent eedd7e4f889f4253f77d9c272ab4e4462c64d1e3

diff --git a/net/ipv4/ip_input.c b/net/ipv4/ip_input.c
--- a/net/ipv4/ip_input.c
+++ b/net/ipv4/ip_input.c
@@ -415,8 +438,15 @@ int __ipt_optimized ip_rcv(struct sk_buf
 
 	iph = ip_hdr(skb);
 
+#ifdef CONFIG_LTQ_TOE_DRIVER
+	if (!skb_is_nonlinear(skb) && !skb_has_frag_list(skb)) {
+		if (unlikely(ip_fast_csum((u8 *)iph, iph->ihl)))
+			goto csum_error;
+	}
+#else
 	if (unlikely(ip_fast_csum((u8 *)iph, iph->ihl)))
 		goto csum_error;
+#endif
 
 	len = ntohs(iph->tot_len);
 	if (skb->len < len) {
