# HG changeset patch
# Parent 47375a012593254aab47f1381cb1814d455aeb13

diff --git a/drivers/net/ethernet/lantiq/Kconfig b/drivers/net/ethernet/lantiq/Kconfig
--- a/drivers/net/ethernet/lantiq/Kconfig
+++ b/drivers/net/ethernet/lantiq/Kconfig
@@ -59,6 +59,14 @@ config LTQ_ETH_OAM
   ---help---
      This feature helps to create a eth oam virtual netdevice and handle the dot1agd control packets.
 
+config LTQ_TOE_DRIVER
+	bool "Enable the TOE (TCP Offload Engine) driver for XRX500"
+        depends on LTQ_ETH_XRX500
+        default n
+        ---help---
+        Lantiq TOE (TCP Offload Engine) driver for XRX500 series.
+
+source "drivers/net/ethernet/lantiq/datapath/Kconfig"
 source "drivers/net/ethernet/lantiq/switch-api/Kconfig"
 source "drivers/net/ethernet/lantiq/tmu/Kconfig"
 source "drivers/net/ethernet/lantiq/cbm/Kconfig"
diff --git a/drivers/net/ethernet/lantiq/Makefile b/drivers/net/ethernet/lantiq/Makefile
--- a/drivers/net/ethernet/lantiq/Makefile
+++ b/drivers/net/ethernet/lantiq/Makefile
@@ -5,6 +5,7 @@ ifneq ($(CONFIG_LTQ_ETH_OAM),)
   obj-y += ltq_eth_oam_netif.o
 endif
 obj-$(CONFIG_LTQ_ETHSW_API) += switch-api/
+obj-$(CONFIG_LTQ_DATAPATH) += datapath/
 obj-$(CONFIG_LTQ_TMU) += tmu/
 obj-$(CONFIG_LTQ_CBM) += cbm/
 obj-$(CONFIG_LTQ_ETH_XRX500) += ltq_eth_drv_xrx500.o
