diff --git a/net/core/skbuff.c b/net/core/skbuff.c
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -892,6 +892,12 @@ EXPORT_SYMBOL(consume_skb);
 
 void __copy_skb_header(struct sk_buff *new, const struct sk_buff *old)
 {
+#ifdef CONFIG_LTQ_CBM
+	new->DW0		= old->DW0;
+	new->DW1		= old->DW1;
+	new->DW2		= old->DW2;
+	new->DW3		= old->DW3;
+#endif
 	new->tstamp		= old->tstamp;
 	new->dev		= old->dev;
 	new->transport_header	= old->transport_header;
