diff -Naur linux-3.10.104.orig/include/net/mac80211.h linux-3.10.104.new/include/net/mac80211.h
--- linux-3.10.104.orig/include/net/mac80211.h	2017-01-19 11:41:22.000000000 +0200
+++ linux-3.10.104.new/include/net/mac80211.h	2017-01-30 10:01:43.000000000 +0200
@@ -1404,6 +1404,7 @@
  *	the station moves to associated state.
  * @smps_mode: current SMPS mode (off, static or dynamic)
  * @tx_rates: rate control selection table
+ * @mfp: indicates whether the STA uses management frame protection or not.
  */
 struct ieee80211_sta {
 	u32 supp_rates[IEEE80211_NUM_BANDS];
@@ -1418,6 +1419,7 @@
 	enum ieee80211_sta_rx_bandwidth bandwidth;
 	enum ieee80211_smps_mode smps_mode;
 	struct ieee80211_sta_rates __rcu *rates;
+	bool mfp;	
 
 	/* must be last */
 	u8 drv_priv[0] __aligned(sizeof(void *));
diff -Naur linux-3.10.104.orig/net/mac80211/cfg.c linux-3.10.104.new/net/mac80211/cfg.c
--- linux-3.10.104.orig/net/mac80211/cfg.c	2016-10-21 13:13:35.000000000 +0300
+++ linux-3.10.104.new/net/mac80211/cfg.c	2017-01-29 10:10:28.000000000 +0200
@@ -1238,6 +1238,7 @@
 	}
 
 	if (mask & BIT(NL80211_STA_FLAG_MFP)) {
+		sta->sta.mfp = !!(set & BIT(NL80211_STA_FLAG_MFP));
 		if (set & BIT(NL80211_STA_FLAG_MFP))
 			set_sta_flag(sta, WLAN_STA_MFP);
 		else
diff -Naur linux-3.10.104.orig/net/mac80211/mlme.c linux-3.10.104.new/net/mac80211/mlme.c
--- linux-3.10.104.orig/net/mac80211/mlme.c	2016-10-21 13:13:35.000000000 +0300
+++ linux-3.10.104.new/net/mac80211/mlme.c	2017-01-29 10:18:48.000000000 +0200
@@ -2721,8 +2721,12 @@
 
 	rate_control_rate_init(sta);
 
-	if (ifmgd->flags & IEEE80211_STA_MFP_ENABLED)
+	if (ifmgd->flags & IEEE80211_STA_MFP_ENABLED){
 		set_sta_flag(sta, WLAN_STA_MFP);
+		sta->sta.mfp = true;
+	} else {
+		sta->sta.mfp = false;
+	}
 
 	if (elems.wmm_param)
 		set_sta_flag(sta, WLAN_STA_WME);
