# HG changeset patch
# Parent 73323f1f9dc5526df4c8a11715f3e66d80b78cf5

diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -391,6 +391,10 @@ int br_add_if(struct net_bridge *br, str
 
 	list_add_rcu(&p->list, &br->port_list);
 
+	dev->features |= (NETIF_F_SOFT_FEATURES | NETIF_F_SG);
+
+	br->dev->features |= (NETIF_F_SOFT_FEATURES | NETIF_F_SG);
+
 	netdev_update_features(br->dev);
 
 	spin_lock_bh(&br->lock);
