# HG changeset patch
# Parent 89966879885244ffc05ed49cd2c9d4cacbed5bcc

diff --git a/net/core/dev.c b/net/core/dev.c
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -183,6 +183,8 @@ EXPORT_SYMBOL(dev_base_lock);
 
 seqcount_t devnet_rename_seq;
 
+struct sk_buff *handle_ing(struct sk_buff *skb, struct packet_type **pt_prev,
+					 int *ret, struct net_device *orig_dev);
 static inline void dev_base_seq_inc(struct net *net)
 {
 	while (++net->dev_base_seq == 0);
@@ -3386,7 +3388,17 @@ static int ing_filter(struct sk_buff *sk
 	return result;
 }
 
-static inline struct sk_buff *handle_ing(struct sk_buff *skb,
+int check_ingress(struct sk_buff *skb)
+{
+	struct netdev_queue *rxq = rcu_dereference(skb->dev->ingress_queue);
+
+	if (!rxq || rxq->qdisc == &noop_qdisc)
+		return 1;
+	return 0;
+}
+EXPORT_SYMBOL(check_ingress);
+
+struct sk_buff *handle_ing(struct sk_buff *skb,
 					 struct packet_type **pt_prev,
 					 int *ret, struct net_device *orig_dev)
 {
@@ -3411,6 +3423,7 @@ out:
 	skb->tc_verd = 0;
 	return skb;
 }
+EXPORT_SYMBOL(handle_ing);
 #endif
 
 /**
