# HG changeset patch
# Parent fad52631dd38dc5e763f98594a60a72b26fdb61a
diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -27,6 +27,10 @@
 
 #include "br_private.h"
 
+#ifdef CONFIG_LTQ_ETH_XRX500
+#define LTQ_PMAC_HDR_LEN 8
+#endif
+
 /*
  * Determine initial path cost based on speed.
  * using recommendations from 802.1d standard
@@ -254,6 +258,9 @@ int br_add_bridge(struct net *net, const
 
 	dev_net_set(dev, net);
 	dev->rtnl_link_ops = &br_link_ops;
+#ifdef CONFIG_LTQ_ETH_XRX500
+	dev->needed_headroom = LTQ_PMAC_HDR_LEN;
+#endif
 
 	res = register_netdev(dev);
 	if (res)
