# HG changeset patch
# Parent f9ecaa5ebf8c336d8f6720e9e6493718857bf446

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -42,6 +42,10 @@
 #include <linux/extmark.h>
 #endif
 
+#ifdef CONFIG_LTQ_DP_MPE_FASTHOOK_TEST
+#include <net/mpe_fast_hook_test.h>
+#endif
+
 /* reccommended as it's faster*/
 #define GET_DATA_FROM_MARK_OPT(mark, mask, pos, value) value = ((mark & mask) >> pos)
 #define SET_DATA_FROM_MARK_OPT(mark, mask, pos, value) mark &= ~mask; mark |= ((value << pos) & mask)
@@ -425,6 +429,10 @@ struct sk_buff {
 #ifdef CONFIG_LTQ_CBM
 	__u32 DW0, DW1, DW2, DW3;
 #endif
+
+#ifdef CONFIG_LTQ_DP_MPE_FASTHOOK_TEST     
+	struct ltq_mpe_fasthook_info *mpe_skb_info; /*for MPE fast hooking test*/
+#endif    
 	ktime_t			tstamp;
 
 	struct sock		*sk;
