diff --git a/net/mac80211/main.c b/net/mac80211/main.c
--- a/net/mac80211/main.c
+++ b/net/mac80211/main.c
@@ -34,6 +34,8 @@
 #include "cfg.h"
 #include "debugfs.h"
 
+static bool k3c = true; //bad idea,but we need it
+
 void ieee80211_configure_filter(struct ieee80211_local *local)
 {
 	u64 mc;
@@ -970,13 +972,30 @@
 		goto fail_rate;
 	}
 
+	// bug?,i cant add this interface in driver,we have to named first sta as wlan3 and second as wlan1
 	/* add one default STA interface if supported */
 	if (local->hw.wiphy->interface_modes & BIT(NL80211_IFTYPE_STATION)) {
-		result = ieee80211_if_add(local, "wlan%d", NULL,
+		if (k3c) {
+			result = ieee80211_if_add(local, "wlan3", NULL,
 					  NL80211_IFTYPE_STATION, NULL);
-		if (result)
+			if (result)
 			wiphy_warn(local->hw.wiphy,
 				   "Failed to add default virtual iface\n");
+			k3c = false;
+		}
+		else {
+			result = ieee80211_if_add(local, "wlan1", NULL,
+					  NL80211_IFTYPE_STATION, NULL);
+			if (result)
+			wiphy_warn(local->hw.wiphy,
+				   "Failed to add default virtual iface\n");
+		}
+		//result = ieee80211_if_add(local, "wlan%d", NULL,
+		//			  NL80211_IFTYPE_STATION, NULL);
+		//if (result)
+		//	wiphy_warn(local->hw.wiphy,
+		//		   "Failed to add default virtual iface\n");
+
 	}
 
 	rtnl_unlock();
