diff -urN oprofile-0.9.4/daemon/Makefile.am oprofile-0.9.4.new/daemon/Makefile.am
--- oprofile-0.9.4/daemon/Makefile.am	2008-04-28 23:23:24.000000000 +0200
+++ oprofile-0.9.4.new/daemon/Makefile.am	2009-07-26 15:23:23.000000000 +0200
@@ -1,4 +1,4 @@
-SUBDIRS = liblegacy .
+SUBDIRS = .
 
 oprofiled_SOURCES = \
 	init.c \
@@ -42,7 +42,6 @@
 bin_PROGRAMS = oprofiled
 
 oprofiled_LDADD = \
-	liblegacy/liblegacy.a \
 	../libabi/libabi.a \
 	../libdb/libodb.a \
 	../libop/libop.a \
diff -urN oprofile-0.9.4/daemon/Makefile.in oprofile-0.9.4.new/daemon/Makefile.in
--- oprofile-0.9.4/daemon/Makefile.in	2008-07-18 01:14:44.000000000 +0200
+++ oprofile-0.9.4.new/daemon/Makefile.in	2009-07-26 15:23:14.000000000 +0200
@@ -68,7 +68,7 @@
 	opd_events.$(OBJEXT) opd_mangling.$(OBJEXT) \
 	opd_perfmon.$(OBJEXT) opd_anon.$(OBJEXT) opd_spu.$(OBJEXT)
 oprofiled_OBJECTS = $(am_oprofiled_OBJECTS)
-oprofiled_DEPENDENCIES = liblegacy/liblegacy.a ../libabi/libabi.a \
+oprofiled_DEPENDENCIES = ../libabi/libabi.a \
 	../libdb/libodb.a ../libop/libop.a ../libutil/libutil.a
 DEFAULT_INCLUDES = -I. -I$(srcdir) -I$(top_builddir)
 depcomp = $(SHELL) $(top_srcdir)/depcomp
@@ -240,7 +240,7 @@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 topdir = @topdir@
-SUBDIRS = liblegacy .
+SUBDIRS = .
 oprofiled_SOURCES = \
 	init.c \
 	oprofiled.c \
@@ -279,7 +279,6 @@
 # -fno-omit-frame-pointer needed for daemon build: see ChangeLog-2004 02-23
 AM_CFLAGS = @OP_CFLAGS@ -fno-omit-frame-pointer
 oprofiled_LDADD = \
-	liblegacy/liblegacy.a \
 	../libabi/libabi.a \
 	../libdb/libodb.a \
 	../libop/libop.a \
diff -urN oprofile-0.9.4/daemon/oprofiled.c oprofile-0.9.4.new/daemon/oprofiled.c
--- oprofile-0.9.4/daemon/oprofiled.c	2008-04-28 23:23:23.000000000 +0200
+++ oprofile-0.9.4.new/daemon/oprofiled.c	2009-07-26 15:25:25.000000000 +0200
@@ -73,7 +73,6 @@
 static char * events;
 static int showvers;
 static struct oprofiled_ops * opd_ops;
-extern struct oprofiled_ops opd_24_ops;
 extern struct oprofiled_ops opd_26_ops;
 
 #define OPD_IMAGE_FILTER_HASH_SIZE 32
@@ -465,9 +464,6 @@
 static struct oprofiled_ops * get_ops(void)
 {
 	switch (op_get_interface()) {
-		case OP_INTERFACE_24:
-			printf("Using 2.4 OProfile kernel interface.\n");
-			return &opd_24_ops;
 		case OP_INTERFACE_26:
 			printf("Using 2.6+ OProfile kernel interface.\n");
 			return &opd_26_ops;
