--- samba-3.0.24.orig/source/Makefile	2011-02-03 20:54:31.985853105 +0530
+++ samba-3.0.24/source/Makefile	2011-02-03 20:56:19.404602337 +0530
@@ -813,7 +813,7 @@
 ######################################################################
 # now the rules...
 ######################################################################
-all : SHOWFLAGS proto_exists bin/smbd  bin/smbpasswd  bin/nmbd
+all : SHOWFLAGS proto_exists bin/smbd  bin/smbpasswd
 
 pam_smbpass : SHOWFLAGS bin/pam_smbpass.so
 
