--- a/pubkeys/wlan.key.pub.pem
+++ b/pubkeys/wlan.key.pub.pem
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0P24he7rvTkaHO+7xzfT
+IrMjZC8hOYkTjpE8Wl8wQvMRVCfoyMgjG6yRonhuRLcN4Wsl6KCL6YBXGfU9YG7j
+qPooWBuUAwN8dEYk0Fh65r1F6A9LvrEorjI2c4AfYNxXpAavcaqpY1abBbEFfcrP
+ere+YGaRZGE0D/mADW1ltcQKbCUXRMwWKIjSLEkwoscZDR7p/pPM/V2PFC0euNAK
+EUxJ6PKeKSS1WQr/2nWg7mtB5ldgYbqZSvk5ovdu94nQ6RTVJQ4LRGvy3l1koAio
+Rw2geoiQQIVv9RlB4L9L7JYxhUYPMQSsNhLyCocpY/J4C8PR7RJs8wuHCr2IXXvx
+swIDAQAB
+-----END PUBLIC KEY-----
