--- mini_httpd-1.19/mini_httpd.c	2011-10-14 23:40:02.986656899 +0530
+++ mini_httpd-1.19.bk/mini_httpd.c	2011-10-14 23:39:40.094659317 +0530
@@ -743,7 +743,7 @@
 	listen4_fd = -1;
 #ifdef CONFIG_FEATURE_CLI_REQUEST
        cli_server = openCLIServer();
-     //  set_cloexec_flag (cli_server,1);
+       set_cloexec_flag (cli_server,1);
 #endif
     /* If we didn't get any valid sockets, fail. */
     if ( listen4_fd == -1 && listen6_fd == -1 )
@@ -1840,7 +1840,7 @@
 	    int count;
 	    int bNewFile = 1;
 	    memset(buf,0x00,MAX_MEMORY_BUFF);
-	    while ((count = my_read(fd, buf,MAX_MEMORY_BUFF-1)) > 0) {
+	    while ((count = read(fd, buf,MAX_MEMORY_BUFF-1)) > 0) {
 		    if ((strstr(fixed_mime_type,"html") != NULL ) || (strstr(file,"asp") != NULL)
 #ifdef CONFIG_FEATURE_CLI_REQUEST
 			  || (strstr(file,"cli") != NULL)
@@ -2048,7 +2048,7 @@
     ** into our buffer.  We also have to do this for all SSL CGIs.
     */
 #ifdef USE_SSL
-    if ( ( method == METHOD_POST && request_len > request_idx ) || do_ssl )
+    if (  method == METHOD_POST  || do_ssl )
 #else /* USE_SSL */
     if ( ( method == METHOD_POST ))//&& request_len > request_idx ) )
 #endif /* USE_SSL */
@@ -2698,10 +2698,11 @@
     send_error_tail();
 
     send_response();
-
+#if 0
 #ifdef USE_SSL
     SSL_free( ssl );
 #endif /* USE_SSL */
+#endif
     /*exit( 1 );*/
     return -1;
     }
