--- mini_httpd-1.19/mini_httpd.c	2012-01-13 20:47:32.705268860 +0530
+++ mini_httpd-1.19.bk/mini_httpd.c	2012-01-13 20:46:46.917269294 +0530
@@ -107,7 +107,7 @@
 static void sendHeaders(int s, char* title, char* extra_header, char* me, char* mt, off_t b, time_t mod,char *redirectURL);
 int (*ifx_psendGoform)(const char *,const char *, const char *, int, const char *,const char *,const char *,int,int) = NULL;
 int (*ifx_psendDHTMLFile)(char *,const char *,int,int,int *) = NULL;
-int (*ifx_AuthHandler_ptr)(const char *,int,int) = NULL;
+int (*ifx_AuthHandler_ptr)(const char *,const char *,int,int) = NULL;
 /*for supporting CGI  */
 
 #if defined(AF_INET6) && defined(IN6_IS_ADDR_V4MAPPED)
@@ -1598,7 +1598,7 @@
 #ifdef CONFIG_FEATURE_CLI_REQUEST
         if(ifx_clireq == 0)
 #endif
-        if((*ifx_AuthHandler_ptr)(file, conn_fd, conn_fd) == 1)
+        if((*ifx_AuthHandler_ptr)(file,referer, conn_fd, conn_fd) == 1)
         {  
             syslog( LOG_INFO, "ifx_AuthHandler_ptr failed"); 
 	    return;
