--- mini_httpd-1.19.old/mini_httpd.c	2012-08-23 19:05:20.000000000 +0530
+++ mini_httpd-1.19.aug23/mini_httpd.c	2012-08-23 18:08:42.000000000 +0530
@@ -1100,7 +1100,11 @@
 #endif
 	    set_cloexec_flag (conn_fd,1);
 	    handle_request();
-	    free(request);   
+	    if(request)	
+	    {
+	    	free(request);
+		request = NULL;   
+	    }
             /* Always chdir to / after a chroot. */
 	    if ( chdir( cwd ) < 0 )
 	    {
@@ -1659,8 +1663,11 @@
 #endif
 	got_one: ;
 	}
-    free(request);
+    if(request)
+    {
+     	free(request);
 	request=NULL;
+    }
 #ifdef USE_SSL
     SSL_free( ssl );
 #endif /* USE_SSL */
@@ -2901,6 +2908,7 @@
     {
     request_size = 0;
     request_idx = 0;
+    request_len = 0;
     }
 
 static void
@@ -2915,6 +2923,11 @@
     int i;
     char c;
    
+    
+    if(request == NULL)
+    {	 
+    	return (char*) 0;
+    }
     for ( i = request_idx; request_idx < request_len; ++request_idx )
 	{
 	c = request[request_idx];
@@ -2942,6 +2955,8 @@
 start_response( void )
     {
     response_size = 0;
+    response_len = 0;
+    
     }
 
 static void
