--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,35 @@
+2013-07-09:
+	001-lq-httpd-support.patch added to remove compilation dependency over ltq packages
+
+2012-08-23:
+	014-lq-minihttpd-peakcpu-fix.patch added - Null check added before free
+
+2012-01-13:
+	013-lq-https-support-mod.patch added
+
+2011-10-31:
+	012-lq-https-crash-fix.patch added to change the return value to -1 instead of 1.
+
+2011-10-14:
+	010-lq-https-support.patch added
+
+2011-04-24:
+	009-lq-post-memory-leakage.patch added to free the memory
+
+2011-04-07:
+	008-lq-memory-leakage.patch added to fix mem leak
+
+2011-02-23:
+	006-lq-mini_httpd-login-issue.patch added to fix login issue
+
+2011-02-10:
+	005-lq-httpd-support.patch added
+
+2011-02-07:
+	004-lq-httpd-support.patch added to support cli request
+
+2011-01-19:
+	003-lq-httpd-support.patch added for error handling
+
+2011-01-11:
+	002-lq-httpd-support.patch added - mini-httpd changes for redirecting to frame_setup.htm when passwd protection is disable.
