diff --git a/package/kernel/linux/modules/input.mk b/package/kernel/linux/modules/input.mk
index 86ff33c..291a553 100644
--- a/package/kernel/linux/modules/input.mk
+++ b/package/kernel/linux/modules/input.mk
@@ -176,7 +176,7 @@ $(eval $(call KernelPackage,input-matrixkmap))
 define KernelPackage/acpi-button
   SUBMENU:=$(INPUT_MODULES_MENU)
   TITLE:=ACPI Button Support
-  DEPENDS:=@(TARGET_x86_generic||TARGET_x86_kvm_guest||TARGET_x86_xen_domu) +kmod-input-evdev
+  DEPENDS:=@(TARGET_x86_generic||TARGET_x86_kvm_guest||TARGET_x86_xen_domu||TARGET_x86_puma) +kmod-input-evdev
   KCONFIG:=CONFIG_ACPI_BUTTON
   FILES:=$(LINUX_DIR)/drivers/acpi/button.ko
   AUTOLOAD:=$(call AutoLoad,06,button)
