UGW: generic fixes to compile UGW

diff --git a/package/base-files/Makefile b/package/base-files/Makefile
--- a/package/base-files/Makefile
+++ b/package/base-files/Makefile
@@ -31,7 +31,7 @@ endif
 define Package/base-files
   SECTION:=base
   CATEGORY:=Base system
-  DEPENDS:=+netifd +libc +procd +jsonfilter +SIGNED_PACKAGES:usign
+  DEPENDS:=+!DEFAULT_LTQBASE:netifd +libc +!DEFAULT_LTQBASE:procd +!DEFAULT_LTQBASE:jsonfilter +SIGNED_PACKAGES:usign
   TITLE:=Base filesystem for OpenWrt
   URL:=http://openwrt.org/
   VERSION:=$(PKG_RELEASE)-$(REVISION)
