add a config option for VERSION_IMAGE

diff --git a/package/base-files/image-config.in b/package/base-files/image-config.in
--- a/package/base-files/image-config.in
+++ b/package/base-files/image-config.in
@@ -218,6 +218,25 @@ if VERSIONOPT
 		help
 			Enable this to include the version number in firmware image, SDK-
 			and Image Builder archive file names
+
+	config VERSION_IMAGE
+		string
+		prompt "Image name"
+		default "%N_%V"
+		help
+			This is the release image name embedded in the image.
+			It may contain the following placeholders:
+			 %R .. Revision number
+			 %V .. Release version or revision number, uppercase
+			 %v .. Release version or revision number, lowercase
+			 %C .. Release version or "Bleeding Edge", uppercase
+			 %c .. Release version or "bleeding_edge", lowercase
+			 %N .. Release name, uppercase
+			 %n .. Release name, lowercase
+			 %D .. Distribution name or "OpenWrt", uppercase
+			 %d .. Distribution name or "openwrt", lowercase
+			 %T .. Target name
+			 %S .. Target/Subtarget name
 endif
 
 
