add support for profile.arch and diag.arch.sh

diff --git a/package/base-files/files/etc/diag.sh b/package/base-files/files/etc/diag.sh
--- a/package/base-files/files/etc/diag.sh
+++ b/package/base-files/files/etc/diag.sh
@@ -2,3 +2,7 @@
 # Copyright (C) 2006-2009 OpenWrt.org
 
 set_state() { :; }
+
+if [ -e /etc/diag.arch.sh ]; then
+	. /etc/diag.arch.sh
+fi
diff --git a/package/base-files/files/etc/profile b/package/base-files/files/etc/profile
--- a/package/base-files/files/etc/profile
+++ b/package/base-files/files/etc/profile
@@ -13,3 +13,5 @@ export PS1='\u@\h:\w\$ '
 
 [ -x /usr/bin/arp ] || arp() { cat /proc/net/arp; }
 [ -x /usr/bin/ldd ] || ldd() { LD_TRACE_LOADED_OBJECTS=1 $*; }
+
+[ -f /etc/profile.arch ] && . /etc/profile.arch
