# HG changeset patch
# Parent 90638bbfef9fb0b7ffbda191329866d3784c857f

diff --git a/package/utils/e2fsprogs/Makefile b/package/utils/e2fsprogs/Makefile
--- a/package/utils/e2fsprogs/Makefile
+++ b/package/utils/e2fsprogs/Makefile
@@ -99,6 +99,7 @@ CONFIGURE_VARS += \
 	ac_cv_lib_pthread_sem_init=no
 
 CONFIGURE_ARGS += \
+	--disable-backtrace \
 	--disable-testio-debug \
 	--enable-elf-shlibs	\
 	--disable-libuuid	\
