build dm9000 driver into kernel (for nfs usage)

diff --git a/package/kernel/linux/modules/netdevices.mk b/package/kernel/linux/modules/netdevices.mk
--- a/package/kernel/linux/modules/netdevices.mk
+++ b/package/kernel/linux/modules/netdevices.mk
@@ -731,6 +731,20 @@ endef
 
 $(eval $(call KernelPackage,dm9000))
 
+define KernelPackage/dm9000-nfs
+  SUBMENU:=$(NETWORK_DEVICES_MENU)
+  TITLE:=Davicom 9000 Ethernet support (in kernel for nfs)
+  KCONFIG:=CONFIG_DM9000=y \
+    CONFIG_DM9000_DEBUGLEVEL=4 \
+    CONFIG_DM9000_FORCE_SIMPLE_PHY_POLL=y
+endef
+
+define KernelPackage/dm9000-nfs/description
+ Kernel driver for Davicom 9000 Ethernet adapters (in kernel for nfs).
+endef
+
+$(eval $(call KernelPackage,dm9000-nfs))
+
 
 define KernelPackage/forcedeth
   SUBMENU:=$(NETWORK_DEVICES_MENU)
