install ipset kmods based on kernel version.

diff --git a/package/kernel/linux/modules/netfilter.mk b/package/kernel/linux/modules/netfilter.mk
--- a/package/kernel/linux/modules/netfilter.mk
+++ b/package/kernel/linux/modules/netfilter.mk
@@ -228,14 +228,14 @@ IPSET_MODULES:= \
 	ipset/ip_set_bitmap_ipmac \
 	ipset/ip_set_bitmap_port \
 	ipset/ip_set_hash_ip \
-	ipset/ip_set_hash_ipmark \
+	$(if $(call kernel_version_cmp,-ge,$(LINUX_VERSION),3.15.0),ipset/ip_set_hash_ipmark) \
 	ipset/ip_set_hash_ipport \
 	ipset/ip_set_hash_ipportip \
 	ipset/ip_set_hash_ipportnet \
-	ipset/ip_set_hash_mac \
-	ipset/ip_set_hash_netportnet \
+	$(if $(call kernel_version_cmp,-ge,$(LINUX_VERSION),3.18.0),ipset/ip_set_hash_mac) \
+	$(if $(call kernel_version_cmp,-ge,$(LINUX_VERSION),3.13.0),ipset/ip_set_hash_netportnet) \
 	ipset/ip_set_hash_net \
-	ipset/ip_set_hash_netnet \
+	$(if $(call kernel_version_cmp,-ge,$(LINUX_VERSION),3.13.0),ipset/ip_set_hash_netnet) \
 	ipset/ip_set_hash_netport \
 	ipset/ip_set_hash_netiface \
 	ipset/ip_set_list_set \
@@ -460,6 +460,7 @@ define KernelPackage/ipt-tproxy
   	CONFIG_NETFILTER_XT_MATCH_SOCKET \
   	CONFIG_NETFILTER_XT_TARGET_TPROXY
   FILES:= \
+	$(if $(call kernel_patchver_lt,3.10.12),$(LINUX_DIR)/net/netfilter/nf_tproxy_core.ko) \
   	$(foreach mod,$(IPT_TPROXY-m),$(LINUX_DIR)/net/$(mod).ko)
   AUTOLOAD:=$(call AutoProbe,$(notdir nf_tproxy_core $(IPT_TPROXY-m)))
   $(call AddDepends/ipt)
