diff -Naur racoon2-20100526a.orig/aclocal.m4 racoon2-20100526a/aclocal.m4
--- racoon2-20100526a.orig/aclocal.m4	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/aclocal.m4	2010-06-15 10:54:01.000000000 -0400
@@ -329,7 +329,7 @@
 		for d in /usr/include "/lib/modules/`uname -r`/build/include"
 		do
 			if test -f "$d/linux/pfkeyv2.h" ; then
-				OPTFLAG="$OPTFLAG -I$d"
+				#OPTFLAG="$OPTFLAG -I$d"
 				echo "$d added to the include path"
 				break
 			fi
diff -Naur racoon2-20100526a.orig/iked/configure racoon2-20100526a/iked/configure
--- racoon2-20100526a.orig/iked/configure	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/iked/configure	2010-06-15 10:54:01.000000000 -0400
@@ -1383,7 +1383,7 @@
 		for d in /usr/include "/lib/modules/`uname -r`/build/include"
 		do
 			if test -f "$d/linux/pfkeyv2.h" ; then
-				OPTFLAG="$OPTFLAG -I$d"
+				#OPTFLAG="$OPTFLAG -I$d"
 				echo "$d added to the include path"
 				break
 			fi
@@ -1758,7 +1758,7 @@
 fi
 
 echo "$ac_t""$ac_cv_func_memcmp_clean" 1>&6
-test $ac_cv_func_memcmp_clean = no && LIBOBJS="$LIBOBJS memcmp.${ac_objext}"
+#test $ac_cv_func_memcmp_clean = no && LIBOBJS="$LIBOBJS memcmp.${ac_objext}"
 
 echo $ac_n "checking return type of signal handlers""... $ac_c" 1>&6
 echo "configure:1765: checking return type of signal handlers" >&5
@@ -2150,19 +2150,19 @@
 	else
 		echo "configure: warning: can't locate libcrypto.a" 1>&2
 	fi
-else
-	for d in /usr/local /usr/local/openssl /usr/pkg /usr '' ; do
-		if test -f "$d/lib/libcrypto.a" ; then
-						echo "$ac_t""using $d/lib/libcrypto.a" 1>&6
-			openssl_libdir="$d/lib"
-			LDFLAGS="$LDFLAGS -L$openssl_libdir"
-			case "$openssl_libdir" in
-			/lib|/usr/lib)	;;
-			*)		CPPFLAGS="$CPPFLAGS -I$d/include";;
-			esac
-			break
-		fi
-	done
+#else
+#	for d in /usr/local /usr/local/openssl /usr/pkg /usr '' ; do
+#		if test -f "$d/lib/libcrypto.a" ; then
+#						echo "$ac_t""using $d/lib/libcrypto.a" 1>&6
+#			openssl_libdir="$d/lib"
+#			LDFLAGS="$LDFLAGS -L$openssl_libdir"
+#			case "$openssl_libdir" in
+#			/lib|/usr/lib)	;;
+#			*)		CPPFLAGS="$CPPFLAGS -I$d/include";;
+#			esac
+#			break
+#		fi
+#	done
 fi
 LIBS="-lcrypto $LIBS"
 case $host_os in
diff -Naur racoon2-20100526a.orig/iked/Makefile.in racoon2-20100526a/iked/Makefile.in
--- racoon2-20100526a.orig/iked/Makefile.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/iked/Makefile.in	2010-06-15 10:54:01.000000000 -0400
@@ -66,16 +66,16 @@
 install: install-prog install-doc
 
 install-prog: $(PROG)
-	$(INSTALL) -d $(sbindir)
-	$(INSTALL_PROGRAM) $(PROG) $(sbindir)
+	$(INSTALL) -d $(DESTDIR)$(sbindir)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)
 
 install-doc:
-	$(INSTALL) -d $(mandir)/man8
+	$(INSTALL) -d $(DESTDIR)$(mandir)/man8
 # not friendly with -n  :-(
 	sysconfdir="$$(echo '$(sysconfdir)' | sed 's/%/\\\%/g')"; \
 	for f in $(MAN); do \
 		sed -e s%\@sysconfdir\@%"$$sysconfdir"%g < $$f > $${f}.tmp; \
-		$(INSTALL_DATA) $${f}.tmp $(mandir)/man$${f##*.}/$$f; \
+		$(INSTALL_DATA) $${f}.tmp $(DESTDIR)$(mandir)/man$${f##*.}/$$f; \
 		rm $${f}.tmp; \
 	done
 
diff -Naur racoon2-20100526a.orig/kinkd/Makefile.in racoon2-20100526a/kinkd/Makefile.in
--- racoon2-20100526a.orig/kinkd/Makefile.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/kinkd/Makefile.in	2010-06-15 10:54:01.000000000 -0400
@@ -49,14 +49,14 @@
 ../lib/libracoon.a:		# check its timestamp only when there is.
 
 install: all
-	$(INSTALL) -d $(sbindir)
-	$(INSTALL_PROGRAM) $(PROG) $(sbindir)
-	$(INSTALL) -d $(mandir)/man8
+	$(INSTALL) -d $(DESTDIR)$(sbindir)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)
+	$(INSTALL) -d $(DESTDIR)$(mandir)/man8
 # not friendly with -n  :-(
 	sysconfdir="$$(echo '$(sysconfdir)' | sed 's/%/\\\%/g')"; \
 	for f in $(MAN); do \
 		sed -e s%\@sysconfdir\@%"$$sysconfdir"%g < $$f > $${f}.tmp; \
-		$(INSTALL_DATA) $${f}.tmp $(mandir)/man$${f##*.}/$$f; \
+		$(INSTALL_DATA) $${f}.tmp $(DESTDIR)$(mandir)/man$${f##*.}/$$f; \
 		rm $${f}.tmp; \
 	done
 
diff -Naur racoon2-20100526a.orig/lib/configure racoon2-20100526a/lib/configure
--- racoon2-20100526a.orig/lib/configure	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/lib/configure	2010-06-15 10:54:01.000000000 -0400
@@ -1608,7 +1608,7 @@
 		for d in /usr/include "/lib/modules/`uname -r`/build/include"
 		do
 			if test -f "$d/linux/pfkeyv2.h" ; then
-				OPTFLAG="$OPTFLAG -I$d"
+				#OPTFLAG="$OPTFLAG -I$d"
 				echo "$d added to the include path"
 				break
 			fi
@@ -1909,7 +1909,7 @@
 fi
 
 echo "$ac_t""$ac_cv_func_memcmp_clean" 1>&6
-test $ac_cv_func_memcmp_clean = no && LIBOBJS="$LIBOBJS memcmp.${ac_objext}"
+#test $ac_cv_func_memcmp_clean = no && LIBOBJS="$LIBOBJS memcmp.${ac_objext}"
 
 echo $ac_n "checking for strftime""... $ac_c" 1>&6
 echo "configure:1916: checking for strftime" >&5
diff -Naur racoon2-20100526a.orig/pskgen/autogen.spmd.pwd.in racoon2-20100526a/pskgen/autogen.spmd.pwd.in
--- racoon2-20100526a.orig/pskgen/autogen.spmd.pwd.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/pskgen/autogen.spmd.pwd.in	2010-06-15 10:54:01.000000000 -0400
@@ -1,11 +1,11 @@
 #!/bin/sh
 
-if [ -f @prefix@/etc/racoon2/spmd.pwd ]
+if [ -f $(DESTDIR)@prefix@/etc/racoon2/spmd.pwd ]
 then
-	mv @prefix@/etc/racoon2/spmd.pwd @prefix@/etc/racoon2/spmd.pwd.bak
+	mv $(DESTDIR)@prefix@/etc/racoon2/spmd.pwd $(DESTDIR)@prefix@/etc/racoon2/spmd.pwd.bak
 fi
 
-if [ -x @prefix@/sbin/pskgen ]
+if [ -x $(DESTDIR)@prefix@/sbin/pskgen ]
 then
-	@prefix@/sbin/pskgen -r -o @prefix@/etc/racoon2/spmd.pwd
+	$(DESTDIR)@prefix@/sbin/pskgen -r -o $(DESTDIR)@prefix@/etc/racoon2/spmd.pwd
 fi
diff -Naur racoon2-20100526a.orig/pskgen/Makefile.in racoon2-20100526a/pskgen/Makefile.in
--- racoon2-20100526a.orig/pskgen/Makefile.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/pskgen/Makefile.in	2010-06-15 10:54:29.000000000 -0400
@@ -17,11 +17,11 @@
 all:
 
 install: all
-	$(INSTALL) -d $(sbindir)
-	$(INSTALL_SCRIPT) $(PROG) $(sbindir)
-	$(INSTALL_DATA) $(PROG).8 $(mandir)/man8
-	$(INSTALL) -d $(prefix)/etc/racoon2
-	sh ./autogen.spmd.pwd
+	$(INSTALL) -d $(DESTDIR)$(sbindir)
+	$(INSTALL_SCRIPT) $(PROG) $(DESTDIR)$(sbindir)
+	$(INSTALL_DATA) $(PROG).8 $(DESTDIR)$(mandir)/man8
+	echo $(INSTALL) -d $(DESTDIR)$(prefix)/etc/racoon2
+	echo sh ./autogen.spmd.pwd
 
 depend:
 
diff -Naur racoon2-20100526a.orig/samples/Makefile.in racoon2-20100526a/samples/Makefile.in
--- racoon2-20100526a.orig/samples/Makefile.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/samples/Makefile.in	2010-06-15 10:54:01.000000000 -0400
@@ -12,60 +12,60 @@
 all:
 
 install: all install-startup-@startup_scripts@ install-samples install-hook
-	$(INSTALL) -d -o 0 -g 0 -m 700 /var/run/racoon2
+	$(INSTALL) -d -m 700 $(DESTDIR)/var/run/racoon2
 
 install-samples:
-	$(INSTALL) -d $(sysconfdir)
-	$(INSTALL) -m 600 racoon2.conf $(sysconfdir)/racoon2.conf.sample
-	$(INSTALL) -m 600 vals.conf $(sysconfdir)/vals.conf.sample
-	$(INSTALL) -m 600 default.conf $(sysconfdir)/default.conf.sample
-	$(INSTALL) -m 600 transport_ike.conf $(sysconfdir)/transport_ike.conf.sample
-	$(INSTALL) -m 600 tunnel_ike.conf $(sysconfdir)/tunnel_ike.conf.sample
-	$(INSTALL) -m 600 tunnel_ike_natt.conf $(sysconfdir)/tunnel_ike_natt.conf.sample
-	$(INSTALL) -m 600 transport_kink.conf $(sysconfdir)/transport_kink.conf.sample
-	$(INSTALL) -m 600 tunnel_kink.conf $(sysconfdir)/tunnel_kink.conf.sample
-	$(INSTALL) -m 600 local-test.conf $(sysconfdir)/local-test.conf.sample
-	$(INSTALL) -d -m 700 $(sysconfdir)/psk
-	$(INSTALL) -d -m 700 $(sysconfdir)/cert
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)
+	$(INSTALL) -m 600 racoon2.conf $(DESTDIR)$(sysconfdir)/racoon2.conf.sample
+	$(INSTALL) -m 600 vals.conf $(DESTDIR)$(sysconfdir)/vals.conf.sample
+	$(INSTALL) -m 600 default.conf $(DESTDIR)$(sysconfdir)/default.conf.sample
+	$(INSTALL) -m 600 transport_ike.conf $(DESTDIR)$(sysconfdir)/transport_ike.conf.sample
+	$(INSTALL) -m 600 tunnel_ike.conf $(DESTDIR)$(sysconfdir)/tunnel_ike.conf.sample
+	$(INSTALL) -m 600 tunnel_ike_natt.conf $(DESTDIR)$(sysconfdir)/tunnel_ike_natt.conf.sample
+	$(INSTALL) -m 600 transport_kink.conf $(DESTDIR)$(sysconfdir)/transport_kink.conf.sample
+	$(INSTALL) -m 600 tunnel_kink.conf $(DESTDIR)$(sysconfdir)/tunnel_kink.conf.sample
+	$(INSTALL) -m 600 local-test.conf $(DESTDIR)$(sysconfdir)/local-test.conf.sample
+	$(INSTALL) -d -m 700 $(DESTDIR)$(sysconfdir)/psk
+	$(INSTALL) -d -m 700 $(DESTDIR)$(sysconfdir)/cert
 
 install-startup-rc-d:
-	$(INSTALL) -d $(sysconfdir)/rc.d
-	$(INSTALL_SCRIPT) rc.d-iked $(sysconfdir)/rc.d/iked
-	$(INSTALL_SCRIPT) rc.d-kinkd $(sysconfdir)/rc.d/kinkd
-	$(INSTALL_SCRIPT) rc.d-spmd $(sysconfdir)/rc.d/spmd
-	$(INSTALL_SCRIPT) rc.d-racoon2 $(sysconfdir)/rc.d/racoon2
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/rc.d
+	$(INSTALL_SCRIPT) rc.d-iked $(DESTDIR)$(sysconfdir)/rc.d/iked
+	$(INSTALL_SCRIPT) rc.d-kinkd $(DESTDIR)$(sysconfdir)/rc.d/kinkd
+	$(INSTALL_SCRIPT) rc.d-spmd $(DESTDIR)$(sysconfdir)/rc.d/spmd
+	$(INSTALL_SCRIPT) rc.d-racoon2 $(DESTDIR)$(sysconfdir)/rc.d/racoon2
 
 install-startup-init-d:
-	$(INSTALL) -d $(sysconfdir)/init.d
-	$(INSTALL_SCRIPT) init.d-iked $(sysconfdir)/init.d/iked
-	$(INSTALL_SCRIPT) init.d-kinkd $(sysconfdir)/init.d/kinkd
-	$(INSTALL_SCRIPT) init.d-spmd $(sysconfdir)/init.d/spmd
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/init.d
+	$(INSTALL_SCRIPT) init.d-iked $(DESTDIR)$(sysconfdir)/init.d/S38-iked
+	$(INSTALL_SCRIPT) init.d-kinkd $(DESTDIR)$(sysconfdir)/init.d/S39-kinkd
+	$(INSTALL_SCRIPT) init.d-spmd $(DESTDIR)$(sysconfdir)/init.d/S37-spmd
 
 install-startup-none:
 
 install-hook:
-	$(INSTALL) -d $(sysconfdir)/hook
-	$(INSTALL_DATA) functions $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/ikesa-up.d
-	$(INSTALL_SCRIPT) ikesa-up $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/ikesa-down.d
-	$(INSTALL_SCRIPT) ikesa-down $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/ikesa-rekey.d
-	$(INSTALL_SCRIPT) ikesa-rekey $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/child-up.d
-	$(INSTALL_SCRIPT) child-up $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/child-down.d
-	$(INSTALL_SCRIPT) child-down $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/child-rekey.d
-	$(INSTALL_SCRIPT) child-rekey $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/migration.d
-	$(INSTALL_SCRIPT) migration $(sysconfdir)/hook
-	$(INSTALL_SCRIPT) 00ikesaup_sample $(sysconfdir)/hook/ikesa-up.d
-	$(INSTALL_SCRIPT) 00childup_sample $(sysconfdir)/hook/child-up.d
-	$(INSTALL) -d $(sysconfdir)/hook/ph1-up.d
-	$(INSTALL_SCRIPT) ph1-up $(sysconfdir)/hook
-	$(INSTALL) -d $(sysconfdir)/hook/ph1-down.d
-	$(INSTALL_SCRIPT) ph1-down $(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL_DATA) functions $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/ikesa-up.d
+	$(INSTALL_SCRIPT) ikesa-up $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/ikesa-down.d
+	$(INSTALL_SCRIPT) ikesa-down $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/ikesa-rekey.d
+	$(INSTALL_SCRIPT) ikesa-rekey $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/child-up.d
+	$(INSTALL_SCRIPT) child-up $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/child-down.d
+	$(INSTALL_SCRIPT) child-down $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/child-rekey.d
+	$(INSTALL_SCRIPT) child-rekey $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/migration.d
+	$(INSTALL_SCRIPT) migration $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL_SCRIPT) 00ikesaup_sample $(DESTDIR)$(sysconfdir)/hook/ikesa-up.d
+	$(INSTALL_SCRIPT) 00childup_sample $(DESTDIR)$(sysconfdir)/hook/child-up.d
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/ph1-up.d
+	$(INSTALL_SCRIPT) ph1-up $(DESTDIR)$(sysconfdir)/hook
+	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/hook/ph1-down.d
+	$(INSTALL_SCRIPT) ph1-down $(DESTDIR)$(sysconfdir)/hook
 
 # empty
 depend:
diff -Naur racoon2-20100526a.orig/spmd/configure racoon2-20100526a/spmd/configure
--- racoon2-20100526a.orig/spmd/configure	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/spmd/configure	2010-06-15 10:54:01.000000000 -0400
@@ -1624,6 +1624,8 @@
 fi
 
 
+ac_cv_sizeof_long_long=0
+ 
 echo $ac_n "checking size of long long""... $ac_c" 1>&6
 echo "configure:1629: checking size of long long" >&5
 if eval "test \"`echo '$''{'ac_cv_sizeof_long_long'+set}'`\" = set"; then
@@ -1970,19 +1972,19 @@
 	else
 		echo "configure: warning: can't locate libcrypto.a" 1>&2
 	fi
-else
-	for d in /usr/local /usr/local/openssl /usr/pkg /usr '' ; do
-		if test -f "$d/lib/libcrypto.a" ; then
-						echo "$ac_t""using $d/lib/libcrypto.a" 1>&6
-			openssl_libdir="$d/lib"
-			LDFLAGS="$LDFLAGS -L$openssl_libdir"
-			case "$openssl_libdir" in
-			/lib|/usr/lib)	;;
-			*)		CPPFLAGS="$CPPFLAGS -I$d/include";;
-			esac
-			break
-		fi
-	done
+#else
+#	for d in /usr/local /usr/local/openssl /usr/pkg /usr '' ; do
+#		if test -f "$d/lib/libcrypto.a" ; then
+#						echo "$ac_t""using $d/lib/libcrypto.a" 1>&6
+#			openssl_libdir="$d/lib"
+#			LDFLAGS="$LDFLAGS -L$openssl_libdir"
+#			case "$openssl_libdir" in
+#			/lib|/usr/lib)	;;
+#			*)		CPPFLAGS="$CPPFLAGS -I$d/include";;
+#			esac
+#			break
+#		fi
+#	done
 fi
 LIBS="-lcrypto $LIBS"
 case $host_os in
diff -Naur racoon2-20100526a.orig/spmd/Makefile.in racoon2-20100526a/spmd/Makefile.in
--- racoon2-20100526a.orig/spmd/Makefile.in	2010-06-15 10:22:14.000000000 -0400
+++ racoon2-20100526a/spmd/Makefile.in	2010-06-15 10:54:01.000000000 -0400
@@ -38,11 +38,11 @@
 spmdctl.o: spmd_internal.h
 
 install: all
-	$(INSTALL_DIR) $(sbindir)
-	$(INSTALL_PROGRAM) $(DAEMON) $(sbindir)
-	$(INSTALL_PROGRAM) $(COMMAND) $(sbindir)
-	$(INSTALL_DIR) $(man8dir)
-	$(INSTALL_DATA) $(MANFILES) $(man8dir)
+	$(INSTALL_DIR) $(DESTDIR)$(sbindir)
+	$(INSTALL_PROGRAM) $(DAEMON) $(DESTDIR)$(sbindir)
+	$(INSTALL_PROGRAM) $(COMMAND) $(DESTDIR)$(sbindir)
+	$(INSTALL_DIR) $(DESTDIR)$(man8dir)
+	$(INSTALL_DATA) $(MANFILES) $(DESTDIR)$(man8dir)
 
 clean:
 	-rm -f $(TARGET) *.o $(COMMAND)
