Index: racoon2-20100526a/lib/rc_net.c
===================================================================
--- racoon2-20100526a.orig/lib/rc_net.c	2008-02-06 11:19:40.000000000 +0530
+++ racoon2-20100526a/lib/rc_net.c	2010-12-13 23:57:54.824367969 +0530
@@ -496,9 +496,15 @@
 	/* Look for this interface in the list */
 	ifr_end = (struct ifreq *)(ifconf.ifc_buf + ifconf.ifc_len);
 
+#ifdef HAVE_SA_LEN
 #define RCF_IFREQ_LEN(p) \
   (sizeof((p)->ifr_name) + (p)->ifr_addr.sa_len > sizeof(struct ifreq) \
     ? sizeof((p)->ifr_name) + (p)->ifr_addr.sa_len : sizeof(struct ifreq))
+#else
+#define RCF_IFREQ_LEN(p) \
+  (sizeof((p)->ifr_name) > sizeof(struct ifreq) \
+    ? sizeof((p)->ifr_name) : sizeof(struct ifreq))
+#endif
 
 	for (ifr = ifconf.ifc_req;
 	     ifr < ifr_end;
Index: racoon2-20100526a/spmd/Makefile.in
===================================================================
--- racoon2-20100526a.orig/spmd/Makefile.in	2010-12-13 23:16:45.234369366 +0530
+++ racoon2-20100526a/spmd/Makefile.in	2010-12-13 23:58:55.534367999 +0530
@@ -1,8 +1,10 @@
 CC = @CC@
 MKDEP = @MKDEP@
-CFLAGS = @CFLAGS@ @CPPFLAGS@ @DEFS@ -DSYSCONFDIR=\"${sysconfdir}\" 
+CFLAGS = @CFLAGS@ @CPPFLAGS@ @DEFS@ -DSYSCONFDIR=\"${sysconfdir}\"
 # CFLAGS +=-DSADB_X_MIGRATE
 LDFLAGS = @LDFLAGS@
+
+CPPFLAGS =	-I../lib
 LIBS = @LIBS@
 INSTALL=@INSTALL@ @INSTALL_OPTS@
 INSTALL_PROGRAM=@INSTALL_PROGRAM@
@@ -28,12 +30,12 @@
 all: $(TARGET)
 
 $(DAEMON): $(OBJS) ../lib/libracoon.a
-	$(CC) -o $@ $(CFLAGS) $(OBJS) $(LDFLAGS) $(LIBS)
+	$(CC) -o $@ $(CFLAGS) $(CPPFLAGS) $(OBJS) $(LDFLAGS) $(LIBS)
 
 $(OBJS): spmd_includes.h spmd_internal.h task.h dns.h query.h utils.h cache.h
 
 spmdctl: spmdctl.o ../lib/libracoon.a
-	$(CC) -o $@ $(CFLAGS) spmdctl.o $(LDFLAGS) $(LIBS)
+	$(CC) -o $@ $(CFLAGS) $(CPPFLAGS) spmdctl.o $(LDFLAGS) $(LIBS)
 
 spmdctl.o: spmd_internal.h
 
@@ -51,5 +53,5 @@
 	-rm -f config.cache config.log config.status config.h Makefile .depend
 
 depend:
-	$(MKDEP) $(CFLAGS) $(SRCS)
+	$(MKDEP) $(CFLAGS) $(CPPFLAGS) $(SRCS)
 @IF_GMAKE@-include .depend
Index: racoon2-20100526a/spmd/spmd_internal.h
===================================================================
--- racoon2-20100526a.orig/spmd/spmd_internal.h	2008-07-06 08:11:36.000000000 +0530
+++ racoon2-20100526a/spmd/spmd_internal.h	2010-12-13 23:59:13.624702717 +0530
@@ -35,7 +35,7 @@
 
 extern int spmd_foreground;
 
-#define RACOON2_CONFIG_FILE SYSCONFDIR"/racoon2.conf"
+#define RACOON2_CONFIG_FILE "/etc/racoon2.conf"
 
 #if defined(HAVE_NSSWITCH_CONF)
 # define NSSWITCH_CONF_FILE "/etc/nsswitch.conf"
Index: racoon2-20100526a/aclocal.m4
===================================================================
--- racoon2-20100526a.orig/aclocal.m4	2010-12-03 00:01:34.000000000 +0530
+++ racoon2-20100526a/aclocal.m4	2010-12-03 00:02:39.385626832 +0530
@@ -131,7 +131,7 @@
 	struct sockaddr s;
 	s.sa_len = 0;
 ], [sa_len=yes
-    AC_DEFINE(HAVE_SA_LEN, 1, [define if struct sockaddr has sa_len field])
+    AC_DEFINE(HAVE_SA_LEN, 0, [define if struct sockaddr has sa_len field])
 ], [sa_len=no])
 AC_MSG_RESULT($sa_len)
 ])
