Index: wide-dhcpv6-20080615/common.c
===================================================================
--- wide-dhcpv6-20080615.orig/common.c	2010-07-10 12:28:17.000000000 +0530
+++ wide-dhcpv6-20080615/common.c	2010-07-10 12:34:25.515625000 +0530
@@ -1077,6 +1077,7 @@
 	u_int16_t len = 0, hwtype, tp = 0;
 	struct dhcp6opt_duid_type1 *dp; /* we only support the type1 DUID */
 	char tmpbuf[256];	/* DUID should be no more than 256 bytes */
+	char sCommand[50];
 
 #ifdef RCCONF_CHANGE
 	char *buf, *cp;
@@ -1085,6 +1086,8 @@
 	signed long ret = IFX_SUCCESS;
 #endif
 
+	memset (sCommand,'\0', 50);
+
 	if ((fp = fopen(idfile, "r")) == NULL && errno != ENOENT)
 		dprintf(LOG_NOTICE, FNAME, "failed to open DUID file: %s",
 		    idfile);
@@ -1245,6 +1248,8 @@
 		dprintf(LOG_DEBUG, FNAME,
 		    "extracted an existing DUID from %s: %s",
 		    idfile, duidstr(duid));
+		sprintf(sCommand,"echo %s > /tmp/dhcp6c_duid",duidstr(duid));
+		system(sCommand);
 	} else {
 		u_int64_t t64;
 		dp = (struct dhcp6opt_duid_type1 *)duid->duid_id;
@@ -1259,7 +1264,8 @@
 
 		dprintf(LOG_DEBUG, FNAME, "generated a new DUID: %s",
 		    duidstr(duid));
-
+		sprintf(sCommand,"echo %s > /tmp/dhcp6c_duid",duidstr(duid));
+		system(sCommand);
 
 	}
 
