diff -Nuar wide-dhcpv6-20080615/addrconf.c wide-dhcpv6-20080615.modif/addrconf.c
--- wide-dhcpv6-20080615/addrconf.c	2010-10-29 12:20:50.764644838 +0530
+++ wide-dhcpv6-20080615.modif/addrconf.c	2010-10-29 14:06:52.404643869 +0530
@@ -409,8 +409,13 @@
 	ret = ifaddrconf(cmd, sa->dhcpif->ifname, &sin6, addr->plen,
 	    addr->pltime, addr->vltime);
 	#else
+#ifdef DHCP6_TAHI_FIX
+	ret = ifaddrconf(cmd, sa->dhcpif->ifname, &sin6, 64,
+	    addr->pltime, addr->vltime);
+#else
 	ret = ifaddrconf(cmd, sa->dhcpif->ifname, &sin6, 128,
 	    addr->pltime, addr->vltime);
+#endif
 	#endif
        
 	dprintf(LOG_NOTICE, FNAME, "RET = %d\n", ret); 
diff -Nuar wide-dhcpv6-20080615/common.h wide-dhcpv6-20080615.modif/common.h
--- wide-dhcpv6-20080615/common.h	2007-03-21 15:22:57.000000000 +0530
+++ wide-dhcpv6-20080615.modif/common.h	2010-10-29 14:15:27.674643459 +0530
@@ -99,6 +99,10 @@
 #define SO_REUSEPORT SO_REUSEADDR
 #endif
 
+#if (ADDR_POOL_PREFIX_SUPPORT && DHCP6_TAHI_FIX)
+#error "ADDR_POOL_PREFIX_SUPPORT & DHCP6_TAHI_FIX both are defined !!"
+#endif
+
 /* s*_len stuff */
 static __inline u_int8_t
 sysdep_sa_len (const struct sockaddr *sa)
diff -Nuar wide-dhcpv6-20080615/config.h wide-dhcpv6-20080615.modif/config.h
--- wide-dhcpv6-20080615/config.h	2010-10-29 12:20:50.694643680 +0530
+++ wide-dhcpv6-20080615.modif/config.h	2010-10-29 14:21:03.004643669 +0530
@@ -42,7 +42,7 @@
 struct dhcp6_range {
 	struct    in6_addr min;
 	struct    in6_addr max;
-	#ifdef ADDR_POOL_PREFIX_SUPPORT /* Added by Sundar */
+	#if (ADDR_POOL_PREFIX_SUPPORT || DHCP6_TAHI_FIX) /* Added by Sundar */
 	u_int8_t  prefix_length;
 	#endif
 };
diff -Nuar wide-dhcpv6-20080615/dhcp6.h wide-dhcpv6-20080615.modif/dhcp6.h
--- wide-dhcpv6-20080615/dhcp6.h	2010-10-29 12:20:50.724643704 +0530
+++ wide-dhcpv6-20080615.modif/dhcp6.h	2010-10-29 12:43:19.647386965 +0530
@@ -355,12 +355,12 @@
 struct dhcp6opt_ia_addr {
 	u_int16_t dh6_ia_addr_type;
 	u_int16_t dh6_ia_addr_len;
+	struct    in6_addr dh6_ia_addr_addr;
 	u_int32_t dh6_ia_addr_preferred_time;
 	u_int32_t dh6_ia_addr_valid_time;
 	#ifdef ADDR_POOL_PREFIX_SUPPORT /* Added by Sundar */
 	u_int8_t  dh6_ia_addr_prefix_len;
 	#endif
-	struct    in6_addr dh6_ia_addr_addr;
 } __attribute__ ((__packed__));
 
 /* IA_PD Prefix */
diff -Nuar wide-dhcpv6-20080615/Makefile.in wide-dhcpv6-20080615.modif/Makefile.in
--- wide-dhcpv6-20080615/Makefile.in	2010-10-29 12:20:50.754644923 +0530
+++ wide-dhcpv6-20080615.modif/Makefile.in	2010-10-29 12:42:01.254643952 +0530
@@ -36,7 +36,7 @@
 user= @user@
 group= @group@
 
-CFLAGS=	@CFLAGS@ @DEFS@ -DADDR_POOL_PREFIX_SUPPORT -DADDR_PREFIX_SUPPORT \
+CFLAGS=	@CFLAGS@ @DEFS@ -DDHCP6_TAHI_FIX -DADDR_PREFIX_SUPPORT \
 	-DADDR_OVERRIDE_HOST_CONF -DUPDATE_PD_IN_RADVD \
 	-DSYSCONFDIR=\"${sysconfdir}\" \
 	-DLOCALDBDIR=\"${localdbdir}\" \
