--- wide-dhcpv6-20080615/prefixconf.c	2010-12-30 16:40:00.088450000 +0530
+++ wide-dhcpv6-20080615_modif/prefixconf.c	2010-12-30 16:43:17.232311000 +0530
@@ -139,8 +139,7 @@
                         in6addr2str(a, 0), pre_len , pinfo->ifname); 
 
 	system("touch /var/radvd_pd");
-	rd = fopen("/var/radvd.conf", "w+");
-	fclose(rd);
+	system("touch /var/radvd.conf");
 
 	buf3 = (char *) malloc (80);
 
