--- wide-dhcpv6-20080615/dhcp6c.c	2011-05-17 13:30:31.205808534 +0530
+++ wide-dhcpv6-20080615/dhcp6c.c	2011-05-18 12:42:50.411809249 +0530
@@ -146,7 +146,7 @@
 int client6_start __P((struct dhcp6_if *));
 static void info_printf __P((const char *, ...));
 
-extern int client6_script __P((char *, int, struct dhcp6_optinfo *));
+extern int client6_script __P((char *, int, struct dhcp6_optinfo *, char *));
 
 #define MAX_ELAPSED_TIME 0xffff
 
@@ -1895,7 +1895,7 @@
 	 */
 	if (ifp->scriptpath != NULL && strlen(ifp->scriptpath) != 0) {
 		dprintf(LOG_DEBUG, FNAME, "executes %s", ifp->scriptpath);
-		client6_script(ifp->scriptpath, state, optinfo);
+		client6_script(ifp->scriptpath, state, optinfo, ifp->ifname); //AMS
 	}
 
 	/*

--- wide-dhcpv6-20080615/dhcp6c_script.c	2007-03-21 15:22:55.000000000 +0530
+++ wide-dhcpv6-20080615/dhcp6c_script.c	2011-05-18 13:23:52.072684276 +0530
@@ -73,10 +73,11 @@
 static char bcmcsname_str[] = "new_bcmcs_name";
 
 int
-client6_script(scriptpath, state, optinfo)
+client6_script(scriptpath, state, optinfo, ifname)
 	char *scriptpath;
 	int state;
 	struct dhcp6_optinfo *optinfo;
+        char *ifname;
 {
 	int i, dnsservers, ntpservers, dnsnamelen, envc, elen, ret = 0;
 	int sipservers, sipnamelen;
@@ -400,11 +401,12 @@
 			    "script \"%s\" terminated", scriptpath);
 		}
 	} else {
-		char *argv[2];
+		char *argv[3];
 		int fd;
 
 		argv[0] = scriptpath;
-		argv[1] = NULL;
+		argv[1] = ifname;
+		argv[2] = NULL;
 
 		if (safefile(scriptpath)) {
 			dprintf(LOG_ERR, FNAME,
