diff -Nuar wide-dhcpv6-20080615_org/dhcp6s.c wide-dhcpv6-20080615_modif/dhcp6s.c
--- wide-dhcpv6-20080615_org/dhcp6s.c	2011-07-25 18:36:15.042511500 +0530
+++ wide-dhcpv6-20080615_modif/dhcp6s.c	2011-07-25 18:37:20.920511276 +0530
@@ -1453,7 +1453,7 @@
 		}
 
 	#ifdef ADDR_OVERRIDE_HOST_CONF /* Added by Sundar */
-		if((!client_conf->pool.name) && (ifp->pool.name)) {
+		if(client_conf && (!client_conf->pool.name) && (ifp->pool.name)) {
 			dprintf(LOG_DEBUG, FNAME, "Add pool name %s to client_conf", 
 				    ifp->pool.name);
 			client_conf->pool.name = strdup(ifp->pool.name);
