diff -Nuar wide-dhcpv6-20080615/common.c wide-dhcpv6-20080615_reconf_accept/common.c
--- wide-dhcpv6-20080615/common.c	2011-08-19 16:35:02.080878527 +0530
+++ wide-dhcpv6-20080615_reconf_accept/common.c	2011-08-19 16:35:38.062878624 +0530
@@ -1676,7 +1676,9 @@
 		goto fail;
 	if (duidcpy(&dst->serverID, &src->serverID))
 		goto fail;
+	
 	dst->rapidcommit = src->rapidcommit;
+	dst->reconfigaccept = src->reconfigaccept;
 
 	if (dhcp6_copy_list(&dst->iapd_list, &src->iapd_list))
 		goto fail;
@@ -1734,7 +1736,7 @@
 	dst->authalgorithm = src->authalgorithm;
 	dst->authrdm = src->authrdm;
 	dst->authrd = src->authrd;
-
+	
 	switch (src->authproto) {
 	case DHCP6_AUTHPROTO_DELAYED:
 		dst->delayedauth_keyid = src->delayedauth_keyid;
@@ -2012,6 +2014,13 @@
 				goto malformed;
 			optinfo->rapidcommit = 1;
 			break;
+
+		case DH6OPT_RECONF_ACCEPT:
+			if (optlen != 0 )
+				goto malformed;
+			optinfo->reconfigaccept = 1;
+			break;
+
 		case DH6OPT_INTERFACE_ID:
 			if ((optinfo->ifidopt_id = malloc(optlen)) == NULL)
 				goto fail;
@@ -2717,6 +2726,13 @@
 		tmpbuf = NULL;
 	}
 
+        if (optinfo->reconfigaccept) {
+                if (copy_option(DH6OPT_RECONF_ACCEPT, 0, NULL, &p,
+                    optep, &len) != 0) {
+                        goto fail;
+                }
+        }
+
 	if (dhcp6_set_domain(DH6OPT_SIP_SERVER_D, &optinfo->sipname_list,
 	    &p, optep, &len) != 0)
 		goto fail;
diff -Nuar wide-dhcpv6-20080615/dhcp6c.c wide-dhcpv6-20080615_reconf_accept/dhcp6c.c
--- wide-dhcpv6-20080615/dhcp6c.c	2011-08-19 16:35:02.081878596 +0530
+++ wide-dhcpv6-20080615_reconf_accept/dhcp6c.c	2011-08-19 16:37:16.789753435 +0530
@@ -1423,6 +1423,15 @@
 		optinfo.rapidcommit = 1;
 	}
 
+/* 		reconfig is set in
+ * 		1. SOLICIT message by default
+ * 		2. if the select server had sent a reconfigaccept option
+ */	
+	if (ev->state == DHCP6S_SOLICIT || (ev->current_server && ev->current_server->optinfo.reconfigaccept == 1 )
+		|| ev->state == DHCP6S_RENEW || ev->state == DHCP6S_INFOREQ) {
+		optinfo.reconfigaccept = 1;
+	}
+
 	/* elapsed time */
 	if (ev->timeouts == 0) {
 		gettimeofday(&ev->tv_start, NULL);
diff -Nuar wide-dhcpv6-20080615/dhcp6.h wide-dhcpv6-20080615_reconf_accept/dhcp6.h
--- wide-dhcpv6-20080615/dhcp6.h	2011-08-19 16:35:02.082878573 +0530
+++ wide-dhcpv6-20080615_reconf_accept/dhcp6.h	2011-08-19 16:35:38.066878430 +0530
@@ -240,6 +240,7 @@
 	int reconf_msg;
 	char reconfig_key[16]; /* receieved in last reply msg */
 #endif
+	int reconfigaccept;
 };
 
 /* DHCP6 base packet format */
@@ -298,6 +299,7 @@
 #define DH6OPT_INTERFACE_ID 18
 #define DH6OPT_RECONF_MSG 19
 
+#define DH6OPT_RECONF_ACCEPT 20
 #define DH6OPT_SIP_SERVER_D 21
 #define DH6OPT_SIP_SERVER_A 22
 #define DH6OPT_DNS 23
