--- wide-dhcpv6-20080615/prefixconf.c	2011-08-23 18:16:59.309642163 +0530
+++ wide-dhcpv6-20080615_routerinfo/prefixconf.c	2011-08-23 15:18:58.951957227 +0530
@@ -706,6 +706,7 @@
 		snprintf(env[i++], sizeof(envbuf[0]), "SRC_INTF=%s", dhcpifp->ifname);
 		snprintf(env[i++], sizeof(envbuf[0]), "EVENT_TYPE=UP");
 		snprintf(env[i++], sizeof(envbuf[0]), "INTF=%s", pconf->ifname);
+		snprintf(env[i++], sizeof(envbuf[0]), "DELEGATED_PREFIX=%s/%d", in6addr2str(&prefix->addr, 0), prefix->plen);
 
 		if(dnsservers && i <= env_max_idx - 2 ) {
 			int c = 0;
