diff -Nur wide-dhcpv6-20080615/addrconf.c wide-dhcpv6-20080615_wrk/addrconf.c
--- wide-dhcpv6-20080615/addrconf.c	2011-09-23 11:52:57.924767022 +0530
+++ wide-dhcpv6-20080615_wrk/addrconf.c	2011-09-21 14:39:21.931767184 +0530
@@ -74,7 +74,7 @@
 #define iacna_reestablish_data common.reestablish_data
 #define iacna_release_data common.release_data
 #define iacna_cleanup common.cleanup
-#define DAD_TIMEOUT 2
+#define DEC_TIMEOUT 1
 
 struct statefuladdr {
 	TAILQ_ENTRY (statefuladdr) link;
@@ -448,7 +448,7 @@
                            }
           }
           /* update the timer */
-          timo.tv_sec = DAD_TIMEOUT;
+          timo.tv_sec = DEC_TIMEOUT;
           timo.tv_usec = 0;
 
           ev->state=DHCP6S_DECLINE; 

diff -Nur wide-dhcpv6-20080615/dhcp6c.c wide-dhcpv6-20080615_wrk/dhcp6c.c
--- wide-dhcpv6-20080615/dhcp6c.c	2011-09-23 11:52:57.932766981 +0530
+++ wide-dhcpv6-20080615_wrk/dhcp6c.c	2011-09-22 12:17:46.770767089 +0530
@@ -166,6 +166,7 @@
 #define INTERFACEINFO "/proc/net/if_inet6"
 #define DAD_FLAGS 0x08
 #define DAD_FOUND 2
+#define DEC_MAX_RC 6
 
 #define IN6_ARE_ADDR_EQUAL(a,b) \
         ((((__const uint32_t *) (a))[0] == ((__const uint32_t *) (b))[0])     \
@@ -1062,6 +1063,7 @@
                     {
                         client6_send(ev);
 		        ev->state=DHCP6S_DECLINE_SENT;
+                        ev->max_retrans_cnt = DEC_MAX_RC;
                         remove_addr(ev->sa);
                     }
                     else
