diff -Nuar wide-dhcpv6-20080615_org/dhcp6c.c wide-dhcpv6-20080615/dhcp6c.c
--- wide-dhcpv6-20080615_org/dhcp6c.c	2012-01-04 14:54:40.809471061 +0530
+++ wide-dhcpv6-20080615/dhcp6c.c	2012-02-16 15:05:13.777736710 +0530
@@ -126,7 +126,7 @@
 
 int dad_parse(const char *, struct dhcp6_event *);
 static void usage __P((void));
-static void client6_init __P((void));
+static void client6_init __P((char *));
 static void client6_startall __P((int));
 void free_resources __P((struct dhcp6_if *));
 static void client6_mainloop __P((void));
@@ -252,7 +252,12 @@
 
 	setloglevel(debug);
 
-	client6_init();
+	if ( argc == 1 ) {
+		client6_init( argv[0]);
+	} else {
+		client6_init( NULL);
+	}
+		
 #ifdef ENABLE_IPV6NM
 	init_ipv6nm(argc);
 #endif
@@ -301,7 +306,8 @@
 /*------------------------------------------------------------*/
 
 void
-client6_init()
+client6_init(intf)
+	char *intf;
 {
 	struct addrinfo hints, *res;
 	static struct sockaddr_in6 sa6_allagent_storage;
@@ -387,6 +393,9 @@
 		exit(1);
 	}
 #endif
+	if (intf != NULL) {
+		setsockopt(sock, SOL_SOCKET, SO_BINDTODEVICE, intf, strlen(intf));
+	}
 
 	/*
 	 * According RFC3315 2.2, only the incoming port should be bound to UDP
