diff -Nuar wide-dhcpv6-20080615/dhcp6.h wide-dhcpv6-20080615_modify/dhcp6.h
--- wide-dhcpv6-20080615/dhcp6.h	2013-08-01 18:07:31.000000000 +0530
+++ wide-dhcpv6-20080615_modify/dhcp6.h	2013-08-01 18:16:17.000000000 +0530
@@ -108,6 +108,14 @@
 #define DHCP6_IRT_DEFAULT 86400	/* 1 day */
 #define DHCP6_IRT_MINIMUM 600
 
+#ifndef HAVE_IN6_PKTINFO
+struct in6_pktinfo {
+	struct in6_addr ipi6_addr; /* src/dst IPv6 address */
+	unsigned int ipi6_ifindex; /* send/recv interface index */
+};
+#endif /* HAVE_IN6_PKTINFO */
+
+
 /* DUID: DHCP unique Identifier */
 struct duid {
 	size_t duid_len;	/* length */
--- wide-dhcpv6-20080615/raw_icmpv6.c	2013-08-01 18:07:31.000000000 +0530
+++ wide-dhcpv6-20080615_modify/raw_icmpv6.c	2013-08-01 18:22:55.000000000 +0530
@@ -13,6 +13,14 @@
 #include <net/if.h>
 #define debug_log printf
 
+#ifndef HAVE_IN6_PKTINFO
+struct in6_pktinfo {
+	struct in6_addr ipi6_addr; /* src/dst IPv6 address */
+	unsigned int ipi6_ifindex; /* send/recv interface index */
+};
+#endif /* HAVE_IN6_PKTINFO */
+
+
 unsigned short int checksum(unsigned short int *, int);
 
 int ipv6nm_sendrs(char *intf)
