diff -Nuar wide-dhcpv6-20080615_org/dhcp6c.h wide-dhcpv6-20080615_modif/dhcp6c.h
--- wide-dhcpv6-20080615_org/dhcp6c.h	2015-08-11 14:37:29.382212475 +0530
+++ wide-dhcpv6-20080615_modif/dhcp6c.h	2015-08-11 17:04:11.873427331 +0530
@@ -28,6 +28,12 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+
+#ifdef CONFIG_NEW_FRAMEWORK
+#undef LOCALDBDIR
+#define LOCALDBDIR "/opt/lantiq/config"
+#endif
+
 #define DHCP6C_CONF SYSCONFDIR "/dhcp6c.conf"
 #define DHCP6C_PIDFILE "/var/run/dhcp6c.pid"
 #define DUID_FILE LOCALDBDIR "/dhcp6c_duid"
diff -Nuar wide-dhcpv6-20080615_org/dhcp6s.c wide-dhcpv6-20080615_modif/dhcp6s.c
--- wide-dhcpv6-20080615_org/dhcp6s.c	2015-08-11 14:37:29.445213631 +0530
+++ wide-dhcpv6-20080615_modif/dhcp6s.c	2015-08-11 17:05:09.902476493 +0530
@@ -77,6 +77,11 @@
 #include <signal.h>
 #include <lease.h>
 
+#ifdef CONFIG_NEW_FRAMEWORK
+#undef LOCALDBDIR
+#define LOCALDBDIR "/opt/lantiq/config"
+#endif
+
 #define DUID_FILE LOCALDBDIR "/dhcp6s_duid"
 #define DHCP6S_CONF SYSCONFDIR "/dhcp6s.conf"
 #define DEFAULT_KEYFILE SYSCONFDIR "/dhcp6sctlkey"
diff -Nuar wide-dhcpv6-20080615/addrconf.c wide-dhcpv6-20080615_modif/addrconf.c
--- wide-dhcpv6-20080615/addrconf.c	2015-09-10 15:54:03.905917942 +0530
+++ wide-dhcpv6-20080615_modif/addrconf.c	2015-09-10 16:20:46.515315934 +0530
@@ -425,9 +425,15 @@
  
         inet_ntop(AF_INET6, &sin6.sin6_addr, source, INET6_ADDRSTRLEN);
             
+#ifndef CONFIG_NEW_FRAMEWORK
         sprintf(sCommand,". /etc/rc.d/bringup_wanstatus %s %s ", sa->dhcpif->ifname, source);
         dprintf(LOG_NOTICE, FNAME, "Scommand = %s\n ", sCommand);
 	system(sCommand);
+#else
+        sprintf(sCommand,"/etc/rc.d/update_and_run_radvd.sh iana %s %s ", sa->dhcpif->ifname, source);
+        dprintf(LOG_NOTICE, FNAME, "Scommand = %s\n ", sCommand);
+	system(sCommand);
+#endif	
 
         if (ev == NULL) {
                 dprintf(LOG_INFO, FNAME, "No event");
