--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,98 @@
+
+2015-08-11:
+	032-new-framework-adaptation.patch added
+
+2013-08-01:
+	031-wide-dhcp-compilation-fix.patch added
+
+2012-08-30:
+	030-ipv6nm_sendrs.patch added
+
+2012-05-10:
+	029-decline-and-reconfigure-ia.patch added
+
+2012-03-13:
+	028-wan-down-event.patch added
+
+2012-02-16:
+	027-multiwan_support.patch patch added
+
+2011-12-28:
+	Support for dslite after name options added
+
+
+2011-11-28:
+	025-ce-router-unh.patch added
+
+2011-09-21:
+	024_decline_fix.patch added
+
+2011-09-09:
+	DUID Support added
+
+2011-08-23:
+	022-Route-information-option.patch added
+
+2011-08-22:
+	020-decline-support.patch added
+
+2011-08-19:
+	Support for reconf option added
+
+2011-08-17:
+	019-reconf_with_auth_support.patch added
+
+2011-07-25:
+	018-dhcpv6_bugfix.patch added
+
+2011-07-20:
+	Support for ipv6nm added
+
+2011-07-20:
+	Support for ipv6helper ds lite added
+
+2011-06-20:
+	KW issues fixed
+
+2011-06-17:
+	Support for ipv6helper added
+
+2011-05-18:
+	013-dns-update-fix.patch added
+
+2011-05-04:
+	011-single_radvd_support.patch added
+
+2010-12-30:
+	Creating radvd.conf file using system command
+
+2010-11-23:
+	009-sla_len_fix.patch added
+
+2010-11-10:
+	Support fro dhcp6 tahi added
+
+2010-10-29:
+	007-dhcp6_tahi_fix_default_prefix_64.patch added
+
+2010-07-21:
+	Support for ipv6 wan status added
+
+2010-07-15:
+	Support for ipv6 wan status added
+
+2010-07-10:
+	004-dhcpv6c_duid_for_ref.patch added
+
+2009-12-18:
+	002-dhcpv6_save_duid_fix added
+
+2009-12-16:
+	003-generate_radvdconf_run_radvd added
+
+2009-11-12:
+	001_no_libfl_a patch added
+
+2009-10-29:
+	Support for address pool prefix added
+
