--- a/configure
+++ b/configure
@@ -3695,7 +3695,7 @@ if test "${ac_cv_lib_fl_yywrap+set}" = s
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lfl  $LIBS"
+LIBS="$LIBS"
 cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
 _ACEOF
--- a/Makefile.in
+++ b/Makefile.in
@@ -65,7 +65,7 @@ INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
 LDFLAGS = @LDFLAGS@
 LEX = @LEX@
-LEXLIB = @LEXLIB@
+LEXLIB =
 LEX_OUTPUT_ROOT = @LEX_OUTPUT_ROOT@
 LIBOBJS = @LIBOBJS@
 LIBOL_DIR = @LIBOL_DIR@
--- a/src/cfg-lex.c
+++ b/src/cfg-lex.c
@@ -668,7 +668,9 @@ static void append_char(char c);
 #ifdef __cplusplus
 extern "C" int yywrap (void );
 #else
-extern int yywrap (void );
+int yywrap (void ) {
+	return 1;
+}
 #endif
 #endif
 
