--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,145 @@
+
+2016-07-01:
+	JIRA: DRVLIB_SW-236 - Workaround for empty page ECC with bitflips in data area
+
+2016-06-17:
+	JIRA: UBOOT_SW-222 - U-boot sends packet storm, disable flow control to prevent sending pause frame when not ready
+
+2016-06-07:
+	JIRA : UBOOT_SW-212 - rollback id check support for bootcore and u-boot.ltq
+
+2016-06-06:
+	JIRA: UBOOT_SW-223 - Fixed wrong cpu-id definition in cache handling
+
+2016-05-11:
+	JIRA : UBOOT_SW-219, fix RoT implementation from changeset aab33f1b97b3, the build fail and loop on another machine
+
+2016-05-10:
+	JIRA : DRVLIB_SW-151 - secure jtag application development
+
+2016-02-12:
+	JIRA GRX500SYS-1198: programing of interAptiv policy manager to set firmware priorities
+
+2016-02-04:
+	JIRA UBOOT_SW-169: Added Latch Clock Delay to support high spi frequency
+
+2016-02-01:
+	JIRA: GRX500SYS-1615 - SPI boot with NAND flash support in u-boot
+
+2015-12-18:
+	1.work around for GRX390 WIFI exception issue. 2. Bugfix for GRX390 menuconfig options. 3. Added menuconfig option for source utility
+
+2015-12-04:
+	1.support bootcore update from web;2.bugfix for build_grx500_asc.pl
+
+2015-11-27:
+	JIRA:GRX500SYS-1438 Secure boot development for A21, add test function for auth image in block or chunk size
+
+2015-11-25:
+	JIRA: UGW_SW-4873 - New DDR settings for GRX390
+
+2015-11-07:
+	added extra image checking feature
+
+2015-10-29:
+	Multiple JIRA commits: 1) JIRA: UBOOT_SW-206 - Tri-colour LED Implementation, 2) JIRA: UBOOT_SW-205 - IPI IRQ switch label fix, 3) JIRA: UBOOT_SW-204 - Bugfix for new bootflow implementation
+
+2015-10-08:
+	JIRA UBOOT_SW-200:bugfix for MXIC SPI NOR flash detection failure
+
+2015-07-29:
+	1.Secure upgrade support for GRX500;2.Hush shell support in uboot;3.Bugfix for ICC communication protocol
+
+2015-07-24:
+	JIRA: UBOOT_SW-194 - Integrate SPI NOR boot with interaptiv/4Kec boot support
+
+2015-06-24:
+	fix compilation problem for GRX500 non-secure boot with patch segregation
+
+2015-03-16:
+	1) Updated I2C commands. 2) Added I2C driver support for GRX500
+
+2014-12-03:
+	get_fpi_clk fix for fused grx390 platform
+
+2014-10-09:
+	initial support for OF fdt boot support in mips
+
+2014-07-24:
+	1.removed extra header checking feature;2.added VRX220_W2 support
+
+2014-05-09:
+	Added PCDU changes to fix AR10/GRX390 RX CRC issue
+
+2014-04-23:
+	ddr autotunning version changed to 1.0
+
+2014-02-18:
+	added winbond spi flash support
+
+2014-01-15:
+	ported dual GMAC support
+
+2013-12-23:
+	bugfix for envcrc.c
+
+2013-11-25:
+	bugfix for watchdog timer menuconfig option
+
+2013-11-17:
+	WAVE500_SW-patch update, vbg500
+
+2013-10-17:
+	merged watchdog timer and chkset patch from AEI branch
+
+2013-05-20:
+	added spansion spi flash S25FL128S/256S support
+
+2013-05-07:
+	env var ver becomes shorter
+
+2013-02-26:
+	1.Added small nand flash support to AR10;2.aes key taken out of secureboot stage1 image
+
+2013-01-03:
+	split vbg400 patches into smaller ones
+
+2012-11-08:
+	1.ubi wear leveling bugfix;2.added u-boot version variable to uboot environment
+
+2012-10-09:
+	1.fixed 1 vr9 ethernet driver bug;2.Added Toshiba 1Gib nand flash definition
+
+2012-10-02:
+	1.fixed one ethernet driver bug;2.Added 256M nand flash menuconfig option
+
+2012-08-01:
+	1.Added more image header definitions in image.h;2.Added perl script to generate asc file from binary image;
+
+2012-07-25:
+	added macro protection in div64.h for ifxutility
+
+2012-07-03:
+	added suppress kernel output options
+
+2012-01-12:
+	1.AR9 added suport for big page nand flash;2.VR9 added support for small page nand flash;3.Bugfix in nand miniloader
+
+2011-11-08:
+	1.bugfix for Calloc;2.AR10 added gphy internal clock options;3.dual image env variables modified
+
+2011-10-28:
+	1.hn1 added support for external gphy firmware;2.code clean up for menuconfig patch;3.ar9 bugfix for external lantiq phy chip
+
+2011-10-07:
+	1.menuconfig system updated;2.added message when erase the serial flash
+
+2011-10-05:
+	1.AR10 added ddr1 ram support;2.AR10 bootup using reset method after change cpu clock
+
+2011-09-30:
+	1.bugfix for MLC nand flash;2.mtdparts bugfix for large nand flash;
+
+2011-07-25:
+	1.patch re-organized.2.vr9 spi mode compilation bugfix
+
