# HG changeset patch
# Parent e31f72fa5fcb26c76b7db1e0c3af7bbd33470775

diff --git a/include/malloc.h b/include/malloc.h
--- a/include/malloc.h
+++ b/include/malloc.h
@@ -831,7 +831,7 @@ Void_t *(*__morecore)() = __default_more
 
 #define MORECORE (*__morecore)
 #define MORECORE_FAILURE 0
-#define MORECORE_CLEARS 1
+#define MORECORE_CLEARS 0
 
 #else /* INTERNAL_LINUX_C_LIB */
 
@@ -850,7 +850,7 @@ extern Void_t*     sbrk();
 #endif
 
 #ifndef MORECORE_CLEARS
-#define MORECORE_CLEARS 1
+#define MORECORE_CLEARS 0
 #endif
 
 #endif /* INTERNAL_LINUX_C_LIB */
