# HG changeset patch
# Parent 447677d816ac80e246165d7a35b6dc57e5e44f1f

diff --git a/tools/envcrc.c b/tools/envcrc.c
--- a/tools/envcrc.c
+++ b/tools/envcrc.c
@@ -35,7 +35,7 @@
 #include <config.h>
 #undef	__ASSEMBLY__
 
-#if defined(CONFIG_ENV_IS_IN_FLASH)
+#if defined(CONFIG_ENV_IS_IN_FLASH) || defined(CONFIG_ENV_IS_IN_NAND)
 # ifndef  CONFIG_ENV_ADDR
 #  define CONFIG_ENV_ADDR	(CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET)
 # endif
@@ -134,7 +134,40 @@ int main (int argc, char **argv)
 		printf ("0x%08X\n", crc);
 	}
 #else
-	printf ("0\n");
+    unsigned char pad = 0x00;
+    uint32_t crc;
+    unsigned char *envptr = NULL;
+    unsigned int filesize = 0;
+	unsigned int datasize = ENV_SIZE;
+    unsigned char buf[ENV_SIZE];
+	unsigned int eoe;
+    int ret;
+	int i;
+	FILE *fp;
+
+
+
+    fp = fopen (argv[1], "r");
+	if (fp == NULL)
+       return -1;
+
+    fseek(fp, 0, SEEK_END);  
+    filesize = ftell(fp);
+    fseek(fp, 0, SEEK_SET);
+  
+    for(i=0;i<ENV_SIZE;i++){
+	     buf[i]=0;
+	}
+
+    ret=fread(buf, filesize, 1, fp );
+    crc = crc32 (0, buf, ENV_SIZE);
+
+    fclose(fp);
+	printf("%08x\n",crc);
 #endif
 	return EXIT_SUCCESS;
 }
+
+
+
+
