# HG changeset patch
# Parent 35a5e82c9c3ece4c182b36d076beffb99f04add1

diff --git a/drivers/mtd/ubi/build.c b/drivers/mtd/ubi/build.c
--- a/drivers/mtd/ubi/build.c
+++ b/drivers/mtd/ubi/build.c
@@ -476,21 +476,22 @@ static int attach_by_scanning(struct ubi
 	if (err)
 		goto out_si;
 
-	err = ubi_wl_init_scan(ubi, si);
-	if (err)
-		goto out_vtbl;
-
 	err = ubi_eba_init_scan(ubi, si);
 	if (err)
 		goto out_wl;
 
-	ubi_scan_destroy_si(si);
+    err = ubi_wl_init_scan(ubi, si);
+    if (err)
+        goto out_vtbl;
+
+
+    ubi_scan_destroy_si(si);
 	return 0;
 
-out_wl:
-	ubi_wl_close(ubi);
 out_vtbl:
 	vfree(ubi->vtbl);
+out_wl:
+    ubi_wl_close(ubi);
 out_si:
 	ubi_scan_destroy_si(si);
 	return err;
diff --git a/include/ubi_uboot.h b/include/ubi_uboot.h
--- a/include/ubi_uboot.h
+++ b/include/ubi_uboot.h
@@ -38,7 +38,7 @@ do {									\
 } while (0)
 
 /* configurable */
-#define CONFIG_MTD_UBI_WL_THRESHOLD	4096
+#define CONFIG_MTD_UBI_WL_THRESHOLD	CONFIG_UBI_WL_THRESHOLD
 #define CONFIG_MTD_UBI_BEB_RESERVE	1
 #define UBI_IO_DEBUG			0
 
