# HG changeset patch
# Parent 8b164f8be97612a0e32da37524caef7c297d8106
diff --git a/examples/standalone/Makefile b/examples/standalone/Makefile
--- a/examples/standalone/Makefile
+++ b/examples/standalone/Makefile
@@ -80,7 +80,7 @@ all:	$(obj).depend $(OBJS) $(LIB) $(SREC
 
 $(ELF):
 $(obj)%:	$(obj)%.o $(LIB)
-		$(LD) -g -Ttext $(STANDALONE_LOAD_ADDR) \
+		$(LD) -g -Ttext $(STANDALONE_LOAD_ADDR) $(PLATFORM_LDFLAGS)\
 			-o $@ -e $(SYM_PREFIX)$(notdir $(<:.o=)) $< $(LIB) \
 			-L$(gcclibdir) -lgcc
 
diff --git a/examples/standalone/mips.lds b/examples/standalone/mips.lds
--- a/examples/standalone/mips.lds
+++ b/examples/standalone/mips.lds
@@ -24,7 +24,7 @@
 /*
 OUTPUT_FORMAT("elf32-bigmips", "elf32-bigmips", "elf32-bigmips")
 */
-OUTPUT_FORMAT("elf32-tradbigmips", "elf32-tradbigmips", "elf32-tradbigmips")
+OUTPUT_FORMAT("elf32-tradbigmips", "elf32-tradbigmips", "elf32-tradlittlemips")
 OUTPUT_ARCH(mips)
 SECTIONS
 {
