diff --git a/common/cmd_upgrade.c b/common/cmd_upgrade.c
--- a/common/cmd_upgrade.c
+++ b/common/cmd_upgrade.c
@@ -254,7 +254,7 @@ int http_upgrade(ulong srcAddr, int srcL
 		srcHeader = (!srcHeader ? srcAddr : srcHeader + curParSize);
 		printf("srcHeader=0x%08x\n",srcHeader);
 		pimg_header = (image_header_t *)srcHeader;		
-		curParSize = sizeof(image_header_t) + pimg_header->ih_size;
+		curParSize = sizeof(image_header_t) + ntohl(pimg_header->ih_size);
 		printf("curParSize=0x%x\n",curParSize);
 		
 		pad = (16 - (curParSize % 16)) % 16;
diff --git a/common/upgrade.c b/common/upgrade.c
--- a/common/upgrade.c
+++ b/common/upgrade.c
@@ -338,7 +338,7 @@ int upgrade_img(ulong srcAddr, ulong src
 	if (FindNPImgLoc(img_addr,&nextStartAddr,&preEndAddr))
 		return 1;
 	pimg_header = (image_header_t *)srcAddr;
-	if (pimg_header->ih_magic == IH_MAGIC) {
+	if (ntohl(pimg_header->ih_magic) == IH_MAGIC) {
 		printf("Image contains header with name [%s]\n",pimg_header->ih_name);
 		if(pimg_header->ih_type != IH_TYPE_KERNEL) {
 			upgrade_debug_printf("This is not kernel image and so removing header\n");
