# HG changeset patch
# Parent 670936c2f30a98f0c2e069a938805ec288858680

diff --git a/common/Kconfig b/common/Kconfig
--- a/common/Kconfig
+++ b/common/Kconfig
@@ -15,6 +15,9 @@ config CMD_BMP
 config CMD_CACHE
      bool "cache"
 
+config CMD_CHKSET
+	 bool "chkset"
+
 config CMD_CONSOLE
      bool "console"
 	 
diff --git a/common/cmd_nvedit.c b/common/cmd_nvedit.c
--- a/common/cmd_nvedit.c
+++ b/common/cmd_nvedit.c
@@ -606,6 +606,31 @@ U_BOOT_CMD(
 #endif
 
 
+int do_chkset(cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
+{
+   char* varname;
+   ulong address;
+   ulong value;
+   
+   if (argc < 4) {
+           cmd_usage(cmdtp);
+           return 1;
+		       }
+
+   varname=argv[1];
+   address=simple_strtoul(argv[2], NULL, 16);
+   value=simple_strtoul(argv[3], NULL, 16);;
+   
+   if(*(ulong*)address==value){
+     setenv(varname, "1");
+   }else{
+     setenv(varname, "0"); 
+   }
+
+   return 0;
+}
+
+
 /************************************************************************
  * Match a name / name=value pair
  *
@@ -680,3 +705,13 @@ U_BOOT_CMD(
 	"    - run the commands in the environment variable(s) 'var'"
 );
 #endif
+
+#if defined(CONFIG_CMD_CHKSET)
+U_BOOT_CMD(
+    chkset, CONFIG_SYS_MAXARGS, 3, do_chkset,
+    "Set var depending on the value of specified address",
+	"chkset var address value\n "
+);
+#endif
+
+
