# HG changeset patch
# Parent 136cf5816640be0e696d30f11ba54217a76e0b44

diff --git a/arch/mips/lib/bootm.c b/arch/mips/lib/bootm.c
--- a/arch/mips/lib/bootm.c
+++ b/arch/mips/lib/bootm.c
@@ -28,6 +28,13 @@
 #include <asm/byteorder.h>
 #include <asm/addrspace.h>
 
+#if defined(CONFIG_OF_LIBFDT)
+#include <fdt.h>
+#include <libfdt.h>
+#include <fdt_support.h>
+#endif
+
+
 DECLARE_GLOBAL_DATA_PTR;
 
 #define	LINUX_MAX_ENVS		256
@@ -100,8 +107,15 @@ int do_bootm_linux(int flag, int argc, c
 	/* we assume that the kernel is in place */
 	printf ("\nStarting kernel ...\n\n");
 
+#if defined(CONFIG_OF_LIBFDT)
+        if (images->ft_addr) { 
+          printf("booting from OF flat tree...\n");  
+          theKernel(FDT_MAGIC, images->ft_addr, NULL, 0);
+        }
+#else
 	theKernel (linux_argc, linux_argv, linux_env, 0);
 	/* does not return */
+#endif
 	return 1;
 }
 
diff --git a/common/cmd_bootm.c b/common/cmd_bootm.c
--- a/common/cmd_bootm.c
+++ b/common/cmd_bootm.c
@@ -301,7 +301,7 @@ static int bootm_start(cmd_tbl_t *cmdtp,
 		}
 
 #if defined(CONFIG_OF_LIBFDT)
-#if defined(CONFIG_PPC) || defined(CONFIG_M68K) || defined(CONFIG_SPARC)
+#if defined(CONFIG_PPC) || defined(CONFIG_M68K) || defined(CONFIG_SPARC) || defined(CONFIG_MIPS)
 		/* find flattened device tree */
 		ret = boot_get_fdt (flag, argc, argv, &images,
 				    &images.ft_addr, &images.ft_len);
