# HG changeset patch
# Parent 9eaccdd01ef221054cae57c5c62ab3d87ba2b349
diff --git a/common/Kconfig b/common/Kconfig
--- a/common/Kconfig
+++ b/common/Kconfig
@@ -235,3 +235,13 @@ config YAFFS2
      bool "YAFFS2"
 	 
 endmenu
+
+config SYS_HUSH_PARSER
+    bool "Hush Parser"
+       default n
+
+config SYS_PROMPT_HUSH_PS2
+    string "Hush PS2 Prompt"
+       default "> "
+       depends on SYS_HUSH_PARSER
+
