# HG changeset patch
# Parent 0190728089a42ccca400427a6e071e2c2fc96ee9

diff --git a/include/configs/lq_extra_env.h b/include/configs/lq_extra_env.h
--- a/include/configs/lq_extra_env.h
+++ b/include/configs/lq_extra_env.h
@@ -65,7 +65,14 @@
   "update_gphyfirmware="CONFIG_UPDATE_GPHYFIRMWARE"\0"
   "gphy_fw_addr="MK_STR(CONFIG_HN1_EXTERN_GPHY_FW_ADDR)"\0"
 #endif
-#ifndef CONFIG_ENV_IS_NOWHERE 
+#if defined(CONFIG_DRIVER_GRX500)
+"update_gphyfirmware="CONFIG_UPDATE_GPHYFIRMWARE"\0"
+"gphy_fw_addr="MK_STR(CONFIG_GRX500_EXTERN_GPHY_FW_ADDR)"\0"
+"update_bootcore="CONFIG_UPDATE_BOOTCORE"\0"
+"run_bootcore="CONFIG_RUN_BOOTCORE"\0"
+"bootcore="CONFIG_BOOT_CORE"\0"
+#endif
+#ifndef CONFIG_ENV_IS_NOWHERE
   "reset_uboot_config="CONFIG_RESET_UBOOT_CONFIG"\0" 
   "reset_ddr_config="CONFIG_RESET_DDR_CONFIG"\0"
   "reset_sysconfig="CONFIG_RESET_SYSCONFIG"\0"
@@ -109,9 +116,9 @@
   "f_ddrconfig_size=" MK_STR(IFX_CFG_FLASH_DDR_CFG_SIZE) "\0"             
   "f_ddrconfig_end=" MK_STR(IFX_CFG_FLASH_DDR_CFG_END_ADDR) "\0"        
 #ifdef CONFIG_OVERLAY
-  "f_overlay_addr=" MK_STR(IFX_CFG_FLASH_OVERLAY_IMAGE_START_ADDR) "\0"       
-  "f_overlay_size=" MK_STR(IFX_CFG_FLASH_OVERLAY_IMAGE_SIZE) "\0"   
-  "f_overlay_end=" MK_STR(IFX_CFG_FLASH_OVERLAY_IMAGE_END_ADDR) "\0"
+  "f_data_addr=" MK_STR(IFX_CFG_FLASH_DATA_IMAGE_START_ADDR) "\0"       
+  "f_data_size=" MK_STR(IFX_CFG_FLASH_DATA_IMAGE_SIZE) "\0"   
+  "f_data_end=" MK_STR(IFX_CFG_FLASH_DATA_IMAGE_END_ADDR) "\0"
 #endif
 #ifdef CONFIG_DRIVER_VBG400
   "wlanm=31M\0"
