Index: irqbalance-1.0.4/classify.c
===================================================================
--- irqbalance-1.0.4.orig/classify.c	2012-08-27 20:56:39.000000000 +0530
+++ irqbalance-1.0.4/classify.c	2012-10-03 22:13:36.000000000 +0530
@@ -353,7 +353,8 @@
 	while(gentry) {
 		ninfo = gentry->data;
 		iinfo = get_irq_info(ninfo->irq);
-		new_irq_list = g_list_remove(gentry, ninfo);
+		new_irq_list = g_list_delete_link(gentry, gentry); /* fix as per http://code.google.com/p/irqbalance/issues/detail?id=39 - g_list_remove(gentry, ninfo); */
+
 		if (!iinfo) {
 			if (debug_mode)
 				printf("Adding untracked IRQ %d to database\n", ninfo->irq);
