From 81dd8975ef98a625497489b8fbcbd28ff9c973f2 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 15:45:55 +0300
Subject: [PATCH 07/69] hostapd: fix max rts and frag threshold

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/drivers/driver_nl80211.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index d23a647..a3d3e53 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -5353,7 +5353,7 @@ static int i802_set_rts(void *priv, int rts)
 	int ret;
 	u32 val;
 
-	if (rts >= 2347)
+	if (rts > 2347)
 		val = (u32) -1;
 	else
 		val = rts;
@@ -5381,7 +5381,7 @@ static int i802_set_frag(void *priv, int frag)
 	int ret;
 	u32 val;
 
-	if (frag >= 2346)
+	if (frag > 2346)
 		val = (u32) -1;
 	else
 		val = frag;
-- 
1.8.3.4

