From 28d0f5e12d099f073bcc82b8345ea05b2a97c1be Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:20 +0300
Subject: [PATCH 12/69] hostapd: acs

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 hostapd/.config | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hostapd/.config b/hostapd/.config
index 0ded826..44f8dd1 100644
--- a/hostapd/.config
+++ b/hostapd/.config
@@ -319,4 +319,4 @@ CONFIG_NO_RANDOM_POOL=y
 # For more details refer to:
 # http://wireless.kernel.org/en/users/Documentation/acs
 #
-#CONFIG_ACS=y
+CONFIG_ACS=y
-- 
1.8.3.4

