From d44e2c79b717fcd5818f22ce9b835c543145695e Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:20 +0300
Subject: [PATCH 19/69] disable pairwise rekeying

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/wpa_auth.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/ap/wpa_auth.c b/src/ap/wpa_auth.c
index 3587086..2d76560 100644
--- a/src/ap/wpa_auth.c
+++ b/src/ap/wpa_auth.c
@@ -813,11 +813,13 @@ static int wpa_receive_error_report(struct wpa_authenticator *wpa_auth,
 		wpa_auth->dot11RSNAStatsTKIPRemoteMICFailures++;
 	}
 
+#if 0 /* disable pairwise rekeying */
 	/*
 	 * Error report is not a request for a new key handshake, but since
 	 * Authenticator may do it, let's change the keys now anyway.
 	 */
 	wpa_request_new_ptk(sm);
+#endif
 	return 0;
 }
 
-- 
1.8.3.4

