From 596bae11dca75b4a91af53c670c52b64a1f8deb0 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:20 +0300
Subject: [PATCH 20/69] hostapd: socket size

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/drivers/driver_nl80211.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 0d01424..4d6a8cd 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -132,11 +132,13 @@ static void nl_destroy_handles(struct nl_handle **handle)
 #define ELOOP_SOCKET_INVALID	(intptr_t) 0x88888889ULL
 #endif
 
+#define MGMT_SOCKET_RX_SIZE (262144)
+
 static void nl80211_register_eloop_read(struct nl_handle **handle,
 					eloop_sock_handler handler,
 					void *eloop_data)
 {
-#ifdef CONFIG_LIBNL20
+#if defined(CONFIG_LIBNL20) || defined(CONFIG_LIBNL32)
 	/*
 	 * libnl uses a pretty small buffer (32 kB that gets converted to 64 kB)
 	 * by default. It is possible to hit that limit in some cases where
@@ -144,7 +146,9 @@ static void nl80211_register_eloop_read(struct nl_handle **handle,
 	 * to hostapd and STA entry deletion. Try to increase the buffer to make
 	 * this less likely to occur.
 	 */
-	if (nl_socket_set_buffer_size(*handle, 262144, 0) < 0) {
+	wpa_printf(MSG_DEBUG, "nl80211: increase nl_socket RX buffer size: %d", 
+		MGMT_SOCKET_RX_SIZE);
+	if (nl_socket_set_buffer_size(*handle, MGMT_SOCKET_RX_SIZE, 0) < 0) {
 		wpa_printf(MSG_DEBUG,
 			   "nl80211: Could not set nl_socket RX buffer size: %s",
 			   strerror(errno));
-- 
1.8.3.4

