From 29122fe6c06046c0c529ca56feec9eb447589f21 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:20 +0300
Subject: [PATCH 21/69] hostapd: send deauthentication after activate ap

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/hostapd.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 8c97958..d9d1e0f 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1836,8 +1836,10 @@ dfs_offload:
 	if (iface->interfaces && iface->interfaces->terminate_on_error > 0)
 		iface->interfaces->terminate_on_error--;
 
-	for (j = 0; j < iface->num_bss; j++)
+	for (j = 0; j < iface->num_bss; j++) {
 		hostapd_set_own_neighbor_report(iface->bss[j]);
+		hostapd_flush_old_stations(iface->bss[j], WLAN_REASON_PREV_AUTH_NOT_VALID);
+	}
 
 	return 0;
 
-- 
1.8.3.4

