From 1cea6dd8a96449e1856d67aaa88c2ea78074af94 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:21 +0300
Subject: [PATCH 24/69] hostapd: extend time limit of signal terminate

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/utils/eloop.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/utils/eloop.c b/src/utils/eloop.c
index 436bc8c..80296b3 100644
--- a/src/utils/eloop.c
+++ b/src/utils/eloop.c
@@ -971,7 +971,7 @@ static void eloop_handle_signal(int sig)
 		 * would not allow the program to be killed. */
 		eloop.pending_terminate = 1;
 		signal(SIGALRM, eloop_handle_alarm);
-		alarm(2);
+		alarm(30);
 	}
 #endif /* CONFIG_NATIVE_WINDOWS */
 
-- 
1.8.3.4

