From 84c88004076c00fac1dcb51f5ecc58c426f99b5e Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:21 +0300
Subject: [PATCH 29/69] hostapd: fix for disable 11ac

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/ap_config.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index f621cf7..9a11033 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -847,6 +847,7 @@ static int hostapd_config_check_bss(struct hostapd_bss_config *bss,
 	if (full_config && conf->ieee80211n &&
 	    conf->hw_mode == HOSTAPD_MODE_IEEE80211B) {
 		bss->disable_11n = 1;
+		bss->disable_11ac = 1;
 		wpa_printf(MSG_ERROR, "HT (IEEE 802.11n) in 11b mode is not "
 			   "allowed, disabling HT capabilities");
 	}
@@ -854,6 +855,7 @@ static int hostapd_config_check_bss(struct hostapd_bss_config *bss,
 	if (full_config && conf->ieee80211n &&
 	    bss->ssid.security_policy == SECURITY_STATIC_WEP) {
 		bss->disable_11n = 1;
+		bss->disable_11ac = 1;
 		wpa_printf(MSG_ERROR, "HT (IEEE 802.11n) with WEP is not "
 			   "allowed, disabling HT capabilities");
 	}
@@ -864,6 +866,7 @@ static int hostapd_config_check_bss(struct hostapd_bss_config *bss,
 				   WPA_CIPHER_CCMP_256 | WPA_CIPHER_GCMP_256)))
 	{
 		bss->disable_11n = 1;
+		bss->disable_11ac = 1;
 		wpa_printf(MSG_ERROR, "HT (IEEE 802.11n) with WPA/WPA2 "
 			   "requires CCMP/GCMP to be enabled, disabling HT "
 			   "capabilities");
-- 
1.8.3.4

