From 54d008603f748234ea1c155a0aa3ef5fee8f52af Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:21 +0300
Subject: [PATCH 33/69] hostapd: assoc cb fix

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/ieee802_11.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 2a6418c..d4ca524 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -2859,6 +2859,13 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 		status = le_to_host16(mgmt->u.reassoc_resp.status_code);
 	else
 		status = le_to_host16(mgmt->u.assoc_resp.status_code);
+		
+	if (!(sta->flags & WLAN_STA_ASSOC_REQ_OK)) {
+		hostapd_logger(hapd, sta->addr, HOSTAPD_MODULE_IEEE80211,
+		       HOSTAPD_LEVEL_INFO,
+		       "STA is not associated, ignoring assoc cb");
+		return;
+	}
 
 	if (!ok) {
 		hostapd_logger(hapd, mgmt->da, HOSTAPD_MODULE_IEEE80211,
-- 
1.8.3.4

