From 1a9b001f9daeb7b1e5612e8578b1a86358ae611e Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Wed, 26 Oct 2016 15:09:39 +0300
Subject: [PATCH 35/69] hostapd: arp proxy

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/ieee802_11_shared.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
index 259413b..ba95790 100644
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -246,6 +246,8 @@ u8 * hostapd_eid_ext_capab(struct hostapd_data *hapd, u8 *eid)
 	if (len < 9 &&
 	    (hapd->conf->ftm_initiator || hapd->conf->ftm_responder))
 		len = 9;
+	if (len < 2 && hapd->conf->proxy_arp)
+		len = 2;
 #ifdef CONFIG_WNM
 	if (len < 4)
 		len = 4;
-- 
2.7.4

