From 5e6a2e82be583220908b91b289cd5c080f830bdb Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Mon, 5 Sep 2016 16:21:22 +0300
Subject: [PATCH 43/69] hostapd: fixed vht chan scan

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 src/ap/hw_features.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/src/ap/hw_features.c b/src/ap/hw_features.c
index 3478e5f..9f6befa 100644
--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -639,12 +639,19 @@ static void ieee80211n_scan_channels_5g(struct hostapd_iface *iface,
 		return;
 
 	pri_freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
-	if (iface->conf->secondary_channel > 0) {
-		affected_start = pri_freq - 10;
-		affected_end = pri_freq + 30;
+  /* If VHT channel used, trigger scan for all channels in chanwidth to calibrate all those channels */
+	if (iface->conf->ieee80211ac && iface->conf->vht_oper_chwidth == VHT_CHANWIDTH_80MHZ) {
+		int low_freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->vht_oper_centr_freq_seg0_idx - 6);
+		affected_start = low_freq - 10;
+		affected_end = low_freq + 70;
 	} else {
-		affected_start = pri_freq - 30;
-		affected_end = pri_freq + 10;
+		if (iface->conf->secondary_channel > 0) {
+			affected_start = pri_freq - 10;
+			affected_end = pri_freq + 30;
+		} else {
+			affected_start = pri_freq - 30;
+			affected_end = pri_freq + 10;
+		}
 	}
 	wpa_printf(MSG_DEBUG, "40 MHz affected channel range: [%d,%d] MHz",
 		   affected_start, affected_end);
-- 
1.8.3.4

