From 50ba5fa9d4b1cd572f9ea5ba7260ae7442533fc9 Mon Sep 17 00:00:00 2001
From: Simon Dinkin <simonx.dinkin@intel.com>
Date: Sun, 7 Aug 2016 12:21:43 +0300
Subject: [PATCH 52/69] disable new implemantation of proxy arp

---
 src/ap/hostapd.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index e7cc6f7..c00af85 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1133,7 +1133,7 @@ static int hostapd_setup_bss(struct hostapd_data *hapd, int first)
 		wpa_printf(MSG_ERROR, "BSS Load initialization failed");
 		return -1;
 	}
-
+#if 0
 	if (conf->proxy_arp) {
 		if (x_snoop_init(hapd)) {
 			wpa_printf(MSG_ERROR,
@@ -1153,7 +1153,7 @@ static int hostapd_setup_bss(struct hostapd_data *hapd, int first)
 			return -1;
 		}
 	}
-
+#endif /* disable proxy arp */
 	if (!hostapd_drv_none(hapd) && vlan_init(hapd)) {
 		wpa_printf(MSG_ERROR, "VLAN initialization failed.");
 		return -1;
-- 
1.8.3.4

