From 54e796caf63e584713fa5e45b8828d9b8aa1c658 Mon Sep 17 00:00:00 2001
From: Simon Dinkin <simonx.dinkin@intel.com>
Date: Tue, 1 Nov 2016 20:29:22 +0200
Subject: [PATCH] 76 hostapd wmm uapsd fix

---
 src/ap/ieee802_11.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index e7b745b..43aa3ab 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -1493,6 +1493,11 @@ static u16 check_wmm(struct hostapd_data *hapd, struct sta_info *sta,
 		sta->flags |= WLAN_STA_WMM;
 		wmm = (struct wmm_information_element *) wmm_ie;
 		sta->qosinfo = wmm->qos_info;
+
+		/* clear UAPSD if it unsupported by BSS */
+		if (!hapd->conf->wmm_uapsd)
+			sta->qosinfo &= ~WMM_QOSINFO_STA_AC_MASK;
+
 	}
 	return WLAN_STATUS_SUCCESS;
 }
-- 
2.7.4

