--- hostap-724fce8_orig2/src/ap/ieee802_11.c	2015-10-12 09:08:13.000000000 +0300
+++ hostap-724fce8_mfp2/src/ap/ieee802_11.c	2015-10-14 15:31:14.000000000 +0300
@@ -1075,6 +1075,11 @@
 			ap_check_sa_query_timeout(hapd, sta);
 		if ((sta->flags & WLAN_STA_MFP) && !sta->sa_query_timed_out &&
 		    (!reassoc || sta->auth_alg != WLAN_AUTH_FT)) {
+
+			if (!wpa_auth_pairwise_set(sta->wpa_sm)) {
+				return WLAN_STATUS_PAIRWISE_CIPHER_NOT_VALID;
+			}
+
 			/*
 			 * STA has already been associated with MFP and SA
 			 * Query timeout has not been reached. Reject the
@@ -1409,6 +1414,14 @@
 	/* followed by SSID and Supported rates; and HT capabilities if 802.11n
 	 * is used */
 	resp = check_assoc_ies(hapd, sta, pos, left, reassoc);
+
+	if (resp == WLAN_STATUS_PAIRWISE_CIPHER_NOT_VALID) {
+		wpa_printf(MSG_INFO, "Pairwise key is not set. Send Deauth");
+		ap_sta_disconnect(hapd, sta, sta->addr, resp);
+		ap_free_sta(hapd, sta);
+		return;
+	}
+
 	if (resp != WLAN_STATUS_SUCCESS)
 		goto fail;
 
