--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -300,6 +300,7 @@ static u8 * hostapd_eid_csa(struct hosta
 
 static u8 * hostapd_eid_ecsa(struct hostapd_data *hapd, u8 *eid)
 {
+#if 0
 	if (!hapd->cs_freq_params.channel || !hapd->iface->cs_oper_class)
 		return eid;
 
@@ -311,6 +312,12 @@ static u8 * hostapd_eid_ecsa(struct host
 	*eid++ = hapd->cs_count;
 
 	return eid;
+#else
+  /* todo: not possible to set capa->max_csa_counters due to missing code in
+   * nl89211 of currently used kernel. Remove this fix, when
+   * NL80211_ATTR_MAX_CSA_COUNTERS becomes available. */
+  return eid;
+#endif
 }
 
 
@@ -429,6 +436,8 @@ static u8 * hostapd_gen_probe_resp(struc
 
 	pos = hostapd_eid_rm_enabled_capab(hapd, pos, epos - pos);
 
+  /* todo: hostapd_eid_ecsa has been patched to exclude this element. Remove
+   * patch once NL80211_ATTR_MAX_CSA_COUNTERS becomes available in kernel. */
 	/* eCSA IE */
 	csa_pos = hostapd_eid_ecsa(hapd, pos);
 	if (csa_pos != pos)
@@ -1090,6 +1099,8 @@ int ieee802_11_build_ap_params(struct ho
 	tailpos = hostapd_eid_bss_load(hapd, tailpos,
 				       tail + BEACON_TAIL_BUF_SIZE - tailpos);
 
+	/* todo: hostapd_eid_ecsa has been patched to exclude this element. Remove
+	 * patch once NL80211_ATTR_MAX_CSA_COUNTERS becomes available in kernel. */
 	/* eCSA IE */
 	csa_pos = hostapd_eid_ecsa(hapd, tailpos);
 	if (csa_pos != tailpos)
