diff -Npur hostap-2.6/src/ap/ieee802_11_vht.c hostap-2.6-patched/src/ap/ieee802_11_vht.c
--- hostap-2.6/src/ap/ieee802_11_vht.c	2016-11-16 11:55:53.000000000 +0200
+++ hostap-2.6-patched/src/ap/ieee802_11_vht.c	2016-11-16 11:48:32.000000000 +0200
@@ -531,8 +531,23 @@ void hostapd_get_vht_capab(struct hostap
 	/* mask out supported MCS set vs our own MCS set */
 	{
 		u16 our_mcs_set[4];
+		struct hostapd_hw_modes *mode = hapd->iface->current_mode;
+
+		if (mode->mode == HOSTAPD_MODE_IEEE80211G && hapd->conf->vendor_vht &&
+		    mode->vht_capab == 0 && hapd->iface->hw_features) {
+			int i;
+
+			for (i = 0; i < hapd->iface->num_hw_features; i++) {
+				if (hapd->iface->hw_features[i].mode ==
+				    HOSTAPD_MODE_IEEE80211A) {
+					mode = &hapd->iface->hw_features[i];
+					break;
+				}
+			}
+		}
+
+		os_memcpy(our_mcs_set, mode->vht_mcs_set, sizeof(our_mcs_set));
 
-		os_memcpy(our_mcs_set, hapd->iface->current_mode->vht_mcs_set, sizeof(our_mcs_set));
 		/* RX highest vs our TX highest*/
 		neg_vht_cap->vht_supported_mcs_set.rx_highest =
 			limit_rx_tx_highest(neg_vht_cap->vht_supported_mcs_set.rx_highest, our_mcs_set[3]);
diff -Npur hostap-2.6/src/common/ieee802_11_common.c hostap-2.6-patched/src/common/ieee802_11_common.c
--- hostap-2.6/src/common/ieee802_11_common.c	2016-11-16 11:55:53.000000000 +0200
+++ hostap-2.6-patched/src/common/ieee802_11_common.c	2016-11-16 11:49:32.000000000 +0200
@@ -138,7 +138,8 @@ static int ieee802_11_parse_vendor_speci
 		case VENDOR_VHT_TYPE:
 			if (elen > 4 &&
 			    (pos[4] == VENDOR_VHT_SUBTYPE ||
-			     pos[4] == VENDOR_VHT_SUBTYPE2)) {
+			     pos[4] == VENDOR_VHT_SUBTYPE2 ||
+			     pos[4] == VENDOR_VHT_SUBTYPE3)) {
 				elems->vendor_vht = pos;
 				elems->vendor_vht_len = elen;
 			} else
diff -Npur hostap-2.6/src/common/ieee802_11_defs.h hostap-2.6-patched/src/common/ieee802_11_defs.h
--- hostap-2.6/src/common/ieee802_11_defs.h	2016-11-16 11:55:53.000000000 +0200
+++ hostap-2.6-patched/src/common/ieee802_11_defs.h	2016-11-16 12:00:08.000000000 +0200
@@ -1383,6 +1383,7 @@ enum plink_action_field {
 #define VENDOR_VHT_TYPE		0x04
 #define VENDOR_VHT_SUBTYPE	0x08
 #define VENDOR_VHT_SUBTYPE2	0x00
+#define VENDOR_VHT_SUBTYPE3	0x18
 
 #define VENDOR_HT_CAPAB_OUI_TYPE 0x33 /* 00-90-4c:0x33 */
 
