diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -556,8 +556,7 @@
 
 #ifdef CONFIG_MBO
 
-u8 * hostapd_eid_mbo(struct hostapd_data *hapd, u8 *eid, size_t len)
-{
+u8 * hostapd_eid_mbo(struct hostapd_data *hapd, u8 *eid, size_t len) {
 	u8 mbo[6], *mbo_pos = mbo;
 	u8 *pos = eid;
 
@@ -572,14 +571,13 @@
 		*mbo_pos++ = MBO_ATTR_ID_ASSOC_DISALLOW;
 		*mbo_pos++ = 1;
 		*mbo_pos++ = hapd->mbo_bss_assoc_disallow;
-	} else {
-		*mbo_pos++ = MBO_ATTR_ID_AP_CAPA_IND;
-		*mbo_pos++ = 1;
-		if (hapd->conf->mbo_cell_aware)
-			*mbo_pos++ = MBO_AP_CAPA_CELL_AWARE;
-		else
-			*mbo_pos++ = 0; /* Not Cellular aware */
 	}
+	*mbo_pos++ = MBO_ATTR_ID_AP_CAPA_IND;
+	*mbo_pos++ = 1;
+	if (hapd->conf->mbo_cell_aware)
+		*mbo_pos++ = MBO_AP_CAPA_CELL_AWARE;
+	else
+		*mbo_pos++ = 0; /* Not Cellular aware */
 
 	pos += mbo_add_ie(pos, len, mbo, mbo_pos - mbo);
 
