diff --git a/src/ap/rrm.c b/src/ap/rrm.c
index a23f300..0f89769 100644
--- a/src/ap/rrm.c
+++ b/src/ap/rrm.c
@@ -1552,7 +1552,7 @@ int hostapd_send_beacon_req(struct hostapd_data *hapd, const u8 *addr,
 	wpabuf_put_u8(buf, WLAN_ACTION_RADIO_MEASUREMENT);
 	wpabuf_put_u8(buf, WLAN_RRM_RADIO_MEASUREMENT_REQUEST);
 	wpabuf_put_u8(buf, hapd->beacon_req_token);
-	wpabuf_put_le16(buf, num_of_repetitions);
+	wpabuf_put_be16(buf, num_of_repetitions);
 
 	/* IEEE P802.11-REVmc/D5.0, 9.4.2.21 */
 	wpabuf_put_u8(buf, WLAN_EID_MEASURE_REQUEST);
@@ -1565,8 +1565,8 @@ int hostapd_send_beacon_req(struct hostapd_data *hapd, const u8 *addr,
 	/* IEEE P802.11-REVmc/D4.0, 8.4.2.20.7 */
 	wpabuf_put_u8(buf, oper_class);
 	wpabuf_put_u8(buf, channel);
-	wpabuf_put_le16(buf, random_interval);
-	wpabuf_put_le16(buf, measurement_duration);
+	wpabuf_put_be16(buf, random_interval);
+	wpabuf_put_be16(buf, measurement_duration);
 	wpabuf_put_u8(buf, mode); /* Measurement Mode */
 	if (!bssid) {
 		/* use wildcard BSSID instead of a specific BSSID */
