diff -Npur --exclude=.cproject --exclude=.project --exclude=.settings --exclude=hostapd_debug --exclude='*.orig' hostapd-2.6-orig/src/ap/hw_features.c hostapd-2.6/src/ap/hw_features.c
--- hostapd-2.6-orig/src/ap/hw_features.c	2017-04-19 11:29:32.717871199 +0300
+++ hostapd-2.6/src/ap/hw_features.c	2017-04-19 11:31:13.985245190 +0300
@@ -1052,7 +1052,6 @@ static int ieee80211ac_supported_vht_cap
 	VHT_CAP_CHECK_MAX(VHT_CAP_RXSTBC_MASK);
 	VHT_CAP_CHECK(VHT_CAP_SU_BEAMFORMER_CAPABLE);
 	VHT_CAP_CHECK(VHT_CAP_SU_BEAMFORMEE_CAPABLE);
-	VHT_CAP_CHECK_MAX(VHT_CAP_BEAMFORMEE_STS_MAX);
 	VHT_CAP_CHECK_MAX(VHT_CAP_SOUNDING_DIMENSION_MAX);
 	VHT_CAP_CHECK(VHT_CAP_MU_BEAMFORMER_CAPABLE);
 	VHT_CAP_CHECK(VHT_CAP_MU_BEAMFORMEE_CAPABLE);
