diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index d1c3c15..7f36c2d 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -612,7 +612,6 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 	if (hapd->csa_in_progress &&
 	    freq == hapd->cs_freq_params.freq) {
 		hostapd_cleanup_cs_params(hapd);
-		ieee802_11_set_beacon(hapd);
 
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_CSA_FINISHED "freq=%d Channel=%d "
 				"OperatingChannelBandwidt=%s ExtensionChannel=%d cf1=%d cf2=%d reason=%s"
@@ -633,6 +632,12 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 
 		hapd->iface->chan_switch_reason = HAPD_CHAN_SWITCH_OTHER;
 	}
+
+	/* set_beacon needs to be called when AP csa is in progress, and also when STA role
+	 * interface initiated the ch switch following Peer AP CSA (the AP the sta is
+	 * connected to is switching channel and AP beacons need to be changed */
+	ieee802_11_set_beacon(hapd);
+
 #endif /* NEED_AP_MLME */
 }
 
