UGW: tools/m4 fix for compilation failure on older distributions

diff --git a/tools/m4/Makefile b/tools/m4/Makefile
--- a/tools/m4/Makefile
+++ b/tools/m4/Makefile
@@ -18,7 +18,7 @@ HOST_BUILD_PARALLEL:=1
 
 include $(INCLUDE_DIR)/host-build.mk
 
-HOST_CONFIGURE_VARS += gl_cv_func_strstr_linear=no
+HOST_CONFIGURE_VARS += gl_cv_func_strstr_linear=no CPPFLAGS="-fgnu89-inline"
 
 define Host/Clean
 	-$(MAKE) -C $(HOST_BUILD_DIR) uninstall
