UGW: Add platform call in image program

diff --git a/include/image.mk b/include/image.mk
--- a/include/image.mk
+++ b/include/image.mk
@@ -253,6 +253,7 @@ endef
 
 define Image/mkfs/prepare
 	$(call Image/mkfs/prepare/default)
+	$(call Image/mkfs/prepare/platform)
 endef
 
 
@@ -479,6 +480,7 @@ define BuildImage
 		$(call Image/Build,$(fs))
 	)
 	$(call Image/mkfs/ubifs)
+	$(call Image/mkfs/platform)
 	$(call Image/Checksum,md5sum --binary,md5sums)
 	$(call Image/Checksum,openssl dgst -sha256,sha256sums)
 
