add support for scripts/localmirrors.default

diff --git a/scripts/download.pl b/scripts/download.pl
--- a/scripts/download.pl
+++ b/scripts/download.pl
@@ -29,6 +29,14 @@ sub localmirrors {
 		}
 		close LM;
 	};
+	open LM_DEF, "$scriptdir/localmirrors.default" and do {
+	    while (<LM_DEF>) {
+			chomp $_;
+			push @mlist, $_;
+		}
+		close LM_DEF;
+	};
+
 	open CONFIG, "<".$ENV{'TOPDIR'}."/.config" and do {
 		while (<CONFIG>) {
 			/^CONFIG_LOCALMIRROR="(.+)"/ and do {
diff --git a/scripts/localmirrors.default b/scripts/localmirrors.default
new file mode 100644
--- /dev/null
+++ b/scripts/localmirrors.default
@@ -0,0 +1,3 @@
+file:///tmp2/wireline_sw_dist/
+http://mts.lantiq.com/~wlnsw/dist
+http://mts.lantiq.com/~wlnsw/dist/updates
