disable check for 'svn'

diff --git a/include/prereq-build.mk b/include/prereq-build.mk
--- a/include/prereq-build.mk
+++ b/include/prereq-build.mk
@@ -141,7 +141,7 @@ PKG_NAME:=Build dependency
 	python2 -V 2>&1 | grep Python, \
 	python -V 2>&1 | grep Python))
 
-$(eval $(call SetupHostCommand,svn,Please install the Subversion client, \
+#$(eval $(call SetupHostCommand,svn,Please install the Subversion client, \
 	svn --version | grep Subversion))
 
 $(eval $(call SetupHostCommand,git,Please install Git (git-core) >= 1.6.5, \
