define CONFIG_DEBUG_SECTION_MISMATCH to show details

diff --git a/include/kernel-defaults.mk b/include/kernel-defaults.mk
--- a/include/kernel-defaults.mk
+++ b/include/kernel-defaults.mk
@@ -9,6 +9,7 @@ KERNEL_MAKEOPTS := -C $(LINUX_DIR) \
 	CROSS_COMPILE="$(KERNEL_CROSS)" \
 	ARCH="$(LINUX_KARCH)" \
 	KBUILD_HAVE_NLS=no \
+	CONFIG_DEBUG_SECTION_MISMATCH=y \
 	CONFIG_SHELL="$(BASH)" \
 	$(if $(findstring c,$(OPENWRT_VERBOSE)),V=1,V='')
 
