tools/mtd-utils: fix PKG_MIRROR_MD5SUM

diff --git a/tools/mtd-utils/Makefile b/tools/mtd-utils/Makefile
--- a/tools/mtd-utils/Makefile
+++ b/tools/mtd-utils/Makefile
@@ -14,7 +14,7 @@ PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$
 PKG_SOURCE_URL:=git://git.infradead.org/mtd-utils.git
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
-PKG_MIRROR_MD5SUM:=154592f5bbb01e1e98bdd26c8a592305
+PKG_MIRROR_MD5SUM:=9fe0518213feddacd1842935dddeb6c8
 PKG_CAT:=zcat
 
 HOST_BUILD_DIR:=$(BUILD_DIR_HOST)/$(PKG_NAME)-$(PKG_VERSION)
