--- odhcpd-2015-09-07-1/src/router.c	2015-09-15 12:40:39.000000000 +0530
+++ odhcpd-2015-09-07-1_modif/src/router.c	2016-11-24 16:28:09.815907726 +0530
@@ -383,11 +383,11 @@
 		uint32_t lifetime;
 		uint32_t addr[4];
 	} routes[RELAYD_MAX_PREFIXES];
-
+	int nValid = 0;
 	for (ssize_t i = 0; i < ipcnt; ++i) {
 		struct odhcpd_ipaddr *addr = &addrs[i];
-		if (addr->dprefix > 64 || addr->dprefix == 0 || addr->valid <= (uint32_t)now ||
-				(addr->dprefix == 64 && addr->prefix == 64)) {
+		nValid = (uint32_t)addr->valid;
+		if (addr->dprefix > 64 ||(addr->valid <= (uint32_t)now && nValid != -1)) {
 			continue; // Address not suitable
 		} else if (addr->dprefix > 32) {
 			addr->addr.s6_addr32[1] &= htonl(~((1U << (64 - addr->dprefix)) - 1));
