diff -Naur wireless_tools.30.pre6/iwconfig.c wireless_tools.30.pre7/iwconfig.c
--- wireless_tools.30.pre6/iwconfig.c	2008-01-17 03:05:08.000000000 +0200
+++ wireless_tools.30.pre7/iwconfig.c	2008-05-16 02:17:52.000000000 +0300
@@ -100,7 +100,7 @@
 #ifndef WE_ESSENTIAL
   /* Get NickName */
   wrq.u.essid.pointer = (caddr_t) info->nickname;
-  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 1;
+  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 2;
   wrq.u.essid.flags = 0;
   if(iw_get_ext(skfd, ifname, SIOCGIWNICKN, &wrq) >= 0)
     if(wrq.u.data.length > 1)
@@ -639,7 +639,7 @@
 	/* Get old essid */
 	memset(essid, '\0', sizeof(essid));
 	wrq.u.essid.pointer = (caddr_t) essid;
-	wrq.u.essid.length = IW_ESSID_MAX_SIZE + 1;
+	wrq.u.essid.length = IW_ESSID_MAX_SIZE + 2;
 	wrq.u.essid.flags = 0;
 	if(iw_get_ext(skfd, ifname, SIOCGIWESSID, &wrq) < 0)
 	  return(IWERR_GET_EXT);
diff -Naur wireless_tools.30.pre6/iwgetid.c wireless_tools.30.pre7/iwgetid.c
--- wireless_tools.30.pre6/iwgetid.c	2007-09-18 00:59:09.000000000 +0200
+++ wireless_tools.30.pre7/iwgetid.c	2008-05-16 02:20:56.000000000 +0300
@@ -84,7 +84,7 @@
 	    int			format)
 {
   struct iwreq		wrq;
-  char			essid[IW_ESSID_MAX_SIZE + 1];	/* ESSID */
+  char			essid[IW_ESSID_MAX_SIZE + 2];	/* ESSID */
   char			pessid[IW_ESSID_MAX_SIZE + 1];	/* Pcmcia format */
   unsigned int		i;
   unsigned int		j;
@@ -94,7 +94,7 @@
 
   /* Get ESSID */
   wrq.u.essid.pointer = (caddr_t) essid;
-  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 1;
+  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 2;
   wrq.u.essid.flags = 0;
   if(iw_get_ext(skfd, ifname, SIOCGIWESSID, &wrq) < 0)
     return(-1);
diff -Naur wireless_tools.30.pre6/iwlib.c wireless_tools.30.pre7/iwlib.c
--- wireless_tools.30.pre6/iwlib.c	2007-12-06 23:44:24.000000000 +0200
+++ wireless_tools.30.pre7/iwlib.c	2008-05-16 02:11:43.000000000 +0300
@@ -716,7 +716,7 @@
 
   /* Get ESSID */
   wrq.u.essid.pointer = (caddr_t) info->essid;
-  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 1;
+  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 2;
   wrq.u.essid.flags = 0;
   if(iw_get_ext(skfd, ifname, SIOCGIWESSID, &wrq) >= 0)
     {
@@ -3131,7 +3131,7 @@
     case SIOCGIWESSID:
       wscan->b.has_essid = 1;
       wscan->b.essid_on = event->u.data.flags;
-      memset(wscan->b.essid, '\0', IW_ESSID_MAX_SIZE+1);
+      memset(wscan->b.essid, '\0', IW_ESSID_MAX_SIZE + 1);
       if((event->u.essid.pointer) && (event->u.essid.length))
 	memcpy(wscan->b.essid, event->u.essid.pointer, event->u.essid.length);
       break;
diff -Naur wireless_tools.30.pre6/iwlib.h wireless_tools.30.pre7/iwlib.h
--- wireless_tools.30.pre6/iwlib.h	2008-01-17 03:20:33.000000000 +0200
+++ wireless_tools.30.pre7/iwlib.h	2008-05-16 02:16:32.000000000 +0300
@@ -184,7 +184,7 @@
   int		key_flags;		/* Various flags */
   int		has_essid;
   int		essid_on;
-  char		essid[IW_ESSID_MAX_SIZE + 1];	/* ESSID (extended network) */
+  char		essid[IW_ESSID_MAX_SIZE + 2];	/* ESSID (extended network) */
   int		has_mode;
   int		mode;			/* Operation mode */
  } wireless_config;
@@ -199,7 +199,7 @@
   int		has_sens;
   iwparam	sens;			/* sensitivity */
   int		has_nickname;
-  char		nickname[IW_ESSID_MAX_SIZE + 1]; /* NickName */
+  char		nickname[IW_ESSID_MAX_SIZE + 2]; /* NickName */
   int		has_ap_addr;
   sockaddr	ap_addr;		/* Access point address */
   int		has_bitrate;