Index: electric-fence-2.1.13/Makefile
===================================================================
--- electric-fence-2.1.13.orig/Makefile	2002-02-20 03:21:44.000000000 +0530
+++ electric-fence-2.1.13/Makefile	2011-07-22 20:20:51.000000000 +0530
@@ -1,10 +1,10 @@
-ASFLAGS= -mips2
-CC= cc
-AR= ar
+#ASFLAGS= -mips2
+#CC= cc
+#AR= ar
 INSTALL= install
 MV= mv
 CHMOD= chmod
-CFLAGS= -g
+CFLAGS += -g
 LIB_INSTALL_DIR= /usr/lib
 MAN_INSTALL_DIR= /usr/man/man3
 
@@ -29,6 +29,8 @@
 OBJECTS= efence.o page.o print.o
 
 all:	libefence.a tstheap eftest
+
+test:
 	@ echo
 	@ echo "Testing Electric Fence."
 	@ echo "After the last test, it should print that the test has PASSED."
Index: electric-fence-2.1.13/page.c
===================================================================
--- electric-fence-2.1.13.orig/page.c	2001-04-11 20:14:04.000000000 +0530
+++ electric-fence-2.1.13/page.c	2011-07-22 20:22:05.000000000 +0530
@@ -7,6 +7,8 @@
 #include <errno.h>
 #include <string.h>
 
+#define _AIX
+
 /*
  * Lots of systems are missing the definition of PROT_NONE.
  */
