--- db-4.7.25.NC/dist/configure	2008-05-16 02:42:41.000000000 +0530
+++ db-4.7.25.NC.new/dist/configure	2016-03-30 10:29:11.297332029 +0530
@@ -33382,7 +33382,7 @@
 
 
 for ac_func in \
-	_fstati64 backtrace backtrace_symbols directio fchmod fclose\
+	_fstati64 directio fchmod fclose\
 	fcntl fdatasync fgetc fgets fopen fwrite getgid\
 	getrusage getuid hstrerror mprotect pstat_getdynamic\
 	pthread_yield sched_yield select setgid setuid sigaction\
