Index: srcBE/Makefile
===================================================================
--- srcBE.orig/Makefile	2011-03-17 22:10:44.000000000 +0530
+++ srcBE/Makefile	2013-02-11 20:22:07.000000000 +0530
@@ -2,37 +2,12 @@
 
 # Written by Zenon Fortuna, zenon@mips.com, zenon@fortuna.org
 
-#
-# Makefile to use the CodeSourcery compiler
-#
-
-BIN=/mips/arch/overflow/codesourcery/mips-linux-gnu/lite/release/4.4-203/Linux/bin
-GCC=$(BIN)/mips-linux-gnu-gcc
-CC=$(GCC)
-CROSS_COMPILE_OPTION=mips-linux-gnu-
-MODDIR=/ext/Linux/modules/2.6.32.15BE/smvp
-KernelVersion=2.6.32.15cmp
-
-# First, test if your compiler is in your PATH
-# and the "build" directory is where specified
-
-COMPILER_VERSION := $(shell $(GCC) --version)
-$(if $(COMPILER_VERSION),, \
-    $(error Cannot find $(GCC) - check your PATH))
-
-BUILD_DIR=$(MODDIR)/lib/modules/$(KernelVersion)/build
-
-BUILD_TEST := $(shell if [ -d $(BUILD_DIR) ]; then echo BUILD_OK; fi)
-$(if $(BUILD_TEST),, \
-    $(error Cannot find $(BUILD_DIR)))
+obj-m += pecostat_noIRQ.o
 
+all: pecostat
 
-obj-m += pecostat_noIRQ.o
+pecostat: pecostat_tool.o
+	$(CC) $(CFLAGS) -o $@ $<
 
-pecostat_noIRQ.ko: Makefile pecostat_noIRQ.c
-	echo "#define PECOSTAT_VERSION \"$(VERSION)\"" > pecostat_version.h
-	make modules -C $(MODDIR)/lib/modules/$(KernelVersion)/build M=$(PWD) \
-CROSS_COMPILE=$(CROSS_COMPILE_OPTION)
-	@/bin/rm -f pecostat.ko
-	@/bin/rm -f pecostat_noIRQ.mod.c pecostat_noIRQ.mod.o pecostat_noIRQ.o
-	@echo Done
+clean:
+	rm -f *.o pecostat *.ko
