Index: srcBE/pecostat_noIRQ.c
===================================================================
--- srcBE.orig/pecostat_noIRQ.c	2011-03-22 05:37:37.000000000 +0530
+++ srcBE/pecostat_noIRQ.c	2013-02-11 20:14:43.000000000 +0530
@@ -45,7 +45,7 @@
 
 #define DEBUG 0
 
-#define APRP 0
+#define APRP 1
 
 #define DRIVER_AUTHOR "Zenon Fortuna <zenon@mips.com,zenon@fortuna.org>"
 #define DRIVER_DESC   "Read the MIPS hardware performance counters"
@@ -109,7 +109,10 @@
 static int VPEmask;
 static int TCVpe1;
 static short Ntc, Npc0, Npc, Pct, Nvp, Nvpe;
-static short Pct_array[MIPS_MAX_VPS], Npc_array[MIPS_MAX_VPS], Nvpe_array[MIPS_MAX_VPS], VPE_ID[MIPS_MAX_VPS], Ntc_array[MIPS_MAX_VPS], Npc_core[MIPS_MAX_VPS];
+static short Pct_array[MIPS_MAX_VPS], Npc_array[MIPS_MAX_VPS], Nvpe_array[MIPS_MAX_VPS], Ntc_array[MIPS_MAX_VPS], Npc_core[MIPS_MAX_VPS];
+#if !APRP
+static short VPE_ID[MIPS_MAX_VPS];
+#endif
 static unsigned char CCRes[MIPS_MAX_VPS];
 static unsigned long Count[MIPS_MAX_VPS];
 static unsigned int Config1[MIPS_MAX_VPS];
