Index: srcBE/pecostat_noIRQ.c
===================================================================
--- srcBE.orig/pecostat_noIRQ.c	2013-07-19 14:29:04.000000000 +0530
+++ srcBE/pecostat_noIRQ.c	2013-07-19 14:31:37.618284000 +0530
@@ -37,6 +37,7 @@
 #include <linux/module.h>
 #include <linux/fs.h>
 #include <linux/miscdevice.h>
+#include <linux/slab.h>
 #include <asm/mipsregs.h>
 #include <asm/mipsmtregs.h>
 #include <asm/cpu.h>
