UGW: use uclibc for external toolchain

diff --git a/toolchain/Config.in b/toolchain/Config.in
--- a/toolchain/Config.in
+++ b/toolchain/Config.in
@@ -63,8 +63,8 @@ menuconfig EXTERNAL_TOOLCHAIN
 		default "aarch64_be-unknown-linux-gnu"  if aarch64_be
 		default "arm-unknown-linux-gnu"      if arm
 		default "armeb-unknown-linux-gnu"    if armeb
-		default "i486-unknown-linux-gnu"     if i386
-		default "mips-unknown-linux-gnu"     if mips
+		default "i486-openwrt-linux-uclibc"     if i386
+		default "mips-openwrt-linux-uclibc"     if mips
 		default "mipsel-unknown-linux-gnu"   if mipsel
 		default "powerpc-unknown-linux-gnu"  if powerpc
 		default "x86_64-unknown-linux-gnu"   if x86_64
@@ -77,8 +77,8 @@ menuconfig EXTERNAL_TOOLCHAIN
 		default "aarch64_be-unknown-linux-gnu"  if aarch64_be
 		default "arm-unknown-linux-gnu-"      if arm
 		default "armeb-unknown-linux-gnu-"    if armeb
-		default "i486-unknown-linux-gnu-"     if i386
-		default "mips-unknown-linux-gnu-"     if mips
+		default "i486-openwrt-linux-uclibc-"     if i386
+		default "mips-openwrt-linux-uclibc-"     if mips
 		default "mipsel-unknown-linux-gnu-"   if mipsel
 		default "powerpc-unknown-linux-gnu-"  if powerpc
 		default "x86_64-unknown-linux-gnu-"   if x86_64
