Index: iproute2-4.0.0/include/linux/if.h
===================================================================
--- iproute2-4.0.0.orig/include/linux/if.h
+++ iproute2-4.0.0/include/linux/if.h
@@ -139,6 +139,9 @@ enum net_device_flags {
 #define IF_PROTO_FR_ETH_PVC 0x200B
 #define IF_PROTO_RAW    0x200C          /* RAW Socket                   */
 
+#define IFF_NOMULTIPATH	0x80000		/* Disable for MPTCP */
+#define IFF_MPMASTER    0x200000	/* Use as a master path for MPTCP */
+
 /* RFC 2863 operational status */
 enum {
 	IF_OPER_UNKNOWN,
Index: iproute2-4.0.0/ip/ipaddress.c
===================================================================
--- iproute2-4.0.0.orig/ip/ipaddress.c
+++ iproute2-4.0.0/ip/ipaddress.c
@@ -105,6 +105,8 @@ static void print_link_flags(FILE *fp, u
 	_PF(BROADCAST);
 	_PF(POINTOPOINT);
 	_PF(MULTICAST);
+	_PF(NOMULTIPATH);
+	_PF(MPMASTER);
 	_PF(NOARP);
 	_PF(ALLMULTI);
 	_PF(PROMISC);
Index: iproute2-4.0.0/ip/iplink.c
===================================================================
--- iproute2-4.0.0.orig/ip/iplink.c
+++ iproute2-4.0.0/ip/iplink.c
@@ -63,6 +63,8 @@ void iplink_usage(void)
 	fprintf(stderr, "	                  [ dynamic { on | off } ]\n");
 	fprintf(stderr, "	                  [ multicast { on | off } ]\n");
 	fprintf(stderr, "	                  [ allmulticast { on | off } ]\n");
+	fprintf(stderr, "	                  [ multipath { on | off } ]\n");
+	fprintf(stderr, "	                  [ multipath { on | off | backup | master | nomaster } ]\n");
 	fprintf(stderr, "	                  [ promisc { on | off } ]\n");
 	fprintf(stderr, "	                  [ trailers { on | off } ]\n");
 	fprintf(stderr, "	                  [ txqueuelen PACKETS ]\n");
@@ -467,6 +469,20 @@ int iplink_parse(int argc, char **argv, 
 				req->i.ifi_flags &= ~IFF_ALLMULTI;
 			} else
 				return on_off("allmulticast", *argv);
+		} else if (strcmp(*argv, "multipath") == 0) {
+			NEXT_ARG();
+			req->i.ifi_change |= IFF_NOMULTIPATH;
+			req->i.ifi_change |= IFF_MPMASTER;
+			if (strcmp(*argv, "on") == 0) {
+				req->i.ifi_flags &= ~IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "off") == 0) {
+				req->i.ifi_flags |= IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "master") == 0) {
+				req->i.ifi_flags |= IFF_MPMASTER;
+			} else if (strcmp(*argv, "nomaster") == 0) {
+				req->i.ifi_flags &= ~IFF_MPMASTER;
+			} else
+				return on_off("multipath", *argv);
 		} else if (strcmp(*argv, "promisc") == 0) {
 			NEXT_ARG();
 			req->i.ifi_change |= IFF_PROMISC;
@@ -1049,6 +1065,20 @@ static int do_set(int argc, char **argv)
 				flags &= ~IFF_ALLMULTI;
 			} else
 				return on_off("allmulticast", *argv);
+		} else if (strcmp(*argv, "multipath") == 0) {
+			NEXT_ARG();
+			mask |= IFF_NOMULTIPATH;
+			mask |= IFF_MPMASTER;
+			if (strcmp(*argv, "on") == 0) {
+				flags &= ~IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "off") == 0) {
+				flags |= IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "master") == 0) {
+				flags |= IFF_MPMASTER;
+			} else if (strcmp(*argv, "nomaster") == 0) {
+				flags &= ~IFF_MPMASTER;
+			} else
+				return on_off("multipath", *argv);
 		} else if (strcmp(*argv, "promisc") == 0) {
 			NEXT_ARG();
 			mask |= IFF_PROMISC;
Index: iproute2-4.0.0/man/man8/ip-link.8.in
===================================================================
--- iproute2-4.0.0.orig/man/man8/ip-link.8.in
+++ iproute2-4.0.0/man/man8/ip-link.8.in
@@ -96,6 +96,8 @@ ip-link \- network device configuration
 .br
 .BR multicast " { " on " | " off " } |"
 .br
+.BR multipath " { " on " | " off " | " master " | " nomaster " } |"
+.br
 .B  txqueuelen
 .IR PACKETS " |"
 .br
@@ -634,6 +636,13 @@ change the
 flag on the device.
 
 .TP
+.BR "multipath on " or " multipath off"
+change the
+.B MULTIPATH
+flag on the device, to enable or disable MPTCP.
+
+.TP
+.TP
 .BR "dynamic on " or " dynamic off"
 change the
 .B DYNAMIC
