# HG changeset patch
# Parent 5d427cb8f1f3b3986ea6df9509b2b9adae02d4f4

diff --git a/tools/Makefile b/tools/Makefile
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -32,7 +32,8 @@ tools-$(CONFIG_TARGET_orion_generic) += 
 tools-$(CONFIG_powerpc) += upx
 tools-$(CONFIG_TARGET_x86) += qemu
 tools-$(CONFIG_TARGET_mxs) += elftosb
-tools-$(CONFIG_TARGET_brcm2708)$(CONFIG_TARGET_sunxi)$(CONFIG_TARGET_mxs) += mtools dosfstools
+tools-$(CONFIG_TARGET_brcm2708)$(CONFIG_TARGET_sunxi)$(CONFIG_TARGET_mxs) += dosfstools
+tools-$(CONFIG_TARGET_brcm2708)$(CONFIG_TARGET_sunxi)$(CONFIG_TARGET_mxs)$(CONFIG_TARGET_x86_puma) += mtools
 tools-$(CONFIG_TARGET_ar71xx) += lzma-old squashfs
 tools-y += lzma squashfs4
 tools-$(BUILD_B43_TOOLS) += b43-tools
diff --git a/tools/mtools/Makefile b/tools/mtools/Makefile
--- a/tools/mtools/Makefile
+++ b/tools/mtools/Makefile
@@ -22,15 +22,17 @@ HOST_CONFIGURE_VARS += \
 	ac_cv_header_iconv_h=no
 
 define Host/Compile
-	$(MAKE) -C $(HOST_BUILD_DIR) mcopy
+	$(MAKE) -C $(HOST_BUILD_DIR) mcopy mmd
 endef
 
 define Host/Install
 	$(INSTALL_BIN) $(HOST_BUILD_DIR)/mcopy $(STAGING_DIR_HOST)/bin/
+	$(INSTALL_BIN) $(HOST_BUILD_DIR)/mmd $(STAGING_DIR_HOST)/bin/
 endef
 
 define Host/Clean
 	rm -f $(STAGING_DIR_HOST)/bin/mcopy
+	rm -f $(STAGING_DIR_HOST)/bin/mmd
 endef
 
 $(eval $(call HostBuild))
