--- a/src/include/atm.h
+++ b/src/include/atm.h
@@ -89,6 +89,10 @@ extern "C" {
 #define MAX_ATM_QOS_LEN 116	/* 5+4+2*(3+3*(7+9)+2)+1 */
 #define MAX_ATM_SAP_LEN	255	/* BHLI(27)+1+3*BLLI(L2=33,L3=41,+1)+2 */
 
+#define ATM_VBR_NRT     ATM_VBR
+#define ATM_VBR_RT      6
+#define ATM_UBR_PLUS    7
+#define ATM_GFR         8
 
 int text2atm(const char *text,struct sockaddr *addr,int length,int flags);
 int atm2text(char *buffer,int length,const struct sockaddr *addr,int flags);
