--- a/ChangeLog
+++ b/ChangeLog
@@ -4,6 +4,15 @@ Version 2.4.1 to 2.5.0 (30-DEC-2007)
 Bug fixes
 ---------
 
+2013-07-15:
+	Feeds fixes and updates for buildsystem update
+
+2009-11-11:
+	002-open_linux_atm.patch added
+
+2009-08-19:
+	Adding support for passing scr and mbs for vbr traffic
+
  - building outside the source tree should now work properly
  - TimeTicks (in ilmid) should be unsigned
  - fix zeppelin's handling of LE_ARP REMOTE_ADDRESS flag
