diff -Naur linux-ftpd-0.17/ftpd/Makefile open_linux_ftpd/ftpd/Makefile
--- linux-ftpd-0.17/ftpd/Makefile	2009-12-08 17:22:09.514777000 +0530
+++ open_linux_ftpd/ftpd/Makefile	2009-12-08 16:54:31.560099000 +0530
@@ -5,7 +5,7 @@
 LDFLAGS+=$(IFX_LDFLAGS)
 
 CFLAGS+=-I../support
-LIBS+=-L../support -lsupport -lIFXAPIs
+LIBS+=-L../support -lsupport 
 
 OBJS=ftpd.o ftpcmd.o logutmp.o logwtmp.o popen.o
 
--- open_linux_ftpd/ftpd/ftpcmd.y	2009-12-08 17:53:33.233980000 +0530
+++ linux-ftpd-0.17/ftpd/ftpcmd.y	2009-12-08 17:51:36.672685000 +0530
@@ -611,7 +611,9 @@
 		{
 			if($2 && $4 != NULL)
 			{
+#ifdef IFX_FTP_UPGRADE
 				upgrade_cmd($4);
+#endif
 			}
 			if($4)
 				free($4);
