diff --git a/modules/luci-base/root/sbin/luci-reload b/modules/luci-base/root/sbin/luci-reload
index cc41da2..ac4b594 100755
--- a/modules/luci-base/root/sbin/luci-reload
+++ b/modules/luci-base/root/sbin/luci-reload
@@ -30,6 +30,7 @@ reload_init() {
 		echo "Reloading $1... "
 		/etc/init.d/$2 reload >/dev/null 2>&1
 		[ -n "$3" -a "$?" != "$3" ] && echo '!!! Failed to reload' $1 '!!!'
+		/sbin/reload_config
 	}
 }
 
