From fbe3559340a2f476810c7a36ae1463b8df2190f6 Mon Sep 17 00:00:00 2001
From: "Brett T. Warden" <brett.t.warden@intel.com>
Date: Wed, 28 Jan 2015 13:59:15 -0800
Subject: [PATCH 001/441] Add support for CGM SATA controller in IDE mode

---
 drivers/ata/ata_piix.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/ata/ata_piix.c
+++ b/drivers/ata/ata_piix.c
@@ -348,6 +348,10 @@ static const struct pci_device_id piix_p
 	{ 0x8086, 0x8c80, PCI_ANY_ID, PCI_ANY_ID, 0, 0, ich8_sata_snb },
 	/* SATA Controller IDE (9 Series) */
 	{ 0x8086, 0x8c81, PCI_ANY_ID, PCI_ANY_ID, 0, 0, ich8_sata_snb },
+	/* SATA Controller IDE (Cougar Mountain).
+	 * Attach iff the controller is in IDE mode. */
+	{ 0x8086, 0x2ba0, PCI_ANY_ID, PCI_ANY_ID,
+	  PCI_CLASS_STORAGE_IDE << 8, 0xffff00, ich8_2port_sata },
 
 	{ }	/* terminate list */
 };
