diff -Npur -x'*.*.rej' -x'*.*.orig' -x'*.patch' b05/net/wireless/core.c b06/net/wireless/core.c
--- b05/net/wireless/core.c	2016-09-27 17:13:04.000000000 +0300
+++ b06/net/wireless/core.c	2016-09-27 17:13:04.000000000 +0300
@@ -740,6 +740,12 @@ void cfg80211_unregister_wdev(struct wir
 }
 EXPORT_SYMBOL(cfg80211_unregister_wdev);
 
+struct workqueue_struct *cfg80211_get_cfg80211_wq(void)
+{
+	return cfg80211_wq;
+}
+EXPORT_SYMBOL(cfg80211_get_cfg80211_wq);
+
 static struct device_type wiphy_type = {
 	.name	= "wlan",
 };
diff -Npur -x'*.*.rej' -x'*.*.orig' -x'*.patch' b05/net/wireless/mlme.c b06/net/wireless/mlme.c
--- b05/net/wireless/mlme.c	2016-09-27 17:13:04.000000000 +0300
+++ b06/net/wireless/mlme.c	2016-09-27 17:13:04.000000000 +0300
@@ -708,7 +708,7 @@ void cfg80211_dfs_channels_update_work(s
 				continue;
 
 			timeout = c->dfs_state_entered +
-				  IEEE80211_DFS_MIN_NOP_TIME_MS;
+				msecs_to_jiffies(IEEE80211_DFS_MIN_NOP_TIME_MS);
 
 			if (time_after_eq(jiffies, timeout)) {
 				c->dfs_state = NL80211_DFS_USABLE;
