diff -Npur -x'*.*.rej' -x'*.*.orig' -x'*.patch' a/include/net/cfg80211.h b00/include/net/cfg80211.h
--- a/include/net/cfg80211.h	2016-09-27 17:13:04.000000000 +0300
+++ b00/include/net/cfg80211.h	2016-09-27 17:13:04.000000000 +0300
@@ -735,6 +735,9 @@ enum station_parameters_apply_mask {
  * @capability: station capability
  * @ext_capab: extended capabilities of the station
  * @ext_capab_len: number of extended capabilities
+ * @resp: last association response frame
+ *	(or NULL for no change)
+ * @resp_len: length of last association response
  */
 struct station_parameters {
 	const u8 *supported_rates;
@@ -754,6 +757,8 @@ struct station_parameters {
 	u16 capability;
 	const u8 *ext_capab;
 	u8 ext_capab_len;
+	u8 *resp;
+	size_t resp_len;
 };
 
 /**
diff -Npur -x'*.*.rej' -x'*.*.orig' -x'*.patch' a/net/wireless/nl80211.c b00/net/wireless/nl80211.c
--- a/net/wireless/nl80211.c	2016-09-27 17:13:04.000000000 +0300
+++ b00/net/wireless/nl80211.c	2016-09-27 17:13:04.000000000 +0300
@@ -4054,6 +4054,11 @@ static int nl80211_new_station(struct sk
 	params.listen_interval =
 		nla_get_u16(info->attrs[NL80211_ATTR_STA_LISTEN_INTERVAL]);
 
+	if (info->attrs[NL80211_ATTR_FRAME]) {
+		params.resp = nla_data(info->attrs[NL80211_ATTR_FRAME]);
+		params.resp_len = nla_len(info->attrs[NL80211_ATTR_FRAME]);
+	}
+
 	if (info->attrs[NL80211_ATTR_PEER_AID])
 		params.aid = nla_get_u16(info->attrs[NL80211_ATTR_PEER_AID]);
 	else
