From b03375aafe39a3da2837a83ec1fec76ecebd6634 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Tue, 10 Jan 2017 20:31:21 +0200
Subject: [PATCH 2/2] mac80211: add get/set net_device_ops from/to wdev

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 include/net/mac80211.h | 22 ++++++++++++++++++++++
 net/mac80211/util.c    | 38 ++++++++++++++++++++++++++++++++++++++
 2 files changed, 60 insertions(+)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 800e0cb8..23bd5d4 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1510,6 +1510,28 @@ struct wireless_dev *ieee80211_vif_to_wdev(struct ieee80211_vif *vif);
 struct ieee80211_vif *net_device_to_ieee80211_vif(struct net_device *dev);
 
 /**
+ * get_net_device_ops_from_wdev - extract the net_device from wdev
+ * @wdev: the wdev to get the net_device for
+ */
+const struct net_device_ops *get_net_device_ops_from_wdev(
+		struct wireless_dev *wdev);
+
+/**
+ * set_net_device_ops_from_wdev - replace the net_device to a wdev
+ * @wdev: the wdev to set the net_device for
+ */
+void set_net_device_ops_in_wdev(struct wireless_dev *wdev,
+		const struct net_device_ops *ops);
+
+/**
+ * copy_net_device_ops_from_wdev - clone a net_device_ops from wdev
+ * @wdev: the wdev to get the net_device_ops for
+ * @ops: the net_device_ops object to clone the ops to
+ */
+void copy_net_device_ops_from_wdev(struct wireless_dev *wdev,
+		struct net_device_ops *ops);
+
+/**
  * ieee80211_vif_to_name - return the vif name
  * @vif: the vif to get the wdev for
  *
diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index 7481bc8c..26ed06d 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -735,6 +735,44 @@ struct ieee80211_vif *net_device_to_ieee80211_vif(struct net_device *dev)
 }
 EXPORT_SYMBOL_GPL(net_device_to_ieee80211_vif);
 
+const struct net_device_ops *get_net_device_ops_from_wdev(
+		struct wireless_dev *wdev)
+{
+	struct ieee80211_sub_if_data *sdata;
+
+	if (!wdev)
+		return NULL;
+	sdata = IEEE80211_WDEV_TO_SUB_IF(wdev);
+	return sdata->dev->netdev_ops;
+}
+EXPORT_SYMBOL_GPL(get_net_device_ops_from_wdev);
+
+void set_net_device_ops_in_wdev(struct wireless_dev *wdev,
+		const struct net_device_ops *ops)
+{
+	struct ieee80211_sub_if_data *sdata;
+
+	if (!wdev || !ops)
+		return;
+	sdata = IEEE80211_WDEV_TO_SUB_IF(wdev);
+	sdata->dev->netdev_ops = ops;
+}
+EXPORT_SYMBOL_GPL(set_net_device_ops_in_wdev);
+
+void copy_net_device_ops_from_wdev(struct wireless_dev *wdev,
+		struct net_device_ops *ops)
+{
+	struct ieee80211_sub_if_data *sdata;
+
+	if (!wdev || !ops)
+		return;
+	sdata = IEEE80211_WDEV_TO_SUB_IF(wdev);
+	if (!sdata->dev->netdev_ops)
+		return;
+	memcpy(ops, sdata->dev->netdev_ops, sizeof(struct net_device_ops));
+}
+EXPORT_SYMBOL_GPL(copy_net_device_ops_from_wdev);
+
 const char *ieee80211_vif_to_name(struct ieee80211_vif *vif)
 {
 	struct ieee80211_sub_if_data *sdata;
-- 
2.8.4.windows.1

