# HG changeset patch
# Parent bfcc30d60482e999cbddb92bec278b2eaa795a46

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4593,4 +4593,13 @@ void ieee80211_report_wowlan_wakeup(stru
 				    struct cfg80211_wowlan_wakeup *wakeup,
 				    gfp_t gfp);
 
+/**
+ * ieee80211_drv_start_sw_scan - allow driver to fallback to sw scan from hw
+ * @hw: pointer as obtained from ieee80211_alloc_hw()
+ *
+ * Note: This function must be called from hw_scan callback only.
+ *
+ */
+void ieee80211_drv_start_sw_scan(struct ieee80211_hw *hw);
+
 #endif /* MAC80211_H */
diff --git a/net/mac80211/scan.c b/net/mac80211/scan.c
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -384,6 +384,17 @@ static int ieee80211_start_sw_scan(struc
 	return 0;
 }
 
+void ieee80211_drv_start_sw_scan(struct ieee80211_hw *hw)
+{
+	struct ieee80211_local *local = hw_to_local(hw);
+	test_and_clear_bit(SCAN_HW_SCANNING, &local->scanning);
+	__set_bit(SCAN_SW_SCANNING, &local->scanning);
+	/* TODO: check remain on channel (scan for current channel requested) */
+	
+	ieee80211_start_sw_scan(local);
+}
+EXPORT_SYMBOL(ieee80211_drv_start_sw_scan);
+
 static bool ieee80211_can_scan(struct ieee80211_local *local,
 			       struct ieee80211_sub_if_data *sdata)
 {
