From b5b05d91fe7c933ce4b29a8e4b634c291b7f470b Mon Sep 17 00:00:00 2001
From: Emmanuel Jillela <emmanuel.jillela@intel.com>
Date: Fri, 15 May 2015 13:46:39 -0700
Subject: [PATCH 081/441] corrected a cougarmountain pin name.

---
 drivers/gpio/gpio-intelce2700.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/gpio/gpio-intelce2700.c
+++ b/drivers/gpio/gpio-intelce2700.c
@@ -284,7 +284,7 @@ enum {
    MMC3_CLK_1P8_3P3                ,    // 0x4C08
    MMC3_D3_I2C2_SDA_1P8_3P3        ,    // 0x4C10
    MMC3_D2_I2C2_SCL_1P8_3P3        ,    // 0x4C18
-   MMC3_CMD_SIG1_9_1P8_3P3         ,    // 0x4C20
+   MMC3_CMD_1P8_3P3                ,    // 0x4C20
    MMC3_D0_1P8_3P3                 ,    // 0x4C28
    JTAG_TRST_B_1P8                 ,    // 0x5000
    CX_PREQ_B_1P8                   ,    // 0x5008
@@ -584,7 +584,7 @@ static struct gpio_pad_info southwest_pa
    [MMC3_CLK_1P8_3P3            ] = { 0x4C08, -1},
    [MMC3_D3_I2C2_SDA_1P8_3P3    ] = { 0x4C10, -1},
    [MMC3_D2_I2C2_SCL_1P8_3P3    ] = { 0x4C18, -1},
-   [MMC3_CMD_SIG1_9_1P8_3P3     ] = { 0x4C20, -1},
+   [MMC3_CMD_1P8_3P3            ] = { 0x4C20, -1},
    [MMC3_D0_1P8_3P3             ] = { 0x4C28, -1},
    [JTAG_TRST_B_1P8             ] = { 0x5000, -1},
    [CX_PREQ_B_1P8               ] = { 0x5008, -1},
