From 7a9b13c23270e40a7bdd0012a443bc03aa8b201f Mon Sep 17 00:00:00 2001
From: Shweta <shweta.s.kulkarni@intel.com>
Date: Tue, 16 Jun 2015 10:41:41 -0700
Subject: [PATCH 133/441] Update pci device table for heci to include CSE

---
 drivers/misc/heci/pci-ish.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/misc/heci/pci-ish.c
+++ b/drivers/misc/heci/pci-ish.c
@@ -62,7 +62,7 @@ MODULE_PARM_DESC(nomsi, "don't use msi (
 static struct pci_dev *heci_pci_device;
 
 static DEFINE_PCI_DEVICE_TABLE(ish_pci_tbl) = {
-	{PCI_DEVICE(PCI_VENDOR_ID_INTEL, 0x22D8)},
+	{PCI_DEVICE(PCI_VENDOR_ID_INTEL, 0x2B98)},
 	{0, }
 };
 
