From 19d7dffa4f384ada738203d9a0c3a915f7752f97 Mon Sep 17 00:00:00 2001
From: Shweta <shweta.s.kulkarni@intel.com>
Date: Tue, 2 Jun 2015 13:09:12 -0700
Subject: [PATCH 134/441] Added ISS, SENSCOL dependencies for HID

---
 drivers/hid/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/hid/Kconfig
+++ b/drivers/hid/Kconfig
@@ -779,7 +779,7 @@ config HID_ZYDACRON
 
 config HID_SENSOR_HUB
 	tristate "HID Sensors framework support"
-	depends on HID
+	depends on HID && ISS_PATH_SENSCOL
 	select MFD_CORE
 	default n
 	---help---
