From 374d12b580593416612ff1c327fd8b336a026a19 Mon Sep 17 00:00:00 2001
From: Emmanuel Jillela <emmanuel.jillela@intel.com>
Date: Thu, 16 Jul 2015 16:21:40 -0700
Subject: [PATCH 140/441] deleted debug prints from HW Mutex lock api.

---
 kernel/hwmutex.c | 2 --
 1 file changed, 2 deletions(-)

--- a/kernel/hwmutex.c
+++ b/kernel/hwmutex.c
@@ -209,8 +209,6 @@ void hw_mutex_lock(uint8_t mutex)
 	hw_mutex_set_owner(hmutex);
 	if (hmutex_ops->lock(hmutex,0)) {
 		spin_unlock_irqrestore(&hmutex->irq_lock,flags);
-                printk("HW Mutex Lock successful\n");	
-                dump_hw_mutex_registers(hmutex);
 		return ;
 	}
         dump_hw_mutex_registers(hmutex);
