From 78c9b2bf53f19103ce4eac89b01364d0e7f55dc4 Mon Sep 17 00:00:00 2001
From: chandrap <prakash1.chandra@intel.com>
Date: Thu, 20 Aug 2015 17:24:52 -0700
Subject: [PATCH 148/441] fix read and write from ATOM_INTC registers

---
 arch/x86/NetIP_SubSystem/avalanche_intc.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

--- a/arch/x86/NetIP_SubSystem/avalanche_intc.c
+++ b/arch/x86/NetIP_SubSystem/avalanche_intc.c
@@ -124,7 +124,7 @@ int avalanche_intc_clear_status( unsigne
         /* Requested interrupt is enabled */
         status_reg =  (status_reg | (1<<irq_bit));
         /* write value back to controller register */
-        __raw_writel( avalanche_hw0_icregs->icestar[irq_index], cpu_to_be32(status_reg));
+        __raw_writel( cpu_to_be32(status_reg), avalanche_hw0_icregs->icestar[irq_index]);
     }
 
     /* else need not to do anything */
@@ -163,11 +163,10 @@ int avalanche_intc_init(void)
     avalanche_hw0_icregs =   (struct avalanche_ictrl_regs *)AVALANCHE_INTC_BASE;
     
     /* read ATOM_INTC revision register */
-    cntrl_ver = *(volatile unsigned int*)(AVALANCHE_INTC_BASE);
     /**ARM11 big endian to CPU  read register and call the corresponding value */
-    cntrl_ver = be32_to_cpu(__raw_readl(cntrl_ver));
+    cntrl_ver = be32_to_cpu(__raw_readl(avalanche_hw0_icregs->icrevr));
     printk("ATOM_INTC Interrupt controller revision : %x\n", cntrl_ver);
-
+    return 0;
 }
 
 void avalanche_intc_cleanup(void)
