From 60f1f35468add211c74e885642aefe87cc497f22 Mon Sep 17 00:00:00 2001
From: "Brett T. Warden" <brett.t.warden@intel.com>
Date: Mon, 28 Sep 2015 11:44:36 -0700
Subject: [PATCH 170/441] Add CONFIG_GMIN_INTEL_MID

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>
---
 arch/x86/Kconfig | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -573,6 +573,13 @@ config X86_MDFLD
 
 endif
 
+config GMIN_INTEL_MID
+       bool "G-Min minimal INTEL_MID support"
+       ---help---
+         Select Y to build a kernel capable of booting on G-Min
+         enabled devices without the full INTEL_MID/SFI firmware
+         stack.
+
 config X86_INTEL_LPSS
 	bool "Intel Low Power Subsystem Support"
 	depends on ACPI
