From d1f3f5e124c138aa07d3acfdbe0ce6ea6dbc4de9 Mon Sep 17 00:00:00 2001
From: Alan Olsen <alanx.olsen@intel.com>
Date: Fri, 20 Nov 2015 11:11:23 -0800
Subject: [PATCH 239/441] Added two lines truncated by hand edited patch

---
 drivers/mtd/spi-nor/spi-nor.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -1153,3 +1153,5 @@ static const struct spi_device_id *spi_n
 
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Huang Shijie <shijie8@gmail.com>");
+MODULE_AUTHOR("Mike Lavender");
+MODULE_DESCRIPTION("framework for SPI NOR");
