From 1452f4af23c30c43317a2b38a5b84c39b84ef61c Mon Sep 17 00:00:00 2001
From: Wade T Cline <wade.t.cline@intel.com>
Date: Thu, 7 Jan 2016 16:45:46 -0800
Subject: [PATCH 251/441] Rename CGM to Puma7.

---
 drivers/usb/host/xhci-pci.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -43,7 +43,7 @@
 #define PCI_DEVICE_ID_INTEL_LYNXPOINT_XHCI	0x8c31
 #define PCI_DEVICE_ID_INTEL_LYNXPOINT_LP_XHCI	0x9c31
 #define PCI_DEVICE_ID_INTEL_CHERRYVIEW_XHCI		0x22b5
-(??)
+#define PCI_DEVICE_ID_INTEL_PUMA7_XHCI			0x2bb5
 #define PCI_DEVICE_ID_INTEL_SUNRISEPOINT_H_XHCI		0xa12f
 #define PCI_DEVICE_ID_INTEL_SUNRISEPOINT_LP_XHCI	0x9d2f
 
@@ -184,7 +184,7 @@ static void xhci_pme_quirk(struct usb_hc
 
 	if (pdev->vendor == PCI_VENDOR_ID_INTEL &&
 		(pdev->device == PCI_DEVICE_ID_INTEL_CHERRYVIEW_XHCI ||
-		 pdev->device == PCI_DEVICE_ID_INTEL_COUGARPARK_XHCI)) {
+		 pdev->device == PCI_DEVICE_ID_INTEL_PUMA7_XHCI)) {
 
 		reg = (void __iomem *) xhci->cap_regs + PORT2_SSIC_CONFIG_REG2;
 
