# HG changeset patch
# Parent 468d2d88e9b2cc21289ee481b1f05b7b9c0fcd3a

--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -181,6 +181,8 @@ static DEFINE_HASHTABLE(napi_hash, 8);
 
 static seqcount_t devnet_rename_seq;
 
+struct sk_buff *handle_ing(struct sk_buff *skb, struct packet_type **pt_prev,
+					 int *ret, struct net_device *orig_dev);
 static inline void dev_base_seq_inc(struct net *net)
 {
 	while (++net->dev_base_seq == 0);
@@ -3457,7 +3459,17 @@ static int ing_filter(struct sk_buff *sk
 	return result;
 }
 
-static inline struct sk_buff *handle_ing(struct sk_buff *skb,
+int check_ingress(struct sk_buff *skb)
+{
+	struct netdev_queue *rxq = rcu_dereference(skb->dev->ingress_queue);
+
+	if (!rxq || rxq->qdisc == &noop_qdisc)
+		return 1;
+	return 0;
+}
+EXPORT_SYMBOL(check_ingress);
+
+struct sk_buff *handle_ing(struct sk_buff *skb,
 					 struct packet_type **pt_prev,
 					 int *ret, struct net_device *orig_dev)
 {
@@ -3482,6 +3494,7 @@ out:
 	skb->tc_verd = 0;
 	return skb;
 }
+EXPORT_SYMBOL(handle_ing);
 #endif
 
 /**
