# HG changeset patch
# Parent a5e5e0e81eae34062bbed8f7c7a03d3a5804ac9d

--- a/arch/x86/NetIP_SubSystem/netip_subsystem_lld.c
+++ b/arch/x86/NetIP_SubSystem/netip_subsystem_lld.c
@@ -79,14 +79,14 @@ void netss_stop_handshake_thread(void);
 
 struct net_subsystem_drv_data *pnetss_drv_data;
 
-static unsigned int netss_get_bit_position (unsigned int irq)
+inline unsigned char netss_get_bit_position (unsigned int irq)
 {
-   return irq % NUM_ARM11_INTR_PER_REG;
+   return ((unsigned char)(irq % NUM_ARM11_INTR_PER_REG));
 }
 
-static unsigned int netss_get_reg_index(unsigned int irq)
+inline unsigned char netss_get_reg_index(unsigned int irq)
 {
-   return irq / NUM_ARM11_INTR_PER_REG;
+   return ((unsigned char)(irq / NUM_ARM11_INTR_PER_REG));
 }
 
 void netss_interrupt_register(netss_interrupt_t intrpt, int arm11_int_id, netss_interrupt_info_t *irq_info)
