# HG changeset patch
# Parent af14f8486e92d5304e14aeded036f53abe277f88

diff --git a/net/8021q/vlan_dev.c b/net/8021q/vlan_dev.c
--- a/net/8021q/vlan_dev.c
+++ b/net/8021q/vlan_dev.c
@@ -77,9 +77,7 @@ vlan_dev_get_egress_qos_mask(struct net_
 {
 #if defined(CONFIG_MACH_PUMA6)
 	return (VLAN_PRIO_MASK & (((u16)(skb->ti_meta_info)) << VLAN_PRIO_SHIFT));
-#elif defined(CONFIG_MACH_PUMA7)
-	return (VLAN_PRIO_MASK & (((u16)(skb->pp_packet_info->pp_session.priority)) << VLAN_PRIO_SHIFT));
-#elif defined(CONFIG_MACH_PUMA5)
+#elif defined(CONFIG_MACH_PUMA5) || defined(CONFIG_MACH_PUMA7)
 
 	struct vlan_priority_tci_mapping *mp;
 
@@ -95,6 +93,8 @@ vlan_dev_get_egress_qos_mask(struct net_
 		mp = mp->next;
 	}
 	return 0;
+#elif defined(CONFIG_MACH_PUMA7)
+	return (VLAN_PRIO_MASK & (((u16)(skb->pp_packet_info->pp_session.priority)) << VLAN_PRIO_SHIFT));
 #else
 	#error Update required for target Puma SoC
 #endif
