From dc51fbbd294a25d44cc96d487644029bb7a57ee1 Mon Sep 17 00:00:00 2001
From: Lemberger Tzviel <tlemberg@blrrsa0024.blr.lantiq.com>
Date: Mon, 4 Jan 2016 20:22:43 +0530
Subject: [PATCH] change isolate mode value

Signed-off-by: Lemberger Tzviel <tlemberg@blrrsa0024.blr.lantiq.com>
---
 net/bridge/br_private.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -181,7 +181,7 @@ struct net_bridge_port
 #define BR_ADMIN_COST		0x00000010
 #define BR_LEARNING		0x00000020
 #define BR_FLOOD		0x00000040
-#define BR_ISOLATE_MODE	0x00000020
+#define BR_ISOLATE_MODE	0x00000080
 
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 	struct bridge_mcast_query	ip4_query;
