--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -287,6 +287,11 @@ rx_handler_result_t br_handle_frame(stru
 forward:
 	switch (p->state) {
 	case BR_STATE_DISABLED:
+		/* For the ethernet interface this should not happened.
+		Because at this point, it is not clear that the bridge port is able to handle the packet or not. */
+		if (p->dev && memcmp(p->dev->name, "eth", 3) == 0) {
+			goto drop;
+		}
 		if (ether_addr_equal(p->br->dev->dev_addr, dest))
 			skb->pkt_type = PACKET_HOST;
 
