# HG changeset patch
# Parent 38101b1aac72b6aeabfadb061ba6c99624dc9c70

--- a/net/ti/pp/ti_hil_core.c
+++ b/net/ti/pp/ti_hil_core.c
@@ -358,14 +358,16 @@ int ti_hil_is_gbe_device(struct net_devi
  **************************************************************************/
 void ti_hil_clone_netdev_pp_info(struct net_device *newDev, struct net_device *origDev)
 {
+    printk("HIL Core: ti_hil_clone_netdev_pp_info: newDv is %s, origDev is %s \n", newDev->name, origDev->name);
     newDev->pid_handle              = origDev->pid_handle;
     memcpy((void *)&newDev->vpid_block, (void *)&origDev->vpid_block, sizeof(origDev->vpid_block));
+    strncpy(newDev->vpid_block.devName, newDev->name, VPID_IF_NAME_SIZE);
     newDev->qos_setup_hook          = origDev->qos_setup_hook;
     newDev->qos_shutdown_hook       = origDev->qos_shutdown_hook;
     newDev->qos_select_hook         = origDev->qos_select_hook;
+    newDev->qos_get_params_hook       = origDev->qos_get_params_hook;
     newDev->qos_virtual_scheme_idx  = origDev->qos_virtual_scheme_idx;
     newDev->devInstance             = origDev->devInstance;
-
 #if PUMA7_OR_NEWER_SOC_TYPE
 #ifdef CONFIG_ARM_AVALANCHE_SOC
     if (ti_hil_is_gbe_device(origDev))
@@ -374,6 +376,7 @@ void ti_hil_clone_netdev_pp_info(struct
         if (origDev->netdev_copy_priv_hook)
         {
             origDev->netdev_copy_priv_hook(newDev, origDev);
+	    newDev->netdev_copy_priv_hook = origDev->netdev_copy_priv_hook;
         }
     }
 #endif
