# HG changeset patch
# Parent c975096ca3b6cbdf8b64faf99d88e799ad295b46

--- a/include/linux/avalanche/generic/avalanche_pp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pp_api.h
@@ -297,10 +297,10 @@ typedef struct // former TI_PP_QOS_QUEUE
     Uint8               flags;                  /* Control how packets in the queue should be handled. Available options: AVALANCHE_PP_QOS_Q_REALTIME - Disable scaling of the credit. */
     Uint16              egr_q;                  /* Queue manager and queue index of forwarding queue */
 
-    Uint32              it_credit_bytes;        /* The amount of forwarding byte ìcreditî that the queue receives every 25us. */
-    Uint16              it_credit_packets;      /* The amount of forwarding packets ìcreditî that the queue receives every 25us. */
-    Uint32              max_credit_bytes;       /* The maximum amount of forwarding byte ìcreditî that the queue is allowed to hold at the end of the 25us iteration. */
-    Uint16              max_credit_packets;     /* The maximum amount of forwarding byte ìcreditî that the queue is allowed to hold at the end of the 25us iteration. */
+    Uint32              it_credit_bytes;        /* The amount of forwarding byte ‚Äúcredit‚Äù that the queue receives every 25us. */
+    Uint16              it_credit_packets;      /* The amount of forwarding packets ‚Äúcredit‚Äù that the queue receives every 25us. */
+    Uint32              max_credit_bytes;       /* The maximum amount of forwarding byte ‚Äúcredit‚Äù that the queue is allowed to hold at the end of the 25us iteration. */
+    Uint16              max_credit_packets;     /* The maximum amount of forwarding byte ‚Äúcredit‚Äù that the queue is allowed to hold at the end of the 25us iteration. */
     Uint32              congst_thrsh_bytes;     /* The size in bytes at which point the QOS queue is considered to be congested. */
     Uint16              congst_thrsh_packets;   /* The maximum number of packets to be kept in QOS queue */
 }
--- a/include/linux/avalanche/puma7/puma7_npcpu_appcpu_defs.h
+++ b/include/linux/avalanche/puma7/puma7_npcpu_appcpu_defs.h
@@ -77,16 +77,16 @@ typedef enum npcpu_appcpu_hw_mbox_tags
     NPCPU_APPCPU_HW_MBOX_TAG_INVALID = -1,
     NPCPU_APPCPU_HW_MBOX_TAG_BOOT,
     NPCPU_APPCPU_HW_MBOX_TAG_PM,
-    NPCPU_APPCPU_HW_MBOX_TAG_MRPC,
+    NPCPU_APPCPU_HW_MBOX_TAG_MRPC,
     NPCPU_APPCPU_HW_MBOX_TAG_RPC,
-    NPCPU_APPCPU_HW_MBOX_TAG_ITSTORE,
-    NPCPU_APPCPU_HW_MBOX_TAG_LAN,
-    NPCPU_APPCPU_HW_MBOX_TAG_MOCA,
-    NPCPU_APPCPU_HW_MBOX_TAG_NONE,  /* reserved */
+    NPCPU_APPCPU_HW_MBOX_TAG_ITSTORE,
+    NPCPU_APPCPU_HW_MBOX_TAG_LAN,
+    NPCPU_APPCPU_HW_MBOX_TAG_MOCA,
+    NPCPU_APPCPU_HW_MBOX_TAG_NONE,  /* reserved */
     NPCPU_APPCPU_HW_MBOX_TAG_CPPI41_MBX,
     NPCPU_APPCPU_HW_MBOX_TAG_DATAPIPE_MBX,
-    NPCPU_APPCPU_HW_MBOX_TAG_WIFI_MBX,
-    NPCPU_APPCPU_HW_MBOX_TAG_DPI_MBX,
+    NPCPU_APPCPU_HW_MBOX_TAG_WIFI_MBX,
+    NPCPU_APPCPU_HW_MBOX_TAG_DPI_MBX,
     NPCPU_APPCPU_HW_LAST_ENTRY
 } npcpu_appcpu_hw_mbox_tags_e;
 
