# HG changeset patch
# Parent c9508985e7d067f056a799f9116fb75a3feea533

--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -407,6 +407,9 @@ static int macvlan_stop(struct net_devic
 	if (dev->flags & IFF_ALLMULTI)
 		dev_set_allmulti(lowerdev, -1);
 
+#if defined(CONFIG_TI_HIL_PROFILE_INTRUSIVE_P7)
+	ti_hil_pp_event (TI_PP_REMOVE_VPID, dev);
+#endif
 	dev_uc_del(lowerdev, dev->dev_addr);
 
 hash_del:
