# HG changeset patch
# Parent 6794c971c422f4250df465ec768fc0724edb1bf3

--- a/include/linux/avalanche/generic/avalanche_pdsp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pdsp_api.h
@@ -81,6 +81,18 @@
 
 #if PUMA7_OR_NEWER_SOC_TYPE
     #define AVALANCHE_PDSP_H_PUMA7
+
+#define PRECMD_INDEX_SHIFT        16
+#define PRECMD_OPTION_SHIFT       8
+#define PRECMD_COMMAND_SHIFT      0
+#define PRECMD_INDEX_MASK         (0xFFFFu << PRECMD_INDEX_SHIFT)
+
+#define PRECMD_COMMAND_MASK       (0xFFu << PRECMD_COMMAND_SHIFT)
+#define PRECMD_COMMAND(x)         (((x) << PRECMD_COMMAND_SHIFT) & PRECMD_COMMAND_MASK)
+#define PRECMD_INDEX(x)           (((x) << PRECMD_INDEX_SHIFT)   & PRECMD_INDEX_MASK)
+#define PRECMD_OPTION_MASK        (0xFFu << PRECMD_OPTION_SHIFT)
+#define PRECMD_OPTION(x)          (((x) << PRECMD_OPTION_SHIFT)  & PRECMD_OPTION_MASK)
+
 #if defined(CONFIG_MACH_PUMA7_FPGA_PP) || defined (CONFIG_MACH_PUMA7_BOARD)
     //#define FPGA_SANITY_UNITEST     0
 #endif
@@ -103,7 +115,7 @@ typedef enum
     // Global PDSP  commands
     PDSP_ENABLE                     = 0x80,
     PDSP_GET_STATUS                 = 0x81,
-    PDSP_PSM                        = 0x82,
+    PDSP_DISABLE                    = 0x82,
     PDSP_SANITY_CHECK_EXECUTE       = 0x90,
 
     // Sequencer    commands
--- a/include/linux/avalanche/puma7/puma7_pp.h
+++ b/include/linux/avalanche/puma7/puma7_pp.h
@@ -8,7 +8,7 @@
 
   GPL LICENSE SUMMARY
 
-  Copyright(c) 2014 Intel Corporation.
+  Copyright(c) 2014 - 2016 Intel Corporation.
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of version 2 of the GNU General Public License as
@@ -33,7 +33,7 @@
 
   BSD LICENSE
 
-  Copyright(c) 2014 Intel Corporation. All rights reserved.
+  Copyright(c) 2014 - 2016 Intel Corporation. All rights reserved.
 
   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
@@ -80,7 +80,7 @@
 #ifndef _INCLUDE_PUMA7_PP_H
 #define _INCLUDE_PUMA7_PP_H
 
-
+#define PP_DEFAULT_MTU_SIZE    (1500)
 #define MAX_IP_PACKET_SIZE      1514
 #define MIN_IP_PACKET_SIZE      64
 
@@ -266,11 +266,11 @@ typedef enum PAL_CPPI41_PP_QOS_CLUSTERS
 #define AVALANCHE_PP_MODIFIER_3_PARAM_RGN_BASE          (IO_ADDRESS(0xF3120004))
 #define AVALANCHE_PP_MODIFIER_MTU_TABLE_BASE            (IO_ADDRESS(0xF3120100))    // Modifier PDSP MTU table
 
-#define AVALANCHE_PP_MODIFIER_MTU_TABLE_UPDATE(qNum, mtuSize)               \
-{                                                                           \
-    volatile Uint16 *mtu_per_qos_q = 0;                                     \
-    mtu_per_qos_q = (Uint16 *)AVALANCHE_PP_MODIFIER_MTU_TABLE_BASE + (qNum);\
-    *mtu_per_qos_q = cpu_to_be16((mtuSize) + ETH_HLEN);                                  \
+#define AVALANCHE_PP_MODIFIER_MTU_TABLE_UPDATE(vpidId, mtuSize)              \
+{                                                                            \
+    volatile Uint16 *mtu_per_vpid = 0;                                       \
+    mtu_per_vpid = (Uint16 *)AVALANCHE_PP_MODIFIER_MTU_TABLE_BASE + (vpidId);\
+    *mtu_per_vpid = cpu_to_be16(mtuSize);                                    \
 }
 
 #define AVALANCHE_PP_DS_RESEQ_CMD_RGN_BASE         (IO_ADDRESS(0xF3ec0000))
@@ -478,11 +478,11 @@ typedef enum PAL_CPPI41_PP_QOS_CLUSTERS
 #define AVALANCHE_PP_MC_PARAM_RGN_BASE                  (IO_ADDRESS(0xF3F30004))
 #define AVALNACHE_PP_MC_MTU_TABLE_BASE                  (IO_ADDRESS(0xF3F30100))    // MC PDSP MTU table
 
-#define AVALANCHE_PP_MC_MTU_TABLE_UPDATE(qNum, mtuSize)                 \
-{                                                                       \
-    volatile Uint16 *mtu_per_qos_q = 0;                                 \
-    mtu_per_qos_q = (Uint16 *)AVALNACHE_PP_MC_MTU_TABLE_BASE + (qNum);  \
-    *mtu_per_qos_q = cpu_to_be16((mtuSize) + ETH_HLEN);                 \
+#define AVALANCHE_PP_MC_MTU_TABLE_UPDATE(vpidId, mtuSize)              \
+{                                                                      \
+    volatile Uint16 *mtu_per_vpid = 0;                                 \
+    mtu_per_vpid = (Uint16 *)AVALNACHE_PP_MC_MTU_TABLE_BASE + (vpidId);\
+    *mtu_per_vpid = cpu_to_be16(mtuSize);                              \
 }
 
 #define AVALANCHE_PP_DPI_CMD_RGN_BASE                   (IO_ADDRESS(0xF3F40000))
@@ -517,9 +517,9 @@ typedef enum PAL_CPPI41_PP_QOS_CLUSTERS
 #define FREE_RUNNING_COUNTER_H_GET()                    be32_to_cpu(*(volatile unsigned int *)AVALANCHE_PP_STATISTICAL_FRC_S2_H_BASE)
 
 /* PP MTU size */
-#define UPDATE_MTU_TABLE_IN_PDSP_DMEM(qNum, mtuSize)        \
-    AVALANCHE_PP_MODIFIER_MTU_TABLE_UPDATE(qNum, mtuSize);  \
-    AVALANCHE_PP_MC_MTU_TABLE_UPDATE(qNum, mtuSize);
+#define UPDATE_MTU_TABLE_IN_PDSP_DMEM(vpidId, mtuSize)        \
+    AVALANCHE_PP_MODIFIER_MTU_TABLE_UPDATE(vpidId, mtuSize);  \
+    AVALANCHE_PP_MC_MTU_TABLE_UPDATE(vpidId, mtuSize);
 
 /* PP PDSPs Offsets in mailbox to get the PDSP and PP versions */
 #define AVALANCHE_PP_PDSPs_PP_VERSION_OFFSET                0xF8
