# HG changeset patch
# Parent dfcefe8592312f85a09af1f35a0ebcb7cdff6664

--- a/include/linux/avalanche/generic/avalanche_pp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pp_api.h
@@ -1548,7 +1548,11 @@ extern AVALANCHE_PP_RET_e    avalanche_p
 
 /* Session Management API */
 extern AVALANCHE_PP_RET_e    avalanche_pp_session_create        ( AVALANCHE_PP_SESSION_INFO_t *  ptr_session, void * pkt_ptr );
+#if PUMA7_OR_NEWER_SOC_TYPE
+extern AVALANCHE_PP_RET_e    avalanche_pp_session_delete        ( Uint32    session_handle,     AVALANCHE_PP_SESSION_STATS_t *  ptr_session_stats, Bool deleteFromFW );
+#else
 extern AVALANCHE_PP_RET_e    avalanche_pp_session_delete        ( Uint32    session_handle,     AVALANCHE_PP_SESSION_STATS_t *  ptr_session_stats );
+#endif
 extern AVALANCHE_PP_RET_e    avalanche_pp_session_get_list      ( Uint8     vpid_handle,        PP_LIST_ID_e   list_id, Uint32 * num_entries, Uint32 * session_handle_list );
 extern AVALANCHE_PP_RET_e    avalanche_pp_session_get_info      ( Uint32    session_handle,     AVALANCHE_PP_SESSION_INFO_t**  ptr_session_info );
 extern AVALANCHE_PP_RET_e    avalanche_pp_session_set_timeout   (Uint8 timeoutSeconds);
--- a/include/linux/avalanche/generic/pp_qos_p7.h
+++ b/include/linux/avalanche/generic/pp_qos_p7.h
@@ -406,6 +406,16 @@ PP_QOS_MGR_RET_e pp_qos_get_cluster_by_i
  *  Returns PP_QOS_RC_SUCCESS if pass.
  */
 PP_QOS_MGR_RET_e pp_qos_get_cluster_stats(Uint8 phy_cluster_id, PP_QOS_MIB_STATS_t *cluster_stats);
+/**
+ *  pp_qos_get_cluster_by_queue_id - Get the cluster
+ *  configuration by a given queue id
+ *  @phy_queue_id: Queue ID
+ *  @phy_cluster_id: Pointer to return the cluster id
+ *  @clst_cfg: Pointer to return the cluster
+ *
+ *  Returns PP_QOS_RC_SUCCESS if pass.
+ */
+PP_QOS_MGR_RET_e pp_qos_get_cluster_by_queue_id(Uint8 phy_queue_id, Uint8 *phy_cluster_id, AVALANCHE_PP_QOS_CLST_CFG_t *clst_cfg);
 /**************/
 /* Queue      */
 /**************/
--- a/include/linux/avalanche/puma7/puma7_cppi_prv.h
+++ b/include/linux/avalanche/puma7/puma7_cppi_prv.h
@@ -1920,7 +1920,7 @@ typedef enum PAL_CPPI_PP_SOURCE_PORTS
 /* tx fw - rx channels */
 /* convert wifi device id to tx fw infra dma rx channel */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_RX_CH(dev_id)                           \
-    ((dev_id) + PAL_CPPI_PP_DMA09_INFRA_TX_CH_0)
+    ((dev_id) + PAL_CPPI_PP_DMA08_WIFI_DEVICE0_TX_FW_INFRA_RX_CH)
 
 /* convert wifi device id to tx fw infra dma rx channel output queue */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_RX_CH_OUT_Q(dev_id)                     \
@@ -1941,11 +1941,11 @@ typedef enum PAL_CPPI_PP_SOURCE_PORTS
 /* tx fw - tx channels */
 /* convert wifi device id to tx fw infra dma tx channel */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_TX_CH(dev_id)                           \
-    ((dev_id) + PAL_CPPI_PP_DMA09_INFRA_RX_CH_0)
+    ((dev_id) + PAL_CPPI_PP_DMA08_WIFI_DEVICE0_TX_FW_INFRA_TX_CH)
 
 /* convert wifi device id to tx fw infra dma tx channel input queue */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_TX_CH_INPUT_Q(dev_id)                   \
-    (((dev_id)*2) + PAL_CPPI_PP_QMGR_G2_CDMA9_INFRA_INPUT_LOW_TX_CH_0_HI_Q_NUM)
+    (((dev_id)*2) + PAL_CPPI_PP_QMGR_G2_CDMA8_INFRA_INPUT_WIFI_DEVICE0_TX_FW_HI_Q_NUM)
 
 /* tx fw infra dma tx channel input queue manager */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_TX_CH_INPUT_Q_MGR                       \
