# HG changeset patch
# Parent ae9eb8b03de6e8e6f2e26cbf73d516006cdaa31e

--- a/drivers/mrpc/examples/server_example.c
+++ b/drivers/mrpc/examples/server_example.c
@@ -74,6 +74,22 @@ static struct servers_database db;
 static unsigned int num_servers = 1;
 static unsigned int delay = 0;
 
+struct server {
+    struct list_head list;
+    char name[MAX_NAME_SIZE];
+    struct mrpc_server *mrpc;
+};
+
+struct servers_database {
+    struct list_head servers;
+    struct mutex lock;
+    struct kobject *kobj;
+};
+
+static struct servers_database db;
+static unsigned int num_servers = 1;
+static unsigned int delay = 0;
+
 /**
  * demo_proc_0 - server side implementation of proc id 0
  * 
@@ -295,6 +311,7 @@ static int __init server_example_init(vo
         goto out_kobject_put;
     }
 
+    ssleep(delay);
     return 0;
 
 out_kobject_put:
--- a/include/linux/mrpc.h
+++ b/include/linux/mrpc.h
@@ -39,8 +39,8 @@ extern int nf_conn_info_event(unsigned l
 /* error codes */
 enum {
     MRPC_SUCCESS = 0,           /* success */
-    MRPC_ERR_INVALID_SIGNATURE,
-    MRPC_ERR_INVALID_HEADER,
+    MRPC_ERR_INVALID_SIGNATURE, /* hw mailbox callback invalid mrpc signature */
+    MRPC_ERR_INVALID_HEADER,    /* hw mailbox callback invalid mrpc header */
     MRPC_ERR_HWMBOX_NOT_READY,  /* hw mailbox driver not ready */
     MRPC_ERR_HWMBOX_ERROR,      /* hw mailbox error */
     MRPC_ERR_SERVER_NOT_READY,  /* remote server not registered */
@@ -84,7 +84,6 @@ enum {
     MRPC_RESERVED_ID_LAST,
 };
 
-#define ID_VALID(id) ((id) > MRPC_RESERVED_ID_LAST)
 
 struct mrpc_client;
 struct mrpc_server;
@@ -108,7 +107,7 @@ struct mrpc_client *mrpc_client_register
  * mrpc_server_register
  * server's registration function
  *
- * @param id        server's uniqueu ID
+ * @param id        server's unique ID
  * @param name      server's name
  * @param procs     server's procs array
  * @param num_procs number of procs
@@ -155,7 +154,7 @@ int mrpc_server_unregister(struct mrpc_s
  * @return 0 for success, mrpc error code otherwise
  */
 int mrpc_call(struct mrpc_client *client, __u8 proc, void *arg,
-                      ssize_t arg_len, void *rep, ssize_t rep_len,
-                      unsigned long timeout, int *errcode);
+              ssize_t arg_len, void *rep, ssize_t rep_len,
+              unsigned long timeout, int *errcode);
 
 #endif /* __HW_MRPC_H__ */
