# HG changeset patch
# Parent 51f7b601025613036e3a94a53c14cbfef7a8ed48

--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_desc.c
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_desc.c
@@ -40,6 +40,10 @@
 extern uint mss_for_tso;
 #endif
 
+#ifdef GBE_DEBUG
+extern uint mss_for_tso;
+#endif
+
 /*!
 * \brief API to free the transmit descriptor memory.
 *
--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_mdio.c
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_mdio.c
@@ -38,7 +38,7 @@ extern struct DWC_ETH_QOS_prv_data *pdat
 
 #define MDIO_ADDR_LANTIQ 31
 
-INT DWC_ETH_QOS_mdio_read_ext(int regaddr, int *phydata)
+int DWC_ETH_QOS_mdio_read_ext(int regaddr, int *phydata)
 {
 	if (!pdata_p){
 		printk(KERN_ALERT "%s:%d: pdata is null\n", __func__, __LINE__);
@@ -49,7 +49,7 @@ INT DWC_ETH_QOS_mdio_read_ext(int regadd
 }
 EXPORT_SYMBOL(DWC_ETH_QOS_mdio_read_ext);
 
-INT DWC_ETH_QOS_mdio_write_ext(int regaddr, int phydata)
+int DWC_ETH_QOS_mdio_write_ext(int regaddr, int phydata)
 {
 	if (!pdata_p) {
 		printk(KERN_ALERT "%s:%d: pdata is null\n", __func__, __LINE__);
--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_yheader.h
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_yheader.h
@@ -160,6 +160,8 @@
 
 #define ONE_SEC_TO_NS   (1000000000)
 
+#define ONE_SEC_TO_NS   (1000000000)
+
 /* The following macros map error macros to POSIX errno values */
 #define ERR_READ_TIMEOUT ETIME
 #define ERR_WRITE_TIMEOUT ETIME
@@ -1318,6 +1320,30 @@ void DWC_ETH_QOS_enable_eee_mode(struct
 
 extern bool config_prints;
 
+#define CFG_PRINT(x...) \
+do { \
+   if(config_prints) { \
+      printk(KERN_INFO x); \
+   } \
+} while (0)
+
+#ifdef GBE_DEBUG
+
+extern bool print_desc;
+extern uint metadata_on_crc;
+
+#define PRINT_TX_DESC(DESC, TYPE) \
+do { \
+   if(print_desc) { \
+      printk(KERN_INFO "[TX_DESC:%s] 0x%08x:0x%08x:0x%08x:0x%08x\n", #TYPE, \
+             (DESC)->TDES0, (DESC)->TDES1, (DESC)->TDES2, (DESC)->TDES3); \
+   } \
+} while (0)
+
+#endif //GBE_DEBUG
+
+extern bool config_prints;
+
 #define CFG_PRINT(x...) \
 do { \
    if(config_prints) { \
