# HG changeset patch
# Parent 7237eff491518e2742bc60b547e10faeda94446b

--- a/include/linux/avalanche/generic/_tistdtypes.h
+++ b/include/linux/avalanche/generic/_tistdtypes.h
@@ -4,9 +4,12 @@
  * Description:
  * TI Standard defines for primitive "C" types only
  *
+  This file is provided under a dual BSD/GPLv2 license.  When using or
+  redistributing this file, you may do so under either license.
+
   GPL LICENSE SUMMARY
 
-  Copyright(c) 2009-2016 Intel Corporation.
+  Copyright(c) 2009-2015 Intel Corporation.
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of version 2 of the GNU General Public License as
@@ -29,6 +32,37 @@
   2200 Mission College Blvd.
   Santa Clara, CA  97052
 
+  BSD LICENSE 
+
+  Copyright(c) 2009-2015 Intel Corporation. All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without 
+  modification, are permitted provided that the following conditions 
+  are met:
+
+    * Redistributions of source code must retain the above copyright 
+      notice, this list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above copyright 
+      notice, this list of conditions and the following disclaimer in 
+      the documentation and/or other materials provided with the 
+      distribution.
+
+    * Neither the name of Intel Corporation nor the names of its 
+      contributors may be used to endorse or promote products derived 
+      from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT 
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, 
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY 
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
 #ifndef CONFIG_ARM_AVALANCHE_SOC /*AP-CPU*/
--- a/include/linux/avalanche/generic/avalanche_pdsp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pdsp_api.h
@@ -93,6 +93,10 @@
 #define PRECMD_OPTION_MASK        (0xFFu << PRECMD_OPTION_SHIFT)
 #define PRECMD_OPTION(x)          (((x) << PRECMD_OPTION_SHIFT)  & PRECMD_OPTION_MASK)
 
+#define MIN_PKTS_FOR_QDA_ACTIVATION_STEP_A0 (0xffffffff)
+#define MIN_PKTS_FOR_QDA_ACTIVATION_STEP_B0 (10)
+#define MIN_PKTS_FOR_QDA_ACTIVATION_ADDR    (0xF3E13200)
+
 #if defined(CONFIG_MACH_PUMA7_FPGA_PP) || defined (CONFIG_MACH_PUMA7_BOARD)
     //#define FPGA_SANITY_UNITEST     0
 #endif
--- a/include/linux/avalanche/generic/avalanche_pp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pp_api.h
@@ -1,4 +1,6 @@
 /*
+  This file is provided under a dual BSD/GPLv2 license.  When using or
+  redistributing this file, you may do so under either license.
 
   GPL LICENSE SUMMARY
 
@@ -25,8 +27,39 @@
   2200 Mission College Blvd.
   Santa Clara, CA  97052
 
-*/
+  BSD LICENSE
+
+  Copyright(c) 2014-2016 Intel Corporation. All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions
+  are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the
+      distribution.
+
+    * Neither the name of Intel Corporation nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+*/
 #ifndef     _AVALANCHE_PP_H
 #define     _AVALANCHE_PP_H
 #ifdef CONFIG_ARM_AVALANCHE_SOC
@@ -892,7 +925,7 @@ AVALANCHE_PP_PSI_t;
  * @class wifi traffic class
  * @vpid vpid of wifi device
  * @wifi_session_info wifi specific session info
- *  
+ *
  */
 struct wifi_proxy_psi_t {
     Uint8  class;
@@ -1095,7 +1128,7 @@ typedef struct // former TI_PP_SESSION
 #endif
 
     /* Flag which indicates the priority of the session.
-     * With the introduction of QoS this will play an important part. 
+     * With the introduction of QoS this will play an important part.
      * priority 0 is the lowest priority, highest priority is depend
      * on the number of queues in the cluster*/
     Uint8                   priority;
@@ -1131,7 +1164,7 @@ typedef struct // former TI_PP_SESSION
     AVALANCHE_PP_SESSION_TDOX_STATS_t   tdox_stats;
 
 #if PUMA7_OR_NEWER_SOC_TYPE
-    Uint8                   is_irreg_moca; 
+    Uint8                   is_irreg_moca;
 #endif
 }
 AVALANCHE_PP_SESSION_INFO_t;
@@ -1224,7 +1257,7 @@ typedef struct
     Uint32  req_int_desc_starvation;
     Uint32  req_ring_full          ;
     Uint32  req_ring_empty         ;
-    Uint32  global_9               ;
+    Uint32  req_ring_ooo           ;
 } avalanche_pp_wifi_tx_stats_t;
 
 typedef struct
@@ -1382,10 +1415,10 @@ typedef struct
 
     Uint32      AQM_pkts_frwrd_to_QoS;
     Uint32      AQM_pkts_discarded;
-    Uint32      AQM_res3;
-    Uint32      AQM_res4;
-    Uint32      AQM_res5;
-    Uint32      AQM_res6;
+    Uint32      AQM_move_to_dls_interrupt;
+    Uint32      AQM_move_to_normal_mode_intrpt;
+    Uint32      AQM_sent_wakeup_upon_timer_expiry;
+    Uint32      AQM_sent_wakeup_upon_bytes_thrshold_crossed;
     Uint32      AQM_res7;
     Uint32      AQM_res8;
 
@@ -1429,13 +1462,19 @@ typedef struct
 
 typedef struct
 {
-    Uint64 ds_received_packets;
-} AVALANCHE_PP_DS_STATS_t;
+    Uint32      NoSessionCounter;
+    Uint32      NoSessionFragCounter;
+    Uint32      MinimumSessionCounter;
+    Uint32      UsDocsisTcpSessionCounter;
+    Uint32      TunnelEncapSessionCounter;
+    Uint32      TunnelDecapSessionCounter;
+    Uint32      TunnelEncapUsDocsisTcpSessionCounter;
+}AVALANCHE_HOP_LIST_STATS_t;
 
 #define AVALANCHE_PP_US_SERVICE_FLOW_COUNTERS 16
 typedef struct
 {
-    // one 64-bit counter per service flow. 16 is a reasonable upper bound on the number of active service flows, although the system supports more (56?). 
+    // one 64-bit counter per service flow. 16 is a reasonable upper bound on the number of active service flows, although the system supports more (56?).
     // In reality, only 1-2 flows are used.
     Uint64 us_sent_packets[AVALANCHE_PP_US_SERVICE_FLOW_COUNTERS];
 } AVALANCHE_PP_US_STATS_t;
@@ -1445,8 +1484,7 @@ typedef struct
     AVALANCHE_PP_STAGE1_STATS_t         stage1;
     AVALANCHE_PP_STAGE2_Part1_STATS_t   stage2_1;
     AVALANCHE_PP_STAGE2_Part2_STATS_t   stage2_2;
-    //AVALANCHE_PP_DS_STATS_t ds;
-    //AVALANCHE_PP_US_STATS_t us;
+    AVALANCHE_HOP_LIST_STATS_t          hopListCounters;                        /* Not PP counters */
 } AVALANCHE_PP_GLOBAL_STATS_t;
 
 #else
@@ -1559,6 +1597,7 @@ extern AVALANCHE_PP_RET_e    avalanche_p
 extern AVALANCHE_PP_RET_e    avalanche_pp_flush_sessions        ( Uint8     vpid_handle, PP_LIST_ID_e   list_id );
 
 extern AVALANCHE_PP_RET_e   avalanche_pp_session_list_execute      ( Uint8     vpid_handle, PP_LIST_ID_e   list_id,    AVALANCHE_EXEC_HOOK_FN_t   handler, Ptr  data );
+extern AVALANCHE_PP_RET_e   avalanche_pp_flush_sessions_per_mac    ( Uint8     mac[6]);
 extern AVALANCHE_PP_RET_e   avalanche_pp_session_pre_action_bind   ( Uint8     vpid_handle,                            AVALANCHE_EXEC_HOOK_FN_t   handler, Ptr  data );
 extern AVALANCHE_PP_RET_e   avalanche_pp_session_post_action_bind  ( Uint8     vpid_handle,                            AVALANCHE_EXEC_HOOK_FN_t   handler, Ptr  data );
 
@@ -1572,6 +1611,7 @@ extern AVALANCHE_PP_RET_e   avalanche_pp
 extern AVALANCHE_PP_RET_e   avalanche_pp_get_stats_vpid                     ( Uint8  vpid_handle, AVALANCHE_PP_VPID_STATS_t* ptr_vpid_stats );
 extern AVALANCHE_PP_RET_e   avalanche_pp_get_stats_global                   ( AVALANCHE_PP_GLOBAL_STATS_t* ptr_stats );
 extern AVALANCHE_PP_RET_e   avalanche_pp_reset_stats_global                 ( void );
+extern void                 avalanche_pp_substract_stats_global_counters(AVALANCHE_PP_GLOBAL_STATS_t* old_stats, AVALANCHE_PP_GLOBAL_STATS_t* new_stats);
 
 #ifdef CONFIG_WIFI_MESH_TUNNEL
 
@@ -1688,16 +1728,16 @@ extern Bool                  avalanche_p
 
 /**
  * Ring configuration structure for the sender entity
- * @base_addr ring base address 
- * @elem_sz_order element size power of 2 order 
+ * @base_addr ring base address
+ * @elem_sz_order element size power of 2 order
  * @sz_mask ring size mask, number of elements can be written to
  *          the ring in a mask form, (4 elements) = 0b11, (8
  *          elements) = 0b111 and so on
  * @hw_cnt_addr HW counter address
- * @rmt_head_ptr_addr remote head pointer address 
- * @note this struct is aligned to wifi fw pdsps mailbox 
+ * @rmt_head_ptr_addr remote head pointer address
+ * @note this struct is aligned to wifi fw pdsps mailbox
  *       commands, DO NOT change it.
- * @note this struct MUST be 4 bytes aligned 
+ * @note this struct MUST be 4 bytes aligned
  */
 struct send_ring_t {
     Uint32  base_addr;
@@ -1710,15 +1750,15 @@ struct send_ring_t {
 /**
  * Ring configuration structure for the receiver entity
  * @base_addr ring base address
- * @elem_sz_order element size power of 2 order 
+ * @elem_sz_order element size power of 2 order
  * @sz_mask ring size mask, number of elements can be written to
  *          the ring in a mask form, (4 elements) = 0b11, (8
  *          elements) = 0b111 and so on
  * @hw_cnt_addr HW counter address
- * @rmt_tail_ptr_addr remote tail pointer address 
- * @note this struct is aligned to wifi fw pdsps mailbox 
+ * @rmt_tail_ptr_addr remote tail pointer address
+ * @note this struct is aligned to wifi fw pdsps mailbox
  *       commands, DO NOT change it.
- * @note this struct MUST be 4 bytes aligned 
+ * @note this struct MUST be 4 bytes aligned
  */
 struct recv_ring_t {
     Uint32  base_addr;
@@ -1732,7 +1772,7 @@ struct recv_ring_t {
  * wifi tx channel rings configuration, FW to wifi SOC
  * @req_ring used by the FW to send packets to wifi SOC
  * @cmp_ring used by the wifi SOC to recycle buffers back to FW
- * @note this struct MUST be 4 bytes aligned 
+ * @note this struct MUST be 4 bytes aligned
  */
 struct wifi_proxy_tx_ch_t {
     struct send_ring_t req_ring;
@@ -1743,7 +1783,7 @@ struct wifi_proxy_tx_ch_t {
  * wifi rx channel rings configuration, wifi SOC to FW
  * @req_ring used by FW to send wifi SOC empty buffers to use
  * @cmp_ring used by wifi SOC to send packets to FW
- * @note this struct MUST be 4 bytes aligned 
+ * @note this struct MUST be 4 bytes aligned
  */
 struct wifi_proxy_rx_ch_t {
     struct send_ring_t cmp_ring;
@@ -1751,10 +1791,10 @@ struct wifi_proxy_rx_ch_t {
 };
 
 /**
- * wifi device channel configuration. 
- * @wifi_dev_id device's index 
- * @tx_channel tx channel configuration 
- * @rx_channel rx channel configuration 
+ * wifi device channel configuration.
+ * @wifi_dev_id device's index
+ * @tx_channel tx channel configuration
+ * @rx_channel rx channel configuration
  */
 struct wifi_proxy_dev_ch_t {
     wifi_dev_id_e       wifi_dev_id;
@@ -1762,20 +1802,20 @@ struct wifi_proxy_dev_ch_t {
     struct wifi_proxy_rx_ch_t rx_channel;
 };
 
-/** 
- * wifi_proxy_fw_init_info_t 
- * @bMgr tx & rx buffers's buffer manager 
+/**
+ * wifi_proxy_fw_init_info_t
+ * @bMgr tx & rx buffers's buffer manager
  * @bPool tx & rx buffers's buffer pool
- * @tx_int_qMgr tx pdsp internal descriptors free queue manager 
- * @tx_int_qNum tx pdsp internal descriptors free queue number 
- * @tx_int_cnt number of desriptors in tx_int free queue 
+ * @tx_int_qMgr tx pdsp internal descriptors free queue manager
+ * @tx_int_qNum tx pdsp internal descriptors free queue number
+ * @tx_int_cnt number of desriptors in tx_int free queue
  * @rx_freeQNum rx pdsp free queue number
  * @rx_freeQMgr rx pdsp free queue manager number
  * @rx_destQNum rx pdsp destination queue number
  * @rx_destQMgr rx pdsp destination queue manager number
  */
 struct wifi_proxy_fw_init_info_t {
-    Uint32 bMgr; 
+    Uint32 bMgr;
     Uint32 bPool;
     Uint32 tx_int_qMgr;
     Uint32 tx_int_qNum;
@@ -1787,25 +1827,25 @@ struct wifi_proxy_fw_init_info_t {
 };
 
 #ifdef CONFIG_WIFI_PROXY_DEBUG
-/** 
- * wifi_proxy_debug_init_info_t 
+/**
+ * wifi_proxy_debug_init_info_t
  * @counter_index PP's debug IP counter index (0-3)
- * @buff_start_addr FW's log messages buffer start address, this 
+ * @buff_start_addr FW's log messages buffer start address, this
  *                  will be configured to the counter start
  *                  address register
- * @buff_end_addr FW's log messages buffer end address, this 
+ * @buff_end_addr FW's log messages buffer end address, this
  *                will be configured to the counter roll over
  *                register
  */
 struct wifi_proxy_debug_init_info_t {
-    Uint32 counter_index; 
+    Uint32 counter_index;
     Uint32 buff_start_addr;
     Uint32 buff_end_addr;
 };
 #endif
 
 /**
- * init info for initializing wifi proxy. 
+ * init info for initializing wifi proxy.
  * @buffers_base_addr base address to use for buffer pool.
  * @buffer_sz buffer size of each buffer.
  * @buffer_cnt number of buffers.
@@ -1816,10 +1856,10 @@ struct wifi_proxy_init_info_t {
     Uint32 buffer_cnt;
 };
 
-/** 
- * wifi_proxy_init_ret_info_t 
- * info returned as reply to wifi_proxy_init_info_t 
- * @addresses addresses to map by atom 
+/**
+ * wifi_proxy_init_ret_info_t
+ * info returned as reply to wifi_proxy_init_info_t
+ * @addresses addresses to map by atom
  * @lengths addresses lengths
  */
 struct wifi_proxy_init_ret_info_t {
@@ -1829,7 +1869,7 @@ struct wifi_proxy_init_ret_info_t {
 
 /**
  * wifi_proxy_add_ch_info_t
- * @dev_id device id 
+ * @dev_id device id
  * @tx_ch tx channel configuration, fw to wifi soc
  * @rx_ch rx channel configuration, wifi soc to fw
  * @author iemergi (2/2/2016)
@@ -1842,13 +1882,13 @@ struct wifi_proxy_add_ch_info_t {
 
 /**
  * wifi proxy add channel return structure
- * @fw_to_dev_req_tail_addr fw to dev request ring tail pointer 
+ * @fw_to_dev_req_tail_addr fw to dev request ring tail pointer
  *                   address in PDSP's local memory (DMEM)
  * @fw_to_dev_cmp_head_addr fw to dev complete ring head pointer
  *                   address in PDSP's local memory (DMEM)
  * @dev_to_fw_cmp_tail_addr dev to fw complete ring tail pointer
  *                   address in PDSP's local memory (DMEM)
- * @dev_to_fw_req_head_addr dev to fw request ring head pointer 
+ * @dev_to_fw_req_head_addr dev to fw request ring head pointer
  *                   address in PDSP's local memory (DMEM)
  */
 struct wifi_proxy_add_ch_ret_info_t {
@@ -1859,14 +1899,14 @@ struct wifi_proxy_add_ch_ret_info_t {
 };
 
 /**
- * wifi device statistics structure. 
- * @tx_req_sent_pkt_num number of packets sent on tx request 
+ * wifi device statistics structure.
+ * @tx_req_sent_pkt_num number of packets sent on tx request
  *                      ring
- * @tx_cmp_recv_pkt_num number of packets receive on tx complete 
+ * @tx_cmp_recv_pkt_num number of packets receive on tx complete
  *                      ring
- * @rx_cmp_sent_pkt_num number of packets sent on rx complete 
+ * @rx_cmp_sent_pkt_num number of packets sent on rx complete
  *                      ring
- * @rx_req_recv_pkt_num number of packets received on rx request 
+ * @rx_req_recv_pkt_num number of packets received on rx request
  *                      ring
  */
 struct pp_hal_wifi_device_stats{
@@ -1889,15 +1929,15 @@ struct pp_hal_wifi_device_stats{
 
 /**
  * proxy tx pdsp devices control register layout
- *  
+ *
  * @suspend specify if the device is in suspend state
- * @valid specify if the device's entry is valid 
- * @stat_counters_off offset to device's statistical counters 
+ * @valid specify if the device's entry is valid
+ * @stat_counters_off offset to device's statistical counters
  * @free_buffers number of buffers the device can use
  * @max_buffers maximum number of buffers the device can use
- *  
- * @note the struct contains reserve fields in order to be align to 
- * the actual memory layout 
+ *
+ * @note the struct contains reserve fields in order to be align to
+ * the actual memory layout
  */
 struct __attribute__((packed)) fw_tx_pdsp_devs_ctrl_reg {
 #if defined(__BIG_ENDIAN_BITFIELD)
@@ -1917,28 +1957,26 @@ struct __attribute__((packed)) fw_tx_pds
 };
 
 /**
- * proxy tx pdsp devices database entry structure 
- *  
+ * proxy tx pdsp devices database entry structure
+ *
  * @req_ring request ring structure
  * @req_ring_head request ring head pointer
  * @req_ring_tail request ring tail pointer
  * @cmp_ring complete request structure
  * @cmp_ring_head complete ring head pointer
  * @cmp_ring_tail complete ring tail pointer
- * @mcdma_page_offset the offset to the page the McDMA is using 
- * @mcdma_page_num_entries number of entries ceing copied to 
+ * @mcdma_page_offset the offset to the page the McDMA is using
+ * @mcdma_page_num_entries number of entries ceing copied to
  *                         mcdma_page_offset
- * @mcdma_done_page_offset the offset to the page the McDMA is 
+ * @mcdma_done_page_offset the offset to the page the McDMA is
  *                         done using
- * @mcdma_done_num_entries number of entries copied to 
+ * @mcdma_done_num_entries number of entries copied to
  *                         mcdma_done_page_offset
- * @pdsp_page_offset offset to next free page the pdsp can use 
- * @mcdma_ch_regs_offset the offset to mcdma channel regs that 
+ * @pdsp_page_offset offset to next free page the pdsp can use
+ * @mcdma_ch_regs_offset the offset to mcdma channel regs that
  *                       the device is using
- * @pagesBase mcdma pages base offset 
- *  
- * @note the struct contains reserve fields in order to be align to 
- * the actual memory layout 
+ * @pagesBase mcdma pages base offset
+ *
  */
 struct __attribute__((packed)) fw_tx_pdsp_devs_db_entry {
     /* request ring */
@@ -1946,11 +1984,12 @@ struct __attribute__((packed)) fw_tx_pds
     Uint32             req_ring_head;
     Uint32             req_ring_tail;
     /* request ring control */
-    Uint16             input_q_off;
     Uint16             internal_q_off;
     Uint16             infra_dma_q;
+    Uint8              seq_num;
+    Uint8              next_exp_seq_num;
     Uint8              internal_desc_use_cnt;
-    Uint8              reserved;
+    Uint8              num_elemments_pushed;
     /* complete ring */
     struct recv_ring_t cmp_ring;
     Uint32             cmp_ring_head;
@@ -1965,16 +2004,16 @@ struct __attribute__((packed)) fw_tx_pds
 
 /**
  * proxy rx pdsp devices control register layout
- *  
+ *
  * @suspend specify if the device is in suspend state
- * @valid specify if the device's entry is valid 
+ * @valid specify if the device's entry is valid
  * @pid device's pid
- * @stat_counters_off offset to device's statistical counters 
+ * @stat_counters_off offset to device's statistical counters
  * @free_buffers number of buffers the device can use
- * @max_buffers maximum number of buffers the device can use 
- *  
- * @note the struct contains reserve fields in order to be align to 
- * the actual memory layout 
+ * @max_buffers maximum number of buffers the device can use
+ *
+ * @note the struct contains reserve fields in order to be align to
+ * the actual memory layout
  */
 struct __attribute__((packed)) fw_rx_pdsp_devs_ctrl_reg {
 #if defined(__BIG_ENDIAN_BITFIELD)
@@ -1992,29 +2031,29 @@ struct __attribute__((packed)) fw_rx_pds
     Uint16  max_buffers ;
 };
 /**
- * proxy rx pdsp devices database entry structure 
- *  
+ * proxy rx pdsp devices database entry structure
+ *
  * @cmp_ring complete ring structure
  * @cmp_ring_head complete ring head pointer
- * @cmp_ring_tail complete ring tail pointer 
- * @devBuffersLack specify how much buffers needs to be send to 
+ * @cmp_ring_tail complete ring tail pointer
+ * @devBuffersLack specify how much buffers needs to be send to
  *                 the device on complete ring
  * @req_ring request ring structure
  * @req_ring_head request ring head pointer
  * @req_ring_tail request ring tail pointer
- * @mcdma_page_offset the offset to the page the McDMA is using 
- * @mcdma_page_num_entries number of entries ceing copied to 
+ * @mcdma_page_offset the offset to the page the McDMA is using
+ * @mcdma_page_num_entries number of entries ceing copied to
  *                         mcdma_page_offset
- * @mcdma_done_page_offset the offset to the page the McDMA is 
+ * @mcdma_done_page_offset the offset to the page the McDMA is
  *                         done using
- * @mcdma_done_num_entries number of entries copied to 
+ * @mcdma_done_num_entries number of entries copied to
  *                         mcdma_done_page_offset
- * @mcdma_ch_regs_offset the offset to mcdma channel regs that 
+ * @mcdma_ch_regs_offset the offset to mcdma channel regs that
  *                       the device is using
- * @pdsp_page_offset offset to next free page the pdsp can use 
- * @pagesBase mcdma pages base offset 
+ * @pdsp_page_offset offset to next free page the pdsp can use
+ * @pagesBase mcdma pages base offset
  * @note the struct contains reserve fields in order to be
- * align to the actual memory layout 
+ * align to the actual memory layout
  */
 struct __attribute__((packed, aligned(64))) fw_rx_pdsp_devs_db_entry {
     /* complete ring */
@@ -2044,7 +2083,7 @@ AVALANCHE_PP_RET_e avalanche_pp_add_wifi
 AVALANCHE_PP_RET_e avalanche_pp_rem_wifi_dev_ch(wifi_dev_id_e wifi_dev_id);
 AVALANCHE_PP_RET_e avalanche_pp_start_wifi_dev_ch(wifi_dev_id_e wifi_dev_id);
 AVALANCHE_PP_RET_e avalanche_pp_stop_wifi_dev_ch(wifi_dev_id_e wifi_dev_id);
-AVALANCHE_PP_RET_e avalanche_pp_get_dev_ch_stats(wifi_dev_id_e dev_id, 
+AVALANCHE_PP_RET_e avalanche_pp_get_dev_ch_stats(wifi_dev_id_e dev_id,
                                                  struct pp_hal_wifi_device_stats *dev_stats);
 
 #endif
@@ -2139,8 +2178,7 @@ typedef     struct
 #define PP_DRIVER_KERNEL_POST_INIT                  _IO   (PP_DRIVER_MODULE_ID, 11)
 #define PP_DRIVER_SET_ACK_SUPP                      _IOWR (PP_DRIVER_MODULE_ID, 12, avalanche_pp_ackSupp_ioctl_param_t)
 #define PP_DRIVER_GET_DB_PARAMS                     _IOWR (PP_DRIVER_MODULE_ID, 13, avalanche_pp_db_address_param_t)
-#define PP_DRIVER_GET_GLOBAL_STATS                  _IOR  (PP_DRIVER_MODULE_ID, 14, AVALANCHE_PP_GLOBAL_STATS_t)
-#define PP_DRIVER_GET_SESSION_STATS                 _IOWR (PP_DRIVER_MODULE_ID, 15, AVALANCHE_PP_EXTENDED_SESSION_STATS_t)
+#define PP_DRIVER_GET_SESSION_STATS                 _IOWR (PP_DRIVER_MODULE_ID, 14, AVALANCHE_PP_EXTENDED_SESSION_STATS_t)
 
 
 #endif //   _AVALANCHE_PP_H
--- a/include/linux/avalanche/generic/pp_qos_p7.h
+++ b/include/linux/avalanche/generic/pp_qos_p7.h
@@ -146,8 +146,10 @@ typedef struct
 {
     Uint8    active_shared_clusters;
     Uint8    active_private_clusters[AVALANCHE_PP_MAX_PID];
+    Uint8    private_clusters_free_pool_cnt[AVALANCHE_PP_MAX_PID];
     Uint8    active_shared_queues;
     Uint8    active_private_queues[AVALANCHE_PP_MAX_PID];
+    Uint8    private_queues_free_pool_cnt[AVALANCHE_PP_MAX_PID];
     Uint32   clusters_starvation_shared;
     Uint32   queues_starvation_shared;
 }
@@ -488,6 +490,22 @@ PP_QOS_MGR_RET_e pp_qos_modify_active_qu
  *  Returns PP_QOS_RC_SUCCESS if pass.
  */
 PP_QOS_MGR_RET_e pp_qos_get_queue_stats(Uint32 qos_qnum, PP_QOS_MIB_STATS_t *q_stats);
+/**
+ *  pp_qos_inc_session_cnt_for_queue - Increment sessions
+ *  counter for QoS queue
+ *  @phy_queue_id: Queue ID
+ *
+ *  Returns PP_QOS_RC_SUCCESS if pass.
+ */
+PP_QOS_MGR_RET_e pp_qos_inc_session_cnt_for_queue(Uint8 phy_queue_id);
+/**
+ *  pp_qos_dec_session_cnt_for_queue - Decrement sessions
+ *  counter for QoS queue
+ *  @phy_queue_id: Queue ID
+ *
+ *  Returns PP_QOS_RC_SUCCESS if pass.
+ */
+PP_QOS_MGR_RET_e pp_qos_dec_session_cnt_for_queue(Uint8 phy_queue_id);
 
 /**************/
 /* VPID       */
--- a/include/linux/avalanche/puma7/puma7_cppi.h
+++ b/include/linux/avalanche/puma7/puma7_cppi.h
@@ -119,6 +119,9 @@ extern Uint8 *PalCppiDsg2qmgrNames[PAL_C
 #define PAL_CPPI41_SR_DOCSIS_TX_QPDSP_QOS_Q_BASE            PAL_CPPI_PP_QMGR_G1_QOS_US_Q_BASE
 #define PAL_CPPI41_SR_DOCSIS_TX_QPDSP_QOS_Q_LAST            PAL_CPPI_PP_QMGR_G1_QOS_US_Q_LAST
 
+#define IS_QOS_Q(q) ((q >= PAL_CPPI_NETDEV_BUILD_Q_INFO(PAL_CPPI_PP_QMGR_G1, PAL_CPPI_PP_QMGR_G1_QOS_Q_BASE))           && (q <= PAL_CPPI_NETDEV_BUILD_Q_INFO(PAL_CPPI_PP_QMGR_G1, PAL_CPPI_PP_QMGR_G1_QOS_Q_LAST)) )
+#define IS_AQM_Q(q) ((q >= PAL_CPPI_NETDEV_BUILD_Q_INFO(PAL_CPPI_PP_QMGR_G2, PAL_CPPI_PP_QMGR_G2_AQM_US_BE0_LOW_Q_NUM)) && (q <= PAL_CPPI_NETDEV_BUILD_Q_INFO(PAL_CPPI_PP_QMGR_G2, PAL_CPPI_PP_QMGR_G2_AQM_US_BE15_HI_Q_NUM)) )
+
 #define PAL_CPPI41_SR_DOCSIS_MGMT_RX_FD_HOST_DESC_SIZE      PAL_CPPI_PP_QMGR_GLOBAL_DEFAULT_DESC_SIZE
 #define PAL_CPPI41_SR_DOCSIS_MGMT_TX_FD_HOST_DESC_SIZE      PAL_CPPI_PP_QMGR_GLOBAL_DEFAULT_DESC_SIZE
 #define PAL_CPPI41_SR_DOCSIS_MGMT_TX_FD_HOST_BUFF_SIZE      PAL_CPPI_PP_QMGR_GLOBAL_DEFAULT_BUFF_SIZE
--- a/include/linux/avalanche/puma7/puma7_cppi_prv.h
+++ b/include/linux/avalanche/puma7/puma7_cppi_prv.h
@@ -801,7 +801,7 @@ typedef enum PAL_CPPI_PP_BUFFER_POOLs
                                                                     PAL_CPPI_PP_SGMII1_RX_LOW_SHORT_PRIVATE_PACKET_RAM_EMB_FD_DESC_COUNT
 #define PAL_CPPI_PP_PACKET_RAM_512B_BUFFER_SIZE                     512
 #define PAL_CPPI_PP_PACKET_RAM_512B_BUFFER_REF_CNT                  0
-#define PAL_CPPI_PP_PACKET_RAM_512B_BUFFER_BASE                     0xF9020000
+#define PAL_CPPI_PP_PACKET_RAM_512B_BUFFER_BASE                     0xF9000000
 
 #define PAL_CPPI_PP_PACKET_RAM_2KB_BUFFER_COUNT                     PAL_CPPI_PP_DOCSIS_RX_LOW_LONG_PRIVATE_PACKET_RAM_EMB_FD_DESC_COUNT                 +   \
                                                                     PAL_CPPI_PP_RX_LOW_LONG_SHARED_PACKET_RAM_EMB_FD_DESC_COUNT                         +   \
@@ -813,7 +813,7 @@ typedef enum PAL_CPPI_PP_BUFFER_POOLs
                                                                     PAL_CPPI_PP_SGMII1_RX_LOW_LONG_PRIVATE_PACKET_RAM_EMB_FD_DESC_COUNT
 #define PAL_CPPI_PP_PACKET_RAM_2KB_BUFFER_SIZE                      (2 * 1024)
 #define PAL_CPPI_PP_PACKET_RAM_2KB_BUFFER_REF_CNT                   0
-#define PAL_CPPI_PP_PACKET_RAM_2KB_BUFFER_BASE                      0xF9000000
+#define PAL_CPPI_PP_PACKET_RAM_2KB_BUFFER_BASE                      0xF9040000
 
 #define PAL_CPPI_PP_VOICE_DSP_BUFFER_COUNT                          512
 #define PAL_CPPI_PP_VOICE_DSP_BUFFER_SIZE                           1024
@@ -1219,10 +1219,10 @@ typedef enum PAL_CPPI_PP_DMA09_INFRA_RX_
     PAL_CPPI_PP_DMA09_INFRA_RX_CH_17,
     PAL_CPPI_PP_DMA09_INFRA_RX_CH_18,
     PAL_CPPI_PP_DMA09_INFRA_RX_CH_19,
-    PAL_CPPI_PP_DMA09_INFRA_RX_CH_20,
-    PAL_CPPI_PP_DMA09_INFRA_RX_CH_21,
-    PAL_CPPI_PP_DMA09_INFRA_RX_CH_22,
-    PAL_CPPI_PP_DMA09_INFRA_RX_CH_23,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE0_TX_FW_INFRA_RX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE1_TX_FW_INFRA_RX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE2_TX_FW_INFRA_RX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE3_TX_FW_INFRA_RX_CH,
     PAL_CPPI_PP_DMA09_INFRA_RX_CHANNELS_NUM
 }PAL_CPPI_PP_DMA09_INFRA_RX_CHANNELS_e;
 // PAL_CPPI_PP_DMA09_INFRA_TX
@@ -1248,10 +1248,10 @@ typedef enum PAL_CPPI_PP_DMA09_INFRA_TX_
     PAL_CPPI_PP_DMA09_INFRA_TX_CH_17,
     PAL_CPPI_PP_DMA09_INFRA_TX_CH_18,
     PAL_CPPI_PP_DMA09_INFRA_TX_CH_19,
-    PAL_CPPI_PP_DMA09_INFRA_TX_CH_20,
-    PAL_CPPI_PP_DMA09_INFRA_TX_CH_21,
-    PAL_CPPI_PP_DMA09_INFRA_TX_CH_22,
-    PAL_CPPI_PP_DMA09_INFRA_TX_CH_23,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE0_TX_FW_INFRA_TX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE1_TX_FW_INFRA_TX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE2_TX_FW_INFRA_TX_CH,
+    PAL_CPPI_PP_DMA09_WIFI_DEVICE3_TX_FW_INFRA_TX_CH,
     PAL_CPPI_PP_DMA09_INFRA_TX_CHANNELS_NUM
 }PAL_CPPI_PP_DMA09_INFRA_TX_CHANNELS_e;
 
@@ -1918,10 +1918,15 @@ typedef enum PAL_CPPI_PP_SOURCE_PORTS
     (PAL_CPPI_PP_QMGR_G2)
 
 /* tx fw - rx channels */
-/* convert wifi device id to tx fw infra dma rx channel */
-#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_RX_CH(dev_id)                           \
+/* convert wifi device id to tx fw infra dma 08 rx channel */
+#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_08_RX_CH(dev_id)                        \
     ((dev_id) + PAL_CPPI_PP_DMA08_WIFI_DEVICE0_TX_FW_INFRA_RX_CH)
 
+/* convert wifi device id to tx fw infra dma 09 rx channel */
+#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_09_RX_CH(dev_id)                        \
+    ((dev_id) + PAL_CPPI_PP_DMA09_WIFI_DEVICE0_TX_FW_INFRA_RX_CH)
+
+
 /* convert wifi device id to tx fw infra dma rx channel output queue */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_RX_CH_OUT_Q(dev_id)                     \
     ((dev_id) + PAL_CPPI_PP_QMGR_G1_WIFI_TX_DEV0_INTERNAL_Q)
@@ -1939,10 +1944,14 @@ typedef enum PAL_CPPI_PP_SOURCE_PORTS
     (PAL_CPPI_PP_QMGR_G1)
 
 /* tx fw - tx channels */
-/* convert wifi device id to tx fw infra dma tx channel */
-#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_TX_CH(dev_id)                           \
+/* convert wifi device id to tx fw infra dma 08 tx channel */
+#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_08_TX_CH(dev_id)                        \
     ((dev_id) + PAL_CPPI_PP_DMA08_WIFI_DEVICE0_TX_FW_INFRA_TX_CH)
 
+/* convert wifi device id to tx fw infra dma 09 tx channel */
+#define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_09_TX_CH(dev_id)                        \
+    ((dev_id) + PAL_CPPI_PP_DMA09_WIFI_DEVICE0_TX_FW_INFRA_TX_CH)
+
 /* convert wifi device id to tx fw infra dma tx channel input queue */
 #define WIFI_DEVICE_TO_WIFI_TX_FW_INFRA_TX_CH_INPUT_Q(dev_id)                   \
     (((dev_id)*2) + PAL_CPPI_PP_QMGR_G2_CDMA8_INFRA_INPUT_WIFI_DEVICE0_TX_FW_HI_Q_NUM)
