# HG changeset patch
# Parent 81e27eb448c8a3189d75e0d3224f21cd83de6cd9

--- a/include/linux/avalanche/generic/avalanche_pp_api.h
+++ b/include/linux/avalanche/generic/avalanche_pp_api.h
@@ -1868,16 +1868,27 @@ struct wifi_proxy_init_ret_info_t {
 };
 
 /**
+ * wifi device indexes enumeration
+ */
+typedef enum {
+    WIFI_DEV_MODE_PARTIAL,
+    WIFI_DEV_MODE_FULL,
+    WIFI_DEV_MODES_CNT
+} wifi_dev_mode_e;
+
+/**
  * wifi_proxy_add_ch_info_t
  * @dev_id device id
  * @tx_ch tx channel configuration, fw to wifi soc
- * @rx_ch rx channel configuration, wifi soc to fw
+ * @rx_ch rx channel configuration, wifi soc to fw 
+ * @f_mode specify if the device is working in full offload mode
  * @author iemergi (2/2/2016)
  */
 struct wifi_proxy_add_ch_info_t {
-    wifi_dev_id_e dev_id;
-    struct wifi_proxy_tx_ch_t tx_ch;
-    struct wifi_proxy_rx_ch_t rx_ch;
+    wifi_dev_id_e               dev_id;
+    struct wifi_proxy_tx_ch_t   tx_ch;
+    struct wifi_proxy_rx_ch_t   rx_ch;
+    wifi_dev_mode_e             mode;
 };
 
 /**
@@ -1941,13 +1952,21 @@ struct pp_hal_wifi_device_stats{
  */
 struct __attribute__((packed)) fw_tx_pdsp_devs_ctrl_reg {
 #if defined(__BIG_ENDIAN_BITFIELD)
-    Uint8   reserve:6,
+    Uint8   reserve:1,
+            pdsp_res:2,
+            mode:1,
+            cmp_irq:1,
+            req_irq:1,
             suspend:1,
             valid:1;
 #elif defined (__LITTLE_ENDIAN_BITFIELD)
     Uint8   valid:1,
             suspend:1,
-            reserve:6;
+            req_irq:1,
+            cmp_irq:1,
+            mode:1,
+            pdsp_res:2,
+            reserve:1;
 #endif
     Uint8   Breserve0;
 
@@ -2017,13 +2036,19 @@ struct __attribute__((packed)) fw_tx_pds
  */
 struct __attribute__((packed)) fw_rx_pdsp_devs_ctrl_reg {
 #if defined(__BIG_ENDIAN_BITFIELD)
-    Uint8   reserve:6,
+    Uint8   reserve:3,
+            mode:1,
+            cmp_irq:1,
+            req_irq:1,
             suspend:1,
             valid:1;
 #elif defined (__LITTLE_ENDIAN_BITFIELD)
     Uint8   valid:1,
             suspend:1,
-            reserve:6;
+            req_irq:1,
+            cmp_irq:1,
+            mode:1,
+            reserve:3;
 #endif
     Uint8   pid;
     Uint16  stat_counters_off;
--- a/include/linux/avalanche/puma7/puma7_cppi_prv.h
+++ b/include/linux/avalanche/puma7/puma7_cppi_prv.h
@@ -1863,8 +1863,6 @@ typedef enum PAL_CPPI_PP_SOURCE_PORTS
 
 
 #define PAL_CPPI41_SR_GBE_INFRA_FD_HOST_DESC_COUNT                              (512)
-#define PAL_CPPI41_SR_GBE_INFRA_FD_HOST_BUFFER_SIZE                             (1600)
-#define PAL_CPPI41_SR_WPD_INFRA_FD_HOST_BUFFER_SIZE                             (1600)
 
 #define PAL_CPPI41_SR_GBE_HOST_RX_Q_BASE                                        (20)
 #define PAL_CPPI41_SR_DMA_FD_TEARDOWN_Q_NUM                                     (PAL_CPPI_PP_QMGR_G2_TEARDOWN_FD_Q_NUM)
--- a/include/linux/ti_hil.h
+++ b/include/linux/ti_hil.h
@@ -49,6 +49,10 @@ struct hil_core_notifier_param
 #define TI_INET     2
 #define TI_PP       3
 
+/* MRPC module id used for module registration */
+#define PP_DEV      1
+#define PP_HIL      2
+
 /**************************************************************************
  **************************** PP Event Identifiers ************************
  **************************************************************************/
