--- a/include/linux/avalanche/puma7/puma7_pp.h
+++ b/include/linux/avalanche/puma7/puma7_pp.h
@@ -407,6 +407,9 @@ typedef enum {
 #define AVALANCHE_PP_AQM_CMD_RGN_BASE                   (IO_ADDRESS(0xF3C10000))
 #define AVALANCHE_PP_AQM_PARAM_RGN_BASE                 (IO_ADDRESS(0xF3C10004))
 
+#define AVALANCHE_PP_RANDOMIZER_AQM_BASE                (IO_ADDRESS(0xF3C04100))
+#define AVALANCHE_PP_RANDOMIZER_QOS_BASE                (IO_ADDRESS(0xF3E08300))
+
 /* MAILBOX */
 #define AVALANCHE_PP_MAILBOX_CLK_CTRL_REG_STAGE1        (IO_ADDRESS(0xF3C20108))
 
@@ -608,6 +611,11 @@ typedef enum {
 #define FREE_RUNNING_COUNTER_L_GET()                    be32_to_cpu(*(volatile unsigned int *)AVALANCHE_PP_STATISTICAL_FRC_S2_L_BASE)
 #define FREE_RUNNING_COUNTER_H_GET()                    be32_to_cpu(*(volatile unsigned int *)AVALANCHE_PP_STATISTICAL_FRC_S2_H_BASE)
 
+/* Randomizer seeding */
+#define RANDOMIZER_AQM_SEED(seed)                      {(*(volatile unsigned int *)AVALANCHE_PP_RANDOMIZER_AQM_BASE) = cpu_to_be32(seed);}
+#define RANDOMIZER_QOS_SEED(seed)                      {(*(volatile unsigned int *)AVALANCHE_PP_RANDOMIZER_QOS_BASE) = cpu_to_be32(seed);}
+
+
 /* PP MTU size */
 #define UPDATE_MTU_TABLE_IN_PDSP_DMEM(vpidId, mtuSize)        \
     AVALANCHE_PP_MODIFIER_MTU_TABLE_UPDATE(vpidId, mtuSize);  \
