# HG changeset patch
# Parent feb70ac50304505c0510c1fb47bed885f88458c3

--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_dev.c
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_dev.c
@@ -2426,7 +2426,7 @@ static void pre_transmit(struct DWC_ETH_
 
       /* DMA uses the MSS value programed in DMA_CR if driver
        * doesn't provided the CONTEXT descriptor */
-      DWC_REG_WR_FIELD(DMA_CR(qInx), DMA_CR_MSS, tx_pkt_features->mss);
+      //DWC_REG_WR_FIELD(DMA_CR(qInx), DMA_CR_MSS, tx_pkt_features->mss);
 
       tx_desc_data->default_mss = tx_pkt_features->mss;
 
