# HG changeset patch
# Parent ed657c7aa8ccf939f016493bb8d375e78d005864

--- a/include/net/ip.h
+++ b/include/net/ip.h
@@ -102,6 +102,7 @@ extern int		ip_build_and_send_pkt(struct
 extern int		ip_rcv(struct sk_buff *skb, struct net_device *dev,
 			       struct packet_type *pt, struct net_device *orig_dev);
 extern int		ip_local_deliver(struct sk_buff *skb);
+extern int		ip_local_deliver_finish(struct sk_buff *skb);
 extern int		ip_mr_input(struct sk_buff *skb);
 extern int		ip_output(struct sk_buff *skb);
 extern int		ip_mc_output(struct sk_buff *skb);
--- a/net/ipv4/ip_input.c
+++ b/net/ipv4/ip_input.c
@@ -190,7 +190,7 @@ bool ip_call_ra_chain(struct sk_buff *sk
 	return false;
 }
 
-static int ip_local_deliver_finish(struct sk_buff *skb)
+int ip_local_deliver_finish(struct sk_buff *skb)
 {
 	struct net *net = dev_net(skb->dev);
 
@@ -241,6 +241,7 @@ static int ip_local_deliver_finish(struc
 
 	return 0;
 }
+EXPORT_SYMBOL(ip_local_deliver_finish);
 
 /*
  * 	Deliver IP Packets to the higher protocol layers.
