From e5998f83974242ce6216196b45164670971d9097 Mon Sep 17 00:00:00 2001
From: tzahiste <tzahi.stern@intel.com>
Date: Sun, 11 Dec 2016 14:15:25 +0200
Subject: [PATCH] Add ACPI support to tmp102

---
 drivers/hwmon/tmp102.c |   16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/hwmon/tmp102.c b/drivers/hwmon/tmp102.c
index d7b47ab..de95fff 100644
--- a/drivers/hwmon/tmp102.c
+++ b/drivers/hwmon/tmp102.c
@@ -26,6 +26,7 @@
 #include <linux/err.h>
 #include <linux/mutex.h>
 #include <linux/device.h>
+#include <linux/acpi.h>
 #include <linux/jiffies.h>
 
 #define	DRIVER_NAME "tmp102"
@@ -144,7 +145,10 @@ static const struct attribute_group tmp102_attr_group = {
 	.attrs = tmp102_attributes,
 };
 
-#define TMP102_CONFIG  (TMP102_CONF_TM | TMP102_CONF_EM | TMP102_CONF_CR1)
+//enable interrupt mode
+//#define TMP102_CONFIG  (TMP102_CONF_TM | TMP102_CONF_EM | TMP102_CONF_CR1)
+//disable interrupt mode
+#define TMP102_CONFIG  (TMP102_CONF_EM | TMP102_CONF_CR1)
 #define TMP102_CONFIG_RD_ONLY (TMP102_CONF_R0 | TMP102_CONF_R1 | TMP102_CONF_AL)
 
 static int tmp102_probe(struct i2c_client *client,
@@ -277,11 +281,21 @@ static const struct i2c_device_id tmp102_id[] = {
 	{ "tmp102", 0 },
 	{ }
 };
+
+#ifdef CONFIG_ACPI
+static const struct acpi_device_id tmp102_acpi_match[] = {
+	{ "TXNW0066", 0 },
+	{ }
+};
+MODULE_DEVICE_TABLE(acpi, tmp102_acpi_match);
+#endif
+
 MODULE_DEVICE_TABLE(i2c, tmp102_id);
 
 static struct i2c_driver tmp102_driver = {
 	.driver.name	= DRIVER_NAME,
 	.driver.pm	= TMP102_DEV_PM_OPS,
+	.driver.acpi_match_table = ACPI_PTR(tmp102_acpi_match),
 	.probe		= tmp102_probe,
 	.remove		= tmp102_remove,
 	.id_table	= tmp102_id,
-- 
1.7.9.5

