# HG changeset patch
# Parent 784b62f560cf5a40f2e4a48d032a8435fb5fcb7b

diff --git a/drivers/mrpc/core/mrpc.c b/drivers/mrpc/core/mrpc.c
--- a/drivers/mrpc/core/mrpc.c
+++ b/drivers/mrpc/core/mrpc.c
@@ -64,7 +64,7 @@
 #endif
 
 #define MAX_NAME_SIZE                       50
-#define DEFAULT_CLIENT_WAIT_TIMEOUT_MS      200
+#define DEFAULT_CLIENT_WAIT_TIMEOUT_MS      1000
 
 #ifdef CONFIG_ARM_AVALANCHE_SOC
 #define MRPC_HW_MBOX_THIS HW_MBOX_MASTER_NP_CPU
