# HG changeset patch
# Parent 6af7dab50fd7464fbd8f3ab83b748b6904d62f7a

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -590,6 +590,7 @@ config X86_PUMA6
 	HW_MUTEXES, MTD_NMYX25, CE_MAILBOX, UDMA, GPIO_INTELCE, GEN3_SPI,
 	GEN3_USB, SPI_CE5XX_SPI_SLAVE, SPI_CE5XX_SPI_FLASH, GEN3_UART,
 	INTELCE_WDT, GEN3_I2C, UIO_PCI_CE2600
+
 config X86_PUMA7
 	bool "Puma7 Architecture Support"
 	select IOSF_MBI
@@ -603,6 +604,7 @@ config X86_PUMA7
 	ARM_AVALANCHE, INTEL_CHT_PHY, INTEL_ISS, MMC_SDHCI_ACPI_POWER_ON,
 	GMIN_INTEL_MID, GPIO_INTELCE2700, INTEL_MEI_TXE, REGULATOR_INT34D9,
 	PUMA_THERMAL, BRIDGE_EBT_FORWARD, SPI_INTEL_SPI_PLATFORM
+	
 config HW_MAILBOX
         tristate "HW Mailbox support for Intel CE2600 platform"
         depends on NET_SUBSYSTEM
@@ -611,6 +613,8 @@ config HW_MAILBOX
 	This is the HW Mailbox support for Intel CE2600 platform.
 	If unsure, say Y.
 
+menu "Intel ce2700"
+
 config X86_INTEL_CE2700
         tristate "Intel CE2700 Platform Support"
         depends on X86_PUMA7
@@ -618,6 +622,22 @@ config X86_INTEL_CE2700
 	---help---
 	Select to build a kernel capable of supporting CE2700 platform drivers.
 
+config X86_INTEL_CE2700_POWER_BUTTON
+	bool "power button exists on board"
+	depends on X86_INTEL_CE2700
+	default n
+	---help---
+		Select to enable power button gpio support in ce2700 driver
+
+config X86_INTEL_CE2700_POWER_BUTTON_GPIO
+	int "Power button GPIO number"
+	depends on X86_INTEL_CE2700_POWER_BUTTON
+	default 63
+	---help---
+		Power button GPIO number
+
+endmenu
+
 config X86_WANT_INTEL_MID
 	bool "Intel MID platform support"
 	depends on X86_32
diff --git a/arch/x86/platform/intel-ce2700/intel-ce2700.c b/arch/x86/platform/intel-ce2700/intel-ce2700.c
--- a/arch/x86/platform/intel-ce2700/intel-ce2700.c
+++ b/arch/x86/platform/intel-ce2700/intel-ce2700.c
@@ -18,6 +18,8 @@
 #include <linux/platform_device.h>
 #include <linux/platform_data/intel-spi.h>
 #include <linux/leds.h>
+#include <linux/gpio_keys.h>
+#include <linux/input.h>
 #include <linux/gpio.h>
 #include <linux/i2c.h>
 
@@ -68,6 +70,38 @@ static struct platform_device_info ce270
 	.size_data = sizeof(ce2700_pcu_spi_plat_data),
 };
 
+#ifdef CONFIG_X86_INTEL_CE2700_POWER_BUTTON
+static struct gpio_keys_button gpio_keys[] = {
+	{
+		.code = KEY_POWER,
+		.gpio = CONFIG_X86_INTEL_CE2700_POWER_BUTTON_GPIO,
+	},
+
+};
+
+static struct gpio_keys_platform_data gpio_keys_data = {
+	.nbuttons = ARRAY_SIZE(gpio_keys),
+	.buttons = gpio_keys,
+};
+
+static struct platform_device power_button_dev = {
+	.name = "gpio-keys",
+	.id = -1,
+	.dev.platform_data = &gpio_keys_data,
+};
+#endif
+
+static struct __initdata platform_device *cgp_devs[] = {
+#ifdef CONFIG_X86_INTEL_CE2700_POWER_BUTTON
+	&power_button_dev,
+#endif
+};
+
+static int __init add_cgp_devs()
+{
+	return platform_add_devices(cgp_devs, ARRAY_SIZE(cgp_devs));
+}
+
 static struct gpio_led havenPark_leds[] = {
 	{
 		.name = "havenPark_red",
@@ -102,9 +136,8 @@ static struct __initdata platform_device
 	&havenPark_leds_dev,
 };
 
-static int __init ce2700_register_led()
+static int __init add_hvp_devs()
 {
-	/* Setup LED control through leds-gpio driver */
 	return platform_add_devices(havenPark_devs, ARRAY_SIZE(havenPark_devs));
 }
 
@@ -187,10 +220,13 @@ static int __init intel_ce_platform_init
 		switch (puma_board_id)
 		{
 		case EASY750_BOARD_ID:
-        	platform_device_register(&ce2700_gpio_leds);
+			platform_device_register(&ce2700_gpio_leds);
 			break;
 		case HVP_BOARD_ID:
-			ce2700_register_led();
+			add_hvp_devs();
+			break;
+		case CGP_BOARD_ID:
+			add_cgp_devs();
 			break;
 		}
 #endif
