# HG changeset patch
# Parent fe34e6b583448918d2d58af2b6f53b15ad458b1c
diff --git a/drivers/mmc/host/sdhci-acpi.c b/drivers/mmc/host/sdhci-acpi.c
--- a/drivers/mmc/host/sdhci-acpi.c
+++ b/drivers/mmc/host/sdhci-acpi.c
@@ -229,8 +229,7 @@ static const struct sdhci_acpi_slot sdhc
 	.caps    = MMC_CAP_8_BIT_DATA | MMC_CAP_NONREMOVABLE | MMC_CAP_HW_RESET
 		| MMC_CAP_1_8V_DDR,
 	.caps2   = MMC_CAP2_HC_ERASE_SZ | MMC_CAP2_POLL_R1B_BUSY |
-		MMC_CAP2_CACHE_CTRL | MMC_CAP2_HS200_1_8V_SDR |
-		MMC_CAP2_CAN_DO_CMDQ,
+		MMC_CAP2_CACHE_CTRL | MMC_CAP2_HS200_1_8V_SDR, 
 	.flags   = SDHCI_ACPI_RUNTIME_PM,
 	.quirks2 = SDHCI_QUIRK2_TUNING_POLL | SDHCI_QUIRK2_PRESET_VALUE_BROKEN,
 	.pm_caps = MMC_PM_TUNING_AFTER_RTRESUME,
@@ -244,7 +243,7 @@ static const struct sdhci_acpi_slot sdhc
 		| MMC_CAP_1_8V_DDR,
 	.caps2   = MMC_CAP2_HC_ERASE_SZ | MMC_CAP2_POLL_R1B_BUSY |
 		MMC_CAP2_CACHE_CTRL | MMC_CAP2_HS200_1_8V_SDR |
-		MMC_CAP2_CAN_DO_CMDQ | MMC_CAP2_NO_SHUTDOWN_BUS,
+		MMC_CAP2_NO_SHUTDOWN_BUS,
 #endif
 	.flags   = SDHCI_ACPI_RUNTIME_PM,
 #ifdef CONFIG_X86_PUMA7
@@ -260,8 +259,7 @@ static const struct sdhci_acpi_slot sdhc
 	.caps    = MMC_CAP_8_BIT_DATA | MMC_CAP_NONREMOVABLE | MMC_CAP_HW_RESET
 		| MMC_CAP_1_8V_DDR,
 	.caps2   = MMC_CAP2_HC_ERASE_SZ | MMC_CAP2_POLL_R1B_BUSY |
-		MMC_CAP2_CACHE_CTRL | MMC_CAP2_HS200_1_8V_SDR |
-		MMC_CAP2_CAN_DO_CMDQ,
+		MMC_CAP2_CACHE_CTRL | MMC_CAP2_HS200_1_8V_SDR,
 	.flags   = SDHCI_ACPI_RUNTIME_PM,
 	.quirks2 = SDHCI_QUIRK2_TUNING_POLL | SDHCI_QUIRK2_PRESET_VALUE_BROKEN,
 	.pm_caps = MMC_PM_TUNING_AFTER_RTRESUME,
