# HG changeset patch
# Parent fe28a7e9111ab5fdfcea631a5e0796ec3646fe48
diff --git a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_drv.c b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_drv.c
--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_drv.c
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_drv.c
@@ -5119,7 +5119,7 @@ static int DWC_ETH_QOS_ioctl(struct net_
 
    default:
       ret = -EOPNOTSUPP;
-      printk(KERN_ALERT "Unsupported IOCTL call\n");
+      DBGPR("Unsupported IOCTL call\n");
    }
    spin_unlock(&pdata->lock);
 
