--- iptables-1.4.21/extensions/libip6t_REJECT.c	2013-11-22 16:48:13.000000000 +0530
+++ iptables-1.4.21_modif/extensions/libip6t_REJECT.c	2016-11-28 16:06:17.072070714 +0530
@@ -35,7 +35,11 @@
 	{"icmp6-port-unreachable", "port-unreach",
 		IP6T_ICMP6_PORT_UNREACH, "ICMPv6 port unreachable"},
 	{"tcp-reset", "tcp-reset",
-		IP6T_TCP_RESET, "TCP RST packet"}
+		IP6T_TCP_RESET, "TCP RST packet"},
+	{"icmp6-policy-fail", "policy-fail",
+		IP6T_ICMP6_POLICY_FAIL, "ICMPv6 policy fail"},
+	{"icmp6-policy-fail-to-gateway", "policy-fail-to-gateway",
+		IP6T_ICMP6_POLICY_FAIL_TO_GATEWAY, "ICMPv6 policy fail to gateway"}
 };
 
 static void
--- iptables-1.4.21/include/linux/netfilter_ipv6/ip6t_REJECT.h	2013-11-22 16:48:13.000000000 +0530
+++ iptables-1.4.21_modif/include/linux/netfilter_ipv6/ip6t_REJECT.h	2016-11-30 18:27:02.359602690 +0530
@@ -10,7 +10,9 @@
 	IP6T_ICMP6_ADDR_UNREACH,
 	IP6T_ICMP6_PORT_UNREACH,
 	IP6T_ICMP6_ECHOREPLY,
-	IP6T_TCP_RESET
+	IP6T_TCP_RESET,
+	IP6T_ICMP6_POLICY_FAIL,
+	IP6T_ICMP6_POLICY_FAIL_TO_GATEWAY
 };
 
 struct ip6t_reject_info {
