# HG changeset patch
# Parent 9abc775a11c0ccd7f7d7419caf16c3d25bb9c3d7

diff --git a/applications/luci-app-firewall/Makefile b/applications/luci-app-firewall/Makefile
--- a/applications/luci-app-firewall/Makefile
+++ b/applications/luci-app-firewall/Makefile
@@ -7,7 +7,8 @@
 include $(TOPDIR)/rules.mk
 
 LUCI_TITLE:=Firewall and Portforwarding application
-LUCI_DEPENDS:=+firewall
+#Remove dependency with fw3 as this will have a conflict wth UGW sl_firewall. Explicitly enabled fw3 for OpenWrt models.
+#LUCI_DEPENDS:=+firewall
 
 include ../../luci.mk
 
