luci samba app to select samba3 if available

diff --git a/applications/luci-app-samba/Makefile b/applications/luci-app-samba/Makefile
index ba599b3..d2599f7 100644
--- a/applications/luci-app-samba/Makefile
+++ b/applications/luci-app-samba/Makefile
@@ -7,7 +7,7 @@
 include $(TOPDIR)/rules.mk
 
 LUCI_TITLE:=Network Shares - Samba SMB/CIFS module
-LUCI_DEPENDS:=+samba36-server
+LUCI_DEPENDS:=+@PACKAGE_samba3:samba3 +@PACKAGE_samba36-server:samba36-server
 
 include ../../luci.mk
 
