--- net-snmp-5.1/include/net-snmp/library/asn1.h	2002-08-27 09:03:16.000000000 +0530
+++ open_snmp/include/net-snmp/library/asn1.h	2009-12-17 12:42:22.000000000 +0530
@@ -50,7 +50,7 @@
 #define MIN_OID_LEN	    2
 #define MAX_OID_LEN	    128 /* max subid's in an oid */
 #ifndef MAX_NAME_LEN            /* conflicts with some libraries */
-#define MAX_NAME_LEN	    MAX_OID_LEN /* obsolete. use MAX_OID_LEN */
+/* #define MAX_NAME_LEN	    MAX_OID_LEN *//* obsolete. use MAX_OID_LEN */
 #endif
 
 #define OID_LENGTH(x)  (sizeof(x)/sizeof(oid))
--- agent/m2m.h	2002-04-20 12:59:59.000000000 +0530
+++ open_snmp/agent/m2m.h	2009-12-17 12:47:23.000000000 +0530
@@ -37,7 +37,7 @@
 /*
  * maximum length for a description field 
  */
-#define MAX_DESCRIPTION_LEN 128
+/*#define MAX_DESCRIPTION_LEN 128*/
 
 /*
  * defines for noting whether the incoming packet is unicast, broadcast,
--- net-snmp-5.1/Makefile.top	2009-12-17 14:04:36.000000000 +0530
+++ open_snmp/Makefile.top	2009-12-17 13:40:35.000000000 +0530
@@ -64,7 +64,7 @@
 # Increment REVISION every release.  
 # Increment AGE if new stuff is available.
 LIB_LD_CMD      = $(LIBTOOL) --mode=link $(CC) $(CFLAGS) -rpath $(libdir) -version-info $(LIBCURRENT):$(LIBREVISION):$(LIBAGE) -o
-LIB_EXTENSION   = la
+LIB_EXTENSION   = a
 LIB_VERSION     =
 LIB_LDCONFIG_CMD = $(LIBTOOL) --mode=finish $(libdir)
 LINK		= $(LIBTOOL) --mode=link $(CC)
--- net-snmp-5.1/agent/snmpd.c	2009-12-17 14:29:15.000000000 +0530
+++ open_snmp/snmpd.c	2009-12-17 14:32:17.000000000 +0530
@@ -1088,7 +1088,7 @@
     signal(SIGPIPE, SIG_IGN);   /* 'Inline' failure of wayward readers */
 #endif
 
-#ifdef IFX_CONFIG_SNMPv3
+#ifdef CONFIG_FEATURE_SNMPv3
 	IFX_create_usm_cmd_file_on_startup();
 	IFX_read_snmpv3_usm_cmd_file();	
 	kill(getpid(), SIGHUP); /* Send a HUP to recreate /var/net-snmp/snmpd.conf */
@@ -1266,12 +1266,12 @@
 	    snmp_log(LOG_INFO, "NET-SNMP version %s restarted\n",
 		     netsnmp_get_version());
             update_config();
-#ifdef IFX_CONFIG_SNMPv3 
+#ifdef CONFIG_FEATURE_SNMPv3 
 			/* v3 user add/del support - IFX */
 			/* First read the config file to determine if anything needs to
 			 * be done. This is written to by our API. */
 			IFX_read_snmpv3_usm_cmd_file();	
-#endif /* IFX_CONFIG_SNMPv3 */
+#endif /* CONFIG_FEATURE_SNMPv3 */
             send_easy_trap(SNMP_TRAP_ENTERPRISESPECIFIC, 3);
         }
 
@@ -1615,7 +1615,7 @@
 
 #endif/*WIN32*/
 
-#ifdef IFX_CONFIG_SNMPv3
+#ifdef CONFIG_FEATURE_SNMPv3
 
 /* This function reads the user command from the cmd file and identifies 
  * whether to add or delete the spcified user*/
@@ -1711,5 +1711,5 @@
 
 }
 
-#endif /* IFX_CONFIG_SNMPv3 */
+#endif /* CONFIG_FEATURE_SNMPv3 */
 
