--- net-snmp-5.1/configure_full	2009-12-17 17:08:05.000000000 +0530
+++ open_snmp/configure_full	2009-12-17 17:07:16.000000000 +0530
@@ -12,8 +12,6 @@
 ## --------------------- ##
 ## M4sh Initialization.  ##
 ## --------------------- ##
-. sysconfig.sh
-. config.sh
 
 # Be Bourne compatible
 if test -n "${ZSH_VERSION+set}" && (emulate sh) >/dev/null 2>&1; then
@@ -9543,7 +9541,7 @@
 
 echo "$as_me:$LINENO: checking for security modules to use" >&5
 echo $ECHO_N "checking for security modules to use... $ECHO_C" >&6
-if test "$IFX_CONFIG_SNMPv3" = "1" ; then
+if test "$CONFIG_FEATURE_SNMPv3" = "y" ; then
 	security_def_list="usm"
 else
 	security_def_list=""
@@ -9578,7 +9576,7 @@
 
 for i in $security_def_list; do
   if echo " $new_with_out_security " | grep " $i " > /dev/null; then
-	if test "$IFX_CONFIG_SNMPv3" = "1" ; then
+	if test "$CONFIG_FEATURE_SNMPv3" = "y" ; then
 	     if test "x$i" = "xusm"; then
 		echo
 		{ { echo "$as_me:$LINENO: error: The USM security module is mandatory" >&5
@@ -10861,7 +10859,7 @@
 #define HAVE_LIBCRYPTO 1
 _ACEOF
 
-if test "$IFX_CONFIG_SNMPv3" = "1" ; then
+if test "$CONFIG_FEATURE_SNMPv3" = "y" ; then
 	LIBS="-lcrypto $LIBS"
 fi
 # LIBS="$USER_OPENSOURCE_DIR/openssl-0.9.7e/libcrypto.a $LIBS"
