diff -Naur net-snmp-5.1/agent/Makefile.in net-snmp-5.1.lq_patched/agent/Makefile.in
--- net-snmp-5.1/agent/Makefile.in	2010-04-29 00:58:48.714640976 +0530
+++ net-snmp-5.1.lq_patched/agent/Makefile.in	2010-04-29 01:33:31.674641037 +0530
@@ -67,7 +67,10 @@
 HELPERLIB	= helpers/libnetsnmphelpers.$(LIB_EXTENSION)$(LIB_VERSION)
 LOCAL_LIBS	= 
 LIBS		= ../snmplib/libnetsnmp.$(LIB_EXTENSION)$(LIB_VERSION) @AGENTLIBS@
-IFX_COMMON_LIBS = -lIFXAPIs $(IFX_LDFLAGS)
+
+ifeq ($(CONFIG_PACKAGE_ltq-mapi),y)
+  IFX_COMMON_LIBS = -lIFXAPIs $(IFX_LDFLAGS)
+endif
 
 OUR_AGENT_LIBS	= $(INSTALLLIBS) $(HELPERLIB) @WRAPLIBS@ $(LIBS) $(IFX_COMMON_LIBS) @DLLIBS@
 CPPFLAGS	= -I../include -I$(top_srcdir)/include \
diff -Naur net-snmp-5.1/agent/mibgroup/Makefile.in net-snmp-5.1.lq_patched/agent/mibgroup/Makefile.in
--- net-snmp-5.1/agent/mibgroup/Makefile.in	2010-04-29 00:58:48.704650218 +0530
+++ net-snmp-5.1.lq_patched/agent/mibgroup/Makefile.in	2010-04-29 01:33:31.664641817 +0530
@@ -4,6 +4,10 @@
 	-I. -I../.. -I$(srcdir) -I$(srcdir)/../.. \
 	-I$(srcdir)/../../snmplib -I$(srcdir)/.. -I.. @CPPFLAGS@ $(LIB_CFLAGS)
 
+ifeq ($(CONFIG_PACKAGE_ltq-mapi),y)
+  CFLAGS += -DCONFIG_PACKAGE_IFX_IFXAPIS
+endif
+
 # Need a special .c -> .o definition here to make sure we place the
 # object files in the sub directory.
 
diff -Naur net-snmp-5.1/agent/mibgroup/mibII/system_mib.c net-snmp-5.1.lq_patched/agent/mibgroup/mibII/system_mib.c
--- net-snmp-5.1/agent/mibgroup/mibII/system_mib.c	2010-04-29 00:58:48.704650218 +0530
+++ net-snmp-5.1.lq_patched/agent/mibgroup/mibII/system_mib.c	2010-04-29 01:33:31.664641817 +0530
@@ -57,7 +57,9 @@
 	 *
 	 *********************/
 /* MIB-2 persistent support */
-#define MIB2_SYSTEM_PERSIST
+#ifdef CONFIG_PACKAGE_IFX_IFXAPIS
+  #define MIB2_SYSTEM_PERSIST
+#endif
 
 #ifdef MIB2_SYSTEM_PERSIST
 #define FILE_RC_CONF	"/flash/rc.conf"
diff -Naur net-snmp-5.1/agent/snmpd.c net-snmp-5.1.lq_patched/agent/snmpd.c
--- net-snmp-5.1/agent/snmpd.c	2010-04-29 00:58:48.694642591 +0530
+++ net-snmp-5.1.lq_patched/agent/snmpd.c	2010-04-29 01:33:31.654642303 +0530
@@ -161,11 +161,14 @@
 #include <net-snmp/agent/table_iterator.h>
 #include "mib_module_includes.h"
 
-#include "ifx_common.h"
-#include "ifx_snmpdefs.h"
+#include <ifx_config.h>
+
+#ifdef CONFIG_PACKAGE_IFX_IFXAPIS
+  #include "ifx_common.h"
+  #include "ifx_snmpdefs.h"
+#endif
 
 //509141:tc.chen added start
-#include <ifx_config.h>
 #ifdef IFX_CONFIG_CPE_DSL_MIB
 #include "../apps/snmptrapd_handlers.h"
 //#include "../apps/notification_log.h"
@@ -375,8 +378,10 @@
      */
     snmp_close(main_session);
 #endif
+#ifdef CONFIG_PACKAGE_IFX_IFXAPIS
     /* Remove the PID file */
     ifx_rm_pid_file(SNMP_PID_FILE_PREFIX);
+#endif
 }
 
 #ifdef SIGHUP
@@ -1027,7 +1032,9 @@
     /*
      * Initialize the PID file for snmp daemon - IFX
      */
+#ifdef CONFIG_PACKAGE_IFX_IFXAPIS
     ifx_create_pid_file(SNMP_PID_FILE_PREFIX);
+#endif
 
     if ((ret = init_master_agent()) != 0) {
         /*
diff -Naur net-snmp-5.1/Makefile.top net-snmp-5.1.lq_patched/Makefile.top
--- net-snmp-5.1/Makefile.top	2010-04-29 00:58:48.694642591 +0530
+++ net-snmp-5.1.lq_patched/Makefile.top	2010-04-29 01:33:31.644642316 +0530
@@ -46,8 +46,10 @@
 #CFLAGS		= @CFLAGS@ @DEVFLAGS@ $(IFX_CFLAGS)
 CFLAGS		= @DEVFLAGS@ $(IFX_CFLAGS)
 EXTRACPPFLAGS	= @EXTRACPPFLAGS@
-#LDFLAGS	= @LDFLAGS@ $(IFX_LDFLAGS) -lIFXAPIs
-LDFLAGS		= $(IFX_LDFLAGS) -lIFXAPIs
+ifeq ($(CONFIG_PACKAGE_ltq-mapi),y)
+  #LDFLAGS	= @LDFLAGS@ $(IFX_LDFLAGS) -lIFXAPIs
+  LDFLAGS	= $(IFX_LDFLAGS) -lIFXAPIs
+endif
 LIBTOOL		= @LIBTOOL@ @LIBTOOLFLAGS@
 EXEEXT		= @EXEEXT@
 
