--- a/setup.py	2016-05-14 23:49:49.000000000 +0530
+++ b/setup.py	2017-03-09 18:12:28.473332000 +0530
@@ -14,6 +14,15 @@
 
 import os
 import sys
+depend_link=""
+def check_connect():
+	global depend_link
+	response = os.system("wget -q -O /dev/null google.com -T 10")
+	if response == 0:
+		depend_link=['https://pypi.python.org/simple/meld3']
+	else:
+		depend_link=['https://mts.lantiq.com/~wlnsw/dist/meld3-1.0.2.tar.gz']
+	return
 
 if sys.version_info[:2] < (2, 4) or sys.version_info[0] > 2:
     msg = ("Supervisor requires Python 2.4 or later but does not work on "
@@ -60,7 +69,7 @@
 version_txt = os.path.join(here, 'supervisor/version.txt')
 supervisor_version = open(version_txt).read().strip()
 
-dist = setup(
+dist = check_connect(), setup(
     name='supervisor',
     version=supervisor_version,
     license='BSD-derived (http://www.repoze.org/LICENSE.txt)',
@@ -72,6 +81,7 @@
     author_email="chrism@plope.com",
     packages=find_packages(),
     install_requires=requires,
+  	dependency_links=depend_link,
     extras_require={'iterparse': ['cElementTree >= 1.0.2']},
     tests_require=['mock >= 0.5.0'],
     include_package_data=True,
