--- ntpclient-2007.orig/Makefile	2007-12-31 22:43:56.000000000 +0530
+++ ntpclient-2007/Makefile	2015-04-10 15:17:17.814101000 +0530
@@ -12,9 +12,9 @@
 # CFLAGS += -DPRECISION_SIOCGSTAMP
 CFLAGS += -DENABLE_DEBUG
 CFLAGS += -DENABLE_REPLAY
-# CFLAGS += -DUSE_OBSOLETE_GETTIMEOFDAY
+CFLAGS += -DUSE_OBSOLETE_GETTIMEOFDAY
 
-LDFLAGS += -lrt
+#LDFLAGS += -lrt
 
 all: ntpclient
 
@@ -22,8 +22,7 @@
 	./ntpclient -d -r <test.dat
 
 ntpclient: ntpclient.o phaselock.o
-
-ntpclient.o phaselock.o: ntpclient.h
+	$(CC) $(CFLAGS) $^ -o $@
 
 adjtimex: adjtimex.o
 
