diff -rupN ntpclient-2007_old/ntpclient.c ntpclient-2007_new/ntpclient.c
--- ntpclient-2007_old/ntpclient.c	2015-08-20 18:41:00.273017344 +0530
+++ ntpclient-2007_new/ntpclient.c	2015-08-20 18:31:00.174120312 +0530
@@ -534,13 +534,48 @@ static void primary_loop(int usd, struct
 		if (pack_len<0) {
 			perror("recvfrom");
 		} else if (pack_len>0 && (unsigned)pack_len<sizeof_incoming){
+			time_t time_before_sync = 0;
+			time_t time_after_sync = 0;
+			struct tm *local_time_before_sync = NULL;
+			struct tm *local_time_after_sync = NULL;
+			long int time_before_sync_sec = 0;
+			long int time_after_sync_sec = 0;
+			long int diff = 0;
 			get_packet_timestamp(usd, &udp_arrival_ntp);
 			if (check_source(pack_len, &sa_xmit, sa_xmit_len, ntpc)!=0) continue;
+			time_before_sync = time(NULL);
+			local_time_before_sync = localtime(&time_before_sync);
+			if (local_time_before_sync != NULL) {
+				time_before_sync_sec = (( ((local_time_before_sync->tm_hour) * 3600) + ((local_time_before_sync->tm_min) * 60)) + (local_time_before_sync->tm_sec));
+			}
+			diff = time_before_sync_sec;
 			if (rfc1305print(incoming_word, &udp_arrival_ntp, ntpc, &error)!=0) continue;
+			time_after_sync = time(NULL);
+			local_time_after_sync = localtime(&time_after_sync);
+			if (local_time_after_sync != NULL) {
+				time_after_sync_sec = (( ((local_time_after_sync->tm_hour) * 3600) + ((local_time_after_sync->tm_min) * 60)) + (local_time_after_sync->tm_sec));
+			}	
+			diff = time_after_sync_sec - diff;
+			if (diff < 0) {
+				diff*=-1;
+			}
+			if ((local_time_before_sync != NULL) && (local_time_after_sync != NULL) && ((local_time_before_sync->tm_year !=  local_time_after_sync->tm_year)||
+			(local_time_before_sync->tm_yday != local_time_after_sync->tm_yday) ||
+		 	(diff > 60)) ){
+				printf("Generating ntp_sync_notify %ld  -  %ld\n",time_before_sync, time_after_sync);
+				{
+					char sBuff[512] = { 0 };
+					sprintf(sBuff, "/opt/lantiq/etc/time_notify.sh NOTIFY_NTP_SYNC %lu %lu", time_before_sync, time_after_sync);
+					system(sBuff);
+				}
+			}
+
 			success++;
 			/* udp_handle(usd,incoming,pack_len,&sa_xmit,sa_xmit_len); */
 		} else {
 			printf("Ooops.  pack_len=%d\n",pack_len);
+			printf("Generating ntp_sync_error\n");
+			system("/opt/lantiq/etc/time_notify.sh NOTIFY_NTP_SYNC_ERROR");
 			fflush(stdout);
 		}
 		/* best rollover option: specify -g, -s, and -l.
