diff -Naur knopflerfish_osgi_3.2.0/osgi/init-tests.xargs knopflerfish_osgi_3.2.0-ltq/osgi/init-tests.xargs
--- knopflerfish_osgi_3.2.0/osgi/init-tests.xargs	2012-02-13 15:02:16.214431441 +0530
+++ knopflerfish_osgi_3.2.0-ltq/osgi/init-tests.xargs	2012-02-13 19:51:14.945679372 +0530
@@ -35,9 +35,9 @@
 -Forg.knopflerfish.framework.debug.certificates=false
 
 # JKSValidator properties
--Forg.knopflerfish.framework.validator=JKSValidator
--Forg.knopflerfish.framework.validator.jks.ca_certs=${user.dir}/test_jars/test.castore
--Forg.knopflerfish.framework.validator.jks.ca_certs_password=catest
+#-Forg.knopflerfish.framework.validator=JKSValidator
+#-Forg.knopflerfish.framework.validator.jks.ca_certs=${user.dir}/test_jars/test.castore
+#-Forg.knopflerfish.framework.validator.jks.ca_certs_password=catest
 
 # Comma-separated list of packges exported by system classloader
 -Forg.osgi.framework.system.packages.extra=
