diff -Npur compat-wireless-2016-05-12/include/net/cfg80211.h compat-wireless-2016-05-12-modif/include/net/cfg80211.h
--- compat-wireless-2016-05-12/include/net/cfg80211.h	2016-10-18 11:42:49.000000000 +0300
+++ compat-wireless-2016-05-12-modif/include/net/cfg80211.h	2016-10-18 12:20:03.000000000 +0300
@@ -790,6 +790,9 @@ enum station_parameters_apply_mask {
  * @capability: station capability
  * @ext_capab: extended capabilities of the station
  * @ext_capab_len: number of extended capabilities
+ * @resp: last association response frame
+ *	(or NULL for no change)
+ * @resp_len: length of last association response
  * @supported_channels: supported channels in IEEE 802.11 format
  * @supported_channels_len: number of supported channels
  * @supported_oper_classes: supported oper classes in IEEE 802.11 format
@@ -816,6 +819,8 @@ struct station_parameters {
 	u16 capability;
 	const u8 *ext_capab;
 	u8 ext_capab_len;
+	u8 *resp;
+	size_t resp_len;
 	const u8 *supported_channels;
 	u8 supported_channels_len;
 	const u8 *supported_oper_classes;
diff -Npur compat-wireless-2016-05-12/net/wireless/nl80211.c compat-wireless-2016-05-12-modif/net/wireless/nl80211.c
--- compat-wireless-2016-05-12/net/wireless/nl80211.c	2016-10-18 11:42:49.000000000 +0300
+++ compat-wireless-2016-05-12-modif/net/wireless/nl80211.c	2016-10-18 12:26:17.000000000 +0300
@@ -4469,6 +4469,11 @@ static int nl80211_new_station(struct sk
 	params.listen_interval =
 		nla_get_u16(info->attrs[NL80211_ATTR_STA_LISTEN_INTERVAL]);
 
+	if (info->attrs[NL80211_ATTR_FRAME]) {
+		params.resp = nla_data(info->attrs[NL80211_ATTR_FRAME]);
+		params.resp_len = nla_len(info->attrs[NL80211_ATTR_FRAME]);
+	}
+
 	if (info->attrs[NL80211_ATTR_STA_SUPPORT_P2P_PS]) {
 		u8 tmp;
 
