diff -Npur compat-wireless-2016-10-08-clean/net/wireless/core.c compat-wireless-2016-10-08/net/wireless/core.c
--- compat-wireless-2016-10-08-clean/net/wireless/core.c	2016-12-22 10:45:16.000000000 +0200
+++ compat-wireless-2016-10-08/net/wireless/core.c	2016-12-22 10:49:10.000000000 +0200
@@ -568,9 +568,9 @@ static int wiphy_verify_combinations(str
 				CFG80211_MAX_NUM_DIFFERENT_CHANNELS))
 			return -EINVAL;
 
-		/* DFS only works on one channel. */
+		/* Two different channels allowed by WLAN driver during CSA */
 		if (WARN_ON(c->radar_detect_widths &&
-			    (c->num_different_channels > 1)))
+			    (c->num_different_channels > 2)))
 			return -EINVAL;
 
 		if (WARN_ON(!c->n_limits))
