diff -Npur compat-wireless-2016-05-12-modif/net/wireless/core.c compat-wireless-2016-05-12-modifi/net/wireless/core.c
--- compat-wireless-2016-05-12-modif/net/wireless/core.c	2016-05-14 15:27:16.000000000 +0300
+++ compat-wireless-2016-05-12-modifi/net/wireless/core.c	2016-10-18 13:31:31.000000000 +0300
@@ -900,6 +900,12 @@ void cfg80211_unregister_wdev(struct wir
 }
 EXPORT_SYMBOL(cfg80211_unregister_wdev);
 
+struct workqueue_struct *cfg80211_get_cfg80211_wq(void)
+{
+	return cfg80211_wq;
+}
+EXPORT_SYMBOL(cfg80211_get_cfg80211_wq);
+
 static const struct device_type wiphy_type = {
 	.name	= "wlan",
 };
