diff -Npur compat-wireless-2016-05-12-modific/net/wireless/wext-compat.h compat-wireless-2016-05-12-modifica/net/wireless/wext-compat.h
--- compat-wireless-2016-05-12-modific/net/wireless/wext-compat.h	2016-10-18 13:34:46.000000000 +0300
+++ compat-wireless-2016-05-12-modifica/net/wireless/wext-compat.h	2016-10-18 14:57:46.000000000 +0300
@@ -4,6 +4,8 @@
 #include <net/iw_handler.h>
 #include <linux/wireless.h>
 
+#define CPTCFG_CFG80211_WEXT_EXPORT
+
 #ifdef CPTCFG_CFG80211_WEXT_EXPORT
 #define EXPORT_WEXT_HANDLER(h) EXPORT_SYMBOL_GPL(h)
 #else
