diff -Npur compat-wireless-2016-05-12-orig/net/wireless/chan.c compat-wireless-2016-05-12/net/wireless/chan.c
--- compat-wireless-2016-05-12-orig/net/wireless/chan.c	2016-10-25 08:58:43.000000000 +0300
+++ compat-wireless-2016-05-12/net/wireless/chan.c	2016-10-25 09:32:26.000000000 +0300
@@ -811,10 +811,15 @@ static bool _cfg80211_reg_can_beacon(str
 	if (check_no_ir)
 		prohibited_flags |= IEEE80211_CHAN_NO_IR;
 
-	if (cfg80211_chandef_dfs_required(wiphy, chandef, iftype) > 0 &&
-	    cfg80211_chandef_dfs_available(wiphy, chandef)) {
-		/* We can skip IEEE80211_CHAN_NO_IR if chandef dfs available */
-		prohibited_flags = IEEE80211_CHAN_DISABLED;
+	if (cfg80211_chandef_dfs_required(wiphy, chandef, iftype) > 0) {
+		if (cfg80211_chandef_dfs_available(wiphy, chandef)) {
+			/* We can skip IEEE80211_CHAN_NO_IR if chandef dfs available */
+			prohibited_flags = IEEE80211_CHAN_DISABLED;
+		}
+		else if (cfg80211_chandef_dfs_usable(wiphy, chandef)) {
+			/* mtlk wlan driver only operates with usable channels */
+			prohibited_flags &= ~IEEE80211_CHAN_RADAR;
+		}
 	}
 
 	res = cfg80211_chandef_usable(wiphy, chandef, prohibited_flags);
