diff -Npur compat-wireless-2016-10-08-orig/net/wireless/nl80211.c compat-wireless-2016-10-08/net/wireless/nl80211.c
--- compat-wireless-2016-10-08-orig/net/wireless/nl80211.c	2017-03-09 09:57:36.499174232 +0200
+++ compat-wireless-2016-10-08/net/wireless/nl80211.c	2017-03-09 10:00:29.352065180 +0200
@@ -7310,9 +7310,6 @@ static int nl80211_start_radar_detection
 	if (err)
 		return err;
 
-	if (netif_carrier_ok(dev))
-		return -EBUSY;
-
 	if (wdev->cac_started)
 		return -EBUSY;
 
