Index: iptables-1.4.6/extensions/libipt_webstr.c
===================================================================
--- iptables-1.4.6.orig/extensions/libipt_webstr.c
+++ iptables-1.4.6/extensions/libipt_webstr.c
@@ -76,7 +76,7 @@ parse(int c, char **argv, int invert, un
 
 	switch (c) {
 	case '1':
-		check_inverse(optarg, &invert, &optind, 0);
+		xtables_check_inverse(optarg, &invert, &optind, 0, argv);
 		parse_string(argv[optind-1], stringinfo);
 		if (invert)
 			stringinfo->invert = 1;
@@ -85,7 +85,7 @@ parse(int c, char **argv, int invert, un
 		break;
 
 	case '2':
-		check_inverse(optarg, &invert, &optind, 0);
+		xtables_check_inverse(optarg, &invert, &optind, 0, argv);
 		parse_string(argv[optind-1], stringinfo);
 		if (invert)
 			stringinfo->invert = 1;
@@ -94,7 +94,7 @@ parse(int c, char **argv, int invert, un
 		break;
 
 	case '3':
-		check_inverse(optarg, &invert, &optind, 0);
+		xtables_check_inverse(optarg, &invert, &optind, 0, argv);
 		parse_string(argv[optind-1], stringinfo);
 		if (invert)
 			stringinfo->invert = 1;
