diff -Nuar iptables-1.4.6_org/extensions/libip6t_REJECT.c iptables-1.4.6/extensions/libip6t_REJECT.c
--- iptables-1.4.6_org/extensions/libip6t_REJECT.c	2009-12-09 19:42:05.000000000 +0530
+++ iptables-1.4.6/extensions/libip6t_REJECT.c	2011-11-30 19:04:49.788881320 +0530
@@ -33,7 +33,9 @@
 	{"icmp6-port-unreachable", "port-unreach",
 		IP6T_ICMP6_PORT_UNREACH, "ICMPv6 port unreachable"},
 	{"tcp-reset", "tcp-reset",
-		IP6T_TCP_RESET, "TCP RST packet"}
+		IP6T_TCP_RESET, "TCP RST packet"},
+	{"icmp6-invalid-source", "invalid-src",
+		IP6T_ICMP6_INVALID_SRC, "ICMPv6 invalid source address"}
 };
 
 static void
diff -Nuar iptables-1.4.6_org/include/linux/netfilter_ipv6/ip6t_REJECT.h iptables-1.4.6/include/linux/netfilter_ipv6/ip6t_REJECT.h
--- iptables-1.4.6_org/include/linux/netfilter_ipv6/ip6t_REJECT.h	2009-12-09 19:42:05.000000000 +0530
+++ iptables-1.4.6/include/linux/netfilter_ipv6/ip6t_REJECT.h	2011-11-30 19:04:04.566849893 +0530
@@ -8,7 +8,8 @@
 	IP6T_ICMP6_ADDR_UNREACH,
 	IP6T_ICMP6_PORT_UNREACH,
 	IP6T_ICMP6_ECHOREPLY,
-	IP6T_TCP_RESET
+	IP6T_TCP_RESET,
+	IP6T_ICMP6_INVALID_SRC
 };
 
 struct ip6t_reject_info {
