Index: iptables-1.4.6/include/linux/netfilter_ipv4/ipt_ECN.h
===================================================================
--- iptables-1.4.6.orig/include/linux/netfilter_ipv4/ipt_ECN.h	2009-12-09 19:42:05.000000000 +0530
+++ iptables-1.4.6/include/linux/netfilter_ipv4/ipt_ECN.h	2012-06-19 08:01:50.000000000 +0530
@@ -8,7 +8,7 @@
 */
 #ifndef _IPT_ECN_TARGET_H
 #define _IPT_ECN_TARGET_H
-#include <linux/netfilter_ipv4/ipt_DSCP.h>
+#include <linux/netfilter/xt_DSCP.h>
 
 #define IPT_ECN_IP_MASK	(~IPT_DSCP_MASK)
 
Index: iptables-1.4.6/include/linux/netfilter_ipv4/ipt_ecn.h
===================================================================
--- iptables-1.4.6.orig/include/linux/netfilter_ipv4/ipt_ecn.h	2009-12-09 19:42:05.000000000 +0530
+++ iptables-1.4.6/include/linux/netfilter_ipv4/ipt_ecn.h	2012-06-19 08:01:19.000000000 +0530
@@ -8,7 +8,7 @@
 */
 #ifndef _IPT_ECN_H
 #define _IPT_ECN_H
-#include <linux/netfilter_ipv4/ipt_dscp.h>
+#include <linux/netfilter/xt_DSCP.h>
 
 #define IPT_ECN_IP_MASK	(~IPT_DSCP_MASK)
 
