# HG changeset patch
# Parent 04c6244039faf7dfa07dddf4e2cd571a66111950

diff --git a/arch/mips/include/asm/uaccess.h b/arch/mips/include/asm/uaccess.h
--- a/arch/mips/include/asm/uaccess.h
+++ b/arch/mips/include/asm/uaccess.h
@@ -1504,7 +1504,7 @@ static inline long
 		"move\t$4, %1\n\t"
 		"move\t$5, %2\n\t"
 		"move\t$6, %3\n\t"
-#ifdef CONFIG_EVA
+#ifndef CONFIG_EVA
 		__MODULE_JAL(__strncpy_from_kernel_nocheck_asm)
 #else
 		__MODULE_JAL(__strncpy_from_user_nocheck_asm)
