Add simple device tree support to dm9000 driver.

--- a/drivers/net/ethernet/davicom/dm9000.c
+++ b/drivers/net/ethernet/davicom/dm9000.c
@@ -1683,11 +1683,17 @@ dm9000_drv_remove(struct platform_device
 	return 0;
 }
 
+static const struct of_device_id dm9000_drv_of_match[] = {
+	{.compatible = "davicom,dm9000" },
+	{},
+};
+
 static struct platform_driver dm9000_driver = {
 	.driver	= {
 		.name    = "dm9000",
 		.owner	 = THIS_MODULE,
 		.pm	 = &dm9000_drv_pm_ops,
+		.of_match_table = dm9000_drv_of_match,
 	},
 	.probe   = dm9000_probe,
 	.remove  = dm9000_drv_remove,
