# HG changeset patch
# Parent 9714295fcdae61dbddacfb3a17ec8a380f9f629e

diff --git a/arch/mips/pci/pci.c b/arch/mips/pci/pci.c
--- a/arch/mips/pci/pci.c
+++ b/arch/mips/pci/pci.c
@@ -116,6 +116,17 @@ static void pcibios_scanbus(struct pci_c
 			pci_enable_bridges(bus);
 		}
 	}
+
+	/* Configure PCI Express settings */
+	if (bus && !pci_has_flag(PCI_PROBE_ONLY)) {
+		struct pci_bus *child;
+		list_for_each_entry(child, &bus->children, node) {
+			struct pci_dev *self = child->self;
+			if (!self)
+				continue;
+			pcie_bus_configure_settings(child, self->pcie_mpss);
+		}
+	}
 }
 
 #ifdef CONFIG_OF
