# HG changeset patch
# Parent 551280a3721956aa90f89fde69a72b72d765f440
diff --git a/drivers/net/ethernet/lantiq/Kconfig b/drivers/net/ethernet/lantiq/Kconfig
--- a/drivers/net/ethernet/lantiq/Kconfig
+++ b/drivers/net/ethernet/lantiq/Kconfig
@@ -32,4 +32,6 @@ config LANTIQ_ETH_FRAMEWORK
 	---help---
 	Lantiq framework for ethernet drivers
 
+source "drivers/net/ethernet/lantiq/switch-api/Kconfig"
+
 endif # NET_VENDOR_LANTIQ
diff --git a/drivers/net/ethernet/lantiq/Makefile b/drivers/net/ethernet/lantiq/Makefile
--- a/drivers/net/ethernet/lantiq/Makefile
+++ b/drivers/net/ethernet/lantiq/Makefile
@@ -1,2 +1,3 @@
 obj-$(CONFIG_LANTIQ_VRX318) += ltq_vrx318.o
 obj-$(CONFIG_LANTIQ_ETH_FRAMEWORK) += lantiq_eth_framework.o
+obj-$(CONFIG_LTQ_ETHSW_API) += switch-api/
