# HG changeset patch
# Parent 89f21b8d8ca0d44ea0ab2961bdf0886075078f9d

diff --git a/drivers/pinctrl/pinctrl-xway.c b/drivers/pinctrl/pinctrl-xway.c
--- a/drivers/pinctrl/pinctrl-xway.c
+++ b/drivers/pinctrl/pinctrl-xway.c
@@ -215,6 +215,7 @@ static const unsigned pins_asc0_cts_rts[
 static const unsigned pins_stp[] = {GPIO4, GPIO5, GPIO6};
 static const unsigned pins_nmi[] = {GPIO8};
 static const unsigned pins_mdio[] = {GPIO42, GPIO43};
+static const unsigned pins_arc_jtag[] = {GPIO9, GPIO10, GPIO11, GPIO14, GPIO19};
 
 static const unsigned pins_ebu_a24[] = {GPIO13};
 static const unsigned pins_ebu_clk[] = {GPIO21};
@@ -355,6 +356,7 @@ static const struct ltq_pin_group xway_g
 	GRP_MUX("nand wr", EBU, pins_nand_wr),
 	GRP_MUX("nand wp", EBU, pins_nand_wp),
 	GRP_MUX("nand se", EBU, pins_nand_se),
+	GRP_MUX("arc_jtag", GPIO, pins_arc_jtag),
 };
 
 static const struct ltq_pin_group ase_grps[] = {
@@ -416,6 +418,7 @@ static const char * const xrx_pci_grps[]
 						"gnt3", "gnt4",
 						"req1", "req2",
 						"req3", "req4"};
+static const char * const xrx_arc_jtag_grps[] = {"arc_jtag"};
 
 /* ase */
 static const char * const ase_exin_grps[] = {"exin0", "exin1", "exin2"};
@@ -453,6 +456,7 @@ static const struct ltq_pmx_func xrx_fun
 	{"pci",		ARRAY_AND_SIZE(xrx_pci_grps)},
 	{"ebu",		ARRAY_AND_SIZE(xrx_ebu_grps)},
 	{"mdio",	ARRAY_AND_SIZE(xrx_mdio_grps)},
+	{"arc_jtag",	ARRAY_AND_SIZE(xrx_arc_jtag_grps)},
 };
 
 static const struct ltq_pmx_func ase_funcs[] = {
