# HG changeset patch
# Parent c5ac39fb1d4e0bc660a80da17fdf011c8731696d
diff --git a/drivers/pinctrl/pinctrl-xway.c b/drivers/pinctrl/pinctrl-xway.c
--- a/drivers/pinctrl/pinctrl-xway.c
+++ b/drivers/pinctrl/pinctrl-xway.c
@@ -87,6 +87,7 @@
 enum xway_mux {
 	XWAY_MUX_GPIO = 0,
 	XWAY_MUX_SPI,
+	XWAY_MUX_USIF,
 	XWAY_MUX_ASC,
 	XWAY_MUX_PCI,
 	XWAY_MUX_CGU,
@@ -118,8 +119,8 @@
 	MFP_XWAY(GPIO7, GPIO,	CGU,	PCI,	NONE),
 	MFP_XWAY(GPIO8, GPIO,	CGU,	NMI,	NONE),
 	MFP_XWAY(GPIO9, GPIO,	ASC,	SPI,	EXIN),
-	MFP_XWAY(GPIO10, GPIO,	ASC,	SPI,	NONE),
-	MFP_XWAY(GPIO11, GPIO,	ASC,	PCI,	SPI),
+	MFP_XWAY(GPIO10, GPIO,	USIF,	SPI,	NONE),
+	MFP_XWAY(GPIO11, GPIO,	USIF,	PCI,	SPI),
 	MFP_XWAY(GPIO12, GPIO,	ASC,	NONE,	NONE),
 	MFP_XWAY(GPIO13, GPIO,	EBU,	SPI,	NONE),
 	MFP_XWAY(GPIO14, GPIO,	CGU,	PCI,	NONE),
@@ -127,7 +128,7 @@
 	MFP_XWAY(GPIO16, GPIO,	SPI,	NONE,	JTAG),
 	MFP_XWAY(GPIO17, GPIO,	SPI,	NONE,	JTAG),
 	MFP_XWAY(GPIO18, GPIO,	SPI,	NONE,	JTAG),
-	MFP_XWAY(GPIO19, GPIO,	PCI,	NONE,	NONE),
+	MFP_XWAY(GPIO19, GPIO,	USIF,	NONE,	NONE),
 	MFP_XWAY(GPIO20, GPIO,	JTAG,	NONE,	NONE),
 	MFP_XWAY(GPIO21, GPIO,	PCI,	EBU,	GPT),
 	MFP_XWAY(GPIO22, GPIO,	SPI,	NONE,	NONE),
@@ -259,6 +260,7 @@
 static const unsigned pins_spi_cs5[] = {GPIO9};
 static const unsigned pins_spi_cs6[] = {GPIO11};
 
+static const unsigned pins_usif[] = {GPIO10, GPIO11, GPIO19};
 static const unsigned pins_gpt1[] = {GPIO28};
 static const unsigned pins_gpt2[] = {GPIO21};
 static const unsigned pins_gpt3[] = {GPIO6};
@@ -317,6 +319,7 @@
 	GRP_MUX("spi_cs4", SPI, pins_spi_cs4),
 	GRP_MUX("spi_cs5", SPI, pins_spi_cs5),
 	GRP_MUX("spi_cs6", SPI, pins_spi_cs6),
+	GRP_MUX("usif", USIF, pins_usif),
 	GRP_MUX("asc0", ASC, pins_asc0),
 	GRP_MUX("asc0 cts rts", ASC, pins_asc0_cts_rts),
 	GRP_MUX("stp", STP, pins_stp),
@@ -384,6 +387,7 @@
 						"spi_cs2", "spi_cs3",
 						"spi_cs4", "spi_cs5",
 						"spi_cs6"};
+static const char * const xway_usif_grps[] = {"usif"};
 static const char * const xway_cgu_grps[] = {"clkout0", "clkout1",
 						"clkout2", "clkout3"};
 static const char * const xway_ebu_grps[] = {"ebu a23", "ebu a24",
@@ -446,6 +450,7 @@
 
 static const struct ltq_pmx_func xrx_funcs[] = {
 	{"spi",		ARRAY_AND_SIZE(xway_spi_grps)},
+	{"usif",	ARRAY_AND_SIZE(xway_usif_grps)},
 	{"asc",		ARRAY_AND_SIZE(xway_asc_grps)},
 	{"cgu",		ARRAY_AND_SIZE(xway_cgu_grps)},
 	{"jtag",	ARRAY_AND_SIZE(xway_jtag_grps)},
