# HG changeset patch
# Parent e2d649c6d4a67082ef51a234f9404d2e360c4134
diff --git a/drivers/pinctrl/pinctrl-xway.c b/drivers/pinctrl/pinctrl-xway.c
--- a/drivers/pinctrl/pinctrl-xway.c
+++ b/drivers/pinctrl/pinctrl-xway.c
@@ -121,7 +121,7 @@
 	MFP_XWAY(GPIO9, GPIO,	ASC,	SPI,	EXIN),
 	MFP_XWAY(GPIO10, GPIO,	USIF,	SPI,	NONE),
 	MFP_XWAY(GPIO11, GPIO,	USIF,	PCI,	SPI),
-	MFP_XWAY(GPIO12, GPIO,	ASC,	NONE,	NONE),
+	MFP_XWAY(GPIO12, GPIO,	USIF,	NONE,	NONE),
 	MFP_XWAY(GPIO13, GPIO,	EBU,	SPI,	NONE),
 	MFP_XWAY(GPIO14, GPIO,	CGU,	PCI,	NONE),
 	MFP_XWAY(GPIO15, GPIO,	SPI,	JTAG,	NONE),
@@ -261,6 +261,7 @@
 static const unsigned pins_spi_cs6[] = {GPIO11};
 
 static const unsigned pins_usif[] = {GPIO10, GPIO11, GPIO19};
+static const unsigned pins_usif_uart[] = {GPIO11, GPIO12};
 static const unsigned pins_gpt1[] = {GPIO28};
 static const unsigned pins_gpt2[] = {GPIO21};
 static const unsigned pins_gpt3[] = {GPIO6};
@@ -320,6 +321,7 @@
 	GRP_MUX("spi_cs5", SPI, pins_spi_cs5),
 	GRP_MUX("spi_cs6", SPI, pins_spi_cs6),
 	GRP_MUX("usif", USIF, pins_usif),
+	GRP_MUX("usif_uart", USIF, pins_usif_uart),
 	GRP_MUX("asc0", ASC, pins_asc0),
 	GRP_MUX("asc0 cts rts", ASC, pins_asc0_cts_rts),
 	GRP_MUX("stp", STP, pins_stp),
@@ -388,6 +390,7 @@
 						"spi_cs4", "spi_cs5",
 						"spi_cs6"};
 static const char * const xway_usif_grps[] = {"usif"};
+static const char * const xway_usif_uart_grps[] = {"usif_uart"};
 static const char * const xway_cgu_grps[] = {"clkout0", "clkout1",
 						"clkout2", "clkout3"};
 static const char * const xway_ebu_grps[] = {"ebu a23", "ebu a24",
@@ -451,6 +454,7 @@
 static const struct ltq_pmx_func xrx_funcs[] = {
 	{"spi",		ARRAY_AND_SIZE(xway_spi_grps)},
 	{"usif",	ARRAY_AND_SIZE(xway_usif_grps)},
+	{"usif_uart",	ARRAY_AND_SIZE(xway_usif_uart_grps)},
 	{"asc",		ARRAY_AND_SIZE(xway_asc_grps)},
 	{"cgu",		ARRAY_AND_SIZE(xway_cgu_grps)},
 	{"jtag",	ARRAY_AND_SIZE(xway_jtag_grps)},
