# HG changeset patch
# Parent 86711fa984f31c36f48b50148372a8c731b1cd56

diff --git a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
--- a/drivers/crypto/Kconfig
+++ b/drivers/crypto/Kconfig
@@ -387,4 +387,27 @@ config CRYPTO_DEV_ATMEL_SHA
 	  To compile this driver as a module, choose M here: the module
 	  will be called atmel-sha.
 
+config CRYPTO_DEV_LANTIQ_DEU
+    tristate "Support for Lantiq hardware crypto engine"
+	depends on LANTIQ && SOC_XWAY
+	select CRYPTO_AES
+	select CRYPTO_DES
+	select CRYPTO_CBC
+	select CRYPTO_CTR
+	select CRYPTO_ECB
+	select CRYPTO_SHA1
+	select CRYPTO_MD5
+	select CRYPTO_HASH
+	select CRYPTO_HMAC
+	select CRYPTO_ARC4
+	select CRYPTO_ALGAPI
+	select CRYPTO_BLKCIPHER
+	help 
+	  Selects Hardware accelerated crypto drivers. 
+	  Supports:
+	    AES/DES/ARC4/SHA-1/MD5 (DANUBE) 
+	    AES/DES/ARC4/SHA-1/MD5/HMAC (xrx200, xrx300, xrx330) 
+    
+source "drivers/crypto/lantiq_deu/Kconfig"
 endif # CRYPTO_HW
+
diff --git a/drivers/crypto/Makefile b/drivers/crypto/Makefile
--- a/drivers/crypto/Makefile
+++ b/drivers/crypto/Makefile
@@ -21,3 +21,4 @@ obj-$(CONFIG_CRYPTO_DEV_NX) += nx/
 obj-$(CONFIG_CRYPTO_DEV_ATMEL_AES) += atmel-aes.o
 obj-$(CONFIG_CRYPTO_DEV_ATMEL_TDES) += atmel-tdes.o
 obj-$(CONFIG_CRYPTO_DEV_ATMEL_SHA) += atmel-sha.o
+obj-$(CONFIG_CRYPTO_DEV_LANTIQ_DEU) += lantiq_deu/
