# HG changeset patch
# Parent d8045a6fca9b863a1cc26f6f7b482ff20083b3fd
diff --git a/drivers/pinctrl/pinmux.c b/drivers/pinctrl/pinmux.c
--- a/drivers/pinctrl/pinmux.c
+++ b/drivers/pinctrl/pinmux.c
@@ -101,10 +101,10 @@ static int pin_request(struct pinctrl_de
 
 	if (gpio_range) {
 		/* There's no need to support multiple GPIO requests */
-		if (desc->gpio_owner) {
+		if (desc->mux_owner) {
 			dev_err(pctldev->dev,
 				"pin %s already requested by %s; cannot claim for %s\n",
-				desc->name, desc->gpio_owner, owner);
+				desc->name, desc->mux_owner, owner);
 			goto out;
 		}
 
