This patch provides GET/SET macros for nfmark

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -34,6 +34,14 @@
 #include <linux/netdev_features.h>
 #include <net/flow_keys.h>
 
+/* reccommended as it's faster*/
+#define GET_DATA_FROM_MARK_OPT(mark, mask, pos, value) value = ((mark & mask) >> pos)
+#define SET_DATA_FROM_MARK_OPT(mark, mask, pos, value) mark &= ~mask; mark |= ((value << pos) & mask)
+
+#define MASK(pos, len) (((1<<len)-1)<<pos)
+#define GET_DATA_FROM_MARK(mark, pos, len, value) GET_DATA_FROM_MARK_OPT(mark, MASK(pos, len), pos, value)
+#define SET_DATA_FROM_MARK(mark, pos, len, value) SET_DATA_FROM_MARK_OPT(mark, MASK(pos, len), pos, value)
+
 /* Don't change this without changing skb_csum_unnecessary! */
 #define CHECKSUM_NONE 0
 #define CHECKSUM_UNNECESSARY 1
