# HG changeset patch
# Parent 3a90a403bc9c594ba611f908f17b262bfebf4531

diff --git a/net/ipv4/Kconfig b/net/ipv4/Kconfig
--- a/net/ipv4/Kconfig
+++ b/net/ipv4/Kconfig
@@ -663,6 +663,12 @@ config TCP_CONG_CUBIC
 	depends on !TCP_CONG_ADVANCED
 	default y
 
+config BOOST_MARK
+        tristate "Support for prioritiy marking of packet for Device prioritization"
+        default n
+        help
+                Enable packet marking for device prioritization to boost downstream traffic
+
 config DEFAULT_TCP_CONG
 	string
 	default "bic" if DEFAULT_BIC
diff --git a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
--- a/net/ipv4/ip_output.c
+++ b/net/ipv4/ip_output.c
@@ -91,6 +91,12 @@
 int sysctl_ip_default_ttl __read_mostly = IPDEFTTL;
 EXPORT_SYMBOL(sysctl_ip_default_ttl);
 
+#ifdef CONFIG_BOOST_MARK
+/*Hook to mark skb for device prioritization */
+int32_t (*dev_boost_hook)(struct sk_buff *) = NULL;
+EXPORT_SYMBOL(dev_boost_hook);
+#endif
+
 /* Generate a checksum for an outgoing IP datagram. */
 void ip_send_check(struct iphdr *iph)
 {
@@ -200,6 +206,12 @@ static inline int ip_finish_output2(stru
 		skb = skb2;
 	}
 
+#ifdef CONFIG_BOOST_MARK
+	if(dev_boost_hook != NULL) {
+		dev_boost_hook(skb);
+	}
+#endif
+
 #if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
 #if defined(CONFIG_LTQ_PPA_MPE_IP97)
     struct iphdr *hdr = ip_hdr(skb);
