Index: linux-3.10.12/drivers/net/Kconfig
===================================================================
--- linux-3.10.12.orig/drivers/net/Kconfig	2014-03-19 12:27:21.000000000 +0530
+++ linux-3.10.12/drivers/net/Kconfig	2014-03-19 12:38:37.000000000 +0530
@@ -491,5 +491,6 @@
 source "drivers/net/lantiq_ptm/Kconfig"
 
 source "drivers/net/lantiq_ppa/Kconfig"
+source "drivers/net/lantiq_ppa/Kconfig.swaccl"
 
 endif # NETDEVICES
