# HG changeset patch
# Parent c908718d5f5e250431b7297ee06b21a2756a675e

diff --git a/net/ipv4/ip_gre.c b/net/ipv4/ip_gre.c
--- a/net/ipv4/ip_gre.c
+++ b/net/ipv4/ip_gre.c
@@ -965,6 +965,17 @@ static struct pernet_operations ipgre_ta
 	.size = sizeof(struct ip_tunnel_net),
 };
 
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+extern uint32_t (*ppa_is_ipv4_gretap_fn)(struct net_device *dev);
+static uint32_t ppa_is_ipv4_gretap(struct net_device *dev)
+{
+
+  if( dev->netdev_ops == (&gre_tap_netdev_ops) )
+    return 1;
+  return 0;
+}
+#endif
+
 static int __init ipgre_init(void)
 {
 	int err;
@@ -993,6 +1004,10 @@ static int __init ipgre_init(void)
 	if (err < 0)
 		goto tap_ops_failed;
 
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+  ppa_is_ipv4_gretap_fn = ppa_is_ipv4_gretap;
+#endif
+
 	return 0;
 
 tap_ops_failed:
@@ -1008,6 +1023,9 @@ pnet_tap_faied:
 
 static void __exit ipgre_fini(void)
 {
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+  ppa_is_ipv4_gretap_fn = NULL;
+#endif
 	rtnl_link_unregister(&ipgre_tap_ops);
 	rtnl_link_unregister(&ipgre_link_ops);
 	if (gre_del_protocol(&ipgre_protocol, GREPROTO_CISCO) < 0)
diff --git a/net/ipv6/ip6_gre.c b/net/ipv6/ip6_gre.c
--- a/net/ipv6/ip6_gre.c
+++ b/net/ipv6/ip6_gre.c
@@ -1643,6 +1643,18 @@ static struct rtnl_link_ops ip6gre_tap_o
 	.fill_info	= ip6gre_fill_info,
 };
 
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+extern uint32_t (*ppa_is_ipv6_gretap_fn)(struct net_device *dev);
+static uint32_t ppa_is_ipv6_gretap(struct net_device *dev)
+{
+
+  if( dev->netdev_ops == (&ip6gre_tap_netdev_ops) )
+    return 1;
+  
+  return 0;
+}
+#endif
+
 /*
  *	And now the modules code and kernel interface.
  */
@@ -1671,6 +1683,10 @@ static int __init ip6gre_init(void)
 	if (err < 0)
 		goto tap_ops_failed;
 
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+  ppa_is_ipv6_gretap_fn = ppa_is_ipv6_gretap;
+#endif
+
 out:
 	return err;
 
@@ -1685,6 +1701,9 @@ add_proto_failed:
 
 static void __exit ip6gre_fini(void)
 {
+#if defined(CONFIG_LTQ_PPA_API) || defined(CONFIG_LTQ_PPA_API_MODULE)
+  ppa_is_ipv6_gretap_fn = NULL;
+#endif
 	rtnl_link_unregister(&ip6gre_tap_ops);
 	rtnl_link_unregister(&ip6gre_link_ops);
 	inet6_del_protocol(&ip6gre_protocol, IPPROTO_GRE);
