# HG changeset patch
# Parent bd880bd67c058368c9f60bb327f9778baf71fc02

diff --git a/net/bridge/br_netlink.c b/net/bridge/br_netlink.c
--- a/net/bridge/br_netlink.c
+++ b/net/bridge/br_netlink.c
@@ -173,6 +173,13 @@ void br_ifinfo_notify(int event, struct 
 	if (!port)
 		return;
 
+/* crash observed during br-lan bring down/up
+ * sometimes port->dev received as null,same passed to 
+ * br_fill_ifinfo.
+*/
+	if (!port->dev)
+		return;
+
 	net = dev_net(port->dev);
 	br_debug(port->br, "port %u(%s) event %d\n",
 		 (unsigned int)port->port_no, port->dev->name, event);
