diff -ru --show-c-function linux-3.10.12-orig/net/wireless/core.c linux-3.10.12/net/wireless/core.c
--- linux-3.10.12-orig/net/wireless/core.c	2013-09-14 16:55:12.000000000 +0300
+++ linux-3.10.12/net/wireless/core.c	2015-10-19 15:05:53.000000000 +0300
@@ -832,6 +832,12 @@ void cfg80211_unregister_wdev(struct wir
 }
 EXPORT_SYMBOL(cfg80211_unregister_wdev);
 
+struct workqueue_struct *cfg80211_get_cfg80211_wq(void)
+{
+	return cfg80211_wq;
+}
+EXPORT_SYMBOL(cfg80211_get_cfg80211_wq);
+
 static struct device_type wiphy_type = {
 	.name	= "wlan",
 };
diff -ru --show-c-function linux-3.10.12-orig/net/wireless/mlme.c linux-3.10.12/net/wireless/mlme.c
--- linux-3.10.12-orig/net/wireless/mlme.c	2015-10-20 15:28:44.000000000 +0300
+++ linux-3.10.12/net/wireless/mlme.c	2015-10-20 11:19:43.000000000 +0300
@@ -861,7 +861,7 @@ void cfg80211_dfs_channels_update_work(s
 				continue;
 
 			timeout = c->dfs_state_entered +
-				  IEEE80211_DFS_MIN_NOP_TIME_MS;
+				msecs_to_jiffies(IEEE80211_DFS_MIN_NOP_TIME_MS);
 
 			if (time_after_eq(jiffies, timeout)) {
 				c->dfs_state = NL80211_DFS_USABLE;
