# HG changeset patch
# Parent a8f0bc838f93a750d3d66912ef9ca257ec87485d
diff --git a/arch/mips/include/asm/time.h b/arch/mips/include/asm/time.h
--- a/arch/mips/include/asm/time.h
+++ b/arch/mips/include/asm/time.h
@@ -55,6 +55,11 @@ extern unsigned int __weak get_c0_compar
 extern int r4k_clockevent_init(void);
 extern int smtc_clockevent_init(void);
 extern int gic_clockevent_init(void);
+#ifdef CONFIG_CEVT_GPTC
+extern int gptc_module_init(void);
+extern int gptc_clockevent_init(void);
+extern int gptc_clocksource_init(void);
+#endif
 
 static inline int mips_clockevent_init(void)
 {
diff --git a/arch/mips/kernel/smp-cmp.c b/arch/mips/kernel/smp-cmp.c
--- a/arch/mips/kernel/smp-cmp.c
+++ b/arch/mips/kernel/smp-cmp.c
@@ -237,6 +237,11 @@ static void cmp_smp_finish(void)
 	/* CDFIXME: remove this? */
 	write_c0_compare(read_c0_count() + (8 * mips_hpt_frequency / HZ));
 
+#ifdef CONFIG_CEVT_GPTC
+	pr_info("GPTC clock event init on cpu: %d\n", smp_processor_id());
+    gptc_clockevent_init();
+#endif
+
 #ifdef CONFIG_MIPS_MT_FPAFF
 	/* If we have an FPU, enroll ourselves in the FPU-full mask */
 	if (cpu_has_fpu)
