# HG changeset patch
# Parent bbcd44ecbe2c8304041fd601f6ce13fe1309c2d8
diff --git a/drivers/net/usb/cdc_ether.c b/drivers/net/usb/cdc_ether.c
--- a/drivers/net/usb/cdc_ether.c
+++ b/drivers/net/usb/cdc_ether.c
@@ -633,6 +633,13 @@ static const struct usb_device_id	produc
 	.driver_info = 0,
 },
 
+/* TP-LINK UE300 Change ethx to wwan0 */
+{
+	USB_DEVICE_AND_INTERFACE_INFO(0x2357, 0x0601, USB_CLASS_COMM,
+			USB_CDC_SUBCLASS_ETHERNET, USB_CDC_PROTO_NONE),
+	.driver_info = (unsigned long)&wwan_info,
+},
+
 /* Realtek RTL8152 Based USB 2.0 Ethernet Adapters */
 #if defined(CONFIG_USB_RTL8152) || defined(CONFIG_USB_RTL8152_MODULE)
 {
