Index: mtd-utils-1.4.6/Makefile
===================================================================
--- mtd-utils-1.4.6.orig/Makefile	2011-09-21 12:05:04.000000000 +0530
+++ mtd-utils-1.4.6/Makefile	2012-10-15 19:26:16.000000000 +0530
@@ -30,13 +30,14 @@
 	ubidetach ubinize ubiformat ubirename mtdinfo ubirsvol
 
 BINS = $(MTD_BINS)
-BINS += mkfs.ubifs/mkfs.ubifs
+#BINS += mkfs.ubifs/mkfs.ubifs
 BINS += $(addprefix ubi-utils/,$(UBI_BINS))
 SCRIPTS = flash_eraseall
 
 TARGETS = $(BINS)
 TARGETS += lib/libmtd.a
 TARGETS += ubi-utils/libubi.a
+TARGETS += crc32.o
 
 OBJDEPS = $(BUILDDIR)/include/version.h
 
Index: mtd-utils-1.4.6/rbtree.h
===================================================================
--- mtd-utils-1.4.6.orig/rbtree.h	2011-09-21 12:05:04.000000000 +0530
+++ mtd-utils-1.4.6/rbtree.h	2012-10-15 19:27:01.000000000 +0530
@@ -94,7 +94,7 @@
 #ifndef	_LINUX_RBTREE_H
 #define	_LINUX_RBTREE_H
 
-#include <linux/kernel.h>
+/*#include <linux/kernel.h>*/
 #include <linux/stddef.h>
 
 struct rb_node
