Index: br2684ctl-20040226.orig/Makefile
===================================================================
--- br2684ctl-20040226.orig.orig/Makefile	2009-12-17 15:11:00.000000000 +0530
+++ br2684ctl-20040226.orig/Makefile	2010-01-11 16:50:07.000000000 +0530
@@ -20,22 +20,22 @@
 	
 #$(CC) $(INCPATH) $(CFLAGS) -o $@  br2684ctl.c -l$(LIBATM) -L$(LIBATMPATH) -l$(LIBIFXCOMMON) -L$(IFXCOMMONPATH)lib
 
-ifeq ($(CONFIG_PACKAGE_KMOD_IFX_NFEXT),y)
+ifeq ($(CONFIG_FEATURE_LTQ_VLAN),y)
 br2684ctl:br2684ctl.c
-	$(CC) $(INCPATH) $(CFLAGS) -DIFX_BR2684CLIENT $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
+	$(CC) $(CFLAGS) -DIFX_BR2684CLIENT $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
 else
 br2684ctl:br2684ctl.c
-	$(CC) $(INCPATH) $(CFLAGS) -DIFX_BR2684CLIENT -DCONFIG_WAN_VLAN_SUPPORT $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
+	$(CC) $(CFLAGS) -DIFX_BR2684CLIENT $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
 endif
 
-ifeq ($(CONFIG_PACKAGE_KMOD_IFX_NFEXT),y)
+ifeq ($(CONFIG_FEATURE_LTQ_VLAN),y)
 br2684ctld:br2684ctl.c
	rm -f br2684ctl.o
-	$(CC) $(INCPATH) $(CFLAGS) -DIFX_BR2684DEAMON $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
+	$(CC) $(CFLAGS) -DIFX_BR2684DEAMON $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
 else
 br2684ctld:br2684ctl.c
 	rm -f br2684ctl.o
-	$(CC) $(INCPATH) $(CFLAGS) -DIFX_BR2684DEAMON -DCONFIG_WAN_VLAN_SUPPORT $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
+	$(CC) $(CFLAGS) -DIFX_BR2684DEAMON $(LDFLAGS) -o $@  br2684ctl.c -l$(LIBATM) #-l$(LIBIFXCOMMON)
 endif
 
 clean:
Index: br2684ctl-20040226.orig/br2684ctl.c
===================================================================
--- br2684ctl-20040226.orig.orig/br2684ctl.c	2009-12-17 15:11:00.000000000 +0530
+++ br2684ctl-20040226.orig/br2684ctl.c	2010-01-11 16:51:45.000000000 +0530
@@ -22,7 +22,7 @@
 #include <string.h>
 #include <syslog.h>
 #include <atm.h>
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT
+#ifdef CONFIG_FEATURE_LTQ_VLAN
 #include <linux/types.h>
 #endif
 //605161:fchang.removed #include <linux/atmdev.h>
@@ -71,7 +71,7 @@
 extern void ifx_rm_pid_file_atexit(void);
 #ifndef IFX_BR2684CLIENT /* [ IFX_BR2684CLIENT */
 
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT /*For DAEMON only */
+#ifdef CONFIG_FEATURE_LTQ_VLAN /*For DAEMON only */
 struct vlan_config_cmd {
 	int vlan_cmd;
 	int vlan_if_num;
@@ -143,7 +143,7 @@
 	return 0;
 }
 
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT
+#ifdef CONFIG_FEATURE_LTQ_VLAN
 int wan_vlan_config(struct vlan_config_cmd *to_send, int enable, int flag)
 {
 	int err=0;
@@ -288,7 +288,7 @@
 	printf("Usage:\n");
 	printf("\tCreate: %s [-b] [-p 0|1] [-s sndbuf] [-q qos] [-c number] [-e 0|1] [-a [itf.]vpi.vci]\n", s);
 	printf("\tClose: %s -k <nas_index>\n", s);
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT
+#ifdef CONFIG_FEATURE_LTQ_VLAN
 	printf("\tGlobal VLAN Tag Flag: %s -v [-g 0|1] \n", s);
 	printf("\tSet VLAN Tag: %s -v -i <nas_index> [-t 0|1] [-u] <vlan_id> <802.1p> \n", s);
 #endif
@@ -492,7 +492,7 @@
 	struct atm_qos reqqos;
 	lastsock=-1;
 	lastitf=0;
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT
+#ifdef CONFIG_FEATURE_LTQ_VLAN
 	char itf_name[20];
 #endif
 	int fd = -1;
@@ -504,7 +504,7 @@
 	optind = 0;
 	if (argc>1) {
 	
-#ifdef CONFIG_PACKAGE_KMOD_IFX_NFEXT
+#ifdef CONFIG_FEATURE_LTQ_VLAN
 		if (strcmp(argv[1], "-v")==0) {
 			syslog(LOG_INFO,"\nEnter VLAN options\n");
 			memset(&g_vlan_cfg, 0x00, sizeof(g_vlan_cfg));
