--- br2684ctl-20040226.orig/br2684ctl.c	2010-08-27 05:19:38.554555378 +0530
+++ br2684ctl-20040226.orig.lq_patched/br2684ctl.c	2010-08-27 05:22:54.714558694 +0530
@@ -121,6 +121,10 @@
 			struct atm_newif_br2684 ni;
 			ni.backend_num = ATM_BACKEND_BR2684;
 			ni.media = BR2684_MEDIA_ETHERNET;
+#ifdef BR2684_FLAG_ROUTED
+			if (payload == 0)
+				ni.media |= BR2684_FLAG_ROUTED; //Point-Point LINK Type
+#endif
 			ni.mtu = 1500;
 			ni.payload = payload;	/* bridged or routed */
 
