Index: br2684ctl-20040226.orig/br2684ctl.c
===================================================================
--- br2684ctl-20040226.orig.orig/br2684ctl.c	2012-07-11 14:31:53.000000000 +0530
+++ br2684ctl-20040226.orig/br2684ctl.c	2012-07-11 14:33:22.000000000 +0530
@@ -312,6 +312,7 @@
 #ifdef IFX_BR2684DEAMON /*[IFX_BR2684DEAMON */
 #define PROGRAM_NAME	"br2684ctl"
 #define CLOSEIF_OPTION	"-k"
+#define CLOSE_ALLIF_OPTION	"-K"
 
 struct br2684_sock {
 	int ifnum;
@@ -342,6 +343,7 @@
 	char *ptemp = NULL;
 	int sockfd = -1;
 	int i = 0;
+	int deamon_pid = -1;
 
 //	syslog(LOG_INFO,"br2684ctld HUP received");
 	cmd_fd = open(BR2684_CMDFILE,O_RDONLY);
@@ -389,6 +391,37 @@
 		}
 		if(brsock == NULL)
 			terminate(SIGTERM);
+
+
+		return;
+	}
+	else if(strstr(strcmd, CLOSE_ALLIF_OPTION)) {
+		struct br2684_sock *ptrcur = NULL;
+		struct br2684_sock *ptrpre = NULL;
+		ptrcur = brsock;
+		while(ptrcur) {
+//		syslog(LOG_INFO,"br2684ctld : Closing socket for nas%d",ptrcur->ifnum);
+			close(ptrcur->sockfd);
+			brsock = ptrcur->next;
+			free(ptrcur);
+			ptrcur = brsock;
+		}
+		if(brsock == NULL) {
+//		terminate(SIGTERM);
+			deamon_pid = ifx_get_process_pid(DEAMON_PIDFILE_NAME);
+
+			if(deamon_pid < 1) {
+				syslog(LOG_ERR,"Could not get pid of %s!!!",BR2684CTL_DEAMON);
+			} else {
+				syslog(LOG_INFO,"br2684_ctl : Sending SIGHUP to %d",deamon_pid);
+				kill(deamon_pid,SIGKILL);
+			}
+
+		}
+		else {
+			/* Can't be possible */
+		}
+
 		return;
 	}
 
