--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,25 @@
+
+2014-08-04:
+	Support for new vlan structure added
+
+2013-07-25:
+	Replace few IFX macros with LTQ
+
+2012-07-11:
+	Included support option to remove all the nas (atm) interfaces in one go
+
+2010-08-27:
+	Support for Point-to-point link type
+
+2009-11-13:
+	Preparing build environment
+
+2009-11-12:
+	Libifx build dependency fix
+
+2009-08-19:
+	The actual br2684ctl application is divided into 2 application : br2684ctld, br2684ctl
+
+2009-08-05:
+	Added PID file process function definitions
+
